From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
index e3d7ff50817e0ea889524ef7221055b5a25beeb1..1acd506f7c0679fa9f69b6ab221002b28d00c3e5 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -401,10 +401,13 @@ public class MineshaftPieces {
                         BlockPos worldPos = this.getWorldPos(1, 0, i8);
                         if (box.isInside(worldPos) && this.isInterior(level, 1, 0, i8, box)) {
                             this.hasPlacedSpider = true;
-                            level.setBlock(worldPos, Blocks.SPAWNER.defaultBlockState(), 2);
-                            if (level.getBlockEntity(worldPos) instanceof SpawnerBlockEntity spawnerBlockEntity) {
-                                spawnerBlockEntity.setEntityId(EntityType.CAVE_SPIDER, random);
-                            }
+                            // CraftBukkit start
+                            // level.setBlock(worldPos, Blocks.SPAWNER.defaultBlockState(), 2);
+                            // if (level.getBlockEntity(worldPos) instanceof SpawnerBlockEntity spawnerBlockEntity) {
+                            //     spawnerBlockEntity.setEntityId(EntityType.CAVE_SPIDER, random);
+                            // }
+                            this.placeCraftSpawner(level, worldPos, org.bukkit.entity.EntityType.CAVE_SPIDER, 2);
+                            // CraftBukkit end
                         }
                     }
                 }
