--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -428,6 +428,8 @@
 
                 if (structureboundingbox.isInside(blockposition_mutableblockposition)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     generatoraccessseed.setBlock(blockposition_mutableblockposition, Blocks.SPAWNER.defaultBlockState(), 2);
                     TileEntity tileentity = generatoraccessseed.getBlockEntity(blockposition_mutableblockposition);
 
@@ -436,6 +438,9 @@
 
                         tileentitymobspawner.setEntityId(EntityTypes.BLAZE, randomsource);
                     }
+                    */
+                    placeCraftSpawner(generatoraccessseed, blockposition_mutableblockposition, org.bukkit.entity.EntityType.BLAZE, 2);
+                    // CraftBukkit end
                 }
             }
 
