From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/entity/vault/VaultBlockEntity.java b/net/minecraft/world/level/block/entity/vault/VaultBlockEntity.java
index e7b08f042f566a6a7bb44cf143880fdea07474a2..517786b0a3002559beabf1673adc7a15f542405c 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultBlockEntity.java
@@ -272,6 +272,11 @@ public class VaultBlockEntity extends BlockEntity {
                     if (!list.isEmpty()) {
                         player.awardStat(Stats.ITEM_USED.get(stack.getItem()));
                         stack.consume(config.keyItem().getCount(), player);
+                        // CraftBukkit start
+                        org.bukkit.event.block.BlockDispenseLootEvent vaultDispenseLootEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDispenseLootEvent(level, pos, player, list);
+                        if (vaultDispenseLootEvent.isCancelled()) return;
+                        list = vaultDispenseLootEvent.getDispensedLoot().stream().map(org.bukkit.craftbukkit.inventory.CraftItemStack::asNMSCopy).toList();
+                        // CraftBukkit end
                         unlock(level, state, pos, config, serverData, sharedData, list);
                         serverData.addToRewardedPlayers(player);
                         sharedData.updateConnectedPlayersWithinRange(level, pos, serverData, config, config.deactivationRange());
@@ -294,6 +299,11 @@ public class VaultBlockEntity extends BlockEntity {
                 ItemStack randomDisplayItemFromLootTable = getRandomDisplayItemFromLootTable(
                     level, pos, config.overrideLootTableToDisplay().orElse(config.lootTable())
                 );
+                // CraftBukkit start
+                org.bukkit.event.block.VaultDisplayItemEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callVaultDisplayItemEvent(level, pos, randomDisplayItemFromLootTable);
+                if (event.isCancelled()) return;
+                randomDisplayItemFromLootTable = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDisplayItem());
+                // CraftBukkit end
                 sharedData.setDisplayItem(randomDisplayItemFromLootTable);
             }
         }
