--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -64,9 +_,22 @@
         return this.rewardedPlayers.contains(player.getUUID());
     }
 
+    // Paper start - Vault API
+    public boolean removeFromRewardedPlayers(UUID uuid) {
+        final boolean result = this.rewardedPlayers.remove(uuid);
+        if (result)
+            this.markChanged();
+
+        return result;
+    }
+
     @VisibleForTesting
-    public void addToRewardedPlayers(Player player) {
-        this.rewardedPlayers.add(player.getUUID());
+    public boolean addToRewardedPlayers(Player player) {
+        return addToRewardedPlayers(player.getUUID());
+    }
+
+    public boolean addToRewardedPlayers(UUID uuid) {
+        final boolean result = this.rewardedPlayers.add(uuid);
         if (this.rewardedPlayers.size() > 128) {
             Iterator<UUID> iterator = this.rewardedPlayers.iterator();
             if (iterator.hasNext()) {
@@ -76,6 +_,8 @@
         }
 
         this.markChanged();
+        return result;
+        // Paper end - Vault API
     }
 
     public long stateUpdatingResumesAt() {
