From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index 29138621d6e56f03aa34a15609f7f8f8c1aca958..9bf95455f93a17c53bf457d90fd30a3595227569 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -66,7 +66,12 @@ public class VaultServerData {
 
     @VisibleForTesting
     public void addToRewardedPlayers(Player player) {
-        this.rewardedPlayers.add(player.getUUID());
+    // Paper start - Vault API
+        addToRewardedPlayers(player.getUUID());
+    }
+    public boolean addToRewardedPlayers(final java.util.UUID player) {
+        final boolean removed = this.rewardedPlayers.add(player);
+    // Paper end - Vault API
         if (this.rewardedPlayers.size() > 128) {
             Iterator<UUID> iterator = this.rewardedPlayers.iterator();
             if (iterator.hasNext()) {
@@ -76,6 +81,7 @@ public class VaultServerData {
         }
 
         this.markChanged();
+        return removed; // Paper - Vault API
     }
 
     public long stateUpdatingResumesAt() {
@@ -131,4 +137,15 @@ public class VaultServerData {
     public float ejectionProgress() {
         return this.totalEjectionsNeeded == 1 ? 1.0F : 1.0F - Mth.inverseLerp((float)this.getItemsToEject().size(), 1.0F, (float)this.totalEjectionsNeeded);
     }
+
+    // Paper start - Vault API
+    public boolean removeFromRewardedPlayers(final UUID uuid) {
+        if (this.rewardedPlayers.remove(uuid)) {
+            this.markChanged();
+            return true;
+        }
+
+        return false;
+    }
+    // Paper end - Vault API
 }
