--- a/net/minecraft/stats/ServerStatisticManager.java
+++ b/net/minecraft/stats/ServerStatisticManager.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.stats;
 
 import com.google.common.collect.Maps;
@@ -158,13 +159,12 @@
     }
 
     private <T> Optional<Statistic<T>> a(StatisticWrapper<T> statisticwrapper, String s) {
-        Optional optional = Optional.ofNullable(MinecraftKey.a(s));
-        IRegistry iregistry = statisticwrapper.getRegistry();
+        // CraftBukkit - decompile error start
+        Optional<MinecraftKey> optional = Optional.ofNullable(MinecraftKey.a(s));
+        IRegistry<T> iregistry = statisticwrapper.getRegistry();
 
-        iregistry.getClass();
-        optional = optional.flatMap(iregistry::getOptional);
-        statisticwrapper.getClass();
-        return optional.map(statisticwrapper::b);
+        return optional.flatMap(iregistry::getOptional).map(statisticwrapper::b);
+        // CraftBukkit - decompile error end
     }
 
     private static NBTTagCompound a(JsonObject jsonobject) {
