--- a/net/minecraft/commands/ICommandListener.java
+++ b/net/minecraft/commands/ICommandListener.java
@@ -23,6 +23,13 @@
         public boolean shouldBroadcastCommands() {
             return false;
         }
+
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandListenerWrapper wrapper) {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // CraftBukkit end
     };
 
     void sendMessage(IChatBaseComponent ichatbasecomponent, UUID uuid);
@@ -32,4 +39,6 @@
     boolean shouldSendFailure();
 
     boolean shouldBroadcastCommands();
+
+    org.bukkit.command.CommandSender getBukkitSender(CommandListenerWrapper wrapper); // CraftBukkit
 }
