--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -23,6 +23,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.ticks.TickPriority;
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public abstract class DiodeBlock extends HorizontalDirectionalBlock {
 
@@ -59,8 +60,18 @@
             boolean flag1 = this.shouldTurnOn(world, pos, state);
 
             if (flag && !flag1) {
+                // CraftBukkit start
+                if (CraftEventFactory.callRedstoneChange(world, pos, 15, 0).getNewCurrent() != 0) {
+                    return;
+                }
+                // CraftBukkit end
                 world.setBlock(pos, (BlockState) state.setValue(DiodeBlock.POWERED, false), 2);
             } else if (!flag) {
+                // CraftBukkit start
+                if (CraftEventFactory.callRedstoneChange(world, pos, 0, 15).getNewCurrent() != 15) {
+                    return;
+                }
+                // CraftBukkit end
                 world.setBlock(pos, (BlockState) state.setValue(DiodeBlock.POWERED, true), 2);
                 if (!flag1) {
                     world.scheduleTick(pos, (Block) this, this.getDelay(state), TickPriority.VERY_HIGH);
