--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
@@ -219,13 +219,18 @@
                                 }
 
                                 entityinsentient.setPersistenceRequired();
-                                Optional optional1 = mobspawnerdata.getEquipment();
+                                Optional<net.minecraft.world.entity.EquipmentTable> optional1 = mobspawnerdata.getEquipment(); // CraftBukkit - decompile error
 
                                 Objects.requireNonNull(entityinsentient);
                                 optional1.ifPresent(entityinsentient::equip);
                             }
 
-                            if (!worldserver.tryAddFreshEntityWithPassengers(entity)) {
+                            // CraftBukkit start
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callTrialSpawnerSpawnEvent(entity, blockposition).isCancelled()) {
+                                return Optional.empty();
+                            }
+                            if (!worldserver.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRIAL_SPAWNER)) {
+                                // CraftBukkit end
                                 return Optional.empty();
                             } else {
                                 TrialSpawner.a trialspawner_a = this.isOminous ? TrialSpawner.a.OMINOUS : TrialSpawner.a.NORMAL;
