--- a/net/minecraft/util/PlaceholderLookupProvider.java
+++ b/net/minecraft/util/PlaceholderLookupProvider.java
@@ -52,6 +_,13 @@
                             )
                         );
                 }
+
+                // Paper start add method to get the value for pre-filling builders in the reg mod API
+                @Override
+                public HolderLookup.Provider lookupForValueCopyViaBuilders() {
+                    return PlaceholderLookupProvider.this.context;
+                }
+                // Paper end - add method to get the value for pre-filling builders in the reg mod API
             }
         );
     }
