--- a/net/minecraft/world/level/levelgen/MobSpawnerPatrol.java
+++ b/net/minecraft/world/level/levelgen/MobSpawnerPatrol.java
@@ -119,7 +119,7 @@
 
                 entitymonsterpatrolling.setPosition((double) blockposition.getX(), (double) blockposition.getY(), (double) blockposition.getZ());
                 entitymonsterpatrolling.prepare(worldserver, worldserver.getDamageScaler(blockposition), EnumMobSpawn.PATROL, (GroupDataEntity) null, (NBTTagCompound) null);
-                worldserver.addAllEntities(entitymonsterpatrolling);
+                worldserver.addAllEntities(entitymonsterpatrolling, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PATROL); // CraftBukkit
                 return true;
             } else {
                 return false;
