--- a/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
@@ -14,7 +_,7 @@
     private static final int MAX_PAYLOAD_SIZE = 32767;
     public static final StreamCodec<FriendlyByteBuf, ServerboundCustomPayloadPacket> STREAM_CODEC = CustomPacketPayload.<FriendlyByteBuf>codec(
             id -> DiscardedPayload.codec(id, 32767),
-            Util.make(Lists.newArrayList(new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC)), list -> {})
+            java.util.Collections.emptyList() // CraftBukkit - treat all packets the same
         )
         .map(ServerboundCustomPayloadPacket::new, ServerboundCustomPayloadPacket::payload);
 
