--- a/net/minecraft/server/network/CommonListenerCookie.java
+++ b/net/minecraft/server/network/CommonListenerCookie.java
@@ -3,8 +_,8 @@
 import com.mojang.authlib.GameProfile;
 import net.minecraft.server.level.ClientInformation;
 
-public record CommonListenerCookie(GameProfile gameProfile, int latency, ClientInformation clientInformation, boolean transferred) {
+public record CommonListenerCookie(GameProfile gameProfile, int latency, ClientInformation clientInformation, boolean transferred, @javax.annotation.Nullable String brandName, java.util.Set<String> channels) { // Paper
     public static CommonListenerCookie createInitial(GameProfile gameProfile, boolean transferred) {
-        return new CommonListenerCookie(gameProfile, 0, ClientInformation.createDefault(), transferred);
+        return new CommonListenerCookie(gameProfile, 0, ClientInformation.createDefault(), transferred, null, new java.util.HashSet<>()); // Paper
     }
 }
