--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -52,12 +_,26 @@
     public ClientInformation clientInformation;
     private @Nullable SynchronizeRegistriesTask synchronizeRegistriesTask;
     private @Nullable PrepareSpawnTask prepareSpawnTask;
+    public io.papermc.paper.connection.PaperPlayerConfigurationConnection paperConnection; // Paper
 
     public ServerConfigurationPacketListenerImpl(final MinecraftServer server, final Connection connection, final CommonListenerCookie cookie) {
         super(server, connection, cookie);
         this.gameProfile = cookie.gameProfile();
         this.clientInformation = cookie.clientInformation();
-    }
+        this.paperConnection = new io.papermc.paper.connection.PaperPlayerConfigurationConnection(this); // Paper
+    }
+
+    // Paper start - configuration phase API
+    @Override
+    public io.papermc.paper.connection.PlayerCommonConnection getApiConnection() {
+        return this.paperConnection;
+    }
+
+    @Override
+    public net.kyori.adventure.audience.Audience getAudience() {
+        return this.paperConnection.getAudience();
+    }
+    // Paper end - configuration phase API
 
     @Override
     protected GameProfile playerProfile() {
@@ -66,6 +_,11 @@
 
     @Override
     public void onDisconnect(final DisconnectionDetails details) {
+        // Paper start - Debugging
+        if (this.server.isDebugging()) {
+            ServerConfigurationPacketListenerImpl.LOGGER.info("{} ({}) lost connection: {}, while in configuration phase {}", this.gameProfile.name(), this.gameProfile.id(), details.reason().getString(), this.currentTask != null ? this.currentTask.type().id() : "null");
+        } else
+        // Paper end
         LOGGER.info("{} ({}) lost connection: {}", this.gameProfile.name(), this.gameProfile.id(), details.reason().getString());
         if (this.prepareSpawnTask != null) {
             this.prepareSpawnTask.close();
@@ -81,10 +_,15 @@
     }
 
     public void startConfiguration() {
+        new io.papermc.paper.event.connection.configuration.PlayerConnectionInitialConfigureEvent(this.paperConnection).callEvent(); // Paper
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
         ServerLinks serverLinks = this.server.serverLinks();
         if (!serverLinks.isEmpty()) {
-            this.send(new ClientboundServerLinksPacket(serverLinks.untrust()));
+            // Paper start
+            org.bukkit.craftbukkit.CraftServerLinks links = new org.bukkit.craftbukkit.CraftServerLinks(serverLinks);
+            new org.bukkit.event.player.PlayerLinksSendEvent(this.paperConnection, links).callEvent();
+            this.send(new ClientboundServerLinksPacket(links.getServerLinks().untrust()));
+            // Paper end
         }
 
         LayeredRegistryAccess<RegistryLayer> registries = this.server.registries();
@@ -97,11 +_,12 @@
         this.synchronizeRegistriesTask = new SynchronizeRegistriesTask(knownPacks, registries);
         this.configurationTasks.add(this.synchronizeRegistriesTask);
         this.addOptionalTasks();
+        this.configurationTasks.add(new io.papermc.paper.connection.PaperConfigurationTask(this)); // Paper
         this.returnToWorld();
     }
 
     public void returnToWorld() {
-        this.prepareSpawnTask = new PrepareSpawnTask(this.server, new NameAndId(this.gameProfile));
+        this.prepareSpawnTask = new PrepareSpawnTask(this.server, this.gameProfile, this); // Paper - pass full GameProfile & listener for events
         this.configurationTasks.add(this.prepareSpawnTask);
         this.configurationTasks.add(new JoinWorldTask());
         this.startNextTask();
@@ -130,12 +_,14 @@
     @Override
     public void handleClientInformation(final ServerboundClientInformationPacket packet) {
         this.clientInformation = packet.information();
+        this.connection.channel.attr(io.papermc.paper.adventure.PaperAdventure.LOCALE_ATTRIBUTE).set(net.kyori.adventure.translation.Translator.parseLocale(packet.information().language())); // Paper
     }
 
     @Override
     public void handleResourcePackResponse(final ServerboundResourcePackPacket packet) {
         super.handleResourcePackResponse(packet);
-        if (packet.action().isTerminal()) {
+        this.connection.resourcePackStatus = org.bukkit.event.player.PlayerResourcePackStatusEvent.Status.values()[packet.action().ordinal()]; // Paper
+        if (packet.action().isTerminal() && packet.id().equals(this.server.getServerResourcePack().map(MinecraftServer.ServerResourcePackInfo::id).orElse(null))) { // Paper - Ignore resource pack requests that are not vanilla
             this.finishCurrentTask(ServerResourcePackConfigurationTask.TYPE);
         }
     }
@@ -146,11 +_,100 @@
         if (this.synchronizeRegistriesTask == null) {
             throw new IllegalStateException("Unexpected response from client: received pack selection, but no negotiation ongoing");
         }
+        // Paper start - Per Player Quick Action & Pause Menu Dialogs
+        var dialogEvent = new io.papermc.paper.event.connection.configuration.PlayerDialogsReceiveEvent(paperConnection);
+        dialogEvent.callEvent();
+        if (dialogEvent.getPauseMenuAdditions().isPresent() || dialogEvent.getQuickAction().isPresent()) {
+            this.synchronizeRegistriesTask.handleResponse(packet.knownPacks(), p -> spoofPacketForPerPlayerDialogs(p, dialogEvent));
+        } else {
+            // Vanilla
+            this.synchronizeRegistriesTask.handleResponse(packet.knownPacks(), this::send);
+        }
+        // Paper end - Per Player Quick Action & Pause Menu Dialogs
 
-        this.synchronizeRegistriesTask.handleResponse(packet.knownPacks(), this::send);
         this.finishCurrentTask(SynchronizeRegistriesTask.TYPE);
     }
 
+    // Paper start - Per Player Quick Action & Pause Menu Dialogs
+    private void spoofPacketForPerPlayerDialogs(final net.minecraft.network.protocol.Packet<?> packet, io.papermc.paper.event.connection.configuration.PlayerDialogsReceiveEvent dialogEvent) {
+        switch (packet) {
+            case net.minecraft.network.protocol.configuration.ClientboundRegistryDataPacket dataPacket -> {
+                // spoof the dialog quick_actions and custom_options for quick actions and pause menu add
+                // I'd rather use a new custom dialog but I couldn't find out how to add a new one.
+                var dialogRegistry = net.minecraft.core.registries.Registries.DIALOG;
+                if (dataPacket.registry().equals(dialogRegistry)) {
+                    dialogEvent.getQuickAction().ifPresent(spoofDialog(net.minecraft.server.dialog.Dialogs.QUICK_ACTIONS, dataPacket));
+                    dialogEvent.getPauseMenuAdditions().ifPresent(spoofDialog(net.minecraft.server.dialog.Dialogs.CUSTOM_OPTIONS, dataPacket));
+                }
+            }
+            case net.minecraft.network.protocol.common.ClientboundUpdateTagsPacket tagsPacket -> {
+                // spoof the quick_actions and pause_menu_additions Tag so it includes the dialog
+                var registry = net.minecraft.core.registries.Registries.DIALOG;
+                tagsPacket.getTags().compute(registry, (_, old) -> {
+                    var registryAccess = this.server.registries().compositeAccess().lookup(registry).get();
+                    var map = new java.util.HashMap<net.minecraft.resources.Identifier, it.unimi.dsi.fastutil.ints.IntList>();
+                    // if old data exists keep it
+                    if (old != null) {
+                        var tags = old.resolve(registryAccess).tags();
+                        tags.forEach((key, val) -> {
+                            map.put(key.location(), val.stream().map(holder -> registryAccess.getId(holder.value())).collect(java.util.stream.Collectors.toCollection(it.unimi.dsi.fastutil.ints.IntArrayList::new)));
+                        });
+                    }
+                    // replace the Tags with the dialogs we want
+                    if (dialogEvent.getQuickAction().isPresent()) {
+                        map.put(net.minecraft.tags.DialogTags.QUICK_ACTIONS.location(), dialogToIdList(registryAccess, net.minecraft.server.dialog.Dialogs.QUICK_ACTIONS.identifier()));
+                    }
+                    if (dialogEvent.getPauseMenuAdditions().isPresent()) {
+                        map.put(net.minecraft.tags.DialogTags.PAUSE_SCREEN_ADDITIONS.location(), dialogToIdList(registryAccess, net.minecraft.server.dialog.Dialogs.CUSTOM_OPTIONS.identifier()));
+                    }
+                    return new net.minecraft.tags.TagNetworkSerialization.NetworkPayload(map);
+                });
+            }
+            default -> {
+            }
+        }
+        this.send(packet);
+    }
+
+    private static it.unimi.dsi.fastutil.ints.IntList dialogToIdList(
+        net.minecraft.core.Registry<net.minecraft.server.dialog.Dialog> registryAccess,
+        net.minecraft.resources.Identifier identifier
+    ) {
+        return it.unimi.dsi.fastutil.ints.IntList.of(registryAccess.getId(registryAccess.getValue(identifier)));
+    }
+
+    private java.util.function.Consumer<net.kyori.adventure.dialog.DialogLike> spoofDialog(
+        net.minecraft.resources.ResourceKey<net.minecraft.server.dialog.Dialog> dialogKey,
+        net.minecraft.network.protocol.configuration.ClientboundRegistryDataPacket dataPacket
+    ) {
+        var registry = net.minecraft.core.registries.Registries.DIALOG;
+        var ops = this.server.registries().compositeAccess().createSerializationContext(net.minecraft.nbt.NbtOps.INSTANCE);
+        var registryData = new net.minecraft.resources.RegistryDataLoader.RegistryData<>(registry, net.minecraft.server.dialog.Dialog.DIRECT_CODEC, net.minecraft.resources.RegistryValidator.none());
+        return dialog -> {
+            // get index so that the network Id stays the same
+            var identifier = dialogKey.identifier();
+            int index = 0;
+            for (var entry : dataPacket.entries()) {
+                if (entry.id().equals(identifier)) {
+                    break;
+                }
+                ++index;
+            }
+            // remove the element
+            dataPacket.entries().remove(index);
+
+            var paperDialog = io.papermc.paper.dialog.PaperDialog.bukkitToMinecraftHolder((io.papermc.paper.dialog.Dialog) dialog);
+            // parse dialog to, see net.minecraft.core.RegistrySynchronization#packRegistry
+            net.minecraft.nbt.Tag encodedElement = registryData.elementCodec()
+                .encodeStart(ops, paperDialog.value())
+                .getOrThrow(s -> new IllegalArgumentException("Failed to serialize " + identifier + ": " + s));
+            var optional = java.util.Optional.of(encodedElement);
+            // readd the element to the list at the same position
+            dataPacket.entries().add(index, new net.minecraft.core.RegistrySynchronization.PackedRegistryEntry(identifier, optional));
+        };
+    }
+    // Paper end - Per Player Quick Action & Pause Menu Dialogs
+
     @Override
     public void handleAcceptCodeOfConduct(final ServerboundAcceptCodeOfConductPacket packet) {
         this.finishCurrentTask(ServerCodeOfConductConfigurationTask.TYPE);
@@ -169,7 +_,7 @@
                 return;
             }
 
-            Component loginError = playerList.canPlayerLogin(this.connection.getRemoteAddress(), new NameAndId(this.gameProfile));
+            Component loginError = org.bukkit.craftbukkit.event.CraftEventFactory.handleLoginResult(playerList.canPlayerLogin(this.connection.getRemoteAddress(), new NameAndId(this.gameProfile)), this.paperConnection, this.connection, this.gameProfile, this.server, false); // Paper - Login event logic
             if (loginError != null) {
                 this.disconnect(loginError);
                 return;
@@ -231,4 +_,29 @@
         this.currentTask = null;
         this.startNextTask();
     }
+
+    // Paper start
+    @Override
+    public void disconnectAsync(final net.minecraft.network.DisconnectionDetails disconnectionInfo) {
+        if (this.cserver.isPrimaryThread()) {
+            this.disconnect(disconnectionInfo);
+            return;
+        }
+
+        this.connection.setReadOnly();
+        this.server.scheduleOnMain(() -> {
+            this.disconnect(disconnectionInfo); // Currently you cannot cancel disconnect during the config stage
+        });
+    }
+
+    @Override
+    public void handleCustomPayload(net.minecraft.network.protocol.common.ServerboundCustomPayloadPacket packet) {
+        super.handleCustomPayload(packet);
+    }
+
+    @Override
+    public io.papermc.paper.connection.PaperPlayerConfigurationConnection paperConnection() {
+        return paperConnection;
+    }
+    // Paper end
 }
