From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/chat/contents/SelectorContents.java b/net/minecraft/network/chat/contents/SelectorContents.java
index a3cd99bcf2b5c4efb245b3ae96ac3e12bdb987b3..06d0674e3e9460b44a63504c2ea5a3b0a277e973 100644
--- a/net/minecraft/network/chat/contents/SelectorContents.java
+++ b/net/minecraft/network/chat/contents/SelectorContents.java
@@ -37,7 +37,7 @@ public record SelectorContents(CompilableString<EntitySelector> selector, Option
         if (source == null) {
             return Component.empty();
         } else {
-            Optional<? extends Component> resolvedSeparator = ComponentUtils.resolve(context, this.separator, recursionDepth);
+            Optional<? extends Component> resolvedSeparator = ComponentUtils.resolveSeparator(context, this.separator, recursionDepth); // Paper - validate separator
             return ComponentUtils.formatList(this.selector.compiled().findEntities(source), resolvedSeparator, Entity::getDisplayName);
         }
     }
