--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -85,11 +85,16 @@
         protected void handleDataMarker(String s, BlockPosition blockposition, WorldAccess worldaccess, RandomSource randomsource, StructureBoundingBox structureboundingbox) {
             if ("chest".equals(s)) {
                 worldaccess.setBlock(blockposition, Blocks.AIR.defaultBlockState(), 3);
+                // CraftBukkit start - ensure block transformation
+                /*
                 TileEntity tileentity = worldaccess.getBlockEntity(blockposition.below());
 
                 if (tileentity instanceof TileEntityChest) {
                     ((TileEntityChest) tileentity).setLootTable(LootTables.IGLOO_CHEST, randomsource.nextLong());
                 }
+                */
+                setCraftLootTable(worldaccess, blockposition.below(), randomsource, LootTables.IGLOO_CHEST);
+                // CraftBukkit end
 
             }
         }
