--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -41,14 +_,37 @@
             d4 = 0.0;
         }
 
+        // Paper start - BlockDispenseEntityEvent
         AbstractBoat abstractBoat = this.type.create(serverLevel, EntitySpawnReason.DISPENSER);
-        if (abstractBoat != null) {
-            abstractBoat.setInitialPos(d1, d2 + d4, d3);
-            EntityType.<AbstractBoat>createDefaultStackConfig(serverLevel, item, null).accept(abstractBoat);
-            abstractBoat.setYRot(direction.toYRot());
-            serverLevel.addFreshEntity(abstractBoat);
-            item.shrink(1);
-        }
+        if (abstractBoat == null) return item;
+
+        abstractBoat.setInitialPos(d1, d2 + d4, d3);
+        abstractBoat.setYRot(direction.toYRot());
+        EntityType.<AbstractBoat>createDefaultStackConfig(serverLevel, item, null).accept(abstractBoat);
+
+        io.papermc.paper.event.block.BlockDispenseEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDispenseEntityEvent(blockSource, abstractBoat, item);
+        if (event == null) {
+            return item;
+        }
+        // Paper end
+
+        // CraftBukkit start
+        ItemStack singleItemStack = item.copyWithCount(1);
+        org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(singleItemStack);
+
+        boolean shrink = true;
+        if (!event.getItem().equals(craftItem)) {
+            shrink = false;
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(blockSource, eventStack);
+            if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
+                dispenseBehavior.dispense(blockSource, eventStack);
+                return item;
+            }
+        }
+
+        if (serverLevel.addFreshEntity(abstractBoat) && shrink) item.shrink(1);
+        // CraftBukkit end
 
         return item;
     }
