--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -41,13 +_,36 @@
             yOffset = 0.0;
         }
 
+        // CraftBukkit start
+        ItemStack singleItemStack = item.copyWithCount(1);
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockSource.pos());
+        org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(singleItemStack);
+
+        org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(spawnX, spawnY + yOffset, spawnZ));
+        serverLevel.getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return item;
+        }
+
+        boolean shrink = true;
+        if (!event.getItem().equals(craftItem)) {
+            shrink = false;
+            // Chain to handler for new item
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(blockSource, eventStack);
+            if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
+                dispenseBehavior.dispense(blockSource, eventStack);
+                return item;
+            }
+        }
+        // CraftBukkit end
         AbstractBoat boat = this.type.create(level, EntitySpawnReason.DISPENSER);
         if (boat != null) {
-            boat.setInitialPos(spawnX, spawnY + yOffset, spawnZ);
+            boat.setInitialPos(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ()); // CraftBukkit
             EntityType.<AbstractBoat>createDefaultStackConfig(level, dispensed, null).accept(boat);
             boat.setYRot(direction.toYRot());
-            level.addFreshEntity(boat);
-            dispensed.shrink(1);
+            if (level.addFreshEntity(boat) && shrink) dispensed.shrink(1); // Paper - if entity add was successful and supposed to shrink
         }
 
         return dispensed;
