--- a/net/minecraft/core/dispenser/ProjectileDispenseBehavior.java
+++ b/net/minecraft/core/dispenser/ProjectileDispenseBehavior.java
@@ -27,17 +_,40 @@
         ServerLevel serverLevel = blockSource.level();
         Direction direction = blockSource.state().getValue(DispenserBlock.FACING);
         Position dispensePosition = this.dispenseConfig.positionFunction().getDispensePosition(blockSource, direction);
-        Projectile.spawnProjectileUsingShoot(
-            this.projectileItem.asProjectile(serverLevel, dispensePosition, item, direction),
-            serverLevel,
-            item,
-            direction.getStepX(),
-            direction.getStepY(),
-            direction.getStepZ(),
-            this.dispenseConfig.power(),
-            this.dispenseConfig.uncertainty()
-        );
-        item.shrink(1);
+        ItemStack singleItemStack = item.copyWithCount(1);
+
+        // Paper start - BlockDispenseEntityEvent
+        Projectile projectile = this.projectileItem.asProjectile(serverLevel, dispensePosition, singleItemStack, direction);
+
+        io.papermc.paper.event.block.BlockDispenseEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockDispenseEntityEvent(blockSource, projectile, item);
+        if (event == null) {
+            return item;
+        }
+        // Paper end
+
+        // CraftBukkit start
+        org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(singleItemStack);
+        boolean shrink = true;
+        if (!event.getItem().equals(craftItem)) {
+            shrink = false;
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(blockSource, eventStack);
+            if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
+                dispenseBehavior.dispense(blockSource, eventStack);
+                return item;
+            }
+        }
+
+        if (!singleItemStack.isEmpty()) {
+            // Paper start - BlockDispenseEntityEvent
+            Projectile.spawnProjectileUsingShoot(projectile, serverLevel, singleItemStack, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), this.dispenseConfig.power(), this.dispenseConfig.uncertainty());
+            projectile.projectileSource = new org.bukkit.craftbukkit.projectiles.CraftBlockProjectileSource(blockSource.blockEntity());
+            // Paper end
+        }
+
+        if (shrink) item.shrink(1);
+        // CraftBukkit end
+
         return item;
     }
 
