--- a/net/minecraft/world/level/block/state/properties/IntegerProperty.java
+++ b/net/minecraft/world/level/block/state/properties/IntegerProperty.java
@@ -28,8 +_,7 @@
         return this.values;
     }
 
-    @Override
-    public boolean equals(Object other) {
+    public boolean equals_unused(Object other) { // Paper - Perf: Optimize hashCode/equals
         return this == other || other instanceof IntegerProperty integerProperty && super.equals(other) && this.values.equals(integerProperty.values);
     }
 
