--- a/net/minecraft/world/item/enchantment/effects/ExplodeEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/ExplodeEffect.java
@@ -56,7 +_,7 @@
     @Override
     public void apply(ServerLevel level, int enchantmentLevel, EnchantedItemInUse item, Entity entity, Vec3 origin) {
         Vec3 vec3 = origin.add(this.offset);
-        level.explode(
+        level.explode0( // Paper - configure explosion for enchantment damage source
             this.attributeToUser ? entity : null,
             this.getDamageSource(entity, vec3),
             new SimpleExplosionDamageCalculator(
@@ -73,7 +_,8 @@
             this.blockInteraction,
             this.smallParticle,
             this.largeParticle,
-            this.sound
+            this.sound,
+            e -> e.optionalEnchantmentSource = org.bukkit.craftbukkit.event.CraftEventFactory.ExplodeEventSourceEnchantmentSourceImpl.from(entity, item) // Paper - configure explosion for enchantment damage source
         );
     }
 
