From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/enchantment/effects/Ignite.java b/net/minecraft/world/item/enchantment/effects/Ignite.java
index b2a35dc9febc4f9bcd59b0066001d1e4ae2b2c5e..0cd5df3f8d30ed124bb7d96290d8b5c5c7c99c6e 100644
--- a/net/minecraft/world/item/enchantment/effects/Ignite.java
+++ b/net/minecraft/world/item/enchantment/effects/Ignite.java
@@ -15,7 +15,21 @@ public record Ignite(LevelBasedValue duration) implements EnchantmentEntityEffec
 
     @Override
     public void apply(ServerLevel level, int enchantmentLevel, EnchantedItemInUse item, Entity entity, Vec3 origin) {
-        entity.igniteForSeconds(this.duration.calculate(enchantmentLevel));
+        // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+        org.bukkit.event.entity.EntityCombustEvent entityCombustEvent;
+        if (item.owner() != null) {
+            entityCombustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(item.owner().getBukkitEntity(), entity.getBukkitEntity(), this.duration.calculate(enchantmentLevel));
+        } else {
+            entityCombustEvent = new org.bukkit.event.entity.EntityCombustEvent(entity.getBukkitEntity(), this.duration.calculate(enchantmentLevel));
+        }
+
+        org.bukkit.Bukkit.getPluginManager().callEvent(entityCombustEvent);
+        if (entityCombustEvent.isCancelled()) {
+            return;
+        }
+
+        entity.igniteForSeconds(entityCombustEvent.getDuration(), false);
+        // CraftBukkit end
     }
 
     @Override
