--- a/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
+++ b/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
@@ -47,7 +_,7 @@
         for (BlockPos blockPos1 : BlockPos.betweenClosed(blockPos.offset(-i, 0, -i), blockPos.offset(i, Math.min(i1 - 1, 0), i))) {
             if (blockPos1.distToCenterSqr(origin.x(), blockPos1.getY() + 0.5, origin.z()) < Mth.square(i)
                 && this.predicate.map(predicate -> predicate.test(level, blockPos1)).orElse(true)
-                && level.setBlockAndUpdate(blockPos1, this.blockState.getState(random, blockPos1))) {
+                && org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, blockPos1, this.blockState.getState(random, blockPos1), 3, entity, true)) { // CraftBukkit - Call EntityBlockFormEvent for Frost Walker
                 this.triggerGameEvent.ifPresent(event -> level.gameEvent(entity, (Holder<GameEvent>)event, blockPos1));
             }
         }
