From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java b/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
index 55dee23a36d2431a7409c257edb3d973920ad82a..0571664aa5ad77ea5d42b97b3fdaed75995ff45d 100644
--- a/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
+++ b/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
@@ -47,7 +47,7 @@ public record ReplaceDisk(
         for (BlockPos blockPos1 : BlockPos.betweenClosed(blockPos.offset(-i, 0, -i), blockPos.offset(i, Math.min(i1 - 1, 0), i))) {
             if (blockPos1.distToCenterSqr(origin.x(), blockPos1.getY() + 0.5, origin.z()) < Mth.square(i)
                 && this.predicate.map(predicate -> predicate.test(level, blockPos1)).orElse(true)
-                && level.setBlockAndUpdate(blockPos1, this.blockState.getState(random, blockPos1))) {
+                && org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, blockPos1, this.blockState.getState(random, blockPos1), entity)) { // CraftBukkit - Call EntityBlockFormEvent for Frost Walker
                 this.triggerGameEvent.ifPresent(event -> level.gameEvent(entity, (Holder<GameEvent>)event, blockPos1));
             }
         }
