--- a/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java
@@ -58,8 +_,13 @@
         Vec3 knownMovement = entity.getKnownMovement();
         float bbWidth = entity.getBbWidth();
         float bbHeight = entity.getBbHeight();
-        level.sendParticles(
+        // Paper start - Hide soul speed particles for vanished players
+        level.sendParticlesSource(
+            entity,
             this.particle,
+            false,
+            false,
+        // Paper end - Hide soul speed particles for vanished players
             this.horizontalPosition.getCoordinate(origin.x(), origin.x(), bbWidth, random),
             this.verticalPosition.getCoordinate(origin.y(), origin.y() + bbHeight / 2.0F, bbHeight, random),
             this.horizontalPosition.getCoordinate(origin.z(), origin.z(), bbWidth, random),
