From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java b/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java
index e3c3ac9519ec9a0098c332def9229768660ed723..a95550a55cd16603d7bb7570332dd3d224ab834e 100644
--- a/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/SpawnParticlesEffect.java
@@ -58,8 +58,13 @@ public record SpawnParticlesEffect(
         Vec3 knownMovement = entity.getKnownMovement();
         float bbWidth = entity.getBbWidth();
         float bbHeight = entity.getBbHeight();
-        level.sendParticles(
+        // Paper start - Hide soul speed particles for vanished players
+        level.sendParticlesSource(
+            entity,
             this.particle,
+            false,
+            false,
+        // Paper end - Hide soul speed particles for vanished players
             this.horizontalPosition.getCoordinate(origin.x(), origin.x(), bbWidth, random),
             this.verticalPosition.getCoordinate(origin.y(), origin.y() + bbHeight / 2.0F, bbHeight, random),
             this.horizontalPosition.getCoordinate(origin.z(), origin.z(), bbWidth, random),
