--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -221,17 +_,17 @@
         }
     }
 
-    public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<Level> level, @Nullable DimensionDataStorage storage) {
-        if (level == Level.OVERWORLD) {
+    public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<net.minecraft.world.level.dimension.LevelStem> level, @Nullable DimensionDataStorage storage) { // CraftBukkit
+        if (level == net.minecraft.world.level.dimension.LevelStem.OVERWORLD) { // CraftBukkit
             return new LegacyStructureDataHandler(
                 storage,
                 ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"),
                 ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument")
             );
-        } else if (level == Level.NETHER) {
+        } else if (level == net.minecraft.world.level.dimension.LevelStem.NETHER) { // CraftBukkit
             List<String> list = ImmutableList.of("Fortress");
             return new LegacyStructureDataHandler(storage, list, list);
-        } else if (level == Level.END) {
+        } else if (level == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit
             List<String> list = ImmutableList.of("EndCity");
             return new LegacyStructureDataHandler(storage, list, list);
         } else {
