--- a/net/minecraft/world/inventory/ChestMenu.java
+++ b/net/minecraft/world/inventory/ChestMenu.java
@@ -9,6 +_,34 @@
 public class ChestMenu extends AbstractContainerMenu {
     private final Container container;
     private final int containerRows;
+    // CraftBukkit start
+    private @javax.annotation.Nullable org.bukkit.craftbukkit.inventory.CraftInventoryView view = null;
+    private final Inventory inventory;
+
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory;
+        if (this.container instanceof Inventory) {
+            inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryPlayer((Inventory) this.container);
+        } else if (this.container instanceof net.minecraft.world.CompoundContainer) {
+            inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest((net.minecraft.world.CompoundContainer) this.container);
+        } else {
+            inventory = new org.bukkit.craftbukkit.inventory.CraftInventory(this.container);
+        }
+
+        this.view = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.inventory.player.getBukkitEntity(), inventory, this);
+        return this.view;
+    }
+
+    @Override
+    public void startOpen() {
+        this.container.startOpen(this.inventory.player);
+    }
+    // CraftBukkit end
 
     private ChestMenu(MenuType<?> type, int containerId, Inventory playerInventory, int rows) {
         this(type, containerId, playerInventory, new SimpleContainer(9 * rows), rows);
@@ -51,7 +_,10 @@
         checkContainerSize(container, rows * 9);
         this.container = container;
         this.containerRows = rows;
-        container.startOpen(playerInventory.player);
+        // container.startOpen(playerInventory.player); // Paper - don't startOpen until menu actually opens
+        // CraftBukkit start - Save player
+        this.inventory = playerInventory;
+        // CraftBukkit end
         int i = 18;
         this.addChestGrid(container, 8, 18);
         int i1 = 18 + this.containerRows * 18 + 13;
@@ -68,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.container.stillValid(player);
     }
 
