--- a/net/minecraft/world/inventory/ContainerListener.java
+++ b/net/minecraft/world/inventory/ContainerListener.java
@@ -5,5 +5,11 @@
 public interface ContainerListener {
     void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack stack);
 
+    // Paper start - Add PlayerInventorySlotChangeEvent
+    default void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack oldStack, ItemStack stack) {
+        slotChanged(handler, slotId, stack);
+    }
+    // Paper end - Add PlayerInventorySlotChangeEvent
+
     void dataChanged(AbstractContainerMenu handler, int property, int value);
 }
