From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/inventory/DispenserMenu.java b/net/minecraft/world/inventory/DispenserMenu.java
index fc98d858c29631960addabd4c76d023cb52b8aa6..ae90aab3556743f9527daad4d6046b3b0ea0e7e6 100644
--- a/net/minecraft/world/inventory/DispenserMenu.java
+++ b/net/minecraft/world/inventory/DispenserMenu.java
@@ -13,6 +13,10 @@ public class DispenserMenu extends AbstractContainerMenu {
     private static final int USE_ROW_SLOT_START = 36;
     private static final int USE_ROW_SLOT_END = 45;
     public final Container dispenser;
+    // CraftBukkit start
+    private org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity = null;
+    private Inventory player;
+    // CraftBukkit end
 
     public DispenserMenu(int containerId, Inventory playerInventory) {
         this(containerId, playerInventory, new SimpleContainer(9));
@@ -20,6 +24,9 @@ public class DispenserMenu extends AbstractContainerMenu {
 
     public DispenserMenu(int containerId, Inventory playerInventory, Container container) {
         super(MenuType.GENERIC_3x3, containerId);
+        // CraftBukkit start - Save player
+        this.player = playerInventory;
+        // CraftBukkit end
         checkContainerSize(container, 9);
         this.dispenser = container;
         container.startOpen(playerInventory.player);
@@ -38,6 +45,7 @@ public class DispenserMenu extends AbstractContainerMenu {
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.dispenser.stillValid(player);
     }
 
@@ -77,4 +85,17 @@ public class DispenserMenu extends AbstractContainerMenu {
         super.removed(player);
         this.dispenser.stopOpen(player);
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (this.bukkitEntity != null) {
+            return this.bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventory(this.dispenser);
+        this.bukkitEntity = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return this.bukkitEntity;
+    }
+    // CraftBukkit end
 }
