--- a/net/minecraft/world/inventory/HopperMenu.java
+++ b/net/minecraft/world/inventory/HopperMenu.java
@@ -9,6 +_,21 @@
 public class HopperMenu extends AbstractContainerMenu {
     public static final int CONTAINER_SIZE = 5;
     private final Container hopper;
+    // CraftBukkit start
+    private @javax.annotation.Nullable org.bukkit.craftbukkit.inventory.CraftInventoryView view = null;
+    private final Inventory inventory;
+
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventory(this.hopper);
+        this.view = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.inventory.player.getBukkitEntity(), inventory, this);
+        return this.view;
+    }
+    // CraftBukkit end
 
     public HopperMenu(final int containerId, final Inventory inventory) {
         this(containerId, inventory, new SimpleContainer(5));
@@ -17,6 +_,7 @@
     public HopperMenu(final int containerId, final Inventory inventory, final Container hopper) {
         super(MenuType.HOPPER, containerId);
         this.hopper = hopper;
+        this.inventory = playerInventory; // CraftBukkit
         checkContainerSize(hopper, 5);
         hopper.startOpen(inventory.player);
 
@@ -29,6 +_,7 @@
 
     @Override
     public boolean stillValid(final Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.hopper.stillValid(player);
     }
 
