--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -109,13 +_,51 @@
     };
     public final Reference2IntOpenHashMap<ResourceKey<Recipe<?>>> recipesUsed = new Reference2IntOpenHashMap<>();
     private final RecipeManager.CachedCheck<SingleRecipeInput, ? extends AbstractCookingRecipe> quickCheck;
+    public final RecipeType<? extends AbstractCookingRecipe> recipeType; // Paper - cook speed multiplier API
+    public double cookSpeedMultiplier = 1.0; // Paper - cook speed multiplier API
 
     protected AbstractFurnaceBlockEntity(
         final BlockEntityType<?> type, final BlockPos worldPosition, final BlockState blockState, final RecipeType<? extends AbstractCookingRecipe> recipeType
     ) {
         super(type, worldPosition, blockState);
         this.quickCheck = RecipeManager.createCheck(recipeType);
-    }
+        this.recipeType = recipeType; // Paper - cook speed multiplier API
+    }
+
+    // CraftBukkit start - add fields and methods
+    private int maxStack = MAX_STACK;
+    public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+
+    @Override
+    public List<ItemStack> getContents() {
+        return this.items;
+    }
+
+    @Override
+    public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.add(player);
+    }
+
+    @Override
+    public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.remove(player);
+    }
+
+    @Override
+    public List<org.bukkit.entity.HumanEntity> getViewers() {
+        return this.transaction;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return this.maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        this.maxStack = size;
+    }
+    // CraftBukkit end
 
     @Override
     protected void loadAdditional(final ValueInput input) {
@@ -128,6 +_,7 @@
         this.litTotalTime = input.getShortOr("lit_total_time", (short)0);
         this.recipesUsed.clear();
         this.recipesUsed.putAll(input.read("RecipesUsed", RECIPES_USED_CODEC).orElse(Map.of()));
+        this.cookSpeedMultiplier = input.getDoubleOr("Paper.CookSpeedMultiplier", 1); // Paper - cook speed multiplier API
     }
 
     @Override
@@ -137,6 +_,7 @@
         output.putShort("cooking_total_time", (short)this.cookingTotalTime);
         output.putShort("lit_time_remaining", (short)this.litTimeRemaining);
         output.putShort("lit_total_time", (short)this.litTotalTime);
+        output.putDouble("Paper.CookSpeedMultiplier", this.cookSpeedMultiplier); // Paper - cook speed multiplier API
         ContainerHelper.saveAllItems(output, this.items);
         output.store("RecipesUsed", RECIPES_USED_CODEC, this.recipesUsed);
     }
@@ -166,24 +_,48 @@
                     int maxStackSize = entity.getMaxStackSize();
                     ItemStack burnResult = recipe.value().assemble(input);
                     if (!burnResult.isEmpty() && canBurn(entity.items, maxStackSize, burnResult)) {
+                        // CraftBukkit start
+                        org.bukkit.event.inventory.FurnaceBurnEvent furnaceBurnEvent = new org.bukkit.event.inventory.FurnaceBurnEvent(
+                            org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
+                            org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(fuel),
+                            entity.getBurnDuration(level.fuelValues(), fuel)
+                        );
+                        if (!furnaceBurnEvent.callEvent()) return;
+                        // CraftBukkit end
                         if (!isLit) {
-                            int newLitTime = entity.getBurnDuration(level.fuelValues(), fuel);
+                            int newLitTime = furnaceBurnEvent.getBurnTime(); // CraftBukkit - respect event output
                             entity.litTimeRemaining = newLitTime;
                             entity.litTotalTime = newLitTime;
-                            if (newLitTime > 0) {
+                            if (newLitTime > 0 && furnaceBurnEvent.isBurning()) { // CraftBukkit - respect event output // TODO - snapshot - double check surrounding diff post squash
+                                if (furnaceBurnEvent.willConsumeFuel()) { // Paper - add consumeFuel to FurnaceBurnEvent
                                 consumeFuel(entity.items, fuel);
+                                } // Paper - add consumeFuel to FurnaceBurnEvent
                                 isLit = true;
                                 changed = true;
                             }
                         }
 
                         if (isLit) {
+                            // CraftBukkit start
+                            if (entity.cookingTimer == 0) {
+                                org.bukkit.event.inventory.FurnaceStartSmeltEvent event = new org.bukkit.event.inventory.FurnaceStartSmeltEvent(
+                                    org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
+                                    org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(ingredient),
+                                    (org.bukkit.inventory.CookingRecipe<?>) recipe.toBukkitRecipe(),
+                                    getTotalCookTime(level, entity, entity.recipeType, entity.cookSpeedMultiplier) // Paper - cook speed multiplier API
+                                );
+                                event.callEvent();
+
+                                entity.cookingTotalTime = event.getTotalCookTime();
+                            }
+                            // CraftBukkit end
                             entity.cookingTimer++;
-                            if (entity.cookingTimer == entity.cookingTotalTime) {
+                            if (entity.cookingTimer >= entity.cookingTotalTime) { // Paper - cook speed multiplier API
                                 entity.cookingTimer = 0;
-                                entity.cookingTotalTime = recipe.value().cookingTime();
-                                burn(entity.items, ingredient, burnResult);
+                                entity.cookingTotalTime = getTotalCookTime(level, entity, entity.recipeType, entity.cookSpeedMultiplier); // Paper - cook speed multiplier API
+                                if (burn(entity.items, ingredient, burnResult, recipe, level, entity.worldPosition)) { // CraftBukkit - add level & pos // Paper - make burn return a boolean again, add recipe
                                 entity.setRecipeUsed(recipe);
+                                } // Paper
                                 changed = true;
                             }
                         } else {
@@ -233,28 +_,63 @@
         }
     }
 
-    private static void burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, final ItemStack result) {
+    private static boolean burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, ItemStack result, RecipeHolder<? extends AbstractCookingRecipe> recipe, final net.minecraft.world.level.Level level, final BlockPos blockPos) { // CraftBukkit
         ItemStack resultItemStack = items.get(2);
+        // CraftBukkit start - fire FurnaceSmeltEvent
+        org.bukkit.craftbukkit.inventory.CraftItemStack apiIngredient = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(inputItemStack);
+        org.bukkit.inventory.ItemStack apiResult = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(result);
+
+        org.bukkit.event.inventory.FurnaceSmeltEvent furnaceSmeltEvent = new org.bukkit.event.inventory.FurnaceSmeltEvent(
+            org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
+            apiIngredient,
+            apiResult,
+            (org.bukkit.inventory.CookingRecipe<?>) recipe.toBukkitRecipe() // Paper - Add recipe to cook events
+        );
+        if (!furnaceSmeltEvent.callEvent()) return false;
+
+        apiResult = furnaceSmeltEvent.getResult();
+        result = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(apiResult);
+
+        if (!result.isEmpty()) {
+            if (resultItemStack.isEmpty()) {
+                items.set(SLOT_RESULT, result.copy());
+            } else if (org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(resultItemStack).isSimilar(apiResult)) {
+                resultItemStack.grow(result.getCount());
+            } else {
+                return false;
+            }
+        }
+        /*
         if (resultItemStack.isEmpty()) {
             items.set(2, result.copy());
         } else {
             resultItemStack.grow(result.getCount());
         }
+        */
+        // CraftBukkit end
 
         if (inputItemStack.is(Items.WET_SPONGE) && !items.get(1).isEmpty() && items.get(1).is(Items.BUCKET)) {
             items.set(1, new ItemStack(Items.WATER_BUCKET));
         }
 
         inputItemStack.shrink(1);
+        return true; // Paper - make burn return a boolean again
     }
 
     protected int getBurnDuration(final FuelValues fuelValues, final ItemStack itemStack) {
         return fuelValues.burnDuration(itemStack);
     }
 
-    public static int getTotalCookTime(final ServerLevel level, final AbstractFurnaceBlockEntity entity) {
+    public static int getTotalCookTime(final @Nullable ServerLevel level, final AbstractFurnaceBlockEntity entity, final RecipeType<? extends AbstractCookingRecipe> recipeType, final double cookSpeedMultiplier) { // Paper - cook speed multiplier API
         SingleRecipeInput input = new SingleRecipeInput(entity.getItem(0));
-        return entity.quickCheck.getRecipeFor(input, level).map(recipeHolder -> recipeHolder.value().cookingTime()).orElse(200);
+        // Paper start - cook speed multiplier API
+        /* Scale the recipe's cooking time to the current cookSpeedMultiplier */
+        int cookTime = level != null
+            ? entity.quickCheck.getRecipeFor(input, level).map(recipeHolder -> recipeHolder.value().cookingTime()).orElse(BURN_TIME_STANDARD)
+            /* passing a null level here is safe. world is only used for map extending recipes which won't happen here */
+            : (net.minecraft.server.MinecraftServer.getServer().getRecipeManager().getRecipeFor(recipeType, input, level).map(recipeHolder -> recipeHolder.value().cookingTime()).orElse(BURN_TIME_STANDARD));
+        return (int) Math.ceil (cookTime / cookSpeedMultiplier);
+        // Paper end - cook speed multiplier API
     }
 
     @Override
@@ -298,7 +_,7 @@
         this.items.set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
         if (slot == 0 && !same && this.level instanceof ServerLevel serverLevel) {
-            this.cookingTotalTime = getTotalCookTime(serverLevel, this);
+            this.cookingTotalTime = getTotalCookTime(serverLevel, this, this.recipeType, this.cookSpeedMultiplier); // Paper - cook speed multiplier API
             this.cookingTimer = 0;
             this.setChanged();
         }
@@ -333,8 +_,8 @@
     public void awardUsedRecipes(final Player player, final List<ItemStack> itemStacks) {
     }
 
-    public void awardUsedRecipesAndPopExperience(final ServerPlayer player) {
-        List<RecipeHolder<?>> recipesToAward = this.getRecipesToAwardAndPopExperience(player.level(), player.position());
+    public void awardUsedRecipesAndPopExperience(final ServerPlayer player, final ItemStack itemStack, final int amount) { // CraftBukkit
+        List<RecipeHolder<?>> recipesToAward = this.getRecipesToAwardAndPopExperience(player.level(), player.position(), this.worldPosition, player, itemStack, amount); // CraftBukkit - overload for exp spawn events
         player.awardRecipes(recipesToAward);
 
         for (RecipeHolder<?> recipe : recipesToAward) {
@@ -345,30 +_,60 @@
     }
 
     public List<RecipeHolder<?>> getRecipesToAwardAndPopExperience(final ServerLevel level, final Vec3 position) {
+        // CraftBukkit start
+        return this.getRecipesToAwardAndPopExperience(level, position, this.worldPosition, null, null, 0);
+    }
+    public List<RecipeHolder<?>> getRecipesToAwardAndPopExperience(final ServerLevel level, final Vec3 position, final BlockPos blockPos, final @Nullable ServerPlayer serverPlayer, final @Nullable ItemStack itemStack, final int amount) {
+        // CraftBukkit end
         List<RecipeHolder<?>> recipesToAward = Lists.newArrayList();
 
         for (Entry<ResourceKey<Recipe<?>>> entry : this.recipesUsed.reference2IntEntrySet()) {
             level.recipeAccess().byKey(entry.getKey()).ifPresent(recipe -> {
+                if (!(recipe.value() instanceof AbstractCookingRecipe)) return; // Paper - don't process non-cooking recipes
                 recipesToAward.add((RecipeHolder<?>)recipe);
-                createExperience(level, position, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience());
+                createExperience(level, position, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience(), blockPos, serverPlayer, itemStack, amount); // Paper - don't process non-cooking recipes
             });
         }
 
         return recipesToAward;
     }
 
-    private static void createExperience(final ServerLevel level, final Vec3 position, final int amount, final float value) {
+    private static void createExperience(final ServerLevel level, final Vec3 position, final int amount, final float value, final BlockPos blockPos, final ServerPlayer player, final ItemStack itemStack, final int removeCount) {
         int xpReward = Mth.floor(amount * value);
         float xpFraction = Mth.frac(amount * value);
         if (xpFraction != 0.0F && level.getRandom().nextFloat() < xpFraction) {
             xpReward++;
         }
 
-        ExperienceOrb.award(level, position, xpReward);
+        // CraftBukkit start - fire FurnaceExtractEvent / BlockExpEvent
+        org.bukkit.event.block.BlockExpEvent event;
+        if (removeCount != 0) {
+            event = new org.bukkit.event.inventory.FurnaceExtractEvent(
+                player.getBukkitEntity(),
+                org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
+                itemStack.asBukkitCopy(),
+                removeCount,
+                xpReward
+            );
+        } else {
+            event = new org.bukkit.event.block.BlockExpEvent(
+                org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
+                xpReward
+            );
+        }
+        event.callEvent();
+        xpReward = event.getExpToDrop();
+        // CraftBukkit end
+
+        ExperienceOrb.awardWithDirection(level, position, net.minecraft.world.phys.Vec3.ZERO, xpReward, org.bukkit.entity.ExperienceOrb.SpawnReason.FURNACE, player, null); // Paper
     }
 
     @Override
     public void fillStackedContents(final StackedItemContents contents) {
+        // Paper start - don't account fuel stack (fixes MC-243057)
+        contents.accountStack(this.items.get(SLOT_INPUT));
+        contents.accountStack(this.items.get(SLOT_RESULT));
+        // Paper end - don't account fuel stack (fixes MC-243057)
         for (ItemStack itemStack : this.items) {
             contents.accountStack(itemStack);
         }
