--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -86,6 +_,7 @@
     public static final int MIN_OCCUPATION_TICKS_NECTARLESS = 600;
     private List<BeehiveBlockEntity.BeeData> stored = Lists.newArrayList();
     public @Nullable BlockPos savedFlowerPos;
+    public int maxBees = MAX_OCCUPANTS; // CraftBukkit - allow setting max amount of bees a hive can hold
 
     public BeehiveBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityType.BEEHIVE, worldPosition, blockState);
@@ -119,7 +_,7 @@
     }
 
     public boolean isFull() {
-        return this.stored.size() == 3;
+        return this.stored.size() == this.maxBees; // CraftBukkit
     }
 
     public void emptyAllLivingFromHive(final @Nullable Player player, final BlockState state, final BeehiveBlockEntity.BeeReleaseStatus releaseReason) {
@@ -128,7 +_,7 @@
             for (Entity released : releasedFromHive) {
                 if (released instanceof Bee bee && player.position().distanceToSqr(released.position()) <= 16.0) {
                     if (!this.isSedated()) {
-                        bee.setTarget(player);
+                        bee.setTarget(player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER); // CraftBukkit
                     } else {
                         bee.setStayOutOfHiveCountdown(400);
                     }
@@ -138,10 +_,15 @@
     }
 
     private List<Entity> releaseAllOccupants(final BlockState state, final BeehiveBlockEntity.BeeReleaseStatus releaseStatus) {
+        // CraftBukkit start - This allows us to bypass the night/rain/emergency check
+        return this.releaseBees(state, releaseStatus, false);
+    }
+    public List<Entity> releaseBees(final BlockState state, final BeehiveBlockEntity.BeeReleaseStatus releaseStatus, final boolean force) {
+        // CraftBukkit end - This allows us to bypass the night/rain/emergency check
         List<Entity> spawned = Lists.newArrayList();
         this.stored
             .removeIf(
-                occupantEntry -> releaseOccupant(this.level, this.worldPosition, state, occupantEntry.toOccupant(), spawned, releaseStatus, this.savedFlowerPos)
+                occupantEntry -> releaseOccupant(this.level, this.worldPosition, state, occupantEntry.toOccupant(), spawned, releaseStatus, this.savedFlowerPos, force) // CraftBukkit - This allows us to bypass the night/rain/emergency check
             );
         if (!spawned.isEmpty()) {
             super.setChanged();
@@ -155,6 +_,12 @@
         return this.stored.size();
     }
 
+    // Paper start - Add EntityBlockStorage clearEntities
+    public void clearBees() {
+        this.stored.clear();
+    }
+    // Paper end - Add EntityBlockStorage clearEntities
+
     public static int getHoneyLevel(final BlockState blockState) {
         return blockState.getValue(BeehiveBlock.HONEY_LEVEL);
     }
@@ -165,7 +_,16 @@
     }
 
     public void addOccupant(final Bee bee) {
-        if (this.stored.size() < 3) {
+        if (this.stored.size() < this.maxBees) { // CraftBukkit
+            // CraftBukkit start
+            if (this.level != null) {
+                org.bukkit.event.entity.EntityEnterBlockEvent event = new org.bukkit.event.entity.EntityEnterBlockEvent(bee.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(this.level, this.getBlockPos()));
+                if (!event.callEvent()) {
+                    bee.setStayOutOfHiveCountdown(MIN_TICKS_BEFORE_REENTERING_HIVE);
+                    return;
+                }
+            }
+            // CraftBukkit end
             bee.stopRiding();
             bee.ejectPassengers();
             bee.dropLeash();
@@ -190,7 +_,7 @@
                 this.level.gameEvent(GameEvent.BLOCK_CHANGE, blockPos, GameEvent.Context.of(bee, this.getBlockState()));
             }
 
-            bee.discard();
+            bee.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.ENTER_BLOCK); // CraftBukkit - add Bukkit remove cause
             super.setChanged();
         }
     }
@@ -199,6 +_,12 @@
         this.stored.add(new BeehiveBlockEntity.BeeData(occupant));
     }
 
+    // CraftBukkit start
+    private static boolean releaseOccupant(final Level level, final BlockPos blockPos, final BlockState state, final BeehiveBlockEntity.Occupant beeData, final @Nullable List<Entity> spawned, final BeehiveBlockEntity.BeeReleaseStatus releaseStatus, final @Nullable BlockPos savedFlowerPos) {
+        return releaseOccupant(level, blockPos, state, beeData, spawned, releaseStatus, savedFlowerPos, false);
+    }
+    // CraftBukkit end
+
     private static boolean releaseOccupant(
         final Level level,
         final BlockPos blockPos,
@@ -207,8 +_,9 @@
         final @Nullable List<Entity> spawned,
         final BeehiveBlockEntity.BeeReleaseStatus releaseStatus,
         final @Nullable BlockPos savedFlowerPos
+        , final boolean force // CraftBukkit
     ) {
-        if (level.environmentAttributes().getValue(EnvironmentAttributes.BEES_STAY_IN_HIVE, blockPos)
+        if (!force && level.environmentAttributes().getValue(EnvironmentAttributes.BEES_STAY_IN_HIVE, blockPos) // CraftBukkit
             && releaseStatus != BeehiveBlockEntity.BeeReleaseStatus.EMERGENCY) {
             return false;
         } else {
@@ -220,6 +_,17 @@
             } else {
                 Entity entity = beeData.createEntity(level, blockPos);
                 if (entity != null) {
+                    // CraftBukkit start
+                    if (entity instanceof Bee) {
+                        float bbWidth = entity.getBbWidth();
+                        double delta = frontBlocked ? 0.0 : 0.55 + bbWidth / 2.0F;
+                        double spawnX = blockPos.getX() + 0.5 + delta * facing.getStepX();
+                        double spawnY = blockPos.getY() + 0.5 - entity.getBbHeight() / 2.0F;
+                        double spawnZ = blockPos.getZ() + 0.5 + delta * facing.getStepZ();
+                        entity.snapTo(spawnX, spawnY, spawnZ, entity.getYRot(), entity.getXRot());
+                    }
+                    if (!level.addFreshEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BEEHIVE)) return false; // CraftBukkit - SpawnReason, moved from below
+                    // CraftBukkit end
                     if (entity instanceof Bee bee) {
                         RandomSource random = level.getRandom();
                         if (savedFlowerPos != null && !bee.hasSavedFlowerPos() && random.nextFloat() < 0.9F) {
@@ -236,7 +_,13 @@
                                         levelIncrease--;
                                     }
 
-                                    level.setBlockAndUpdate(blockPos, state.setValue(BeehiveBlock.HONEY_LEVEL, honeyLevel + levelIncrease));
+                                    // Paper start - Fire EntityChangeBlockEvent in more places
+                                    BlockState newBlockState = state.setValue(BeehiveBlock.HONEY_LEVEL, honeyLevel + levelIncrease);
+
+                                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockPos, newBlockState)) {
+                                        level.setBlockAndUpdate(blockPos, newBlockState);
+                                    }
+                                    // Paper end - Fire EntityChangeBlockEvent in more places
                                 }
                             }
                         }
@@ -245,17 +_,19 @@
                             spawned.add(bee);
                         }
 
+                        /* CraftBukkit start - move up
                         float bbWidth = entity.getBbWidth();
                         double delta = frontBlocked ? 0.0 : 0.55 + bbWidth / 2.0F;
                         double spawnX = blockPos.getX() + 0.5 + delta * facing.getStepX();
                         double spawnY = blockPos.getY() + 0.5 - entity.getBbHeight() / 2.0F;
                         double spawnZ = blockPos.getZ() + 0.5 + delta * facing.getStepZ();
                         entity.snapTo(spawnX, spawnY, spawnZ, entity.getYRot(), entity.getXRot());
+                        */ // CraftBukkit end
                     }
 
                     level.playSound(null, blockPos, SoundEvents.BEEHIVE_EXIT, SoundSource.BLOCKS, 1.0F, 1.0F);
                     level.gameEvent(GameEvent.BLOCK_CHANGE, blockPos, GameEvent.Context.of(entity, level.getBlockState(blockPos)));
-                    return level.addFreshEntity(entity);
+                    return true; // CraftBukkit - moved up
                 } else {
                     return false;
                 }
@@ -283,6 +_,11 @@
                     changed = true;
                     iterator.remove();
                 }
+                // Paper start - Fix bees aging inside; use exitTickCounter to keep actual bee life
+                else if (level.paperConfig().entities.behavior.cooldownFailedBeehiveReleases) {
+                    data.exitTickCounter = data.occupant.minTicksInHive / 2;
+                }
+                // Paper end - Fix bees aging inside; use exitTickCounter to keep actual bee life
             }
         }
 
@@ -304,9 +_,10 @@
     @Override
     protected void loadAdditional(final ValueInput input) {
         super.loadAdditional(input);
-        this.stored.clear();
+        this.stored = Lists.newArrayList(); // CraftBukkit - SPIGOT-7790: create new copy (may be modified in physics event triggered by honey change)
         input.read("bees", BeehiveBlockEntity.Occupant.LIST_CODEC).orElse(List.of()).forEach(this::storeBee);
         this.savedFlowerPos = input.read("flower_pos", BlockPos.CODEC).orElse(null);
+        this.maxBees = input.getIntOr("Bukkit.MaxEntities", MAX_OCCUPANTS); // Paper - persist max bukkit occupants
     }
 
     @Override
@@ -314,12 +_,13 @@
         super.saveAdditional(output);
         output.store("bees", BeehiveBlockEntity.Occupant.LIST_CODEC, this.getBees());
         output.storeNullable("flower_pos", BlockPos.CODEC, this.savedFlowerPos);
+        output.putInt("Bukkit.MaxEntities", this.maxBees); // Paper - persist max bukkit occupants
     }
 
     @Override
     protected void applyImplicitComponents(final DataComponentGetter components) {
         super.applyImplicitComponents(components);
-        this.stored.clear();
+        this.stored = Lists.newArrayList(); // CraftBukkit - SPIGOT-7790: create new copy (may be modified in physics event triggered by honey change)
         List<BeehiveBlockEntity.Occupant> bees = components.getOrDefault(DataComponents.BEES, Bees.EMPTY).bees();
         bees.forEach(this::storeBee);
     }
@@ -347,15 +_,18 @@
 
     private static class BeeData {
         private final BeehiveBlockEntity.Occupant occupant;
+        private int exitTickCounter; // Paper - Fix bees aging inside hives; separate counter for checking if bee should exit to reduce exit attempts
         private int ticksInHive;
 
         private BeeData(final BeehiveBlockEntity.Occupant occupant) {
             this.occupant = occupant;
             this.ticksInHive = occupant.ticksInHive();
+            this.exitTickCounter = this.ticksInHive; // Paper - Fix bees aging inside hives
         }
 
         public boolean tick() {
-            return this.ticksInHive++ > this.occupant.minTicksInHive;
+            this.ticksInHive++; // Paper - Fix bees aging inside hives
+            return this.exitTickCounter++ > this.occupant.minTicksInHive; // Paper - Fix bees aging inside hives
         }
 
         public BeehiveBlockEntity.Occupant toOccupant() {
