--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -35,8 +_,18 @@
         this.catalystListener = new SculkCatalystBlockEntity.CatalystListener(blockState, new BlockPositionSource(worldPosition));
     }
 
+    // Paper start - Fix NPE in SculkBloomEvent world access
+    @Override
+    public void setLevel(Level level) {
+        super.setLevel(level);
+        this.catalystListener.sculkSpreader.level = level;
+    }
+    // Paper end - Fix NPE in SculkBloomEvent world access
+
     public static void serverTick(final Level level, final BlockPos pos, final BlockState state, final SculkCatalystBlockEntity entity) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = entity.getBlockPos(); // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPos up to five methods deep.
         entity.catalystListener.getSculkSpreader().updateCursors(level, pos, level.getRandom(), true);
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = null; // CraftBukkit
     }
 
     @Override
@@ -87,6 +_,13 @@
         public boolean handleGameEvent(final ServerLevel level, final Holder<GameEvent> event, final GameEvent.Context context, final Vec3 sourcePosition) {
             if (event.is(GameEvent.ENTITY_DIE) && context.sourceEntity() instanceof LivingEntity mob) {
                 if (!mob.wasExperienceConsumed()) {
+                    // Paper start - improve GameEvent events
+                    // TODO(future) add a sub event to capture more info about this happening, xp amount, amount to spread, etc.
+                    io.papermc.paper.event.ReceiveGameEvent blockEvent = org.bukkit.craftbukkit.event.CraftEventFactory.createReceiveGameEvent(level, this.positionSource.getPosition(level).orElseThrow(), event, context);
+                    if (!blockEvent.callEvent()) {
+                        return false;
+                    }
+                    // Paper end - improve GameEvent events
                     DamageSource lastDamageSource = mob.getLastDamageSource();
                     int experienceWouldDrop = mob.getExperienceReward(level, Optionull.map(lastDamageSource, DamageSource::getEntity));
                     if (mob.shouldDropExperience() && experienceWouldDrop > 0) {
