--- a/net/minecraft/world/level/block/entity/TestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestBlockEntity.java
@@ -31,6 +_,7 @@
 
     @Override
     protected void saveAdditional(final ValueOutput output) {
+        super.saveAdditional(output); // Paper - save the PDC
         output.store("mode", TestBlockMode.CODEC, this.mode);
         output.putString("message", this.message);
         output.putBoolean("powered", this.powered);
@@ -38,6 +_,7 @@
 
     @Override
     protected void loadAdditional(final ValueInput input) {
+        super.loadAdditional(input); // Paper - load the PDC
         this.mode = input.read("mode", TestBlockMode.CODEC).orElse(TestBlockMode.FAIL);
         this.message = input.getStringOr("message", "");
         this.powered = input.getBooleanOr("powered", false);
