From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index f224c1626a03ed9854eda456c02377e85ce356d7..8bb0926130a14351507b2b74feca42278c080381 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -129,7 +129,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
 
     @Nullable
     public Vec3 getPortalPosition(ServerLevel level, BlockPos pos) {
-        if (this.exitPortal == null && level.dimension() == Level.END) {
+        if (this.exitPortal == null && level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit - work in alternate worlds
             BlockPos blockPos = findOrCreateValidTeleportPos(level, pos);
             blockPos = blockPos.above(10);
             LOGGER.debug("Creating portal at {}", blockPos);
