--- a/net/minecraft/server/BlockMobSpawner.java
+++ b/net/minecraft/server/BlockMobSpawner.java
@@ -22,9 +22,19 @@
 
     public void dropNaturally(World world, BlockPosition blockposition, IBlockData iblockdata, float f, int i) {
         super.dropNaturally(world, blockposition, iblockdata, f, i);
+        /* CraftBukkit start - Delegate to getExpDrop
         int j = 15 + world.random.nextInt(15) + world.random.nextInt(15);
 
         this.dropExperience(world, blockposition, j);
+        */
+    }
+
+    @Override
+    public int getExpDrop(World world, IBlockData iblockdata, int enchantmentLevel) {
+        int j = 15 + world.random.nextInt(15) + world.random.nextInt(15);
+
+        return j;
+        // CraftBukkit end
     }
 
     public boolean b(IBlockData iblockdata) {
