--- a/net/minecraft/server/BlockJukeBox.java
+++ b/net/minecraft/server/BlockJukeBox.java
@@ -32,7 +32,7 @@
         }
     }
 
-    private void dropRecord(World world, BlockPosition blockposition, IBlockData iblockdata) {
+    public void dropRecord(World world, BlockPosition blockposition, IBlockData iblockdata) { // CraftBukkit - public
         if (!world.isClientSide) {
             TileEntity tileentity = world.getTileEntity(blockposition);
 
@@ -136,6 +136,11 @@
         }
 
         public void setRecord(ItemStack itemstack) {
+            // CraftBukkit start - There can only be one
+            if (itemstack != null) {
+                itemstack.count = 1;
+            }
+            // CraftBukkit end
             this.record = itemstack;
             this.update();
         }
