--- ../work/decompile-8eb82bde//net/minecraft/server/BlockTallPlant.java	2015-02-07 10:08:04.520216186 +0000
+++ src/main/java/net/minecraft/server/BlockTallPlant.java	2015-02-07 10:08:04.520216186 +0000
@@ -55,16 +55,20 @@
             Object object1 = flag ? world.getType(blockposition2).getBlock() : this;
 
             if (object == this) {
-                world.setTypeAndData(blockposition1, Blocks.AIR.getBlockData(), 3);
+                world.setTypeAndData(blockposition1, Blocks.AIR.getBlockData(), 2); // CraftBukkit
             }
 
             if (object1 == this) {
-                world.setTypeAndData(blockposition2, Blocks.AIR.getBlockData(), 3);
+                world.setTypeAndData(blockposition2, Blocks.AIR.getBlockData(), 2); // CraftBukkit
                 if (!flag) {
                     this.b(world, blockposition2, iblockdata, 0);
                 }
             }
 
+            // CraftBukkit start - delay updates
+            world.update(blockposition1, this);
+            world.update(blockposition2, this);
+            // CraftBukkit end
         }
     }
 
