--- ../work/decompile-8eb82bde//net/minecraft/server/CrashReport.java	2014-11-28 17:43:43.029707436 +0000
+++ src/main/java/net/minecraft/server/CrashReport.java	2014-11-28 17:38:17.000000000 +0000
@@ -40,6 +40,7 @@
         this.d.a("Memory", (Callable) (new CrashReportMemory(this)));
         this.d.a("JVM Flags", (Callable) (new CrashReportJVMFlags(this)));
         this.d.a("IntCache", (Callable) (new CrashReportIntCacheSize(this)));
+        this.d.a("CraftBukkit Information", (Callable) (new org.bukkit.craftbukkit.CraftCrashReport())); // CraftBukkit
     }
 
     public String a() {
