--- a/net/minecraft/server/CrashReport.java
+++ b/net/minecraft/server/CrashReport.java
@@ -63,6 +63,7 @@
 
             return String.format("%d total; %s", new Object[] { list.size(), list.stream().collect(Collectors.joining(" "))});
         });
+        this.d.a("CraftBukkit Information", (CrashReportCallable) new org.bukkit.craftbukkit.CraftCrashReport()); // CraftBukkit
     }
 
     public String a() {
