--- a/net/minecraft/server/DamageSource.java
+++ b/net/minecraft/server/DamageSource.java
@@ -36,6 +36,18 @@
     private boolean E;
     private boolean F;
     public final String translationIndex;
+    // CraftBukkit start
+    private boolean sweep;
+
+    public boolean isSweep() {
+        return sweep;
+    }
+
+    public DamageSource sweep() {
+        this.sweep = true;
+        return this;
+    }
+    // CraftBukkit end
 
     public static DamageSource a(EntityLiving entityliving) {
         return new EntityDamageSource("sting", entityliving);
