--- a/net/minecraft/server/DataWatcher.java
+++ b/net/minecraft/server/DataWatcher.java
@@ -70,7 +70,7 @@
         }
     }
 
-    public <T> void register(DataWatcherObject<T> datawatcherobject, T t0) {
+    public <T> void register(DataWatcherObject<T> datawatcherobject, Object t0) { // CraftBukkit T -> Object
         int i = datawatcherobject.a();
 
         if (i > 254) {
@@ -84,7 +84,7 @@
         }
     }
 
-    private <T> void registerObject(DataWatcherObject<T> datawatcherobject, T t0) {
+    private <T> void registerObject(DataWatcherObject<T> datawatcherobject, Object t0) { // CraftBukkit Object
         DataWatcher.Item datawatcher_item = new DataWatcher.Item(datawatcherobject, t0);
 
         this.e.writeLock().lock();
