--- ../work/decompile-8eb82bde//net/minecraft/server/Enchantment.java	2014-11-28 17:43:43.057707436 +0000
+++ src/main/java/net/minecraft/server/Enchantment.java	2014-11-28 17:38:20.000000000 +0000
@@ -8,6 +8,7 @@
 
 public abstract class Enchantment {
 
+    // CraftBukkit - update CraftEnchant.getName(i) if this changes
     private static final Enchantment[] byId = new Enchantment[256];
     public static final Enchantment[] b;
     private static final Map E = Maps.newHashMap();
@@ -55,6 +56,8 @@
             Enchantment.byId[i] = this;
             Enchantment.E.put(minecraftkey, this);
         }
+        
+        org.bukkit.enchantments.Enchantment.registerEnchantment(new org.bukkit.craftbukkit.enchantments.CraftEnchantment(this)); // CraftBukkit
     }
 
     public static Enchantment getByName(String s) {
