--- a/net/minecraft/server/EntityLlamaSpit.java
+++ b/net/minecraft/server/EntityLlamaSpit.java
@@ -6,7 +6,7 @@
 
 public class EntityLlamaSpit extends Entity implements IProjectile {
 
-    public EntityLlama shooter;
+    public EntityLiving shooter; // CraftBukkit - type
     private NBTTagCompound c;
 
     public EntityLlamaSpit(EntityTypes<? extends EntityLlamaSpit> entitytypes, World world) {
@@ -91,6 +91,7 @@
     }
 
     public void a(MovingObjectPosition movingobjectposition) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition); // CraftBukkit - Call event
         MovingObjectPosition.EnumMovingObjectType movingobjectposition_enummovingobjecttype = movingobjectposition.getType();
 
         if (movingobjectposition_enummovingobjecttype == MovingObjectPosition.EnumMovingObjectType.ENTITY && this.shooter != null) {
