--- /home/matt/mc-dev-private//net/minecraft/server/EntityCreature.java	2015-02-26 22:40:22.495608140 +0000
+++ src/main/java/net/minecraft/server/EntityCreature.java	2015-02-26 22:40:22.495608140 +0000
@@ -2,6 +2,10 @@
 
 import java.util.UUID;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityUnleashEvent;
+// CraftBukkit end
+
 public abstract class EntityCreature extends EntityInsentient {
 
     public static final UUID bk = UUID.fromString("E199AD21-BA8A-4C53-8D13-6182D5C69D3A");
@@ -69,6 +73,7 @@
 
             if (this instanceof EntityTameableAnimal && ((EntityTameableAnimal) this).isSitting()) {
                 if (f > 10.0F) {
+                    this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                     this.unleash(true, true);
                 }
 
@@ -100,6 +105,7 @@
             }
 
             if (f > 10.0F) {
+                this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                 this.unleash(true, true);
             }
         } else if (!this.cc() && this.bm) {
