--- a/net/minecraft/server/EntityPhantom.java
+++ b/net/minecraft/server/EntityPhantom.java
@@ -197,7 +197,7 @@
                         EntityHuman entityhuman = (EntityHuman) iterator.next();
 
                         if (PathfinderGoalTarget.a(EntityPhantom.this, entityhuman, false, false)) {
-                            EntityPhantom.this.setGoalTarget(entityhuman);
+                            EntityPhantom.this.setGoalTarget(entityhuman, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                             return true;
                         }
                     }
