--- a/net/minecraft/server/EntityRaider.java
+++ b/net/minecraft/server/EntityRaider.java
@@ -399,7 +399,7 @@
             while (iterator.hasNext()) {
                 EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                entityraider.setGoalTarget(this.c.getGoalTarget());
+                entityraider.setGoalTarget(this.c.getGoalTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
             }
 
         }
@@ -416,7 +416,7 @@
                 while (iterator.hasNext()) {
                     EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                    entityraider.setGoalTarget(entityliving);
+                    entityraider.setGoalTarget(this.c.getGoalTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                     entityraider.q(true);
                 }
 
@@ -490,7 +490,7 @@
 
         private final T b;
 
-        public b(EntityRaider entityraider) {
+        public b(T entityraider) { // CraftBukkit - decompile error
             this.b = entityraider;
             this.a(EnumSet.of(PathfinderGoal.Type.MOVE));
         }
