--- ../work/decompile-bb26c12b/net/minecraft/server/EntityDamageSourceIndirect.java	2014-11-27 08:59:46.661422023 +1100
+++ src/main/java/net/minecraft/server/EntityDamageSourceIndirect.java	2014-11-27 08:42:10.164850887 +1100
@@ -24,5 +24,11 @@
         String s1 = s + ".item";
 
         return itemstack != null && itemstack.hasName() && LocaleI18n.c(s1) ? new ChatMessage(s1, new Object[] { entityliving.getScoreboardDisplayName(), ichatbasecomponent, itemstack.C()}) : new ChatMessage(s, new Object[] { entityliving.getScoreboardDisplayName(), ichatbasecomponent});
+    }    
+    
+    // CraftBukkit start
+    public Entity getProximateDamageSource() {
+        return super.getEntity();
     }
+    // CraftBukkit end
 }
