--- ../work/decompile-bb26c12b/net/minecraft/server/EntityThrownExpBottle.java	2014-11-27 08:59:46.737421688 +1100
+++ src/main/java/net/minecraft/server/EntityThrownExpBottle.java	2014-11-27 08:42:10.112850989 +1100
@@ -28,8 +28,17 @@
 
     protected void a(MovingObjectPosition movingobjectposition) {
         if (!this.world.isStatic) {
-            this.world.triggerEffect(2002, new BlockPosition(this), 0);
-            int i = 3 + this.world.random.nextInt(5) + this.world.random.nextInt(5);
+            // CraftBukkit - moved to after event
+            // this.world.triggerEffect(2002, new BlockPosition(this), 0);
+            int i = 3 + this.world.random.nextInt(5) + this.world.random.nextInt(5);            
+            
+            // CraftBukkit start
+            org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExpBottleEvent(this, i);
+            i = event.getExperience();
+            if (event.getShowEffect()) {
+                this.world.triggerEffect(2002, new BlockPosition(this), 0);
+            }
+            // CraftBukkit end
 
             while (i > 0) {
                 int j = EntityExperienceOrb.getOrbValue(i);
