--- a/net/minecraft/server/EntityArrow.java
+++ b/net/minecraft/server/EntityArrow.java
@@ -7,6 +7,13 @@
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityCombustEvent;
+import org.bukkit.event.player.PlayerPickupArrowEvent;
+// CraftBukkit end
+
 public abstract class EntityArrow extends Entity implements IProjectile {
 
     private static final Predicate<Entity> g = IEntitySelector.f.and(IEntitySelector.a.and(Entity::isInteractable));
@@ -250,6 +257,7 @@
     }
 
     protected void a(MovingObjectPosition movingobjectposition) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition); // CraftBukkit - Call event
         if (movingobjectposition.entity != null) {
             this.b(movingobjectposition);
         } else {
@@ -299,7 +307,13 @@
         }
 
         if (this.isBurning() && !(entity instanceof EntityEnderman)) {
-            entity.setOnFire(5);
+            // CraftBukkit start
+            EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+            org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+            if (!combustEvent.isCancelled()) {
+                entity.setOnFire(combustEvent.getDuration());
+            }
+            // CraftBukkit end
         }
 
         if (entity.damageEntity(damagesource, (float) i)) {
@@ -444,6 +458,7 @@
 
     public void setShooter(@Nullable Entity entity) {
         this.shooter = entity == null ? null : entity.getUniqueID();
+        this.projectileSource = entity == null ? null : (LivingEntity) entity.getBukkitEntity(); // CraftBukkit
     }
 
     @Nullable
@@ -453,9 +468,22 @@
 
     public void d(EntityHuman entityhuman) {
         if (!this.world.isClientSide && (this.inGround || this.q()) && this.shake <= 0) {
+            // CraftBukkit start
+            ItemStack itemstack = this.getItemStack();
+            EntityItem item = new EntityItem(this.world, this.locX, this.locY, this.locZ, itemstack);
+            if (this.fromPlayer == PickupStatus.ALLOWED && entityhuman.inventory.canHold(itemstack) > 0) {
+                PlayerPickupArrowEvent event = new PlayerPickupArrowEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), new org.bukkit.craftbukkit.entity.CraftItem(this.world.getServer(), this, item), (org.bukkit.entity.Arrow) this.getBukkitEntity());
+                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+            }
             boolean flag = this.fromPlayer == EntityArrow.PickupStatus.ALLOWED || this.fromPlayer == EntityArrow.PickupStatus.CREATIVE_ONLY && entityhuman.abilities.canInstantlyBuild || this.q() && this.getShooter().getUniqueID() == entityhuman.getUniqueID();
 
-            if (this.fromPlayer == EntityArrow.PickupStatus.ALLOWED && !entityhuman.inventory.pickup(this.getItemStack())) {
+            if (this.fromPlayer == EntityArrow.PickupStatus.ALLOWED && !entityhuman.inventory.pickup(item.getItemStack())) {
+                // CraftBukkit end
                 flag = false;
             }
 
@@ -528,7 +556,14 @@
         }
 
         if (EnchantmentManager.a(Enchantments.ARROW_FIRE, entityliving) > 0) {
-            this.setOnFire(100);
+            // CraftBukkit start - call EntityCombustEvent
+            EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 100);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                this.setOnFire(event.getDuration());
+            }
+            // CraftBukkit end
         }
 
     }
