--- a/net/minecraft/server/ItemCooldown.java
+++ b/net/minecraft/server/ItemCooldown.java
@@ -7,8 +7,8 @@
 
 public class ItemCooldown {
 
-    private final Map<Item, ItemCooldown.Info> a = Maps.newHashMap();
-    private int b;
+    public final Map<Item, ItemCooldown.Info> a = Maps.newHashMap(); // PAIL: private->public
+    public int b; // PAIL: currentTick, private->public
 
     public ItemCooldown() {}
 
@@ -55,10 +55,10 @@
 
     protected void c(Item item) {}
 
-    class Info {
+    public class Info { // PAIL: private->public
 
         final int a;
-        final int b;
+        public final int b; // PAIL: endTick, private->public
 
         private Info(int i, int j) {
             this.a = i;
