--- a/net/minecraft/server/ItemFishingRod.java
+++ b/net/minecraft/server/ItemFishingRod.java
@@ -1,9 +1,13 @@
 package net.minecraft.server;
 
+import org.bukkit.event.player.PlayerFishEvent; // CraftBukkit
+
 public class ItemFishingRod extends Item {
 
     public ItemFishingRod(Item.Info item_info) {
         super(item_info);
+        // CraftBukkit start - obfuscator went a little crazy
+        /*
         this.a(new MinecraftKey("cast"), (itemstack, world, entityliving) -> {
             if (entityliving == null) {
                 return 0.0F;
@@ -18,6 +22,8 @@
                 return (flag || flag1) && entityliving instanceof EntityHuman && ((EntityHuman) entityliving).hookedFish != null ? 1.0F : 0.0F;
             }
         });
+        */
+        // CraftBukkit end
     }
 
     @Override
@@ -35,12 +41,23 @@
 
             world.playSound((EntityHuman) null, entityhuman.locX(), entityhuman.locY(), entityhuman.locZ(), SoundEffects.ENTITY_FISHING_BOBBER_RETRIEVE, SoundCategory.NEUTRAL, 1.0F, 0.4F / (ItemFishingRod.i.nextFloat() * 0.4F + 0.8F));
         } else {
-            world.playSound((EntityHuman) null, entityhuman.locX(), entityhuman.locY(), entityhuman.locZ(), SoundEffects.ENTITY_FISHING_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemFishingRod.i.nextFloat() * 0.4F + 0.8F));
+            // world.playSound((EntityHuman) null, entityhuman.locX(), entityhuman.locY(), entityhuman.locZ(), SoundEffects.ENTITY_FISHING_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemFishingRod.i.nextFloat() * 0.4F + 0.8F));
             if (!world.isClientSide) {
                 i = EnchantmentManager.c(itemstack);
                 int j = EnchantmentManager.b(itemstack);
 
-                world.addEntity(new EntityFishingHook(entityhuman, world, j, i));
+                // CraftBukkit start
+                EntityFishingHook entityfishinghook = new EntityFishingHook(entityhuman, world, j, i);
+                PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), null, (org.bukkit.entity.FishHook) entityfishinghook.getBukkitEntity(), PlayerFishEvent.State.FISHING);
+                world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+                if (playerFishEvent.isCancelled()) {
+                    entityhuman.hookedFish = null;
+                    return new InteractionResultWrapper(EnumInteractionResult.PASS, itemstack);
+                }
+                world.playSound((EntityHuman) null, entityhuman.locX(), entityhuman.locY(), entityhuman.locZ(), SoundEffects.ENTITY_FISHING_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemFishingRod.i.nextFloat() * 0.4F + 0.8F));
+                world.addEntity(entityfishinghook);
+                // CraftBukkit end
             }
 
             entityhuman.b(StatisticList.ITEM_USED.b(this));
