--- a/net/minecraft/server/PacketPlayOutNamedSoundEffect.java
+++ b/net/minecraft/server/PacketPlayOutNamedSoundEffect.java
@@ -38,7 +38,17 @@
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.d(SoundEffect.a.a((Object) this.a));
+        // CraftBukkit start
+        if (packetdataserializer.version > 107) {
+            packetdataserializer.d(SoundEffect.a.a(this.a));
+        } else {
+            int id = SoundEffect.a.a(this.a);
+            if (id >= SoundEffect.a.a(SoundEffect.a.get(new MinecraftKey("item.elytra.flying")))) {
+                id--;
+            }
+            packetdataserializer.d(id);
+        }
+        // CraftBukkit end
         packetdataserializer.a((Enum) this.b);
         packetdataserializer.writeInt(this.c);
         packetdataserializer.writeInt(this.d);
