--- a/net/minecraft/server/PathfinderGoalRemoveBlock.java
+++ b/net/minecraft/server/PathfinderGoalRemoveBlock.java
@@ -3,6 +3,11 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class PathfinderGoalRemoveBlock extends PathfinderGoalGotoTarget {
 
     private final Block g;
@@ -81,6 +86,14 @@
             }
 
             if (this.i > 60) {
+                // CraftBukkit start - Step on eggs
+                EntityInteractEvent event = new EntityInteractEvent(this.entity.getBukkitEntity(), CraftBlock.at(world, blockposition1));
+                world.getServer().getPluginManager().callEvent((EntityInteractEvent) event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+                // CraftBukkit end
                 world.a(blockposition1, false);
                 if (!world.isClientSide) {
                     for (int i = 0; i < 20; ++i) {
