--- ../work/decompile-bb26c12b/net/minecraft/server/PathfinderGoalEndermanPickupBlock.java	2014-11-27 08:59:46.825421301 +1100
+++ src/main/java/net/minecraft/server/PathfinderGoalEndermanPickupBlock.java	2014-11-27 08:42:10.164850887 +1100
@@ -25,8 +25,12 @@
         Block block = iblockdata.getBlock();
 
         if (EntityEnderman.co().contains(block)) {
-            this.enderman.setCarried(iblockdata);
-            world.setTypeUpdate(blockposition, Blocks.AIR.getBlockData());
+            // CraftBukkit start - Pickup event
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.enderman, this.enderman.world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), org.bukkit.Material.AIR).isCancelled()) {
+                this.enderman.setCarried(iblockdata);
+                world.setTypeUpdate(blockposition, Blocks.AIR.getBlockData());
+            }
+            // CraftBukkit end
         }
 
     }
