--- a/net/minecraft/server/PathfinderGoalOwnerHurtTarget.java
+++ b/net/minecraft/server/PathfinderGoalOwnerHurtTarget.java
@@ -34,7 +34,7 @@
 
     @Override
     public void c() {
-        this.e.setGoalTarget(this.b);
+        this.e.setGoalTarget(this.b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit - reason
         EntityLiving entityliving = this.a.getOwner();
 
         if (entityliving != null) {
