--- ../work/decompile-8eb82bde//net/minecraft/server/PathfinderGoalMakeLove.java	2014-11-28 17:43:43.293707431 +0000
+++ src/main/java/net/minecraft/server/PathfinderGoalMakeLove.java	2014-11-28 17:38:22.000000000 +0000
@@ -87,7 +87,7 @@
         this.b.o(false);
         entityvillager.setAgeRaw(-24000);
         entityvillager.setPositionRotation(this.b.locX, this.b.locY, this.b.locZ, 0.0F, 0.0F);
-        this.d.addEntity(entityvillager);
+        this.d.addEntity(entityvillager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
         this.d.broadcastEntityEffect(entityvillager, (byte) 12);
     }
 }
