--- ../work/decompile-bb26c12b/net/minecraft/server/RecipeMapClone.java	2014-11-27 08:59:46.873421089 +1100
+++ src/main/java/net/minecraft/server/RecipeMapClone.java	2014-11-27 08:42:10.152850911 +1100
@@ -1,8 +1,12 @@
 package net.minecraft.server;
 
-public class RecipeMapClone implements IRecipe {
+public class RecipeMapClone extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
 
-    public RecipeMapClone() {}
+    // CraftBukkit start - Delegate to new parent class
+    public RecipeMapClone() {
+        super(new ItemStack(Items.MAP, 0, -1), java.util.Arrays.asList(new ItemStack(Items.MAP, 0, 0)));
+    }
+    // CraftBukkit end
 
     public boolean a(InventoryCrafting inventorycrafting, World world) {
         int i = 0;
