--- ../work/decompile-8eb82bde//net/minecraft/server/RecipesBannerInnerClass2.java	2014-11-28 17:43:43.357707429 +0000
+++ src/main/java/net/minecraft/server/RecipesBannerInnerClass2.java	2014-11-28 17:38:18.000000000 +0000
@@ -1,8 +1,12 @@
 package net.minecraft.server;
 
-class RecipesBannerInnerClass2 implements IRecipe {
+class RecipesBannerInnerClass2 extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
 
-    private RecipesBannerInnerClass2() {}
+    // CraftBukkit start - Delegate to new parent class with bogus info
+    private RecipesBannerInnerClass2() {
+        super(new ItemStack(Items.BANNER, 0, 0), java.util.Arrays.asList(new ItemStack(Items.DYE, 0, 5)));
+    }
+    // CraftBukkit end
 
     public boolean a(InventoryCrafting inventorycrafting, World world) {
         boolean flag = false;
