--- ../work/decompile-8eb82bde//net/minecraft/server/RemoteControlCommandListener.java	2014-11-28 17:43:43.365707429 +0000
+++ src/main/java/net/minecraft/server/RemoteControlCommandListener.java	2014-11-28 17:38:22.000000000 +0000
@@ -26,6 +26,12 @@
     public IChatBaseComponent getScoreboardDisplayName() {
         return new ChatComponentText(this.getName());
     }
+ 
+    // CraftBukkit start - Send a String
+    public void sendMessage(String message) {
+        this.b.append(message);
+    }
+    // CraftBukkit end
 
     public void sendMessage(IChatBaseComponent ichatbasecomponent) {
         this.b.append(ichatbasecomponent.c());
