--- ../work/decompile-8eb82bde//net/minecraft/server/ShapeDetectorBlock.java	2015-02-02 20:26:40.044135504 +0000
+++ src/main/java/net/minecraft/server/ShapeDetectorBlock.java	2015-02-02 20:26:40.056135504 +0000
@@ -16,8 +16,10 @@
     }
 
     public IBlockData a() {
-        if (this.c == null && this.a.isLoaded(this.b)) {
-            this.c = this.a.getType(this.b);
+        // CraftBukkit start - fix null pointer
+        if (this.c == null) {
+            this.c = this.a.isLoaded(this.b) ? this.a.getType(this.b) : Blocks.AIR.getBlockData();
+        // CraftBukkit end
         }
 
         return this.c;
