--- ../work/decompile-8eb82bde//net/minecraft/server/TileEntityPiston.java	2014-11-28 17:43:43.409707428 +0000
+++ src/main/java/net/minecraft/server/TileEntityPiston.java	2014-11-28 17:38:23.000000000 +0000
@@ -104,6 +104,7 @@
     }
 
     public void c() {
+        if (this.world == null) return; // CraftBukkit
         this.j = this.i;
         if (this.j >= 1.0F) {
             this.a(1.0F, 0.25F);
