--- /home/matt/mc-dev-private//net/minecraft/server/VillageSiege.java	2015-02-26 22:40:23.175608133 +0000
+++ src/main/java/net/minecraft/server/VillageSiege.java	2015-02-26 22:40:23.175608133 +0000
@@ -140,7 +140,7 @@
             }
 
             entityzombie.setPositionRotation(vec3d.a, vec3d.b, vec3d.c, this.a.random.nextFloat() * 360.0F, 0.0F);
-            this.a.addEntity(entityzombie);
+            this.a.addEntity(entityzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_INVASION); // CraftBukkit
             BlockPosition blockposition = this.f.a();
 
             entityzombie.a(blockposition, this.f.b());
