--- a/net/minecraft/server/WorldGenVillagePieces.java
+++ b/net/minecraft/server/WorldGenVillagePieces.java
@@ -1559,7 +1559,7 @@
                     entityvillager.setPositionRotation((double) j1 + 0.5D, (double) k1, (double) l1 + 0.5D, 0.0F, 0.0F);
                     entityvillager.prepare(world.E(new BlockPosition(entityvillager)), (GroupDataEntity) null);
                     entityvillager.setProfession(this.c(i1, entityvillager.getProfession()));
-                    world.addEntity(entityvillager);
+                    world.addEntity(entityvillager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
                 }
 
             }
