--- a/net/minecraft/server/WorldGenMegaTreeProvider.java
+++ b/net/minecraft/server/WorldGenMegaTreeProvider.java
@@ -30,6 +30,7 @@
             return false;
         } else {
             ((WorldGenFeatureTreeConfiguration) worldgenfeatureconfigured.f).b();
+            setTreeType(worldgenfeatureconfigured); // CraftBukkit
             IBlockData iblockdata1 = Blocks.AIR.getBlockData();
 
             worldserver.setTypeAndData(blockposition.b(i, 0, j), iblockdata1, 4);
