--- ../work/decompile-bb26c12b/net/minecraft/server/WorldGenWitchHut.java	2014-11-27 08:59:46.933420825 +1100
+++ src/main/java/net/minecraft/server/WorldGenWitchHut.java	2014-11-27 08:42:10.152850911 +1100
@@ -77,7 +77,7 @@
 
                     entitywitch.setPositionRotation((double) i1 + 0.5D, (double) j1, (double) k1 + 0.5D, 0.0F, 0.0F);
                     entitywitch.prepare(world.E(new BlockPosition(i1, j1, k1)), (GroupDataEntity) null);
-                    world.addEntity(entitywitch);
+                    world.addEntity(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
                 }
             }
 
