--- a/net/minecraft/world/entity/raid/EntityRaider.java
+++ b/net/minecraft/world/entity/raid/EntityRaider.java
@@ -167,7 +167,7 @@
                     MobEffect mobeffect1 = new MobEffect(MobEffects.BAD_OMEN, 120000, i, false, false, true);
 
                     if (!this.level.getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS)) {
-                        entityhuman.addEffect(mobeffect1);
+                        entityhuman.addEffect(mobeffect1, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                     }
                 }
             }
@@ -307,7 +307,7 @@
 
         private final T mob;
 
-        public b(EntityRaider entityraider) {
+        public b(T entityraider) { // CraftBukkit - decompile error
             this.mob = entityraider;
             this.a(EnumSet.of(PathfinderGoal.Type.MOVE));
         }
@@ -523,7 +523,7 @@
             while (iterator.hasNext()) {
                 EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                entityraider.setGoalTarget(this.mob.getGoalTarget());
+                entityraider.setGoalTarget(this.mob.getGoalTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
             }
 
         }
@@ -540,7 +540,7 @@
                 while (iterator.hasNext()) {
                     EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                    entityraider.setGoalTarget(entityliving);
+                    entityraider.setGoalTarget(this.mob.getGoalTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                     entityraider.setAggressive(true);
                 }
 
