--- a/net/minecraft/world/entity/raid/EntityRaider.java
+++ b/net/minecraft/world/entity/raid/EntityRaider.java
@@ -43,6 +43,10 @@
 import net.minecraft.world.level.WorldAccess;
 import net.minecraft.world.phys.Vec3D;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public abstract class EntityRaider extends EntityMonsterPatrolling {
 
     protected static final DataWatcherObject<Boolean> IS_CELEBRATING = DataWatcher.defineId(EntityRaider.class, DataWatcherRegistry.BOOLEAN);
@@ -165,7 +169,7 @@
                     MobEffect mobeffect1 = new MobEffect(MobEffects.BAD_OMEN, 120000, i, false, false, true);
 
                     if (!this.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS)) {
-                        entityhuman.addEffect(mobeffect1);
+                        entityhuman.addEffect(mobeffect1, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                     }
                 }
             }
@@ -256,7 +260,7 @@
             this.onItemPickup(entityitem);
             this.setItemSlot(enumitemslot, itemstack);
             this.take(entityitem, itemstack.getCount());
-            entityitem.discard();
+            entityitem.discard(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             this.getCurrentRaid().setLeader(this.getWave(), this);
             this.setPatrolLeader(true);
         } else {
@@ -305,7 +309,7 @@
 
         private final T mob;
 
-        public b(EntityRaider entityraider) {
+        public b(T entityraider) { // CraftBukkit - decompile error
             this.mob = entityraider;
             this.setFlags(EnumSet.of(PathfinderGoal.Type.MOVE));
         }
@@ -521,7 +525,7 @@
             while (iterator.hasNext()) {
                 EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                entityraider.setTarget(this.mob.getTarget());
+                entityraider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
             }
 
         }
@@ -538,7 +542,7 @@
                 while (iterator.hasNext()) {
                     EntityRaider entityraider = (EntityRaider) iterator.next();
 
-                    entityraider.setTarget(entityliving);
+                    entityraider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                     entityraider.setAggressive(true);
                 }
 
