From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/util/PlaceholderLookupProvider.java b/net/minecraft/util/PlaceholderLookupProvider.java
index c78d8b2a6a84dd27e9e74c97db1145edeca95458..5f2eabacb3bbbcad480a293b7e2ea4355e665436 100644
--- a/net/minecraft/util/PlaceholderLookupProvider.java
+++ b/net/minecraft/util/PlaceholderLookupProvider.java
@@ -52,6 +52,13 @@ public class PlaceholderLookupProvider implements HolderGetter.Provider {
                             )
                         );
                 }
+
+                // Paper start add method to get the value for pre-filling builders in the reg mod API
+                @Override
+                public HolderLookup.Provider lookupForValueCopyViaBuilders() {
+                    return PlaceholderLookupProvider.this.context;
+                }
+                // Paper end - add method to get the value for pre-filling builders in the reg mod API
             }
         );
     }
