--- a/net/minecraft/network/DisconnectionDetails.java
+++ b/net/minecraft/network/DisconnectionDetails.java
@@ -5,7 +_,13 @@
 import java.util.Optional;
 import net.minecraft.network.chat.Component;
 
-public record DisconnectionDetails(Component reason, Optional<Path> report, Optional<URI> bugReportLink) {
+// Paper start - Configuration API: add support for enhanced disconnection causes
+public record DisconnectionDetails(Component reason, Optional<Path> report, Optional<URI> bugReportLink, Optional<Component> quitMessage, Optional<io.papermc.paper.connection.DisconnectionReason> disconnectionReason) {
+    public DisconnectionDetails(Component reason, Optional<Path> report, Optional<URI> bugReportLink) {
+        this(reason, report, bugReportLink, Optional.empty(), Optional.empty());
+    }
+// Paper end - Configuration API: add support for enhanced disconnection causes
+
     public DisconnectionDetails(Component reason) {
         this(reason, Optional.empty(), Optional.empty());
     }
