--- a/net/minecraft/network/RegistryFriendlyByteBuf.java
+++ b/net/minecraft/network/RegistryFriendlyByteBuf.java
@@ -12,6 +_,15 @@
         this.registryAccess = registryAccess;
     }
 
+    // Paper start - Track codec depth
+    public RegistryFriendlyByteBuf(final ByteBuf source, final RegistryFriendlyByteBuf parent) {
+        super(source);
+        this.registryAccess = parent.registryAccess;
+        this.trackCodecDepth = parent.trackCodecDepth;
+        this.codecDepth = parent.codecDepth;
+    }
+    // Paper end - Track codec depth
+
     public RegistryAccess registryAccess() {
         return this.registryAccess;
     }
