From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
index 2ae0015d6c00640a72864091235b71ad0d4aaac3..a4999a82fa5ed58b88cc11b42524d2513a3ed37a 100644
--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -85,7 +85,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
             boolean attachedMob = false;
 
             for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
-                if (leashable.canHaveALeashAttachedTo(this)) {
+                if (leashable.canHaveALeashAttachedTo(this) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(leashable, this, player, hand)) { // Paper - leash event
                     leashable.setLeashedTo(this, true);
                     attachedMob = true;
                 }
@@ -94,7 +94,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
             boolean anyDropped = false;
             if (!attachedMob && !player.isSecondaryUseActive()) {
                 for (Leashable mob : Leashable.leashableLeashedTo(this)) {
-                    if (mob.canHaveALeashAttachedTo(player)) {
+                    if (mob.canHaveALeashAttachedTo(player) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(mob, player, player, hand)) { // Paper - leash event
                         mob.setLeashedTo(player, true);
                         anyDropped = true;
                     }
@@ -114,7 +114,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
     @Override
     public void notifyLeasheeRemoved(final Leashable entity) {
         if (Leashable.leashableLeashedTo(this).isEmpty()) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
         }
     }
 
