--- a/net/minecraft/world/waypoints/WaypointTransmitter.java
+++ b/net/minecraft/world/waypoints/WaypointTransmitter.java
@@ -20,6 +_,7 @@
     Waypoint.Icon waypointIcon();
 
     static boolean doesSourceIgnoreReceiver(final LivingEntity source, final ServerPlayer receiver) {
+        if (!receiver.getBukkitEntity().canSee(source.getBukkitEntity())) return true; // Paper - ignore if entity is hidden from player
         if (receiver.isSpectator()) {
             return false;
         } else if (!source.isSpectator() && !source.hasIndirectPassenger(receiver)) {
