--- a/net/minecraft/server/commands/CommandDifficulty.java
+++ b/net/minecraft/server/commands/CommandDifficulty.java
@@ -42,11 +42,12 @@
 
     public static int setDifficulty(CommandListenerWrapper commandlistenerwrapper, EnumDifficulty enumdifficulty) throws CommandSyntaxException {
         MinecraftServer minecraftserver = commandlistenerwrapper.getServer();
+        net.minecraft.server.level.WorldServer worldServer = commandlistenerwrapper.getLevel(); // CraftBukkit
 
-        if (minecraftserver.getWorldData().getDifficulty() == enumdifficulty) {
+        if (worldServer.getDifficulty() == enumdifficulty) { // CraftBukkit
             throw CommandDifficulty.ERROR_ALREADY_DIFFICULT.create(enumdifficulty.getKey());
         } else {
-            minecraftserver.setDifficulty(enumdifficulty, true);
+            worldServer.serverLevelData.setDifficulty(enumdifficulty); // CraftBukkit
             commandlistenerwrapper.sendSuccess(IChatBaseComponent.translatable("commands.difficulty.success", enumdifficulty.getDisplayName()), true);
             return 0;
         }
