--- a/net/minecraft/commands/arguments/ArgumentEntity.java
+++ b/net/minecraft/commands/arguments/ArgumentEntity.java
@@ -95,9 +95,15 @@
     }
 
     public EntitySelector parse(StringReader stringreader) throws CommandSyntaxException {
+        // CraftBukkit start
+        return parse(stringreader, false);
+    }
+
+    public EntitySelector parse(StringReader stringreader, boolean overridePermissions) throws CommandSyntaxException {
+        // CraftBukkit end
         boolean flag = false;
         ArgumentParserSelector argumentparserselector = new ArgumentParserSelector(stringreader);
-        EntitySelector entityselector = argumentparserselector.parse();
+        EntitySelector entityselector = argumentparserselector.parse(overridePermissions); // CraftBukkit
 
         if (entityselector.a() > 1 && this.single) {
             if (this.playersOnly) {
