--- a/net/minecraft/commands/arguments/ArgumentEntity.java
+++ b/net/minecraft/commands/arguments/ArgumentEntity.java
@@ -102,9 +102,15 @@
     }
 
     private EntitySelector parse(StringReader stringreader, boolean flag) throws CommandSyntaxException {
+        // CraftBukkit start
+        return parse(stringreader, flag, false);
+    }
+
+    public EntitySelector parse(StringReader stringreader, boolean flag, boolean overridePermissions) throws CommandSyntaxException {
+        // CraftBukkit end
         boolean flag1 = false;
         ArgumentParserSelector argumentparserselector = new ArgumentParserSelector(stringreader, flag);
-        EntitySelector entityselector = argumentparserselector.parse();
+        EntitySelector entityselector = argumentparserselector.parse(overridePermissions); // CraftBukkit
 
         if (entityselector.getMaxResults() > 1 && this.single) {
             if (this.playersOnly) {
