From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/animal/squid/Squid.java b/net/minecraft/world/entity/animal/squid/Squid.java
index ad6e0e895f9ecf7b68fd183cd8726aaf773f7c31..637cdcf459aaf78dc381d8a990b0e2c571c06de8 100644
--- a/net/minecraft/world/entity/animal/squid/Squid.java
+++ b/net/minecraft/world/entity/animal/squid/Squid.java
@@ -51,7 +51,7 @@ public class Squid extends AgeableWaterCreature {
 
     public Squid(final EntityType<? extends Squid> type, final Level level) {
         super(type, level);
-        this.random.setSeed(this.getId());
+        // this.random.setSeed(this.getId()); // Paper - Share random for entities to make them more random
         this.tentacleSpeed = 1.0F / (this.random.nextFloat() + 1.0F) * 0.2F;
     }
 
