--- a/net/minecraft/nbt/ByteArrayTag.java
+++ b/net/minecraft/nbt/ByteArrayTag.java
@@ -23,6 +_,7 @@
         private static byte[] readAccounted(final DataInput input, final NbtAccounter accounter) throws IOException {
             accounter.accountBytes(24L);
             int length = input.readInt();
+            com.google.common.base.Preconditions.checkArgument(length < 1 << 24); // Spigot
             accounter.accountBytes(1L, length);
             byte[] data = new byte[length];
             input.readFully(data);
