From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/nbt/IntArrayTag.java b/net/minecraft/nbt/IntArrayTag.java
index fb443d2187d479a08bc96af122c329dc91b4a1ce..a8ea2aeb5a02903a37376fb78b49c10745147411 100644
--- a/net/minecraft/nbt/IntArrayTag.java
+++ b/net/minecraft/nbt/IntArrayTag.java
@@ -23,6 +23,7 @@ public final class IntArrayTag implements CollectionTag {
         private static int[] readAccounted(DataInput input, NbtAccounter accounter) throws IOException {
             accounter.accountBytes(24L);
             int _int = input.readInt();
+            com.google.common.base.Preconditions.checkArgument(_int < 1 << 24); // Spigot
             accounter.accountBytes(4L, _int);
             int[] ints = new int[_int];
 
