From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/DoubleBlockCombiner.java b/net/minecraft/world/level/block/DoubleBlockCombiner.java
index fe08e68f9bd0cafdaf74afdeb685e755e57d65a2..865b239095aca8672af3f33a6b3720cbaf6df577 100644
--- a/net/minecraft/world/level/block/DoubleBlockCombiner.java
+++ b/net/minecraft/world/level/block/DoubleBlockCombiner.java
@@ -34,7 +34,12 @@ public class DoubleBlockCombiner {
                 return new DoubleBlockCombiner.NeighborCombineResult.Single<>(blockEntity);
             } else {
                 BlockPos neighborPos = pos.relative(connectionResolver.apply(state));
-                BlockState neighbourState = level.getBlockState(neighborPos);
+                // Paper start - Don't load Chunks from Hoppers and other things
+                BlockState neighbourState = level.getBlockStateIfLoaded(neighborPos);
+                if (neighbourState == null) {
+                    return new DoubleBlockCombiner.NeighborCombineResult.Single<>(blockEntity);
+                }
+                // Paper end - Don't load Chunks from Hoppers and other things
                 if (neighbourState.is(state.getBlock())) {
                     DoubleBlockCombiner.BlockType neighbourType = typeResolver.apply(neighbourState);
                     if (neighbourType != DoubleBlockCombiner.BlockType.SINGLE
