From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/DragonEggBlock.java b/net/minecraft/world/level/block/DragonEggBlock.java
index 75bfdea8a179e69e6e404489da554ee67af2e319..9e193d0ed9b2d31eec15a0d552a713379445fb6d 100644
--- a/net/minecraft/world/level/block/DragonEggBlock.java
+++ b/net/minecraft/world/level/block/DragonEggBlock.java
@@ -61,6 +61,16 @@ public class DragonEggBlock extends FallingBlock {
                 && !level.getBlockState(testPos.below()).isAir()
                 && worldBorder.isWithinBounds(testPos)
                 && level.isInsideBuildHeight(testPos)) {
+                // CraftBukkit start
+                org.bukkit.block.Block from = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
+                org.bukkit.block.Block to = org.bukkit.craftbukkit.block.CraftBlock.at(level, testPos);
+                org.bukkit.event.block.BlockFromToEvent event = new org.bukkit.event.block.BlockFromToEvent(from, to);
+                if (!event.callEvent()) {
+                    return;
+                }
+
+                testPos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+                // CraftBukkit end
                 if (level.isClientSide()) {
                     for (int j = 0; j < 128; j++) {
                         double d = random.nextDouble();
