From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/HopperBlock.java b/net/minecraft/world/level/block/HopperBlock.java
index f92f5939bd2d3bf9f81f1815dde95636087888d2..e032b5289abe413009fe7c9068e71e788fab512a 100644
--- a/net/minecraft/world/level/block/HopperBlock.java
+++ b/net/minecraft/world/level/block/HopperBlock.java
@@ -106,8 +106,7 @@ public class HopperBlock extends BaseEntityBlock {
     protected InteractionResult useWithoutItem(
         final BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult
     ) {
-        if (!level.isClientSide() && level.getBlockEntity(pos) instanceof HopperBlockEntity hopper) {
-            player.openMenu(hopper);
+        if (!level.isClientSide() && level.getBlockEntity(pos) instanceof HopperBlockEntity hopper && player.openMenu(hopper).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
             player.awardStat(Stats.INSPECT_HOPPER);
         }
 
@@ -167,6 +166,7 @@ public class HopperBlock extends BaseEntityBlock {
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         BlockEntity blockEntity = level.getBlockEntity(pos);
         if (blockEntity instanceof HopperBlockEntity) {
             HopperBlockEntity.entityInside(level, pos, state, entity, (HopperBlockEntity)blockEntity);
