From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/LightningRodBlock.java b/net/minecraft/world/level/block/LightningRodBlock.java
index a7f7f5a1a4aff79e66536ac7ae9c849819ab4dc4..9b696d6ffeaa5302c713c011d74f40c67aa59473 100644
--- a/net/minecraft/world/level/block/LightningRodBlock.java
+++ b/net/minecraft/world/level/block/LightningRodBlock.java
@@ -83,6 +83,13 @@ public class LightningRodBlock extends RodBlock implements SimpleWaterloggedBloc
     }
 
     public void onLightningStrike(final BlockState state, final Level level, final BlockPos pos) {
+        // Paper start - Call BlockRedstoneEvent
+        if (!state.getValue(POWERED)) {
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, true)) {
+                return;
+            }
+        }
+        // Paper end - Call BlockRedstoneEvent
         level.setBlock(pos, state.setValue(POWERED, true), Block.UPDATE_ALL);
         this.updateNeighbours(state, level, pos);
         level.scheduleTick(pos, this, 8);
@@ -96,6 +103,13 @@ public class LightningRodBlock extends RodBlock implements SimpleWaterloggedBloc
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        // Paper start - Call BlockRedstoneEvent
+        if (state.getValue(POWERED)) {
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, false)) {
+                return;
+            }
+        }
+        // Paper end - Call BlockRedstoneEvent
         level.setBlock(pos, state.setValue(POWERED, false), Block.UPDATE_ALL);
         this.updateNeighbours(state, level, pos);
     }
