From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/PoweredRailBlock.java b/net/minecraft/world/level/block/PoweredRailBlock.java
index fafeda63f87b5cf46447c135d3662f3f41de25a7..f33b53257a231dccf16740f1e78a3cdfa6e70726 100644
--- a/net/minecraft/world/level/block/PoweredRailBlock.java
+++ b/net/minecraft/world/level/block/PoweredRailBlock.java
@@ -133,6 +133,13 @@ public class PoweredRailBlock extends BaseRailBlock {
             || this.findPoweredRailSignal(level, pos, state, true, 0)
             || this.findPoweredRailSignal(level, pos, state, false, 0);
         if (flag != poweredValue) {
+            // CraftBukkit start
+            int power = flag ? 15 : 0;
+            int newPower = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, power, 15 - power).getNewCurrent();
+            if (newPower == power) {
+                return;
+            }
+            // CraftBukkit end
             level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(flag)), 3);
             level.updateNeighborsAt(pos.below(), this);
             if (state.getValue(SHAPE).isSlope()) {
