From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/SugarCaneBlock.java b/net/minecraft/world/level/block/SugarCaneBlock.java
index 500f59f2442178f7fbce277641783c348444bfb5..4437fd5724e094e8d2edb97ca9bf494ef57f5b79 100644
--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -55,12 +55,13 @@ public class SugarCaneBlock extends Block {
                 i++;
             }
 
-            if (i < 3) {
+            if (i < level.paperConfig().maxGrowthHeight.reeds) { // Paper - Configurable cactus/bamboo/reed growth height
                 int ageValue = state.getValue(AGE);
-                if (ageValue == 15) {
-                    level.setBlockAndUpdate(pos.above(), this.defaultBlockState());
+                int modifier = level.spigotConfig.caneModifier; // Spigot - SPIGOT-7159: Better modifier resolution
+                if (ageValue >= 15 || (modifier != 100 && random.nextFloat() < (modifier / (100.0F * 16)))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos.above(), this.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
                     level.setBlock(pos, state.setValue(AGE, 0), Block.UPDATE_NONE);
-                } else {
+                } else if (modifier == 100 || random.nextFloat() < (modifier / (100.0F * 16))) { // Spigot - SPIGOT-7159: Better modifier resolution
                     level.setBlock(pos, state.setValue(AGE, ageValue + 1), Block.UPDATE_NONE);
                 }
             }
