--- a/net/minecraft/world/entity/ai/behavior/warden/Roar.java
+++ b/net/minecraft/world/entity/ai/behavior/warden/Roar.java
@@ -64,7 +_,7 @@
             entity.setPose(Pose.STANDING);
         }
 
-        entity.getBrain().getMemory(MemoryModuleType.ROAR_TARGET).ifPresent(entity::setAttackTarget);
+        entity.getBrain().getMemory(MemoryModuleType.ROAR_TARGET).ifPresent(livingEntity -> entity.setAttackTarget(livingEntity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY)); // Paper
         entity.getBrain().eraseMemory(MemoryModuleType.ROAR_TARGET);
     }
 }
