From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/projectile/Arrow.java b/net/minecraft/world/entity/projectile/Arrow.java
index 76ff9044a4f6cc4698e748345983c536e621b7c7..c1e09e701757a300183b62d343ded03033e63aa7 100644
--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -121,12 +121,13 @@ public class Arrow extends AbstractArrow {
                         mobEffectInstance.isVisible()
                     ),
                     effectSource
+                    , org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW // CraftBukkit
                 );
             }
         }
 
         for (MobEffectInstance mobEffectInstance : potionContents.customEffects()) {
-            living.addEffect(mobEffectInstance, effectSource);
+            living.addEffect(mobEffectInstance, effectSource, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }
     }
 
