--- a/net/minecraft/world/RandomizableContainer.java
+++ b/net/minecraft/world/RandomizableContainer.java
@@ -26,7 +_,7 @@
 
     void setLootTable(final @Nullable ResourceKey<LootTable> lootTable);
 
-    default void setLootTable(final ResourceKey<LootTable> lootTable, final long seed) {
+    default void setLootTable(final @Nullable ResourceKey<LootTable> lootTable, final long seed) { // Paper - add nullable
         this.setLootTable(lootTable);
         this.setLootTableSeed(seed);
     }
@@ -50,8 +_,9 @@
     default boolean tryLoadLootTable(final ValueInput base) {
         ResourceKey<LootTable> lootTable = base.read("LootTable", LootTable.KEY_CODEC).orElse(null);
         this.setLootTable(lootTable);
+        if (this.lootableData() != null && lootTable != null) this.lootableData().loadNbt(base); // Paper - LootTable API
         this.setLootTableSeed(base.getLongOr("LootTableSeed", 0L));
-        return lootTable != null;
+        return lootTable != null && this.lootableData() == null; // Paper - only track the loot table if there is chance for replenish
     }
 
     default boolean trySaveLootTable(final ValueOutput base) {
@@ -60,26 +_,42 @@
             return false;
         } else {
             base.store("LootTable", LootTable.KEY_CODEC, lootTable);
+            if (this.lootableData() != null) this.lootableData().saveNbt(base); // Paper - LootTable API
             long lootTableSeed = this.getLootTableSeed();
             if (lootTableSeed != 0L) {
                 base.putLong("LootTableSeed", lootTableSeed);
             }
 
-            return true;
+            return this.lootableData() == null; // Paper - only track the loot table if there is chance for replenish
         }
     }
 
     default void unpackLootTable(final @Nullable Player player) {
+        // Paper start - LootTable API
+        this.unpackLootTable(player, false);
+    }
+    default void unpackLootTable(@Nullable final Player player, final boolean forceClearLootTable) {
+        // Paper end - LootTable API
         Level level = this.getLevel();
         BlockPos worldPosition = this.getBlockPos();
         ResourceKey<LootTable> lootTableKey = this.getLootTable();
-        if (lootTableKey != null && level != null && level.getServer() != null) {
+        // Paper start - LootTable API
+        lootReplenish: if (lootTableKey != null && level != null && level.getServer() != null) {
+            if (this.lootableData() != null && !this.lootableData().shouldReplenish(this, com.destroystokyo.paper.loottable.PaperLootableInventoryData.CONTAINER, player)) {
+                if (forceClearLootTable) {
+                    this.setLootTable(null);
+                }
+                break lootReplenish;
+            }
+            // Paper end - LootTable API
             LootTable lootTable = level.getServer().reloadableRegistries().getLootTable(lootTableKey);
             if (player instanceof ServerPlayer) {
                 CriteriaTriggers.GENERATE_LOOT.trigger((ServerPlayer)player, lootTableKey);
             }
 
+            if (forceClearLootTable || this.lootableData() == null || this.lootableData().shouldClearLootTable(this, com.destroystokyo.paper.loottable.PaperLootableInventoryData.CONTAINER, player)) { // Paper - LootTable API
             this.setLootTable(null);
+            } // Paper - LootTable API
             LootParams.Builder params = new LootParams.Builder((ServerLevel)level).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(worldPosition));
             if (player != null) {
                 params.withLuck(player.getLuck()).withParameter(LootContextParams.THIS_ENTITY, player);
@@ -88,4 +_,16 @@
             lootTable.fill(this, params.create(LootContextParamSets.CHEST), this.getLootTableSeed());
         }
     }
+
+    // Paper start - LootTable API
+    @org.jetbrains.annotations.Contract(pure = true)
+    default com.destroystokyo.paper.loottable.@Nullable PaperLootableInventoryData lootableData() {
+        return null; // some containers don't really have a "replenish" ability like decorated pots
+    }
+
+    default com.destroystokyo.paper.loottable.PaperLootableInventory getLootableInventory() {
+        final org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(java.util.Objects.requireNonNull(this.getLevel(), "Cannot manage loot tables on block entities not in world"), this.getBlockPos());
+        return (com.destroystokyo.paper.loottable.PaperLootableInventory) block.getState(false);
+    }
+    // Paper end - LootTable API
 }
