From 3acbb3aac1c9465f406b4dedbff3f070fa3d7974 Mon Sep 17 00:00:00 2001
From: "Evan A. Haskell" <eah2119@gmail.com>
Date: Thu, 26 Jun 2014 18:37:29 -0400
Subject: [PATCH] Fix Null Tile Entity Worlds


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index db025bf..ebba06e 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2012,6 +2012,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
 
+                tileentity.a(this); // PaperSpigot - No null worlds
                 this.a.add(tileentity);
             } else {
                 this.tileEntityList.add(tileentity);
-- 
1.9.1

