From 172822275bb7cf1ea56a306a5d08b4daa652e8bb Mon Sep 17 00:00:00 2001
From: DoctorDark <doctordark11@gmail.com>
Date: Wed, 16 Mar 2016 02:21:39 -0500
Subject: [PATCH] Configurable end credits


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 8e3a0f3..2e88385 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -161,4 +161,10 @@ public class PaperWorldConfig {
         queueLightUpdates = getBoolean("queue-light-updates", false);
         log("Lighting Queue enabled: " + queueLightUpdates);
     }
+
+    public boolean disableEndCredits;
+    private void disableEndCredits() {
+        disableEndCredits = getBoolean("game-mechanics.disable-end-credits", false);
+        log("End credits disabled: " + disableEndCredits);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 3e33f49..8d04c82 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -508,6 +508,15 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         return this.world.pvpMode;
     }
 
+    /**
+     * Paper - Give "theEnd2" achievement if the player doesn't already have it
+     */
+    private void giveTheEnd2() {
+        if (!this.a(AchievementList.D)) {
+            this.b(AchievementList.D);
+        }
+    }
+
     public Entity c(int i) {
         //this.cj = true; // CraftBukkit - Moved down and into PlayerList#changeDimension
         if (this.dimension == 1 && i == 1) {
@@ -515,7 +524,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.world.kill(this);
             if (!this.viewingCredits) {
                 this.viewingCredits = true;
-                if (this.a(AchievementList.D)) {
+                // Paper start - Allow configurable end portal credits
+                if (world.paperConfig.disableEndCredits || this.a(AchievementList.D)) {
+                    this.giveTheEnd2();
+                    // Paper end
                     this.playerConnection.sendPacket(new PacketPlayOutGameStateChange(4, 0.0F));
                 } else {
                     this.b((Statistic) AchievementList.D);
-- 
2.7.4

