From 344248bcdb3a81dca026e29d05ac86f141bb20b0 Mon Sep 17 00:00:00 2001
From: Dmck2b <suddenly@suddenly.coffee>
Date: Sun, 31 Aug 2014 18:12:10 +0100
Subject: [PATCH] Temporary fix for rails dupe


diff --git a/src/main/java/net/minecraft/server/BlockMinecartTrackAbstract.java b/src/main/java/net/minecraft/server/BlockMinecartTrackAbstract.java
index d36aacb..b6b4146 100644
--- a/src/main/java/net/minecraft/server/BlockMinecartTrackAbstract.java
+++ b/src/main/java/net/minecraft/server/BlockMinecartTrackAbstract.java
@@ -105,8 +105,12 @@ public abstract class BlockMinecartTrackAbstract extends Block {
             }
 
             if (flag) {
-                this.b(world, i, j, k, world.getData(i, j, k), 0);
-                world.setAir(i, j, k);
+                // PaperSpigot start - Rails dupe workaround
+                if (world.getType(i, j, k).getMaterial() != Material.AIR) {
+                    this.b(world, i, j, k, world.getData(i, j, k), 0);
+                    world.setAir(i, j, k);
+                }
+                // PaperSpigot end
             } else {
                 this.a(world, i, j, k, l, i1, block);
             }
-- 
1.9.1

