From 2bed385eff3e829e3d2d9de35b5093d6f0853512 Mon Sep 17 00:00:00 2001
From: Minecrell <dev@minecrell.net>
Date: Mon, 1 Sep 2014 15:33:57 -0500
Subject: [PATCH] Fix for ProtocolLib compatibility


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 6d579e3..98b0979 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -250,7 +250,14 @@ public class NetworkManager extends SimpleChannelInboundHandler {
 
     // Spigot start - protocol patch
     public void enableCompression() {
-        m.pipeline().addBefore( "decoder", "decompress", new SpigotDecompressor() );
+        // PaperSpigot start - Fix ProtocolLib compatibility
+        if ( m.pipeline().get("protocol_lib_decoder") != null ) {
+            m.pipeline().addBefore( "protocol_lib_decoder", "decompress", new SpigotDecompressor() );
+        } else {
+            m.pipeline().addBefore( "decoder", "decompress", new SpigotDecompressor() );
+        }
+        // PaperSpigot end
+
         m.pipeline().addBefore( "encoder", "compress", new SpigotCompressor() );
     }
     // Spigot end
-- 
1.9.1

