From 7ea0b75ec2b73d46cffd36cc8bf2db862735b16e Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Tue, 14 Jul 2015 10:03:45 -0700
Subject: [PATCH] Optimize item movement


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index eacd4b8..62b8b9e 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1222,6 +1222,8 @@ public abstract class World implements IBlockAccess {
         }
         // Spigot end
 
+        if (entity instanceof EntityItem) return arraylist; // PaperSpigot - Optimize item movement
+
         double d0 = 0.25D;
         List list = this.getEntities(entity, axisalignedbb.grow(d0, d0, d0));
 
-- 
1.9.5.msysgit.1

