From 17106878d61a4014f256fb538e80ec6efdadc788 Mon Sep 17 00:00:00 2001
From: William <admin@domnian.com>
Date: Fri, 18 Mar 2016 03:30:17 -0400
Subject: [PATCH] Allow Reloading of Custom Permissions

https://github.com/PaperMC/Paper/issues/49

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 4db5aacc4..1c9637ff7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1960,5 +1960,20 @@ public final class CraftServer implements Server {
             return false;
         }
     }
+
+    @Override
+    public void reloadPermissions() {
+        ((SimplePluginManager) pluginManager).clearPermissions();
+        loadCustomPermissions();
+        for (Plugin plugin : pluginManager.getPlugins()) {
+            plugin.getDescription().getPermissions().forEach((perm) -> {
+                try {
+                    pluginManager.addPermission(perm);
+                } catch (IllegalArgumentException ex) {
+                    getLogger().log(Level.WARNING, "Plugin " + plugin.getDescription().getFullName() + " tried to register permission '" + perm.getName() + "' but it's already registered", ex);
+                }
+            });
+        }
+    }
     // Paper end
 }
-- 
2.18.0

