From e9deaf4147f5d5cff346d3af8e68a91e67682322 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 15 Nov 2015 19:35:31 -0600
Subject: [PATCH] Don't damage already dead ArmorStands


diff --git a/src/main/java/net/minecraft/server/EntityArmorStand.java b/src/main/java/net/minecraft/server/EntityArmorStand.java
index cd57d69..2ea4a5f 100644
--- a/src/main/java/net/minecraft/server/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/server/EntityArmorStand.java
@@ -382,9 +382,9 @@ public class EntityArmorStand extends EntityLiving {
             return false;
         }
         // CraftBukkit end
-        if (this.world.isClientSide) {
+        if (this.world.isClientSide || this.dead) {
             return false;
-        } else if (DamageSource.OUT_OF_WORLD.equals(damagesource)) {
+        } else if (DamageSource.OUT_OF_WORLD.equals(damagesource)) { // PaperSpigot
             this.die();
             return false;
         } else if (!this.isInvulnerable(damagesource) && !this.h && !this.s()) {
-- 
2.6.3

