From dd7c164ead24c4f2b19d7d110e07ab96c037e83d Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 19 Mar 2016 15:16:54 -0400
Subject: [PATCH] Pass world to Village creation

fixes NPE bug #95

diff --git a/src/main/java/net/minecraft/server/PersistentVillage.java b/src/main/java/net/minecraft/server/PersistentVillage.java
index a5b79ecad..01f7cee38 100644
--- a/src/main/java/net/minecraft/server/PersistentVillage.java
+++ b/src/main/java/net/minecraft/server/PersistentVillage.java
@@ -238,7 +238,7 @@ public class PersistentVillage extends PersistentBase {
 
         for (int i = 0; i < nbttaglist.size(); ++i) {
             NBTTagCompound nbttagcompound1 = nbttaglist.get(i);
-            Village village = new Village();
+            Village village = new Village(world); // Paper
 
             village.a(nbttagcompound1);
             this.villages.add(village);
diff --git a/src/main/java/net/minecraft/server/Village.java b/src/main/java/net/minecraft/server/Village.java
index fb52785b9..2eb33a986 100644
--- a/src/main/java/net/minecraft/server/Village.java
+++ b/src/main/java/net/minecraft/server/Village.java
@@ -24,7 +24,7 @@ public class Village {
     private final List<Village.Aggressor> k;
     private int l;
 
-    public Village() {
+    private Village() { // Paper - Nothing should call this - world needs to be set.
         this.c = BlockPosition.ZERO;
         this.d = BlockPosition.ZERO;
         this.j = Maps.newHashMap();
-- 
2.18.0

