From 54d988e5c406fb6f6cacac327c994b2d6132d0b9 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 3 Apr 2016 17:48:50 -0400
Subject: [PATCH] Fix Cancelling BlockPlaceEvent triggering physics


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1519e9e6b..4870264c6 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -460,6 +460,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     public void b(BlockPosition blockposition, IBlockData iblockdata, IBlockData iblockdata1) {}
 
     public void applyPhysics(BlockPosition blockposition, Block block) {
+        if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
         this.a(blockposition.west(), block, blockposition);
         this.a(blockposition.east(), block, blockposition);
         this.a(blockposition.down(), block, blockposition);
-- 
2.24.0

