From 114866a0d3665b3f45a5e3e5b95d9ba34ec895fa Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 13 Feb 2016 19:28:50 -0600
Subject: [PATCH] Optimize getBlockData

Hot method, so reduce # of instructions for the method.

diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 7efacfa..c636da3 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -489,7 +489,18 @@ public class Chunk {
         }
     }
 
+    // PaperSpigot start - Optimize getBlockData
     public IBlockData getBlockData(final BlockPosition blockposition) {
+        if (blockposition.getY() >= 0 && blockposition.getY() >> 4 < this.sections.length) {
+            ChunkSection chunksection = this.sections[blockposition.getY() >> 4];
+            if (chunksection != null) {
+                return chunksection.getType(blockposition.getX() & 15, blockposition.getY() & 15, blockposition.getZ() & 15);
+            }
+        }
+        return Blocks.AIR.getBlockData();
+    }
+    public IBlockData getBlockDataSlow(final BlockPosition blockposition) {
+        // PaperSpigot end
         if (this.world.G() == WorldType.DEBUG_ALL_BLOCK_STATES) {
             IBlockData iblockdata = null;
 
-- 
2.7.1

