From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Isaac Moore <rmsy@me.com>
Date: Tue, 19 Apr 2016 14:09:31 -0500
Subject: [PATCH] Implement PlayerLocaleChangeEvent


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 46754524a7a4a3087f56ab5460e3dac665c06052..d82643dd24fbd4b35dc526292d6f66234617857f 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1561,16 +1561,17 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         return s;
     }
 
-    public String locale = "en_us"; // CraftBukkit - add, lowercase
+    public String locale = null; // CraftBukkit - lowercase // Paper - default to null
     public void a(PacketPlayInSettings packetplayinsettings) {
         // CraftBukkit start
         if (getMainHand() != packetplayinsettings.getMainHand()) {
             PlayerChangedMainHandEvent event = new PlayerChangedMainHandEvent(getBukkitEntity(), getMainHand() == EnumMainHand.LEFT ? MainHand.LEFT : MainHand.RIGHT);
             this.server.server.getPluginManager().callEvent(event);
         }
-        if (!this.locale.equals(packetplayinsettings.locale)) {
+        if (this.locale == null || !this.locale.equals(packetplayinsettings.locale)) { // Paper - check for null
             PlayerLocaleChangeEvent event = new PlayerLocaleChangeEvent(getBukkitEntity(), packetplayinsettings.locale);
             this.server.server.getPluginManager().callEvent(event);
+            new com.destroystokyo.paper.event.player.PlayerLocaleChangeEvent(this.getBukkitEntity(), this.locale, packetplayinsettings.locale).callEvent(); // Paper
         }
         this.locale = packetplayinsettings.locale;
         this.clientViewDistance = packetplayinsettings.viewDistance;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a4c4c08ad8fdd03de0b9ac29ff736dd87e6221c6..f4910df73d64d1fc36598982bf20de3607ffd0ab 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1816,8 +1816,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public String getLocale() {
-        return getHandle().locale;
-
+        // Paper start - Locale change event
+        final String locale = getHandle().locale;
+        return locale != null ? locale : "en_us";
+        // Paper end
     }
 
     // Paper start
