From e0c21d538164861f973f9ac5aa0d3b5f32f29ad7 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 3 Apr 2016 17:48:50 -0400
Subject: [PATCH] Fix Cancelling BlockPlaceEvent triggering physics


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 2c6774082..5b2d67260 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -540,6 +540,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
     }
 
     public void applyPhysics(BlockPosition blockposition, Block block) {
+        if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
         this.a(blockposition.west(), block, blockposition);
         this.a(blockposition.east(), block, blockposition);
         this.a(blockposition.down(), block, blockposition);
-- 
2.18.0

