From 36f8bbc61f94489ba77a4a775807e75b716fe735 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 16 Dec 2016 16:03:19 -0600
Subject: [PATCH] Don't let fishinghooks use portals


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index d017cc8e6..7700652e1 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -156,7 +156,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     public boolean af;
     public boolean impulse;
     public int portalCooldown;
-    protected boolean ai;
+    protected boolean ai; public final boolean inPortal() { return this.ai; } // Paper - OBFHELPER
     protected int aj;
     public DimensionManager dimension;
     protected BlockPosition al;
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index 9cbcce446..3b10043ba 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -163,6 +163,12 @@ public class EntityFishingHook extends Entity {
 
             this.setMot(this.getMot().a(0.92D));
             this.setPosition(this.locX, this.locY, this.locZ);
+
+            // Paper start - These shouldn't be going through portals
+            if (this.inPortal()) {
+                this.die();
+            }
+            // Paper end
         }
     }
 
-- 
2.21.0

