From 9ea05aab2d38b0135649160310d6b46a1133d409 Mon Sep 17 00:00:00 2001
From: Brokkonaut <hannos17@gmx.de>
Date: Sun, 4 Sep 2016 16:35:43 -0500
Subject: [PATCH] Fix AIOOBE in inventory handling


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 078230521f..def0db5fea 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2096,7 +2096,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     case CLONE:
                         if (packetplayinwindowclick.d() == 2) {
                             click = ClickType.MIDDLE;
-                            if (packetplayinwindowclick.c() == -999) {
+                            if (packetplayinwindowclick.c() < 0) { // Paper - GH-404
                                 action = InventoryAction.NOTHING;
                             } else {
                                 Slot slot = this.player.activeContainer.getSlot(packetplayinwindowclick.c());
-- 
2.18.0

