From 77f4d31973ad9ff87d36f5d422d824eee7b38393 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 10 Apr 2016 22:36:10 -0400
Subject: [PATCH] SPIGOT-1903: Update PlayerConnection on world change


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index f0c04e5..d267a44 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -205,7 +205,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
 
     }
 
-    private void d() {
+    public void captureCurrentPosition() { d(); } private void d() { // Paper // OBFHELPER
         this.l = this.player.locX;
         this.m = this.player.locY;
         this.n = this.player.locZ;
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 94f122d..8842eb1 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -653,6 +653,7 @@ public abstract class PlayerList {
         }
         WorldServer worldserver = ((CraftWorld) location.getWorld()).getHandle();
         entityplayer1.setLocation(location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch());
+        entityplayer1.playerConnection.captureCurrentPosition(); // Paper
         // CraftBukkit end
 
         worldserver.getChunkProviderServer().getChunkAt((int) entityplayer1.locX >> 4, (int) entityplayer1.locZ >> 4);
-- 
2.8.1

