From 2d1892fb8670508fe15fc346c2cb749ab4feefb4 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Searchndstroy@users.noreply.github.com>
Date: Sun, 24 Apr 2016 01:37:40 -0700
Subject: [PATCH] Implement Add proper Entity/Player/OfflinePlayer equals and
 hashCode specifications to Paper API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 5934be1..d7ad0c9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -389,6 +389,8 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean equals(Object obj) {
+        // Paper start - comply with Entity.equals
+        /*
         if (obj == null) {
             return false;
         }
@@ -397,6 +399,14 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         }
         final CraftEntity other = (CraftEntity) obj;
         return (this.getEntityId() == other.getEntityId());
+        */
+        if (obj instanceof org.bukkit.entity.Entity) {
+            final org.bukkit.entity.Entity entity = (org.bukkit.entity.Entity) obj;
+            return entity.getEntityId() == getEntityId()
+                    && entity.getUniqueId().equals(getUniqueId());
+        }
+        return false;
+        // Paper end
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 003f700..a3f1ccb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -296,8 +296,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         boolean uuidEquals = this.getUniqueId().equals(other.getUniqueId());
         boolean idEquals = true;
 
-        if (other instanceof CraftPlayer) {
-            idEquals = this.getEntityId() == ((CraftPlayer) other).getEntityId();
+        if (other instanceof Player) { // Paper - change CraftPlayer to Player
+            idEquals = this.getEntityId() == ((Player) other).getEntityId(); // Paper - change CraftPlayer to Player
         }
 
         return uuidEquals && idEquals;
-- 
2.5.0

