From db3336329ce3d39657824be9a6af3f84b4123532 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 31 Jul 2016 16:33:03 -0500
Subject: [PATCH] Re-track players that dismount from other players


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index ad1bb7bab..4e8efc549 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -654,6 +654,13 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         if (entity1 != entity && this.playerConnection != null) {
             this.playerConnection.a(this.locX, this.locY, this.locZ, this.yaw, this.pitch);
         }
+        // Paper start - "Fixes" an issue in which the vehicle player would not be notified that the passenger dismounted
+        if (entity instanceof EntityPlayer) {
+            WorldServer worldServer = (WorldServer) entity.getWorld();
+            worldServer.tracker.untrackEntity(this);
+            worldServer.tracker.track(this);
+        }
+        // Paper end
 
     }
 
-- 
2.12.2.windows.2

