From 0e52dbcd333bbcf4cb0245b8b9faa64584c984b9 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 16 Jun 2016 00:17:23 -0400
Subject: [PATCH] Remove FishingHook reference on Craft Entity removal


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
index ecfc316bb..3f909c174 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
@@ -64,6 +64,16 @@ public class CraftFish extends AbstractProjectile implements Fish {
         this.biteChance = chance;
     }
 
+    // Paper start
+    @Override
+    public void remove() {
+        super.remove();
+        if (getHandle().owner != null) {
+            getHandle().owner.hookedFish = null;
+        }
+    }
+    // Paper end
+
     @Deprecated
     public LivingEntity _INVALID_getShooter() {
         return (LivingEntity) getShooter();
-- 
2.14.2

