From 6a546b2872c9a2453e19103de1ed5486393914d8 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 11 Jun 2017 21:01:18 +0100
Subject: [PATCH] provide a configurable option to disable creeper lingering
 effect spawns


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 54d67994..4d0d4cbe 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -327,4 +327,10 @@ public class PaperWorldConfig {
         parrotsHangOnBetter = getBoolean("parrots-are-unaffected-by-player-movement", false);
         log("Parrots are unaffected by player movement: " + parrotsHangOnBetter);
     }
+
+    public boolean disableCreeperLingeringEffect;
+    private void setDisableCreeperLingeringEffect() {
+        disableCreeperLingeringEffect = getBoolean("disable-creeper-lingering-effect", false);
+        log("Creeper lingering effect: " + disableCreeperLingeringEffect);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityCreeper.java b/src/main/java/net/minecraft/server/EntityCreeper.java
index e1374d23..0503af5a 100644
--- a/src/main/java/net/minecraft/server/EntityCreeper.java
+++ b/src/main/java/net/minecraft/server/EntityCreeper.java
@@ -225,7 +225,7 @@ public class EntityCreeper extends EntityMonster {
     private void createEffectCloud() {
         Collection<MobEffect> collection = this.getEffects();
 
-        if (!collection.isEmpty()) {
+        if (!collection.isEmpty() && !world.paperConfig.disableCreeperLingeringEffect) { // Paper
             EntityAreaEffectCloud entityareaeffectcloud = new EntityAreaEffectCloud(this.world, this.locX, this.locY, this.locZ);
 
             entityareaeffectcloud.setSource(this); // CraftBukkit
-- 
2.22.0

