From 7df8809967eac31f51ea02f50066971479c04742 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 11 Jun 2016 14:59:15 -0500
Subject: [PATCH] MC-80966 - Always send chunk sections


diff --git a/src/main/java/net/minecraft/server/ChunkSection.java b/src/main/java/net/minecraft/server/ChunkSection.java
index fc5c895..b2e81f0 100644
--- a/src/main/java/net/minecraft/server/ChunkSection.java
+++ b/src/main/java/net/minecraft/server/ChunkSection.java
@@ -64,7 +64,9 @@ public class ChunkSection {
     }
 
     public boolean a() {
-        return this.nonEmptyBlockCount == 0;
+        // Paper - MC-80966
+        // Even if there are no blocks, there may be other information associated with the chunk, always send it.
+        return false;
     }
 
     public boolean shouldTick() {
-- 
2.8.3.windows.1

