From 2863f1d882d22061cdb85376800881f0b9c992b8 Mon Sep 17 00:00:00 2001
From: Gabriele C <sgdc3.mail@gmail.com>
Date: Tue, 7 Jun 2016 21:17:22 +0200
Subject: [PATCH] Add disable-usercache-saving setting


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 058720f..cfee2da 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -228,4 +228,9 @@ public class PaperConfig {
     private static void saveEmptyScoreboardTeams() {
         saveEmptyScoreboardTeams = getBoolean("settings.save-empty-scoreboard-teams", false);
     }
+
+    public static boolean disableUserCacheSaving = false;
+    private static void disableCacheSaving() {
+        disableUserCacheSaving = getBoolean("settings.disable-usercache-saving", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/UserCache.java b/src/main/java/net/minecraft/server/UserCache.java
index 2420112..c29530f 100644
--- a/src/main/java/net/minecraft/server/UserCache.java
+++ b/src/main/java/net/minecraft/server/UserCache.java
@@ -193,8 +193,9 @@ public class UserCache {
     }
 
     public void b() {
-        BufferedReader bufferedreader = null;
+        if (com.destroystokyo.paper.PaperConfig.disableUserCacheSaving) return; // Paper
 
+        BufferedReader bufferedreader = null;
         try {
             bufferedreader = Files.newReader(this.h, Charsets.UTF_8);
             List list = (List) this.b.fromJson(bufferedreader, UserCache.i);
@@ -234,6 +235,8 @@ public class UserCache {
     }
     public void c(boolean asyncSave) {
         // Paper end
+        if (com.destroystokyo.paper.PaperConfig.disableUserCacheSaving) return; // Paper
+
         String s = this.b.toJson(this.a(org.spigotmc.SpigotConfig.userCacheCap));
         Runnable save = () -> {
 
-- 
2.5.2.windows.1

