From ddc177a9b7773ff2f78d79d5cc92d24344a146c9 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 26 Sep 2016 01:51:30 -0400
Subject: [PATCH] Disable Vanilla Chunk GC

Bukkit has its own system for this.

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 639f14d..eb05984 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1037,7 +1037,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             chunkproviderserver.a(flag);
             timings.worldSaveChunks.stopTiming(); // Paper
             // CraftBukkit - ArrayList -> Collection
-            Collection arraylist = chunkproviderserver.a();
+            /* //Paper start Collection arraylist = chunkproviderserver.a();
             Iterator iterator = arraylist.iterator();
 
             while (iterator.hasNext()) {
@@ -1046,7 +1046,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 if (chunk != null && !this.manager.a(chunk.locX, chunk.locZ)) {
                     chunkproviderserver.unload(chunk);
                 }
-            }
+            }*/
+            // Paper end
             timings.worldSave.stopTiming(); // Paper
         }
     }
-- 
2.10.2

