From 0de0dd7bad529706df1d5266a83f8fa7162eda1a Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 26 Sep 2016 01:51:30 -0400
Subject: [PATCH] Disable Vanilla Chunk GC

Bukkit has its own system for this.

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index ed3805d..a1350b0 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1028,7 +1028,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
             chunkproviderserver.a(flag);
             // CraftBukkit - ArrayList -> Collection
-            Collection arraylist = chunkproviderserver.a();
+            /* //Paper start Collection arraylist = chunkproviderserver.a();
             Iterator iterator = arraylist.iterator();
 
             while (iterator.hasNext()) {
@@ -1037,7 +1037,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 if (chunk != null && !this.manager.a(chunk.locX, chunk.locZ)) {
                     chunkproviderserver.unload(chunk);
                 }
-            }
+            }*/
+            // Paper end
 
         }
     }
-- 
2.10.0

