From 2ea5cf00fc211719d121bb700ab995996d490eee Mon Sep 17 00:00:00 2001
From: AlphaBlend <whizkid3000@hotmail.com>
Date: Mon, 21 Nov 2016 22:21:53 -0800
Subject: [PATCH] Move ProjectileHitEvent call to proper place


diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index 653b0fd..a594a24 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -265,8 +265,9 @@ public class EntityFishingHook extends Entity {
         }
 
         if (movingobjectposition != null && movingobjectposition.type != MovingObjectPosition.EnumMovingObjectType.MISS) {
-            org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition.entity); // Craftbukkit - Call event // Paper - Include entity in movingobjectposition
+            //org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition.entity); // Craftbukkit - Call event // Paper - Wrong location, moved down to proper location
             if (movingobjectposition.type == MovingObjectPosition.EnumMovingObjectType.ENTITY) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition.entity); // Craftbukkit - Call event // Paper - Include entity in movingobjectposition
                 this.hooked = movingobjectposition.entity;
                 this.s();
             } else {
-- 
2.10.2

