From d7a249718a819f8e00a9546693f5b74596feea59 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 16 Dec 2016 16:03:19 -0600
Subject: [PATCH] Don't let fishinghooks use portals


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index a3685ec..6656358 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -143,7 +143,7 @@ public abstract class Entity implements ICommandListener {
     public boolean ah;
     public boolean impulse;
     public int portalCooldown;
-    protected boolean ak;
+    protected boolean ak; public boolean inPortal() { return ak; } // Paper - OBFHELPER
     protected int al;
     public int dimension;
     protected BlockPosition an;
diff --git a/src/main/java/net/minecraft/server/EntityFishingHook.java b/src/main/java/net/minecraft/server/EntityFishingHook.java
index 0c528c6..606160a 100644
--- a/src/main/java/net/minecraft/server/EntityFishingHook.java
+++ b/src/main/java/net/minecraft/server/EntityFishingHook.java
@@ -185,6 +185,12 @@ public class EntityFishingHook extends Entity {
             this.motY *= 0.92D;
             this.motZ *= 0.92D;
             this.setPosition(this.locX, this.locY, this.locZ);
+
+            // Paper start - These shouldn't be going through portals
+            if (this.inPortal()) {
+                this.die();
+            }
+            // Paper end
         }
     }
 
-- 
2.9.3

