From 4eb6bc6c01b70d2d77fa96f5e57f5b5646a30fcc Mon Sep 17 00:00:00 2001
From: Alfie Cleveland <alfeh@me.com>
Date: Sun, 27 Nov 2016 22:17:37 +0000
Subject: [PATCH] Don't process packets from offline players


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index fd08195..eb68807 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -12,6 +12,7 @@ import javax.annotation.Nullable;
 // CraftBukkit start
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.entity.CraftItem;
+import org.bukkit.craftbukkit.entity.CraftPlayer; // Paper
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.player.PlayerBedEnterEvent;
@@ -292,7 +293,7 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     protected boolean isFrozen() {
-        return this.getHealth() <= 0.0F || this.isSleeping();
+        return this.getHealth() <= 0.0F || this.isSleeping() || (this.bukkitEntity != null && this.bukkitEntity instanceof CraftPlayer && !((CraftPlayer) this.bukkitEntity).isOnline()); // Paper - Pull #517
     }
 
     public void closeInventory() {
-- 
2.10.2

