From df7bf328a6c36551758f3a52f2ec36c3a380afb1 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 7 May 2017 06:26:09 -0500
Subject: [PATCH] PlayerPickupItemEvent#setFlyAtPlayer


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 3e70a06397..d9d52a63d8 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -308,6 +308,7 @@ public class EntityItem extends Entity {
             // CraftBukkit start - fire PlayerPickupItemEvent
             int canHold = entityhuman.inventory.canHold(itemstack);
             int remaining = i - canHold;
+            boolean flyAtPlayer = false; // Paper
 
             if (this.pickupDelay <= 0 && canHold > 0) {
                 itemstack.setCount(canHold);
@@ -315,7 +316,13 @@ public class EntityItem extends Entity {
                 PlayerPickupItemEvent playerEvent = new PlayerPickupItemEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
                 playerEvent.setCancelled(!entityhuman.canPickUpLoot);
                 this.world.getServer().getPluginManager().callEvent(playerEvent);
+                flyAtPlayer = playerEvent.getFlyAtPlayer(); // Paper
                 if (playerEvent.isCancelled()) {
+                    // Paper Start
+                    if (flyAtPlayer) {
+                        entityhuman.receive(this, i);
+                    }
+                    // Paper End
                     return;
                 }
 
@@ -336,6 +343,11 @@ public class EntityItem extends Entity {
 
             if (this.pickupDelay == 0 && (this.g == null || 6000 - this.age <= 200 || this.g.equals(entityhuman.getUniqueID())) && entityhuman.inventory.pickup(itemstack)) {
                 entityhuman.receive(this, i);
+                // Paper Start
+                if (flyAtPlayer) {
+                    entityhuman.receive(this, i);
+                }
+                // Paper End
                 if (itemstack.isEmpty()) {
                     this.die();
                     itemstack.setCount(i);
-- 
2.19.0

