From 2fb6c517d2ebeb3622759755b53c1e29712c4c93 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 00:56:51 -0500
Subject: [PATCH] Change Outdated Build Download URL to Paper CI


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 507390c4b..dab983481 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -200,7 +200,7 @@ public class Main {
                     deadline.add(Calendar.DAY_OF_YEAR, -3);
                     if (buildDate.before(deadline.getTime())) {
                         System.err.println("*** Error, this build is outdated ***");
-                        System.err.println("*** Please download a new build as per instructions from https://www.spigotmc.org/ ***");
+                        System.err.println("*** Please download a new build as per instructions from https://paperci.emc.gs/ ***"); // Paper
                         System.err.println("*** Server will start in 15 seconds ***");
                         Thread.sleep(TimeUnit.SECONDS.toMillis(15));
                     }
-- 
2.13.0

