From 0e4da6acee7d0008d196ac64e1b03b95de26e4c2 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 14 Jan 2017 01:22:07 -0600
Subject: [PATCH] Assign the World in WorldGenStronghold


diff --git a/src/main/java/net/minecraft/server/WorldGenStronghold.java b/src/main/java/net/minecraft/server/WorldGenStronghold.java
index c93754704..a3b958e01 100644
--- a/src/main/java/net/minecraft/server/WorldGenStronghold.java
+++ b/src/main/java/net/minecraft/server/WorldGenStronghold.java
@@ -56,6 +56,7 @@ public class WorldGenStronghold extends StructureGenerator {
     }
 
     public BlockPosition getNearestGeneratedFeature(World world, BlockPosition blockposition, boolean flag) {
+        this.g = world; // Paper
         if (!this.b) {
             this.c();
             this.b = true;
-- 
2.13.0.windows.1

