From 814e728cc82bfcac24ab95113048459a2c8840f3 Mon Sep 17 00:00:00 2001
From: willies952002 <admin@domnian.com>
Date: Sat, 29 Jul 2017 12:13:51 -0400
Subject: [PATCH] Check Verison when using CLI Version Flag


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d3d848f8c..c3edb95a2 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -15,6 +15,9 @@ import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import net.minecraft.server.MinecraftServer;
 import net.minecrell.terminalconsole.TerminalConsoleAppender; // Paper
+import org.bukkit.Bukkit;
+import org.bukkit.command.defaults.VersionCommand;
+import org.bukkit.craftbukkit.util.Versioning;
 
 public class Main {
     public static boolean useJline = true;
@@ -160,7 +163,28 @@ public class Main {
                 Logger.getLogger(Main.class.getName()).log(Level.SEVERE, null, ex);
             }
         } else if (options.has("v")) {
-            System.out.println(CraftServer.class.getPackage().getImplementationVersion());
+            // Paper start - Check version when doing --version
+            String version = CraftServer.class.getPackage().getImplementationVersion();
+            if (version == null) version = "Custom";
+            System.out.println("Current Version: " + version);
+            if (version.startsWith("git-Paper-")) {
+                String[] parts = version.substring("git-Paper-".length()).split("[-\\s]");
+                int distance = VersionCommand.getDistance(null, parts[0]);
+                switch (distance) {
+                    case -1:
+                        System.out.println("Error obtaining version information");
+                        break;
+                    case 0:
+                        System.out.println("You are running the latest version");
+                        break;
+                    case -2:
+                        System.out.println("Unknown version");
+                        break;
+                    default:
+                        System.out.println("You are " + distance + " version(s) behind");
+                }
+            }
+            // Paper end
         } else {
             // Do you love Java using + and ! as string based identifiers? I sure do!
             String path = new File(".").getAbsolutePath();
-- 
2.11.0

