From 743cdc806b5d81634c4da8a7cda5b906487dc3db Mon Sep 17 00:00:00 2001
From: June <junehanabi@gmail.com>
Date: Sat, 21 Oct 2017 20:36:14 -0500
Subject: [PATCH] Allow biome replacement in config file


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index f5cb9799b..0b1bd882d 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -23,6 +23,7 @@ import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
 import co.aikar.timings.Timings;
 import co.aikar.timings.TimingsManager;
+import org.bukkit.configuration.ConfigurationSection;
 
 public class PaperConfig {
 
@@ -276,4 +277,13 @@ public class PaperConfig {
     private static void authenticationServersDownKickMessage() {
         authenticationServersDownKickMessage = Strings.emptyToNull(getString("messages.kick.authentication-servers-down", authenticationServersDownKickMessage));
     }
+    
+    public static ConfigurationSection biomeReplacements;
+
+    private static void biomeReplacements() {
+        // Set a default for the users to know about it and its structure
+        getString("biome-replacements.ocean", "ocean");
+
+        biomeReplacements = config.getConfigurationSection("biome-replacements");
+    }
 }
diff --git a/src/main/java/net/minecraft/server/Biomes.java b/src/main/java/net/minecraft/server/Biomes.java
index 4165febc5..1faa8ae45 100644
--- a/src/main/java/net/minecraft/server/Biomes.java
+++ b/src/main/java/net/minecraft/server/Biomes.java
@@ -1,72 +1,85 @@
 package net.minecraft.server;
 
+// Paper start
+import com.destroystokyo.paper.PaperConfig;
+// Paper end
+
 public abstract class Biomes {
 
-    public static final BiomeBase a;
-    public static final BiomeBase b;
-    public static final BiomeBase c;
-    public static final BiomeBase d;
-    public static final BiomeBase e;
-    public static final BiomeBase f;
-    public static final BiomeBase g;
-    public static final BiomeBase h;
-    public static final BiomeBase i;
-    public static final BiomeBase j;
-    public static final BiomeBase k;
-    public static final BiomeBase l;
-    public static final BiomeBase m;
-    public static final BiomeBase n;
-    public static final BiomeBase o;
-    public static final BiomeBase p;
-    public static final BiomeBase q;
-    public static final BiomeBase r;
-    public static final BiomeBase s;
-    public static final BiomeBase t;
-    public static final BiomeBase u;
-    public static final BiomeBase v;
-    public static final BiomeBase w;
-    public static final BiomeBase x;
-    public static final BiomeBase y;
-    public static final BiomeBase z;
-    public static final BiomeBase A;
-    public static final BiomeBase B;
-    public static final BiomeBase C;
-    public static final BiomeBase D;
-    public static final BiomeBase E;
-    public static final BiomeBase F;
-    public static final BiomeBase G;
-    public static final BiomeBase H;
-    public static final BiomeBase I;
-    public static final BiomeBase J;
-    public static final BiomeBase K;
-    public static final BiomeBase L;
-    public static final BiomeBase M;
-    public static final BiomeBase N;
-    public static final BiomeBase O;
-    public static final BiomeBase P;
-    public static final BiomeBase Q;
-    public static final BiomeBase R;
-    public static final BiomeBase S;
-    public static final BiomeBase T;
-    public static final BiomeBase U;
-    public static final BiomeBase V;
-    public static final BiomeBase W;
-    public static final BiomeBase X;
-    public static final BiomeBase Y;
-    public static final BiomeBase Z;
-    public static final BiomeBase aa;
-    public static final BiomeBase ab;
-    public static final BiomeBase ac;
-    public static final BiomeBase ad;
-    public static final BiomeBase ae;
-    public static final BiomeBase af;
-    public static final BiomeBase ag;
-    public static final BiomeBase ah;
-    public static final BiomeBase ai;
-    public static final BiomeBase aj;
-    public static final BiomeBase ak;
+    // Paper start - Remove final on biome members
+    public static BiomeBase a;
+    public static BiomeBase b;
+    public static BiomeBase c;
+    public static BiomeBase d;
+    public static BiomeBase e;
+    public static BiomeBase f;
+    public static BiomeBase g;
+    public static BiomeBase h;
+    public static BiomeBase i;
+    public static BiomeBase j;
+    public static BiomeBase k;
+    public static BiomeBase l;
+    public static BiomeBase m;
+    public static BiomeBase n;
+    public static BiomeBase o;
+    public static BiomeBase p;
+    public static BiomeBase q;
+    public static BiomeBase r;
+    public static BiomeBase s;
+    public static BiomeBase t;
+    public static BiomeBase u;
+    public static BiomeBase v;
+    public static BiomeBase w;
+    public static BiomeBase x;
+    public static BiomeBase y;
+    public static BiomeBase z;
+    public static BiomeBase A;
+    public static BiomeBase B;
+    public static BiomeBase C;
+    public static BiomeBase D;
+    public static BiomeBase E;
+    public static BiomeBase F;
+    public static BiomeBase G;
+    public static BiomeBase H;
+    public static BiomeBase I;
+    public static BiomeBase J;
+    public static BiomeBase K;
+    public static BiomeBase L;
+    public static BiomeBase M;
+    public static BiomeBase N;
+    public static BiomeBase O;
+    public static BiomeBase P;
+    public static BiomeBase Q;
+    public static BiomeBase R;
+    public static BiomeBase S;
+    public static BiomeBase T;
+    public static BiomeBase U;
+    public static BiomeBase V;
+    public static BiomeBase W;
+    public static BiomeBase X;
+    public static BiomeBase Y;
+    public static BiomeBase Z;
+    public static BiomeBase aa;
+    public static BiomeBase ab;
+    public static BiomeBase ac;
+    public static BiomeBase ad;
+    public static BiomeBase ae;
+    public static BiomeBase af;
+    public static BiomeBase ag;
+    public static BiomeBase ah;
+    public static BiomeBase ai;
+    public static BiomeBase aj;
+    public static BiomeBase ak;
+    // Paper end
 
     private static BiomeBase a(String s) {
+        
+        // Paper start
+        if (PaperConfig.biomeReplacements != null) {
+            s = PaperConfig.biomeReplacements.getString(s, s).trim().toLowerCase();
+        }
+        // Paper end
+
         BiomeBase biomebase = (BiomeBase) BiomeBase.REGISTRY_ID.get(new MinecraftKey(s));
 
         if (biomebase == null) {
@@ -75,8 +88,9 @@ public abstract class Biomes {
             return biomebase;
         }
     }
-
-    static {
+    
+    // Paper start
+    public static void SetupBiomes() {
         if (!DispenserRegistry.a()) {
             throw new RuntimeException("Accessed Biomes before Bootstrap!");
         } else {
@@ -145,4 +159,9 @@ public abstract class Biomes {
             ak = a("mutated_mesa_clear_rock");
         }
     }
+    // Paper end
+
+    static {
+        SetupBiomes(); // Paper
+    }
 }
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 854455711..3492204e9 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -203,6 +203,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
             // Paper start
             com.destroystokyo.paper.PaperConfig.init((File) options.valueOf("paper-settings"));
             com.destroystokyo.paper.PaperConfig.registerCommands();
+            net.minecraft.server.Biomes.SetupBiomes(); // Ensure these are setup now that PaperConfig has loaded
             // Paper end
 
             DedicatedServer.LOGGER.info("Generating keypair");
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b42f4e9ee..3eb227fbb 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -751,6 +751,7 @@ public final class CraftServer implements Server {
             world.spigotConfig.init(); // Spigot
             world.paperConfig.init(); // Paper
         }
+        net.minecraft.server.Biomes.SetupBiomes(); // Paper - Ensure these are setup now that PaperConfig has loaded
 
         Plugin[] pluginClone = pluginManager.getPlugins().clone(); // Paper
         pluginManager.clearPlugins();
-- 
2.14.2

