From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 3 Aug 2018 00:04:54 -0400
Subject: [PATCH] MC-135506: Experience should save as Integers


diff --git a/src/main/java/net/minecraft/server/EntityExperienceOrb.java b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
index 4df7df781d12e38885ecfbc89dd1d42d9de14248..c2be0c2bc315876f120cff207e5516dda2bd55d7 100644
--- a/src/main/java/net/minecraft/server/EntityExperienceOrb.java
+++ b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
@@ -202,7 +202,7 @@ public class EntityExperienceOrb extends Entity {
     public void saveData(NBTTagCompound nbttagcompound) {
         nbttagcompound.setShort("Health", (short) this.e);
         nbttagcompound.setShort("Age", (short) this.c);
-        nbttagcompound.setShort("Value", (short) this.value);
+        nbttagcompound.setInt("Value", this.value); // Paper - save as Integer
         this.savePaperNBT(nbttagcompound); // Paper
     }
 
@@ -210,7 +210,7 @@ public class EntityExperienceOrb extends Entity {
     public void loadData(NBTTagCompound nbttagcompound) {
         this.e = nbttagcompound.getShort("Health");
         this.c = nbttagcompound.getShort("Age");
-        this.value = nbttagcompound.getShort("Value");
+        this.value = nbttagcompound.getInt("Value"); // Paper - load as Integer
         this.loadPaperNBT(nbttagcompound); // Paper
     }
 
