From 050b9d851f88b5d429c2af21566ad1c5a00c42de Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 23 Jul 2018 22:18:31 -0400
Subject: [PATCH] Mark chunk dirty anytime entities change to guarantee it
 saves


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 7382a47d8c..8a8c8f4e88 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -411,6 +411,7 @@ public class Chunk implements IChunkAccess {
         entity.chunkY = k;
         entity.chunkZ = this.loc.z;
         this.entitySlices[k].add(entity);
+        this.markDirty(); // Paper
     }
 
     @Override
@@ -437,6 +438,7 @@ public class Chunk implements IChunkAccess {
             return;
         }
         entityCounts.decrement(entity.getMinecraftKeyString());
+        this.markDirty(); // Paper
         // Paper end
     }
 
-- 
2.21.0

