From ef57c5ff0c40fb3c195d1fb50ea6a887a6eeb81c Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 10 Mar 2018 13:03:49 +0000
Subject: [PATCH] Fix NPE when getting location from players EnderChest


diff --git a/src/main/java/net/minecraft/server/InventoryEnderChest.java b/src/main/java/net/minecraft/server/InventoryEnderChest.java
index cd7de2b53..85ad93d08 100644
--- a/src/main/java/net/minecraft/server/InventoryEnderChest.java
+++ b/src/main/java/net/minecraft/server/InventoryEnderChest.java
@@ -5,7 +5,7 @@ import org.bukkit.inventory.InventoryHolder;
 
 public class InventoryEnderChest extends InventorySubcontainer {
 
-    private TileEntityEnderChest a;
+    private TileEntityEnderChest a; public TileEntityEnderChest getTileEntity() { return a; } // Paper - OBFHELPER
     // CraftBukkit start
     private final EntityHuman owner;
 
@@ -15,6 +15,7 @@ public class InventoryEnderChest extends InventorySubcontainer {
 
     @Override
     public Location getLocation() {
+        if (getTileEntity() == null) return owner.getBukkitEntity().getLocation(); // Paper - return Player location if there is no TE
         return new Location(this.a.getWorld().getWorld(), this.a.getPosition().getX(), this.a.getPosition().getY(), this.a.getPosition().getZ());
     }
 
-- 
2.16.2

