From f0db6278e491b4a845bf8199c36b1dc5ce1bf0aa Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 31 Mar 2018 22:49:05 -0400
Subject: [PATCH] Easter and April Fools fall on the same day
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

¯\_(ツ)_/¯

diff --git a/src/main/java/net/minecraft/server/EntityEgg.java b/src/main/java/net/minecraft/server/EntityEgg.java
index 08131afc..82c37b87 100644
--- a/src/main/java/net/minecraft/server/EntityEgg.java
+++ b/src/main/java/net/minecraft/server/EntityEgg.java
@@ -45,6 +45,13 @@ public class EntityEgg extends EntityProjectile {
                 }
                 EntityType hatchingType = EntityType.CHICKEN;
 
+                // Paper start - 2018-04-01 is easter and April fools day... come on now how can we not
+                java.time.LocalDate today = java.time.LocalDate.now();
+                if (today.getDayOfMonth() == 1 && today.getMonth() == java.time.Month.APRIL && today.getYear() == 2018) {
+                    hatchingType = EntityType.RABBIT;
+                }
+                // Paper end
+
                 Entity shooter = this.getShooter();
                 if (shooter instanceof EntityPlayer) {
                     PlayerEggThrowEvent event = new PlayerEggThrowEvent((Player) shooter.getBukkitEntity(), (org.bukkit.entity.Egg) this.getBukkitEntity(), hatching, b0, hatchingType);
-- 
2.16.2

