From 3fd22e06ffb4a1530ad73701c571c190b7d2dc7b Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 31 Mar 2018 22:49:05 -0400
Subject: [PATCH] Easter and April Fools fall on the same day
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

¯\_(ツ)_/¯

diff --git a/src/main/java/net/minecraft/server/EntityEgg.java b/src/main/java/net/minecraft/server/EntityEgg.java
index 08131afc..7366c4b4 100644
--- a/src/main/java/net/minecraft/server/EntityEgg.java
+++ b/src/main/java/net/minecraft/server/EntityEgg.java
@@ -45,6 +45,13 @@ public class EntityEgg extends EntityProjectile {
                 }
                 EntityType hatchingType = EntityType.CHICKEN;
 
+                // Paper start - 2018-04-01 is easter and April fools day... come on now how can we not
+                java.util.Date rightNow = java.util.Calendar.getInstance().getTime();
+                if (rightNow.getDate() == 1 && rightNow.getMonth() == 3 && rightNow.getYear() == 118) {
+                    hatchingType = EntityType.RABBIT;
+                }
+                // Paper end
+
                 Entity shooter = this.getShooter();
                 if (shooter instanceof EntityPlayer) {
                     PlayerEggThrowEvent event = new PlayerEggThrowEvent((Player) shooter.getBukkitEntity(), (org.bukkit.entity.Egg) this.getBukkitEntity(), hatching, b0, hatchingType);
-- 
2.16.2

