From cbcb5513c2dd25b08035877ac014282b934d84f0 Mon Sep 17 00:00:00 2001
From: Hugo Manrique <hugmanrique@gmail.com>
Date: Mon, 16 Jul 2018 12:42:20 +0200
Subject: [PATCH] Avoid item merge if stack size above max stack size


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 85f80741c..6752d0b8c 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -173,6 +173,10 @@ public class EntityItem extends Entity {
     }
 
     private void v() {
+        // Paper start - avoid item merge if stack size above max stack size
+        ItemStack stack = getItemStack();
+        if (stack.getCount() >= stack.getMaxStackSize()) return;
+        // Paper end
         // Spigot start
         double radius = world.spigotConfig.itemMerge;
         Iterator iterator = this.world.a(EntityItem.class, this.getBoundingBox().grow(radius, radius, radius)).iterator();
-- 
2.21.0

