From 807b6b37d1576b5cf82fb5c8a9423c69647fca98 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 21 Jul 2018 01:51:27 -0500
Subject: [PATCH] EnderDragon Events


diff --git a/src/main/java/net/minecraft/server/DragonControllerLandedFlame.java b/src/main/java/net/minecraft/server/DragonControllerLandedFlame.java
index 1906bcfc96..50d291139c 100644
--- a/src/main/java/net/minecraft/server/DragonControllerLandedFlame.java
+++ b/src/main/java/net/minecraft/server/DragonControllerLandedFlame.java
@@ -63,7 +63,9 @@ public class DragonControllerLandedFlame extends AbstractDragonControllerLanded
             this.d.setDuration(200);
             this.d.setParticle(Particles.j);
             this.d.a(new MobEffect(MobEffects.HARM));
+            if (new com.destroystokyo.paper.event.entity.EnderDragonFlameEvent((org.bukkit.entity.EnderDragon) this.a.getBukkitEntity(), (org.bukkit.entity.AreaEffectCloud) this.d.getBukkitEntity()).callEvent()) // Paper
             this.a.world.addEntity(this.d);
+            else this.removeAreaEffect(); // Paper
         }
 
     }
@@ -73,6 +75,7 @@ public class DragonControllerLandedFlame extends AbstractDragonControllerLanded
         ++this.c;
     }
 
+    public void removeAreaEffect() { this.e(); } // Paper - OBFHELPER
     public void e() {
         if (this.d != null) {
             this.d.die();
diff --git a/src/main/java/net/minecraft/server/DragonControllerStrafe.java b/src/main/java/net/minecraft/server/DragonControllerStrafe.java
index e742586765..141ba1e5e6 100644
--- a/src/main/java/net/minecraft/server/DragonControllerStrafe.java
+++ b/src/main/java/net/minecraft/server/DragonControllerStrafe.java
@@ -67,7 +67,9 @@ public class DragonControllerStrafe extends AbstractDragonController {
                         EntityDragonFireball entitydragonfireball = new EntityDragonFireball(this.a.world, this.a, d9, d10, d11);
 
                         entitydragonfireball.setPositionRotation(d6, d7, d8, 0.0F, 0.0F);
+                        if (new com.destroystokyo.paper.event.entity.EnderDragonShootFireballEvent((org.bukkit.entity.EnderDragon) a.getBukkitEntity(), (org.bukkit.entity.DragonFireball) entitydragonfireball.getBukkitEntity()).callEvent()) // Paper
                         this.a.world.addEntity(entitydragonfireball);
+                        else entitydragonfireball.die(); // Paper
                         this.c = 0;
                         if (this.d != null) {
                             while (!this.d.b()) {
diff --git a/src/main/java/net/minecraft/server/EntityDragonFireball.java b/src/main/java/net/minecraft/server/EntityDragonFireball.java
index 7df8fd4d74..c0d66a0233 100644
--- a/src/main/java/net/minecraft/server/EntityDragonFireball.java
+++ b/src/main/java/net/minecraft/server/EntityDragonFireball.java
@@ -39,8 +39,10 @@ public class EntityDragonFireball extends EntityFireball {
                     }
                 }
 
+                if (new com.destroystokyo.paper.event.entity.EnderDragonFireballHitEvent((org.bukkit.entity.DragonFireball) this.getBukkitEntity(), list.stream().map(EntityLiving::getBukkitLivingEntity).collect(java.util.stream.Collectors.toList()), (org.bukkit.entity.AreaEffectCloud) entityareaeffectcloud.getBukkitEntity()).callEvent()) { // Paper
                 this.world.triggerEffect(2006, new BlockPosition(this.locX, this.locY, this.locZ), 0);
                 this.world.addEntity(entityareaeffectcloud);
+                } else entityareaeffectcloud.die(); // Paper
                 this.die();
             }
 
-- 
2.21.0

