From 9e35ddf0782e89383c33333b36d2640d7cf216ce Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 21 Jul 2018 01:59:59 -0500
Subject: [PATCH] PlayerElytraBoostEvent


diff --git a/src/main/java/net/minecraft/server/ItemFireworks.java b/src/main/java/net/minecraft/server/ItemFireworks.java
index d971bbc76c..4ce9c60895 100644
--- a/src/main/java/net/minecraft/server/ItemFireworks.java
+++ b/src/main/java/net/minecraft/server/ItemFireworks.java
@@ -28,9 +28,15 @@ public class ItemFireworks extends Item {
             if (!world.isClientSide) {
                 EntityFireworks entityfireworks = new EntityFireworks(world, itemstack, entityhuman);
                 entityfireworks.spawningEntity = entityhuman.getUniqueID(); // Paper
-                world.addEntity(entityfireworks);
-                if (!entityhuman.abilities.canInstantlyBuild) {
-                    itemstack.subtract(1);
+                // Paper start
+                com.destroystokyo.paper.event.player.PlayerElytraBoostEvent event = new com.destroystokyo.paper.event.player.PlayerElytraBoostEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Firework) entityfireworks.getBukkitEntity());
+                if (event.callEvent() && world.addEntity(entityfireworks)) {
+                    if (event.shouldConsume() && !entityhuman.abilities.canInstantlyBuild) {
+                        itemstack.subtract(1);
+                    } else ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
+                } else if (entityhuman instanceof EntityPlayer) {
+                    ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
+                    // Paper end
                 }
             }
 
-- 
2.19.0

