From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 6 Oct 2018 21:47:44 -0500
Subject: [PATCH] Allow setting the vex's summoner


diff --git a/src/main/java/net/minecraft/server/EntityVex.java b/src/main/java/net/minecraft/server/EntityVex.java
index a46730460e5d9f6e52d63ba1950d1e7195fe87cd..cf274666c70695ef597b2f5651c18c60702c2dc1 100644
--- a/src/main/java/net/minecraft/server/EntityVex.java
+++ b/src/main/java/net/minecraft/server/EntityVex.java
@@ -133,6 +133,7 @@ public class EntityVex extends EntityMonster {
         this.a(1, flag);
     }
 
+    public void setOwner(EntityInsentient entityinsentient) { a(entityinsentient); } // Paper - OBFHELPER
     public void a(EntityInsentient entityinsentient) {
         this.c = entityinsentient;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
index 169c951ec30fc5b78eb4ba29352ba009b7fa5efd..2f7df3074f8fd66fe2ea1cced08db910078e1848 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
@@ -23,6 +23,10 @@ public class CraftVex extends CraftMonster implements Vex {
         EntityInsentient owner = getHandle().getOwner();
         return owner != null ? (Mob) owner.getBukkitEntity() : null;
     }
+
+    public void setSummoner(Mob summoner) {
+        getHandle().setOwner(summoner == null ? null : ((CraftMob) summoner).getHandle());
+    }
     // Paper end
 
     @Override
