From 678c19922f25678469691fadc0cbe2a52203e27d Mon Sep 17 00:00:00 2001
From: Hugo Manrique <hugmanrique@gmail.com>
Date: Mon, 23 Jul 2018 19:18:37 +0200
Subject: [PATCH] MC-94186 Fix dragon egg falling in lazy chunks


diff --git a/src/main/java/net/minecraft/server/BlockDragonEgg.java b/src/main/java/net/minecraft/server/BlockDragonEgg.java
index bfee7ec2..ead8864f 100644
--- a/src/main/java/net/minecraft/server/BlockDragonEgg.java
+++ b/src/main/java/net/minecraft/server/BlockDragonEgg.java
@@ -53,7 +53,7 @@ public class BlockDragonEgg extends BlockFalling {
                         world.addParticle(Particles.K, d1, d2, d3, (double) f, (double) f1, (double) f2);
                     }
                 } else {
-                    world.setTypeAndData(blockposition1, iblockdata, 2);
+                    world.setTypeAndData(blockposition1.up(), iblockdata, 2); // Paper - fix dragon egg falling in lazy chunks
                     world.setAir(blockposition);
                 }
 
-- 
2.18.0.windows.1

