From e5409f10fa941e455b5b24dbf61405519dcc1ef1 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 15 Aug 2018 01:19:04 -0400
Subject: [PATCH] Pass World to Tile Entity Snapshots


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
index d4d9c5fc50..a72bbdbedb 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
@@ -50,7 +50,7 @@ public class CraftBlockEntityState<T extends TileEntity> extends CraftBlockState
         if (DISABLE_SNAPSHOT) {
             this.snapshot = this.tileEntity;
         } else {
-            this.snapshot = this.createSnapshot(this.tileEntity, null);
+            this.snapshot = this.createSnapshot(this.tileEntity, tileEntity.getWorld());
         }
         // copy tile entity data:
         if(this.snapshot != null) {
-- 
2.18.0

