From 11621f2b919b38c0c58fa036707ec17e313141dd Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 18 Aug 2018 12:43:16 -0400
Subject: [PATCH] Restore vanlla default mob-spawn-range


diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 06d84a0b66..15abde4f5d 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -130,7 +130,7 @@ public class SpigotWorldConfig
     public byte mobSpawnRange;
     private void mobSpawnRange()
     {
-        mobSpawnRange = (byte) getInt( "mob-spawn-range", 6 );
+        mobSpawnRange = (byte) getInt( "mob-spawn-range", 8 ); // Paper - Vanilla
         log( "Mob Spawn Range: " + mobSpawnRange );
     }
 
-- 
2.19.0

