From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach@zachbr.io>
Date: Mon, 4 Mar 2019 02:23:28 -0500
Subject: [PATCH] Set Zombie last tick at start of drowning process

Fixes GH-1887

diff --git a/src/main/java/net/minecraft/server/EntityZombie.java b/src/main/java/net/minecraft/server/EntityZombie.java
index 02a845f99bb445fb4aa66a90273469a848195747..45c470ebeca1552e0a849b607fde95de437e8bb4 100644
--- a/src/main/java/net/minecraft/server/EntityZombie.java
+++ b/src/main/java/net/minecraft/server/EntityZombie.java
@@ -162,6 +162,7 @@ public class EntityZombie extends EntityMonster {
                     ++this.bt;
                     if (this.bt >= 600) {
                         this.startDrownedConversion(300);
+                        this.lastTick = MinecraftServer.currentTick; // Paper - Make sure this is set at start of process - GH-1887
                     }
                 } else {
                     this.bt = -1;
