From 9537f5342de2b00331c786c63c228df41139f040 Mon Sep 17 00:00:00 2001
From: GreenMeanie <GreenMeanieMC@gmail.com>
Date: Sat, 20 Oct 2018 22:34:02 -0400
Subject: [PATCH] Reset players airTicks on respawn


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 57bc805cf..0871e185e 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2254,7 +2254,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
 
     }
 
-    public int bo() {
+    public int getMaxAirTicks() { return bo(); } public int bo() {
         return 300;
     }
 
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 8d18c51c9..c71e4dfbe 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1835,6 +1835,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         }
 
         this.setHealth(this.getMaxHealth());
+        this.setAirTicks(this.getMaxAirTicks()); // Paper
         this.fireTicks = 0;
         this.fallDistance = 0;
         this.foodData = new FoodMetaData(this);
-- 
2.21.0

