From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Mon, 27 May 2019 17:35:39 -0500
Subject: [PATCH] MC-114618 - Fix EntityAreaEffectCloud from going negative
 size


diff --git a/src/main/java/net/minecraft/server/EntityAreaEffectCloud.java b/src/main/java/net/minecraft/server/EntityAreaEffectCloud.java
index f04204b8722a061a09e2b3cad6c67ab5fdd8f071..5b0ee830b45a677bf354b02cf4fd311d842e7678 100644
--- a/src/main/java/net/minecraft/server/EntityAreaEffectCloud.java
+++ b/src/main/java/net/minecraft/server/EntityAreaEffectCloud.java
@@ -175,6 +175,12 @@ public class EntityAreaEffectCloud extends Entity {
         super.tick();
         boolean flag = this.k();
         float f = this.getRadius();
+        // Paper start - fix MC-114618
+        if (f < 0.0F) {
+            this.die();
+            return;
+        }
+        // Paper end
 
         if (this.world.isClientSide) {
             ParticleParam particleparam = this.getParticle();
