From 9f7b472195e42d864192775e950eb31029bf588b Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Tue, 4 Sep 2018 00:43:23 -0500
Subject: [PATCH] Add creative-arrow-despawn-rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 24f353d6f..6e7e64d17 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -222,6 +222,15 @@ public class PaperWorldConfig {
         log("Non Player Arrow Despawn Rate: " + nonPlayerArrowDespawnRate);
     }
 
+    public int creativeArrowDespawnRate = -1;
+    private void creativeArrowDespawnRate() {
+        creativeArrowDespawnRate = getInt("creative-arrow-despawn-rate", -1);
+        if (creativeArrowDespawnRate == -1) {
+            creativeArrowDespawnRate = spigotConfig.arrowDespawnRate;
+        }
+        log("Creative Arrow Despawn Rate: " + creativeArrowDespawnRate);
+    }
+
     public double skeleHorseSpawnChance;
     private void skeleHorseSpawnChance() {
         skeleHorseSpawnChance = getDouble("skeleton-horse-thunder-spawn-chance", 0.01D);
diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 187f02d4e..a82dc0a94 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -272,7 +272,7 @@ public abstract class EntityArrow extends Entity implements IProjectile {
 
     protected void f() {
         ++this.despawnCounter;
-        if (this.despawnCounter >= (fromPlayer != PickupStatus.DISALLOWED ? world.spigotConfig.arrowDespawnRate : world.paperConfig.nonPlayerArrowDespawnRate)) { // Spigot // Paper
+        if (this.despawnCounter >= (fromPlayer == PickupStatus.CREATIVE_ONLY ? world.paperConfig.creativeArrowDespawnRate : (fromPlayer == PickupStatus.DISALLOWED ? world.paperConfig.nonPlayerArrowDespawnRate : world.spigotConfig.arrowDespawnRate))) { // Spigot // Paper
             this.die();
         }
 
-- 
2.18.0

