From f5ffd1e9a4b8fbb7ca801c7c02f429a4906badb0 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Fri, 28 Sep 2018 12:27:54 -0500
Subject: [PATCH] Fix turtle lag


diff --git a/src/main/java/net/minecraft/server/EntityTurtle.java b/src/main/java/net/minecraft/server/EntityTurtle.java
index 1b09f577..4f5592d1 100644
--- a/src/main/java/net/minecraft/server/EntityTurtle.java
+++ b/src/main/java/net/minecraft/server/EntityTurtle.java
@@ -599,7 +599,7 @@ public class EntityTurtle extends EntityAnimal {
                 ++this.d;
             }
 
-            if (this.a.dA() || this.a.getNavigation().p()) {
+            if (/*this.a.dA() ||*/ this.a.getNavigation().p()) { // Paper - Fix GH-1501
                 Vec3D vec3d = RandomPositionGenerator.a((EntityCreature) this.a, 16, 3, new Vec3D((double) blockposition.getX(), (double) blockposition.getY(), (double) blockposition.getZ()), 0.3141592741012573D);
 
                 if (vec3d == null) {
@@ -656,7 +656,7 @@ public class EntityTurtle extends EntityAnimal {
         }
 
         public void e() {
-            if (this.a.dA() || this.a.getNavigation().p()) {
+            if (/*this.a.dA() ||*/ this.a.getNavigation().p()) { // Paper - Fix GH-1501
                 BlockPosition blockposition = this.a.dC();
                 Vec3D vec3d = RandomPositionGenerator.a((EntityCreature) this.a, 16, 3, new Vec3D((double) blockposition.getX(), (double) blockposition.getY(), (double) blockposition.getZ()), 0.3141592741012573D);
 
-- 
2.19.1

