From 56358236f603519affbb0d600cd4efc6147763af Mon Sep 17 00:00:00 2001
From: Caleb Bassham <caleb.bassham@gmail.com>
Date: Fri, 28 Sep 2018 02:32:19 -0500
Subject: [PATCH] Call player spectator target events


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 5ab4e01ed..3283ef215 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1373,15 +1373,34 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         return (Entity) (this.cv == null ? this : this.cv);
     }
 
-    public void setSpectatorTarget(Entity entity) {
+    public void setSpectatorTarget(Entity newSpectatorTarget) {
         Entity entity1 = this.getSpecatorTarget();
 
-        this.cv = (Entity) (entity == null ? this : entity);
-        if (entity1 != this.cv) {
+        if (newSpectatorTarget == null) {
+            this.cv = this;
+            newSpectatorTarget = this;
+        }
+
+        if (entity1 != newSpectatorTarget) {
+            if (newSpectatorTarget == this) {
+                com.destroystokyo.paper.event.player.PlayerStopSpectatingEntityEvent playerStopSpectatingEntityEvent = new com.destroystokyo.paper.event.player.PlayerStopSpectatingEntityEvent((org.bukkit.entity.Player) this, (org.bukkit.entity.Entity) entity1);
+
+                if (!playerStopSpectatingEntityEvent.callEvent()) {
+                    return;
+                }
+            } else {
+                com.destroystokyo.paper.event.player.PlayerStartSpectatingEntityEvent playerStartSpectatingEntityEvent = new com.destroystokyo.paper.event.player.PlayerStartSpectatingEntityEvent((org.bukkit.entity.Player) this, (org.bukkit.entity.Entity) entity1, (org.bukkit.entity.Entity) newSpectatorTarget);
+
+                if (!playerStartSpectatingEntityEvent.callEvent()) {
+                    return;
+                }
+            }
+
             this.playerConnection.sendPacket(new PacketPlayOutCamera(this.cv));
             this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
         }
 
+        this.cv = newSpectatorTarget;
     }
 
     protected void E() {
-- 
2.19.0

