From 891c9fdcaa582d956c5278a2b7fb58a0ae433dd2 Mon Sep 17 00:00:00 2001
From: Caleb Bassham <caleb.bassham@gmail.com>
Date: Wed, 17 Oct 2018 15:56:35 -0500
Subject: [PATCH] Add OBFHELPER for setting spectator target


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 3283ef215..972d1cf81 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -62,7 +62,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     private EntityHuman.EnumChatVisibility cs;
     private boolean ct = true;
     private long cu = SystemUtils.b();
-    private Entity cv;
+    private Entity cv; private void setSpectatorTargetField(Entity e) { this.cv = e; } // Paper - OBFHELPER
     public boolean worldChangeInvuln;
     private boolean cx; private void setHasSeenCredits(boolean has) { this.cx = has; } // Paper - OBFHELPER
     private final RecipeBookServer cy;
@@ -1377,7 +1377,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         Entity entity1 = this.getSpecatorTarget();
 
         if (newSpectatorTarget == null) {
-            this.cv = this;
+            setSpectatorTargetField(this);
             newSpectatorTarget = this;
         }
 
@@ -1400,7 +1400,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
         }
 
-        this.cv = newSpectatorTarget;
+        setSpectatorTargetField(newSpectatorTarget);
     }
 
     protected void E() {
-- 
2.19.1

