From 7ac07ac07ac07ac07ac07ac07ac07ac07ac07ac0 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Fri, 26 Oct 2018 02:39:29 -0500
Subject: [PATCH] Add more Evoker API


diff --git a/src/main/java/net/minecraft/server/EntityEvoker.java b/src/main/java/net/minecraft/server/EntityEvoker.java
index 7ac07ac07ac0..7ac07ac07ac0 100644
--- a/src/main/java/net/minecraft/server/EntityEvoker.java
+++ b/src/main/java/net/minecraft/server/EntityEvoker.java
@@ -89,12 +89,13 @@ public class EntityEvoker extends EntityIllagerWizard {
         return SoundEffects.ENTITY_EVOKER_HURT;
     }
 
+    public void setWololoTarget(@Nullable EntitySheep sheep) { a(sheep); } // Paper - OBFHELPER
     private void a(@Nullable EntitySheep entitysheep) {
         this.c = entitysheep;
     }
 
-    @Nullable
-    private EntitySheep dD() {
+    @Nullable public EntitySheep getWololoTarget() { return dD(); } // Paper - OBFHELPER
+    @Nullable private EntitySheep dD() {
         return this.c;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
index 7ac07ac07ac0..7ac07ac07ac0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
@@ -6,6 +6,12 @@ import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Evoker;
 
+// Paper start
+import net.minecraft.server.EntitySheep;
+import org.bukkit.entity.Sheep;
+import javax.annotation.Nullable;
+// Paper end
+
 public class CraftEvoker extends CraftSpellcaster implements Evoker {
 
     public CraftEvoker(CraftServer server, EntityEvoker entity) {
@@ -36,4 +42,17 @@ public class CraftEvoker extends CraftSpellcaster implements Evoker {
     public void setCurrentSpell(Evoker.Spell spell) {
         getHandle().setSpell(spell == null ? EntityIllagerWizard.Spell.NONE : EntityIllagerWizard.Spell.a(spell.ordinal()));
     }
+
+    // Paper start
+    @Nullable
+    public Sheep getWololoTarget() {
+        EntitySheep target = getHandle().getWololoTarget();
+        return target == null ? null : (Sheep) target.getBukkitEntity();
+    }
+
+    @Override
+    public void setWololoTarget(@Nullable Sheep sheep) {
+        getHandle().setWololoTarget(sheep == null ? null : (EntitySheep) ((CraftEntity) sheep).getHandle());
+    }
+    // Paper end
 }
-- 
2.19.1

