From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach@zachbr.io>
Date: Sat, 8 Feb 2020 18:02:24 -0600
Subject: [PATCH] Allow overriding the java version check

-DPaper.IgnoreJavaVersion=true

diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 463c41454d96e823d1ffd62619fc3b83af3bf168..c3e2d9c4d0ef3de80f5206c6bd234121ee1316cc 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -181,7 +181,7 @@ public class Main {
             float javaVersion = Float.parseFloat(System.getProperty("java.class.version"));
             if (javaVersion > 60.0) {
                 System.err.println("Unsupported Java detected (" + javaVersion + "). Only up to Java 16 is supported.");
-                return;
+                if (!Boolean.getBoolean("Paper.IgnoreJavaVersion")) return; // Paper
             }
 
             try {
