From 96e47b61a7b56b3593c1fd8058c970ee09efb490 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Wed, 26 Jun 2019 03:04:45 +0100
Subject: [PATCH] Fix Raids CME


diff --git a/src/main/java/net/minecraft/server/Raid.java b/src/main/java/net/minecraft/server/Raid.java
index 6c1bbefd71..72f76d6c02 100644
--- a/src/main/java/net/minecraft/server/Raid.java
+++ b/src/main/java/net/minecraft/server/Raid.java
@@ -137,14 +137,14 @@ public class Raid {
     private void y() {
         Collection<EntityPlayer> collection = this.r.getPlayers();
         Set<EntityPlayer> set = Sets.newHashSet(collection);
-        Iterator iterator = collection.iterator();
+        Iterator iterator = set.iterator(); // Paper - use the set, this doesn't get modified out of scope
 
         while (iterator.hasNext()) {
             EntityPlayer entityplayer = (EntityPlayer) iterator.next();
             BlockPosition blockposition = new BlockPosition(entityplayer);
 
             if (this.k.c_(blockposition) != this) {
-                set.remove(entityplayer);
+                iterator.remove(); // Paper
                 this.r.removePlayer(entityplayer);
             }
         }
-- 
2.22.0

