From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 3 Mar 2020 05:26:40 +0000
Subject: [PATCH] Prevent teleporting dead entities


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 1d0345f904de6061e5cfee3ab7c2dbfd94e3fa9e..93877b45f380e57f2dc9656463e333ef422a4b95 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1309,6 +1309,10 @@ public class PlayerConnection implements PacketListenerPlayIn {
     }
 
     private void internalTeleport(double d0, double d1, double d2, float f, float f1, Set<PacketPlayOutPosition.EnumPlayerTeleportFlags> set) {
+        if (player.dead) {
+            LOGGER.info("Attempt to teleport dead player {} restricted", player.getName());
+            return;
+        }
         // CraftBukkit start
         if (Float.isNaN(f)) {
             f = 0;
