From 7b6131237760dc3eced1d858a97a2d73bd9cdfd4 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Thu, 29 Nov 2018 16:32:03 -0800
Subject: [PATCH] Don't fill list with items on startup


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index bec7748ea..1afe018ff 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -367,7 +367,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
                     thread1.start();
                 }
 
-                Items.AIR.a(CreativeModeTab.g, NonNullList.a());
+                // Items.AIR.a(CreativeModeTab.g, NonNullList.a()); // Paper - comment
                 return true;
             }
         }
-- 
2.19.2

