From 425b99a5b8af5ae83a09c9798a43ee80c742176f Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Wed, 9 Oct 2019 21:39:35 -0500
Subject: [PATCH] Add option to limit pillager spawns at outposts


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 7ddcaeaa7..4beeed0d5 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -630,4 +630,9 @@ public class PaperWorldConfig {
     private void generatorSettings() {
         generateFlatBedrock = getBoolean("generator-settings.flat-bedrock", false);
     }
+
+    public int limitPillagerOutpostSpawns = 10;
+    private void limitPillagerOutpostSpawns() {
+        limitPillagerOutpostSpawns = getInt("limit-pillager-outpost-spawns", limitPillagerOutpostSpawns);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/ChunkProviderGenerate.java b/src/main/java/net/minecraft/server/ChunkProviderGenerate.java
index b794a39d3..95249e10b 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderGenerate.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderGenerate.java
@@ -142,6 +142,7 @@ public class ChunkProviderGenerate extends ChunkGeneratorAbstract<GeneratorSetti
             }
         } else if (enumcreaturetype == EnumCreatureType.MONSTER) {
             if (WorldGenerator.PILLAGER_OUTPOST.a(this.a, blockposition)) {
+                if (getWorld().paperConfig.limitPillagerOutpostSpawns > 0 && getWorld().a(EntityPillager.class, new AxisAlignedBB(blockposition).grow(128)).size() < getWorld().paperConfig.limitPillagerOutpostSpawns) // Paper
                 return WorldGenerator.PILLAGER_OUTPOST.e();
             }
 
-- 
2.23.0.rc1

