From 5c0b03a3feaf90167439833ca5dbd478a3589e33 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sat, 28 Sep 2019 15:02:09 -0700
Subject: [PATCH] fixup! incremental chunk saving


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index d2c0299730..ed262161a4 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1088,7 +1088,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
 
         //if (autosavePeriod > 0 && this.ticks % autosavePeriod == 0) { // CraftBukkit // Paper - move down
-            MinecraftServer.LOGGER.debug("Autosave started");
+            //MinecraftServer.LOGGER.debug("Autosave started"); // Paper
             serverAutoSave = (autosavePeriod > 0 && this.ticks % autosavePeriod == 0); // Paper
             this.methodProfiler.enter("save");
             if (autosavePeriod > 0 && this.ticks % autosavePeriod == 0) { // Paper
@@ -1107,7 +1107,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
             // Paper end
 
             this.methodProfiler.exit();
-            MinecraftServer.LOGGER.debug("Autosave finished");
+            //MinecraftServer.LOGGER.debug("Autosave finished"); // Paper
         //} // Paper
 
         this.methodProfiler.enter("snooper");
-- 
2.22.1

