From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 31 Mar 2020 03:50:42 -0400
Subject: [PATCH] Remote Connections shouldn't hold up shutdown

Bugs in the connection logic appears to leave stale connections even, preventing shutdown

diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 8fed3e9380a69cf7973c85456473e6983d1352f7..cf7dff14eba53c51f2b299e7144d15eae35d48fc 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -359,11 +359,11 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         }
 
         if (this.remoteControlListener != null) {
-            this.remoteControlListener.b();
+            //this.remoteControlListener.b(); // Paper - don't wait for remote connections
         }
 
         if (this.remoteStatusListener != null) {
-            this.remoteStatusListener.b();
+            //this.remoteStatusListener.b(); // Paper - don't wait for remote connections
         }
 
         System.exit(0); // CraftBukkit
