From 0a9b9138d7008ff443b302945deb282e2a50db86 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 13 Apr 2019 14:22:02 -0500
Subject: [PATCH] Add setting to suppress unknown attribute warnings


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 9f240c35d..401f58d44 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -462,4 +462,9 @@ public class PaperConfig {
         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
     }
+
+    public static boolean suppressUnknownAttributeWarnings = false;
+    private static void suppressUnknownAttributeWarnings() {
+        suppressUnknownAttributeWarnings = getBoolean("settings.suppress-unknown-attribute-warnings", suppressUnknownAttributeWarnings);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/GenericAttributes.java b/src/main/java/net/minecraft/server/GenericAttributes.java
index 90367a53e..553fe9e6c 100644
--- a/src/main/java/net/minecraft/server/GenericAttributes.java
+++ b/src/main/java/net/minecraft/server/GenericAttributes.java
@@ -4,6 +4,7 @@ import java.util.Collection;
 import java.util.Iterator;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -78,7 +79,7 @@ public class GenericAttributes {
             AttributeInstance attributeinstance = attributemapbase.a(nbttagcompound.getString("Name"));
 
             if (attributeinstance == null) {
-                GenericAttributes.k.warn("Ignoring unknown attribute '{}'", nbttagcompound.getString("Name"));
+                if (!com.destroystokyo.paper.PaperConfig.suppressUnknownAttributeWarnings) GenericAttributes.k.warn("Ignoring unknown attribute '{}'", nbttagcompound.getString("Name")); // Paper
             } else {
                 a(attributeinstance, nbttagcompound);
             }
-- 
2.20.1

