From 1e3ef18dc74b7a6214a7eb11a18b3bf4f3f9b2f5 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Fri, 14 Feb 2020 02:45:10 -0800
Subject: [PATCH] fixup! MC Utils


diff --git a/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java b/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
index 5a44bc644..e1dcbf571 100644
--- a/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
+++ b/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
@@ -318,61 +318,61 @@ public abstract class AreaMap<E> {
         int minZ; // inclusive
 
         if (dx != 0) {
-            // handle right addition
+            // handle left removal
 
-            maxX = toX + (oldViewDistance * right) + right; // exclusive
-            minX = fromX + (oldViewDistance * right) + right; // inclusive
+            maxX = toX - (oldViewDistance * right); // exclusive
+            minX = fromX - (oldViewDistance * right); // inclusive
             maxZ = fromZ + (oldViewDistance * up) + up; // exclusive
             minZ = toZ - (oldViewDistance * up); // inclusive
 
             for (int currX = minX; currX != maxX; currX += right) {
                 for (int currZ = minZ; currZ != maxZ; currZ += up) {
-                    this.addObjectTo(object, currX, currZ, toX, toZ, fromX, fromZ);
+                    this.removeObjectFrom(object, currX, currZ, toX, toZ, fromX, fromZ);
                 }
             }
         }
 
         if (dz != 0) {
-            // handle up addition
+            // handle down removal
 
-            maxX = toX + (oldViewDistance * right) + right; // exclusive
-            minX = toX - (oldViewDistance * right); // inclusive
-            maxZ = toZ + (oldViewDistance * up) + up; // exclusive
-            minZ = fromZ + (oldViewDistance * up) + up; // inclusive
+            maxX = fromX + (oldViewDistance * right) + right; // exclusive
+            minX = fromX - (oldViewDistance * right); // inclusive
+            maxZ = toZ - (oldViewDistance * up); // exclusive
+            minZ = fromZ - (oldViewDistance * up); // inclusive
 
             for (int currX = minX; currX != maxX; currX += right) {
                 for (int currZ = minZ; currZ != maxZ; currZ += up) {
-                    this.addObjectTo(object, currX, currZ, toX, toZ, fromX, fromZ);
+                    this.removeObjectFrom(object, currX, currZ, toX, toZ, fromX, fromZ);
                 }
             }
         }
 
         if (dx != 0) {
-            // handle left removal
+            // handle right addition
 
-            maxX = toX - (oldViewDistance * right); // exclusive
-            minX = fromX - (oldViewDistance * right); // inclusive
+            maxX = toX + (oldViewDistance * right) + right; // exclusive
+            minX = fromX + (oldViewDistance * right) + right; // inclusive
             maxZ = fromZ + (oldViewDistance * up) + up; // exclusive
             minZ = toZ - (oldViewDistance * up); // inclusive
 
             for (int currX = minX; currX != maxX; currX += right) {
                 for (int currZ = minZ; currZ != maxZ; currZ += up) {
-                    this.removeObjectFrom(object, currX, currZ, toX, toZ, fromX, fromZ);
+                    this.addObjectTo(object, currX, currZ, toX, toZ, fromX, fromZ);
                 }
             }
         }
 
         if (dz != 0) {
-            // handle down removal
+            // handle up addition
 
-            maxX = fromX + (oldViewDistance * right) + right; // exclusive
-            minX = fromX - (oldViewDistance * right); // inclusive
-            maxZ = toZ - (oldViewDistance * up); // exclusive
-            minZ = fromZ - (oldViewDistance * up); // inclusive
+            maxX = toX + (oldViewDistance * right) + right; // exclusive
+            minX = toX - (oldViewDistance * right); // inclusive
+            maxZ = toZ + (oldViewDistance * up) + up; // exclusive
+            minZ = fromZ + (oldViewDistance * up) + up; // inclusive
 
             for (int currX = minX; currX != maxX; currX += right) {
                 for (int currZ = minZ; currZ != maxZ; currZ += up) {
-                    this.removeObjectFrom(object, currX, currZ, toX, toZ, fromX, fromZ);
+                    this.addObjectTo(object, currX, currZ, toX, toZ, fromX, fromZ);
                 }
             }
         }
-- 
2.25.0

