From 901cd905f4f9966045765dd861383cf07e42963c Mon Sep 17 00:00:00 2001
From: Zach Brown <zach@zachbr.io>
Date: Tue, 24 Sep 2019 19:54:39 -0500
Subject: [PATCH] Allow running on Java 13


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index b145b55bc..387ed31db 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -172,7 +172,7 @@ public class Main {
             }
 
             float javaVersion = Float.parseFloat(System.getProperty("java.class.version"));
-            if (javaVersion > 56.0) {
+            if (!Boolean.getBoolean("Paper.IgnoreJavaVersion") && javaVersion > 57.0) { // Paper - Don't complain about Java 13
                 System.err.println("Unsupported Java detected (" + javaVersion + "). Only up to Java 12 is supported.");
                 return;
             }
-- 
2.23.0

