From aaf5284b5c839db6bc8c328e38430460901cd4c5 Mon Sep 17 00:00:00 2001
From: Trigary <trigary0@gmail.com>
Date: Mon, 6 Apr 2020 15:23:42 +0200
Subject: [PATCH] track complex entity parts


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index e1e4ea793..48dc947ac 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -1505,6 +1505,13 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
                     PlayerChunkMap.EntityTracker playerchunkmap_entitytracker = new PlayerChunkMap.EntityTracker(entity, i, j, entitytypes.isDeltaTracking());
 
                     entity.tracker = playerchunkmap_entitytracker; // Paper - Fast access to tracker
+                    // Paper start - track complex entity parts
+                    if (entity instanceof EntityEnderDragon) {
+                        for (EntityComplexPart entityPart : ((EntityEnderDragon) entity).children) {
+                            entityPart.tracker = entity.tracker;
+                        }
+                    }
+                    // Paper end
                     this.trackedEntities.put(entity.getId(), playerchunkmap_entitytracker);
                     playerchunkmap_entitytracker.track(this.world.getPlayers());
                     if (entity instanceof EntityPlayer) {
-- 
2.25.1.windows.1

