From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom <cryptite@gmail.com>
Date: Wed, 13 May 2020 08:44:47 -0500
Subject: [PATCH] BlockBreakNaturallyEvent


diff --git a/src/main/java/net/minecraft/server/Block.java b/src/main/java/net/minecraft/server/Block.java
index e40375b67a4a321048c87002a07fde5c5d2395db..24d729c4974d818e5a87243fe54ce1e0773303d4 100644
--- a/src/main/java/net/minecraft/server/Block.java
+++ b/src/main/java/net/minecraft/server/Block.java
@@ -12,6 +12,8 @@ import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockBreakNaturallyEvent;
 
 public class Block implements IMaterial {
 
@@ -191,6 +193,17 @@ public class Block implements IMaterial {
     public static void a(IBlockData iblockdata, IBlockData iblockdata1, GeneratorAccess generatoraccess, BlockPosition blockposition, int i) {
         if (iblockdata1 != iblockdata) {
             if (iblockdata1.isAir()) {
+
+                // Paper start
+                World world = generatoraccess.getMinecraftWorld();
+                if (world.currentBlockEventBlock != null) {
+                    BlockBreakNaturallyEvent breakByBlockEvent = new BlockBreakNaturallyEvent(CraftBlock.at(world, blockposition), world.currentBlockEventPlayer, world.currentBlockEventBlock);
+                    world.getServer().getPluginManager().callEvent(breakByBlockEvent);
+
+                    if (breakByBlockEvent.isCancelled()) return;
+                }
+                // Paper end
+
                 if (!generatoraccess.p_()) {
                     generatoraccess.b(blockposition, (i & 32) == 0);
                 }
