From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Wed, 27 Sep 2017 18:46:14 -0700
Subject: [PATCH] Monumenta - Disable potion has no item logspam


diff --git a/src/main/java/net/minecraft/server/EntityPotion.java b/src/main/java/net/minecraft/server/EntityPotion.java
index b56b021b4cb61bd69536f3aba50ab681dc2a4ea0..061e9d484d50c539d4eed1d3629a1163aaed5f7b 100644
--- a/src/main/java/net/minecraft/server/EntityPotion.java
+++ b/src/main/java/net/minecraft/server/EntityPotion.java
@@ -41,10 +41,6 @@ public class EntityPotion extends EntityProjectile {
         ItemStack itemstack = (ItemStack) this.getDataWatcher().get(EntityPotion.f);
 
         if (itemstack.getItem() != Items.SPLASH_POTION && itemstack.getItem() != Items.LINGERING_POTION) {
-            if (this.world != null) {
-                EntityPotion.LOGGER.error("ThrownPotion entity {} has no item?!", this.getId());
-            }
-
             return new ItemStack(Items.SPLASH_POTION);
         } else {
             return itemstack;
