From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Thu, 5 Oct 2017 21:45:16 -0700
Subject: [PATCH] Monumenta - Removed test for monsters when sleeping in beds


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 7df24be46e4471ae0ddad5cded5eef4937fd37a3..fda891dc5449e3e0e3ed304fc9b1ace20e4a87fa 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1327,16 +1327,6 @@ public abstract class EntityHuman extends EntityLiving {
             }
 
             if (!this.isCreative()) {
-                double d0 = 8.0D;
-                double d1 = 5.0D;
-                Vec3D vec3d = new Vec3D((double) blockposition.getX() + 0.5D, (double) blockposition.getY(), (double) blockposition.getZ() + 0.5D);
-                List<EntityMonster> list = this.world.a(EntityMonster.class, new AxisAlignedBB(vec3d.getX() - 8.0D, vec3d.getY() - 5.0D, vec3d.getZ() - 8.0D, vec3d.getX() + 8.0D, vec3d.getY() + 5.0D, vec3d.getZ() + 8.0D), (entitymonster) -> {
-                    return entitymonster.e(this);
-                });
-
-                if (!list.isEmpty()) {
-                    return Either.left(EntityHuman.EnumBedResult.NOT_SAFE);
-                }
             }
         }
         return Either.right(Unit.INSTANCE);
