From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Tue, 31 Oct 2017 17:44:53 -0700
Subject: [PATCH] Monumenta - Charged creepers no longer cause mob head drops


diff --git a/src/main/java/net/minecraft/server/EntityCreeper.java b/src/main/java/net/minecraft/server/EntityCreeper.java
index 45dfc8104a75f9893c6ee03d507cac80b893249c..481606feef87025e9b268db83225c1ee9d3ae5f6 100644
--- a/src/main/java/net/minecraft/server/EntityCreeper.java
+++ b/src/main/java/net/minecraft/server/EntityCreeper.java
@@ -271,7 +271,7 @@ public class EntityCreeper extends EntityMonster {
     }
 
     public boolean canCauseHeadDrop() {
-        return this.isPowered() && this.bA < 1;
+        return false;
     }
 
     public void setCausedHeadDrop() {
