From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Fri, 14 Aug 2020 23:41:19 +0200
Subject: [PATCH] Don't mark null chunk sections for block updates


diff --git a/src/main/java/net/minecraft/server/PlayerChunk.java b/src/main/java/net/minecraft/server/PlayerChunk.java
index 92a5a5cefb048d4a65ba64e154ce2a3459b28b28..e02ce1c93a6b0a1789212a60a789f7dada82a0fe 100644
--- a/src/main/java/net/minecraft/server/PlayerChunk.java
+++ b/src/main/java/net/minecraft/server/PlayerChunk.java
@@ -363,6 +363,7 @@ public class PlayerChunk {
 
         if (chunk != null) {
             byte b0 = (byte) SectionPosition.a(blockposition.getY());
+            if (chunk.getSections()[b0] == null) return; // Paper - Don't mark null chunk sections for block updates
 
             if (this.dirtyBlocks[b0] == null) {
                 this.p = true;
