From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bermudalocket <github@eigenstuff.net>
Date: Wed, 19 Aug 2020 12:28:11 -0400
Subject: [PATCH] Implement PlayerTradeWithVillagerEvent


diff --git a/src/main/java/net/minecraft/server/EntityVillagerAbstract.java b/src/main/java/net/minecraft/server/EntityVillagerAbstract.java
index 81823b5d5ef17479583fda0121c95091175fdf1e..f4c494a56fd0ca2257c421a3157e48b0905f4fa3 100644
--- a/src/main/java/net/minecraft/server/EntityVillagerAbstract.java
+++ b/src/main/java/net/minecraft/server/EntityVillagerAbstract.java
@@ -106,6 +106,7 @@ public abstract class EntityVillagerAbstract extends EntityAgeable implements NP
         this.b(merchantrecipe);
         if (this.tradingPlayer instanceof EntityPlayer) {
             CriterionTriggers.s.a((EntityPlayer) this.tradingPlayer, this, merchantrecipe.getSellingItem());
+            Bukkit.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerTradeWithVillagerEvent(((EntityPlayer) this.tradingPlayer).getBukkitEntity(), (org.bukkit.entity.AbstractVillager) this.getBukkitEntity(), merchantrecipe.asBukkit())); // Paper
         }
 
     }
