From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Fri, 21 Aug 2020 12:51:25 -0700
Subject: [PATCH] Fix MC-197883 by specifying the typeKey


diff --git a/src/main/java/net/minecraft/server/WorldGenFeatureConfigured.java b/src/main/java/net/minecraft/server/WorldGenFeatureConfigured.java
index d0ee03c1923d81b50a451d4853245154858088ac..75b1d026b4fe08ca2d55de3ced7526c669eeec17 100644
--- a/src/main/java/net/minecraft/server/WorldGenFeatureConfigured.java
+++ b/src/main/java/net/minecraft/server/WorldGenFeatureConfigured.java
@@ -10,7 +10,7 @@ import org.apache.logging.log4j.Logger;
 
 public class WorldGenFeatureConfigured<FC extends WorldGenFeatureConfiguration, F extends WorldGenerator<FC>> implements IDecoratable<WorldGenFeatureConfigured<?, ?>> {
 
-    public static final Codec<WorldGenFeatureConfigured<?, ?>> a = IRegistry.FEATURE.dispatch((worldgenfeatureconfigured) -> {
+    public static final Codec<WorldGenFeatureConfigured<?, ?>> a = IRegistry.FEATURE.dispatch("name", (worldgenfeatureconfigured) -> { // Paper - MC-197883 - specify "name" as the typeKey
         return worldgenfeatureconfigured.e;
     }, WorldGenerator::a);
     public static final Codec<Supplier<WorldGenFeatureConfigured<?, ?>>> b = RegistryFileCodec.a(IRegistry.au, WorldGenFeatureConfigured.a);
