From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Sat, 23 Mar 2019 01:51:24 -0400
Subject: [PATCH] Monumenta - drowned no longer have biome restrictions

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/server/EntityDrowned.java b/src/main/java/net/minecraft/server/EntityDrowned.java
index 4bdd21047cd8f4a9590229c8d1c5d258c9aadb82..dbf5a989376233f61cf8811341484a8ce0af0a72 100644
--- a/src/main/java/net/minecraft/server/EntityDrowned.java
+++ b/src/main/java/net/minecraft/server/EntityDrowned.java
@@ -48,7 +48,7 @@ public class EntityDrowned extends EntityZombie implements IRangedEntity {
         BiomeBase biomebase = generatoraccess.getBiome(blockposition);
         boolean flag = generatoraccess.getDifficulty() != EnumDifficulty.PEACEFUL && a(generatoraccess, blockposition, random) && (enummobspawn == EnumMobSpawn.SPAWNER || generatoraccess.getFluid(blockposition).a(TagsFluid.WATER));
 
-        return biomebase != Biomes.RIVER && biomebase != Biomes.FROZEN_RIVER ? random.nextInt(40) == 0 && a(generatoraccess, blockposition) && flag : random.nextInt(15) == 0 && flag;
+        return random.nextInt(15) == 0 && flag;
     }
 
     private static boolean a(GeneratorAccess generatoraccess, BlockPosition blockposition) {
