From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: simpleauthority <jacob@algorithmjunkie.com>
Date: Mon, 21 Sep 2020 16:51:32 -0700
Subject: [PATCH] Fix villager trade menu not closing when unable to trade


diff --git a/src/main/java/net/minecraft/server/EntityVillager.java b/src/main/java/net/minecraft/server/EntityVillager.java
index c10ff09dd005c0c8b3cd374d9de82e77efcd9097..94e56eeafbe20583b1e3643967d4a160b54c70a8 100644
--- a/src/main/java/net/minecraft/server/EntityVillager.java
+++ b/src/main/java/net/minecraft/server/EntityVillager.java
@@ -968,6 +968,12 @@ public class EntityVillager extends EntityVillagerAbstract implements Reputation
         this.bg.setMemory(MemoryModuleType.LAST_SLEPT, this.world.getTime()); // CraftBukkit - decompile error
         this.bg.removeMemory(MemoryModuleType.WALK_TARGET);
         this.bg.removeMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE);
+        // Paper start
+        if (getTrader() != null ) {
+            getTrader().closeInventory();
+            setTradingPlayer(null);
+        }
+        // Paper end
     }
 
     @Override
