From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JRoy <joshroy126@gmail.com>
Date: Sun, 18 Oct 2020 18:21:46 -0400
Subject: [PATCH] Add option to hide server gui in config


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index efa8fd5972b5e115509e58d3286c2f1ad6c37224..51a4f01518676fc65d42ac0b50f7418997eb9376 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -464,4 +464,9 @@ public class PaperConfig {
     private static void trackPluginScoreboards() {
         trackPluginScoreboards = getBoolean("settings.track-plugin-scoreboards", false);
     }
+
+    public static boolean hideServerGui;
+    private static void hideServerGui() {
+        hideServerGui = getBoolean("settings.hide-server-gui", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index e8ef0b114a9c9fc1cc669f087b1ec621b35bfb4e..3c53a4a27316e5e777615afd35603c74e4039ff6 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -200,7 +200,7 @@ public class Main {
                 dedicatedserver1.c(optionset.has(optionspec2));
                 dedicatedserver1.b((String) optionset.valueOf(optionspec12));
                 */
-                boolean flag1 = !optionset.has("nogui") && !optionset.nonOptionArguments().contains("nogui");
+                boolean flag1 = !optionset.has("nogui") && !optionset.nonOptionArguments().contains("nogui") && !com.destroystokyo.paper.PaperConfig.hideServerGui; // Paper - Add option to hide server gui in config
 
                 if (flag1 && !GraphicsEnvironment.isHeadless()) {
                     dedicatedserver1.bc();
