From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ysl3000 <yannicklamprecht@live.de>
Date: Mon, 5 Oct 2020 21:25:16 +0200
Subject: [PATCH] Player Chunk Load/Unload Events


diff --git a/src/main/java/net/minecraft/server/level/EntityPlayer.java b/src/main/java/net/minecraft/server/level/EntityPlayer.java
index 8fa8d2e1e85fa382af9e9d65322bd98a87a6e3a9..a2971272602f5c2556ccb46042926617138a1d55 100644
--- a/src/main/java/net/minecraft/server/level/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/level/EntityPlayer.java
@@ -136,6 +136,8 @@ import net.minecraft.world.scores.ScoreboardScore;
 import net.minecraft.world.scores.ScoreboardTeam;
 import net.minecraft.world.scores.ScoreboardTeamBase;
 import net.minecraft.world.scores.criteria.IScoreboardCriteria;
+import io.papermc.paper.event.packet.PlayerChunkLoadEvent; // Paper
+import io.papermc.paper.event.packet.PlayerChunkUnloadEvent; // Paper
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -2085,11 +2087,21 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public void a(ChunkCoordIntPair chunkcoordintpair, Packet<?> packet, Packet<?> packet1) {
         this.playerConnection.sendPacket(packet1);
         this.playerConnection.sendPacket(packet);
+        // Paper start
+        if(PlayerChunkLoadEvent.getHandlerList().getRegisteredListeners().length > 0){
+            new PlayerChunkLoadEvent(this.getBukkitEntity().getWorld().getChunkAt(chunkcoordintpair.longKey), this.getBukkitEntity()).callEvent();
+        }
+        // Paper end
     }
 
     public void a(ChunkCoordIntPair chunkcoordintpair) {
         if (this.isAlive()) {
             this.playerConnection.sendPacket(new PacketPlayOutUnloadChunk(chunkcoordintpair.x, chunkcoordintpair.z));
+            // Paper start
+            if(PlayerChunkUnloadEvent.getHandlerList().getRegisteredListeners().length > 0){
+                new PlayerChunkUnloadEvent(this.getBukkitEntity().getWorld().getChunkAt(chunkcoordintpair.longKey), this.getBukkitEntity()).callEvent();
+            }
+            // Paper end
         }
 
     }
