From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BrodyBeckwith <brody@beckwith.dev>
Date: Fri, 9 Oct 2020 20:30:12 -0400
Subject: [PATCH] Allow disabling mob spawner spawn egg transformation


diff --git a/src/main/java/io/papermc/paper/PaperWorldConfig.java b/src/main/java/io/papermc/paper/PaperWorldConfig.java
index 1b469cb2a07640d6424fa7aecfec4c7b66e719ce..247945946c73f78f7c3bc7bcfa6578c4fafa53c9 100644
--- a/src/main/java/io/papermc/paper/PaperWorldConfig.java
+++ b/src/main/java/io/papermc/paper/PaperWorldConfig.java
@@ -720,4 +720,9 @@ public class PaperWorldConfig {
     private void fixCuringExploit() {
         fixCuringZombieVillagerDiscountExploit = getBoolean("game-mechanics.fix-curing-zombie-villager-discount-exploit", fixCuringZombieVillagerDiscountExploit);
     }
+
+    public boolean disableMobSpawnerSpawnEggTransformation = false;
+    private void disableMobSpawnerSpawnEggTransformation() {
+        disableMobSpawnerSpawnEggTransformation = getBoolean("game-mechanics.disable-mob-spawner-spawn-egg-transformation", disableMobSpawnerSpawnEggTransformation);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/item/ItemMonsterEgg.java b/src/main/java/net/minecraft/world/item/ItemMonsterEgg.java
index c170d2141504d80624e3c1a7f78f7968ea8a80ee..4d965e504a40eb52777575df839856c825a0900a 100644
--- a/src/main/java/net/minecraft/world/item/ItemMonsterEgg.java
+++ b/src/main/java/net/minecraft/world/item/ItemMonsterEgg.java
@@ -60,7 +60,7 @@ public class ItemMonsterEgg extends Item {
             EnumDirection enumdirection = itemactioncontext.getClickedFace();
             IBlockData iblockdata = world.getType(blockposition);
 
-            if (iblockdata.a(Blocks.SPAWNER)) {
+            if (!world.paperConfig.disableMobSpawnerSpawnEggTransformation && iblockdata.a(Blocks.SPAWNER)) { // Paper
                 TileEntity tileentity = world.getTileEntity(blockposition);
 
                 if (tileentity instanceof TileEntityMobSpawner) {
