From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Sat, 20 Feb 2021 16:43:07 +0000
Subject: [PATCH] Skip POI finding if stuck in vehicle


diff --git a/src/main/java/net/minecraft/server/BehaviorFindPosition.java b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
index 63a761ebef80d4af09cdc2682e496d78492c4a3a..86349bea3ddb931ed56da7b02b611b45b5aa1a1d 100644
--- a/src/main/java/net/minecraft/server/BehaviorFindPosition.java
+++ b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
@@ -45,6 +45,11 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
     }
 
     protected boolean a(WorldServer worldserver, EntityCreature entitycreature) {
+        // Paper start - don't calculate POI if they're trapped in a vehicle
+        if (entitycreature.getVehicle() != null) {
+            return false;
+        }
+        // Paper end
         if (this.d && entitycreature.isBaby()) {
             return false;
         } else if (this.f == 0L) {
@@ -84,7 +89,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
                 villageplace.a(this.b.c(), (blockposition1) -> {
                     return blockposition1.equals(blockposition);
                 }, blockposition, 1);
-                entitycreature.getBehaviorController().setMemory(this.c, (Object) GlobalPos.create(worldserver.getDimensionKey(), blockposition));
+                entitycreature.getBehaviorController().setMemory(this.c, GlobalPos.create(worldserver.getDimensionKey(), blockposition)); // Paper - decompile fix
                 this.e.ifPresent((obyte) -> {
                     worldserver.broadcastEntityEffect(entitycreature, obyte);
                 });
