From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tofik <slodkitofik@gmail.com>
Date: Mon, 22 Mar 2021 14:23:58 +0100
Subject: [PATCH] Added report command


diff --git a/src/main/java/com/destroystokyo/paper/PaperCommand.java b/src/main/java/com/destroystokyo/paper/PaperCommand.java
index 12313a37ceeb6a0b6a539c38fdba67e5e43d7413..11cbd574b1eb0e1d916f1865d727482bd0210b64 100644
--- a/src/main/java/com/destroystokyo/paper/PaperCommand.java
+++ b/src/main/java/com/destroystokyo/paper/PaperCommand.java
@@ -1,5 +1,7 @@
 package com.destroystokyo.paper;
 
+import net.kyroi.adventure.text.TextComponent;
+import net.kyroi.adventure.text.Component;
 import com.destroystokyo.paper.io.chunk.ChunkTaskManager;
 import com.destroystokyo.paper.io.SyncLoadFinder;
 import com.google.common.base.Functions;
@@ -62,7 +64,7 @@ import java.util.stream.Collectors;
 
 public class PaperCommand extends Command {
     private static final String BASE_PERM = "bukkit.command.paper.";
-    private static final ImmutableSet<String> SUBCOMMANDS = ImmutableSet.<String>builder().add("heap", "entity", "reload", "version", "debug", "chunkinfo", "dumpwaiting", "syncloadinfo", "fixlight", "dumpitem").build();
+    private static final ImmutableSet<String> SUBCOMMANDS = ImmutableSet.<String>builder().add("heap", "entity", "reload", "version", "debug", "chunkinfo", "dumpwaiting", "syncloadinfo", "fixlight", "dumpitem", "report").build();
 
     public PaperCommand(String name) {
         super(name);
@@ -165,6 +167,9 @@ public class PaperCommand extends Command {
             case "heap":
                 dumpHeap(sender);
                 break;
+            case "report":
+                doReport(sender);
+                break;
             case "entity":
                 listEntities(sender, args);
                 break;
@@ -558,6 +563,12 @@ public class PaperCommand extends Command {
         }
     }
 
+    private void DoReport(CommandSender sender) {
+    final TextComponent textComponent = Component.text("To open an issue report , please click the following link")
+            .color(TextColor.color(0x5635925))
+            .append(Component.text("https://github.com/PaperMC/Paper/issues/new/choose", NamedTextColor.GREEN));
+    }
+    
     private void doReload(CommandSender sender) {
         Command.broadcastCommandMessage(sender, ChatColor.RED + "Please note that this command is not supported and may cause issues.");
         Command.broadcastCommandMessage(sender, ChatColor.RED + "If you encounter any issues please use the /stop command to restart your server.");
