From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 29 Aug 2021 20:24:17 +0100
Subject: [PATCH] Remove duplicate code


diff --git a/src/main/java/net/minecraft/world/inventory/ContainerFurnace.java b/src/main/java/net/minecraft/world/inventory/ContainerFurnace.java
index 14cc1f15ae3af5b7e0996eab7ecc895546a4ed8c..4fbc84be10ec7f9ecd6df25e96b974b9e121c39f 100644
--- a/src/main/java/net/minecraft/world/inventory/ContainerFurnace.java
+++ b/src/main/java/net/minecraft/world/inventory/ContainerFurnace.java
@@ -89,11 +89,6 @@ public abstract class ContainerFurnace extends ContainerRecipeBook<IInventory> {
         this.furnace.clear();
     }
 
-    @Override
-    public void a(boolean flag, IRecipe<?> irecipe, EntityPlayer entityplayer) {
-        (new AutoRecipeFurnace(this)).a(entityplayer, irecipe, flag); // CraftBukkit - decompile error
-    }
-
     @Override
     public boolean a(IRecipe<? super IInventory> irecipe) {
         return irecipe.a(this.furnace, this.c);
