From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Aug 2018 20:56:26 -0400
Subject: [PATCH] Don't double add golems to world

spawnCreature adds to world now

diff --git a/src/main/java/net/minecraft/server/Village.java b/src/main/java/net/minecraft/server/Village.java
index 5b56ae1818..afb79bb641 100644
--- a/src/main/java/net/minecraft/server/Village.java
+++ b/src/main/java/net/minecraft/server/Village.java
@@ -0,0 +0,0 @@ public class Village {
 
                 if (entityirongolem != null) {
                     if (entityirongolem.a((GeneratorAccess) this.a, false) && entityirongolem.a((IWorldReader) this.a)) {
-                        this.a.addEntity(entityirongolem);
+                        //this.a.addEntity(entityirongolem); // Paper - already is in the world
                         return entityirongolem;
                     }
 
--