From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 19 Dec 2016 01:23:03 -0500
Subject: [PATCH] Fix worldborder bug

typo in variable name

diff --git a/src/main/java/net/minecraft/server/WorldBorder.java b/src/main/java/net/minecraft/server/WorldBorder.java
index 2252fb03f..215e6f5d0 100644
--- a/src/main/java/net/minecraft/server/WorldBorder.java
+++ b/src/main/java/net/minecraft/server/WorldBorder.java
@@ -0,0 +0,0 @@ public class WorldBorder {
 
     // Inlined the getters from ChunkCoordIntPair
     public boolean isInBounds(int x, int z) {
-        return (double) ((x  << 4) + 15) > this.b() && (double) (x << 4) < this.d() && (double) ((z  << 4) + 15) > this.c() && (double) (x << 4) < this.e();
+        return (double) ((x  << 4) + 15) > this.b() && (double) (x << 4) < this.d() && (double) ((z  << 4) + 15) > this.c() && (double) (z << 4) < this.e(); // Paper - fix CraftBukkit bug
     }
 
     public boolean a(AxisAlignedBB axisalignedbb) {
--