From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <1254957+zachbr@users.noreply.github.com>
Date: Sat, 29 Apr 2017 20:41:58 -0500
Subject: [PATCH] Raise book max title length to 32


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index dfeeb54e6..0c7ae36b4 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -0,0 +0,0 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     static final ItemMetaKey GENERATION = new ItemMetaKey("generation");
     static final int MAX_PAGES = 50;
     static final int MAX_PAGE_LENGTH = 256;
-    static final int MAX_TITLE_LENGTH = 16;
+    static final int MAX_TITLE_LENGTH = 32; // Paper
 
     protected String title;
     protected String author;
--