From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java b/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java
index 289681fcfafae8395acf289b519caf6e2255be4a..fbb9840e853052368012d13d287b62defe6cfcb1 100644
--- a/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java
@@ -69,6 +69,12 @@ public abstract class AbstractChestedHorse extends AbstractHorse {
         super.dropEquipment(level);
         if (this.hasChest()) {
             this.spawnAtLocation(level, Blocks.CHEST);
+            // Paper start - moved to post death logic
+        }
+    }
+    protected void postDeathDropItems(org.bukkit.event.entity.EntityDeathEvent event) {
+        if (this.hasChest() && !event.isCancelled()) {
+            // Paper end - moved to post death logic
             this.setChest(false);
         }
     }
