--- a/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
+++ b/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
@@ -122,7 +_,7 @@
     public void aiStep() {
         super.aiStep();
         if (this.isTrap() && this.trapTime++ >= 18000) {
-            this.discard();
+            this.discard(com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent.Cause.DESPAWN); // Paper - EntityRemoveFromWorldEvent
         }
     }
 
