From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/EmptyBlockAndTintGetter.java b/net/minecraft/world/level/EmptyBlockAndTintGetter.java
index dbad9e6d9466d512f48eadc20a084539f2e33c49..d27024f9ac20c61bbb23ef713eafc8fdd4b7ddcf 100644
--- a/net/minecraft/world/level/EmptyBlockAndTintGetter.java
+++ b/net/minecraft/world/level/EmptyBlockAndTintGetter.java
@@ -38,6 +38,18 @@ public enum EmptyBlockAndTintGetter implements BlockAndTintGetter {
         return Blocks.AIR.defaultBlockState();
     }
 
+    // Paper start
+    @Override
+    public @Nullable BlockState getBlockStateIfLoaded(final BlockPos pos) {
+        return null;
+    }
+
+    @Override
+    public @Nullable FluidState getFluidIfLoaded(final BlockPos pos) {
+        return null;
+    }
+    // Paper end
+
     @Override
     public FluidState getFluidState(BlockPos pos) {
         return Fluids.EMPTY.defaultFluidState();
