--- a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -30,7 +_,8 @@
         if (explosionRadius != null) {
             RandomSource random = context.getRandom();
             float probability = 1.0F / explosionRadius;
-            return random.nextFloat() <= probability;
+            // CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+            return random.nextFloat() < probability;
         } else {
             return true;
         }
