From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
index 8d6dd7306db0bdc251d3397dcfa528ac4ddc570f..7ab5e009e3ea1a75a1f185c80133854e85068ad8 100644
--- a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -30,7 +30,8 @@ public class ExplosionCondition implements LootItemCondition {
         if (_float != null) {
             RandomSource random = context.getRandom();
             float f = 1.0F / _float;
-            return random.nextFloat() <= f;
+            // CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+            return random.nextFloat() < f;
         } else {
             return true;
         }
