--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -27,8 +27,10 @@
     }
 
     public ClientboundInitializeBorderPacket(WorldBorder worldborder) {
-        this.newCenterX = worldborder.getCenterX();
-        this.newCenterZ = worldborder.getCenterZ();
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldborder.getCenterX() * worldborder.world.getDimensionManager().getCoordinateScale();
+        this.newCenterZ = worldborder.getCenterZ() * worldborder.world.getDimensionManager().getCoordinateScale();
+        // CraftBukkit end
         this.oldSize = worldborder.getSize();
         this.newSize = worldborder.k();
         this.lerpTime = worldborder.j();
