--- a/net/minecraft/network/protocol/game/PacketPlayInChat.java
+++ b/net/minecraft/network/protocol/game/PacketPlayInChat.java
@@ -25,7 +25,7 @@
     }
 
     public PacketPlayInChat(PacketDataSerializer packetdataserializer) {
-        this.message = packetdataserializer.readUtf(256);
+        this.message = org.apache.commons.lang3.StringUtils.normalizeSpace(packetdataserializer.readUtf(256)); // CraftBukkit - see PlayerConnection
         this.timeStamp = packetdataserializer.readInstant();
         this.saltSignature = new MinecraftEncryption.b(packetdataserializer);
         this.signedPreview = packetdataserializer.readBoolean();
