From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/npc/villager/VillagerTrades.java b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
index ed9450f380c5108286a93bf345f6e57c5d384087..32a3b26ff9f654329052e3333d3b43bfec836401 100644
--- a/net/minecraft/world/entity/npc/villager/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
@@ -1737,7 +1737,8 @@ public class VillagerTrades {
 
         @Override
         public @Nullable MerchantOffer getOffer(ServerLevel level, Entity trader, RandomSource random) {
-            BlockPos blockPos = level.findNearestMapStructure(this.destination, trader.blockPosition(), 100, true);
+            if (!level.paperConfig().environment.treasureMaps.enabled) return null; // Paper - Configurable cartographer treasure maps
+            BlockPos blockPos = level.findNearestMapStructure(this.destination, trader.blockPosition(), 100, !level.paperConfig().environment.treasureMaps.findAlreadyDiscoveredVillager); // Paper - Configurable cartographer treasure maps
             if (blockPos != null) {
                 ItemStack itemStack = MapItem.create(level, blockPos.getX(), blockPos.getZ(), (byte)2, true, true);
                 MapItem.renderBiomePreviewMap(level, itemStack);
