--- a/net/minecraft/world/entity/animal/frog/ShootTongue.java
+++ b/net/minecraft/world/entity/animal/frog/ShootTongue.java
@@ -94,7 +_,7 @@
             if (target.isAlive()) {
                 body.doHurtTarget(level, target);
                 if (!target.isAlive()) {
-                    target.remove(Entity.RemovalReason.KILLED);
+                    target.remove(Entity.RemovalReason.KILLED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
                 }
             }
         }
