From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/effect/WeavingMobEffect.java b/net/minecraft/world/effect/WeavingMobEffect.java
index f1b2d890437e60668b2f576076d0685ec62544f7..28fbe47577f52522f20d64be4f02c3fdc2a6424f 100644
--- a/net/minecraft/world/effect/WeavingMobEffect.java
+++ b/net/minecraft/world/effect/WeavingMobEffect.java
@@ -27,11 +27,11 @@ class WeavingMobEffect extends MobEffect {
     @Override
     public void onMobRemoved(final ServerLevel level, final LivingEntity mob, final int amplifier, final Entity.RemovalReason reason) {
         if (reason == Entity.RemovalReason.KILLED && (mob instanceof Player || level.getGameRules().get(GameRules.MOB_GRIEFING))) {
-            this.spawnCobwebsRandomlyAround(level, mob.getRandom(), mob.blockPosition());
+            this.spawnCobwebsRandomlyAround(level, mob.getRandom(), mob.blockPosition(), mob); // Paper - Fire EntityChangeBlockEvent in more places
         }
     }
 
-    private void spawnCobwebsRandomlyAround(final ServerLevel level, final RandomSource random, final BlockPos pos) {
+    private void spawnCobwebsRandomlyAround(final ServerLevel level, final RandomSource random, final BlockPos pos, LivingEntity entity) { // Paper - Fire EntityChangeBlockEvent in more places
         Set<BlockPos> positionsToTransform = Sets.newHashSet();
         int cobwebCount = this.maxCobwebs.applyAsInt(random);
 
@@ -48,6 +48,7 @@ class WeavingMobEffect extends MobEffect {
         }
 
         for (BlockPos blockPos : positionsToTransform) {
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockPos, Blocks.COBWEB.defaultBlockState())) continue; // Paper - Fire EntityChangeBlockEvent in more places
             level.setBlock(blockPos, Blocks.COBWEB.defaultBlockState(), Block.UPDATE_ALL);
             level.levelEvent(LevelEvent.ANIMATION_SPAWN_COBWEB, blockPos, 0);
         }
