From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
index 1c72fb91a3992c1e9ffdeafad3f7e4c7fe2747b2..be8ba1425b55a4948afb3f86c0feb3e2b1c2b561 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -18,6 +18,11 @@ public class ClientboundLevelChunkWithLightPacket implements Packet<ClientGamePa
     private final int z;
     private final ClientboundLevelChunkPacketData chunkData;
     private final ClientboundLightUpdatePacketData lightData;
+    // Paper start - Anti-Xray
+    public void setReady(final boolean ready) {
+        // Empty hook, updated by feature patch
+    }
+    // Paper end - Anti-Xray
 
     public ClientboundLevelChunkWithLightPacket(
         final LevelChunk levelChunk,
