From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundCommandSuggestionPacket.java b/net/minecraft/network/protocol/game/ServerboundCommandSuggestionPacket.java
index e34dc1451706696af9db1d29e35438f0a7d50daa..a65beb9b14be7d68a6993c42087312bdb64590d2 100644
--- a/net/minecraft/network/protocol/game/ServerboundCommandSuggestionPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundCommandSuggestionPacket.java
@@ -19,7 +19,7 @@ public class ServerboundCommandSuggestionPacket implements Packet<ServerGamePack
 
     private ServerboundCommandSuggestionPacket(final FriendlyByteBuf input) {
         this.id = input.readVarInt();
-        this.command = input.readUtf(32500);
+        this.command = input.readUtf(2048); // Paper
     }
 
     private void write(final FriendlyByteBuf output) {
