--- a/net/minecraft/world/level/storage/loot/predicates/LootItemConditionRandomChanceWithLooting.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemConditionRandomChanceWithLooting.java
@@ -41,6 +41,11 @@
         if (entity instanceof EntityLiving) {
             i = EnchantmentManager.g((EntityLiving) entity);
         }
+        // CraftBukkit start - only use lootingModifier if set by Bukkit
+        if (loottableinfo.hasContextParameter(LootContextParameters.LOOTING_MOD)) {
+            i = loottableinfo.getContextParameter(LootContextParameters.LOOTING_MOD);
+        }
+        // CraftBukkit end
 
         return loottableinfo.a().nextFloat() < this.percent + (float) i * this.lootingMultiplier;
     }
