From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/storage/loot/LootDataType.java b/net/minecraft/world/level/storage/loot/LootDataType.java
index 8648d0a47ab8c82a268809c52314d992c98670ba..180033a66299e3278e38105fecccbf1353f65103 100644
--- a/net/minecraft/world/level/storage/loot/LootDataType.java
+++ b/net/minecraft/world/level/storage/loot/LootDataType.java
@@ -32,9 +32,14 @@ public record LootDataType<T>(ResourceKey<Registry<T>> registryKey, Codec<T> cod
     }
 
     private static LootDataType.Validator<LootTable> createLootTableValidator() {
-        return (context, key, value) -> value.validate(
-            context.setContextKeySet(value.getParamSet()).enterElement(new ProblemReporter.RootElementPathElement(key), key)
-        );
+        // CraftBukkit start
+        return (context, key, value) -> {
+            value.validate(
+                context.setContextKeySet(value.getParamSet()).enterElement(new ProblemReporter.RootElementPathElement(key), key)
+            );
+            value.craftLootTable = new org.bukkit.craftbukkit.CraftLootTable(org.bukkit.craftbukkit.util.CraftNamespacedKey.fromMinecraft(key.identifier()), value);
+            // CraftBukkit end
+        };
     }
 
     @FunctionalInterface
