--- a/net/minecraft/world/effect/MobEffect.java
+++ b/net/minecraft/world/effect/MobEffect.java
@@ -76,6 +_,7 @@
     public void applyInstantenousEffect(
         ServerLevel level, @Nullable Entity source, @Nullable Entity indirectSource, LivingEntity entity, int amplifier, double health
     ) {
+        if (!new io.papermc.paper.event.entity.EntityPotionTickEvent(entity.getBukkitLivingEntity(), org.bukkit.craftbukkit.potion.CraftPotionEffectType.minecraftToBukkit(this), amplifier).callEvent()) { return; } // Paper - Add EntityPotionTickEvent
         this.applyEffectTick(level, entity, amplifier);
     }
 
