--- a/net/minecraft/world/level/block/AbstractSkullBlock.java
+++ b/net/minecraft/world/level/block/AbstractSkullBlock.java
@@ -76,6 +_,11 @@
         if (!level.isClientSide()) {
             boolean signal = level.hasNeighborSignal(pos);
             if (signal != state.getValue(POWERED)) {
+                // Paper start - Call BlockRedstoneEvent
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, signal)) {
+                    return;
+                }
+                // Paper end - Call BlockRedstoneEvent
                 level.setBlock(pos, state.setValue(POWERED, signal), Block.UPDATE_CLIENTS);
             }
         }
