--- a/net/minecraft/world/level/block/AbstractCauldronBlock.java
+++ b/net/minecraft/world/level/block/AbstractCauldronBlock.java
@@ -63,7 +_,7 @@
         final BlockHitResult hitResult
     ) {
         CauldronInteraction behavior = this.interactions.get(itemStack);
-        return behavior.interact(state, level, pos, player, hand, itemStack);
+        return behavior.interact(state, level, pos, player, hand, itemStack, hitResult.getDirection()); // Paper - pass hit direction
     }
 
     @Override
