--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -90,6 +_,7 @@
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!level.isClientSide()) {
             int signal = this.getSignalForState(state);
             if (signal == 0) {
@@ -102,6 +_,16 @@
         int signal = this.getSignalStrength(level, pos);
         boolean wasPressed = oldSignal > 0;
         boolean isPressed = signal > 0;
+
+        // Paper start - Call BlockRedstoneEvent
+        if (oldSignal != signal) {
+            org.bukkit.event.block.BlockRedstoneEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldSignal, signal);
+
+            signal = event.getNewCurrent();
+            isPressed = signal > 0;
+        }
+        // Paper end - Call BlockRedstoneEvent
+
         if (oldSignal != signal) {
             BlockState newState = this.setSignalForState(state, signal);
             level.setBlock(pos, newState, Block.UPDATE_CLIENTS);
@@ -150,7 +_,13 @@
     }
 
     protected static int getEntityCount(final Level level, final AABB entityDetectionBox, final Class<? extends Entity> entityClass) {
-        return level.getEntitiesOfClass(entityClass, entityDetectionBox, EntitySelector.NO_SPECTATORS.and(e -> !e.isIgnoringBlockTriggers())).size();
+        // CraftBukkit start
+        return BasePressurePlateBlock.getEntities(level, entityDetectionBox, entityClass).size();
+    }
+
+    protected static <T extends Entity> java.util.List<T> getEntities(Level level, AABB box, Class<T> entityClass) {
+        return level.getEntitiesOfClass(entityClass, box, EntitySelector.NO_SPECTATORS.and(e -> !e.isIgnoringBlockTriggers()));
+        // CraftBukkit end
     }
 
     protected abstract int getSignalStrength(Level level, BlockPos pos);
