From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/BaseRailBlock.java b/net/minecraft/world/level/block/BaseRailBlock.java
index 81053797b73b7b0308005bf0c2977d015854ae57..4e13f5b233018d648f8ff83447624353fd62683b 100644
--- a/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/net/minecraft/world/level/block/BaseRailBlock.java
@@ -71,6 +71,7 @@ public abstract class BaseRailBlock extends Block implements SimpleWaterloggedBl
         state = this.updateDir(level, pos, state, true);
         if (this.isStraight) {
             level.neighborChanged(state, pos, this, null, movedByPiston);
+            state = level.getBlockState(pos); // Paper - Fix some rails connecting improperly
         }
 
         return state;
