--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -91,7 +_,7 @@
             }
 
             BedRule bedRule = level.environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, pos);
-            if (bedRule.explodes()) {
+            if (false && bedRule.explodes()) { // CraftBukkit - moved world and biome check into Player
                 bedRule.errorMessage().ifPresent(component -> player.displayClientMessage(component, true));
                 level.removeBlock(pos, false);
                 BlockPos blockPos = pos.relative(state.getValue(FACING).getOpposite());
@@ -103,22 +_,61 @@
                 level.explode(null, level.damageSources().badRespawnPointExplosion(center), null, center, 5.0F, true, Level.ExplosionInteraction.BLOCK);
                 return InteractionResult.SUCCESS_SERVER;
             } else if (state.getValue(OCCUPIED)) {
+                if (bedRule.explodes()) return this.explodeBed(state, level, pos); // Paper - check explode first
                 if (!this.kickVillagerOutOfBed(level, pos)) {
                     player.displayClientMessage(Component.translatable("block.minecraft.bed.occupied"), true);
                 }
 
                 return InteractionResult.SUCCESS_SERVER;
             } else {
+                // CraftBukkit start
+                final BlockState finalBlockState = state;
+                final BlockPos finalBlockPos = pos;
+                // CraftBukkit end
                 player.startSleepInBed(pos).ifLeft(bedSleepingProblem -> {
-                    if (bedSleepingProblem.message() != null) {
+                    // Paper start - PlayerBedFailEnterEvent
+                    if (false && bedSleepingProblem.message() != null) { // Moved down
                         player.displayClientMessage(bedSleepingProblem.message(), true);
                     }
+                    if (bedSleepingProblem != null) {
+                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBedFailEnterEvent(player, finalBlockPos, bedSleepingProblem);
+                        if (event.isCancelled()) {
+                            return;
+                        }
+                        // Paper end - PlayerBedFailEnterEvent
+                        // CraftBukkit start - handling bed explosion from below here
+                        if (event.getWillExplode()) { // Paper - PlayerBedFailEnterEvent
+                            this.explodeBed(finalBlockState, level, finalBlockPos);
+                        }
+                        // CraftBukkit end
+                        // Paper start - PlayerBedFailEnterEvent
+                        final net.kyori.adventure.text.Component message = event.getMessage();
+                        if (message != null) {
+                            player.displayClientMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(message), true);
+                        }
+                    }
+                    // Paper end - PlayerBedFailEnterEvent
                 });
                 return InteractionResult.SUCCESS_SERVER;
             }
         }
     }
 
+    // CraftBukkit start - Copied from the above method
+    private InteractionResult explodeBed(BlockState state, Level level, BlockPos pos) {
+        org.bukkit.block.BlockState blockState = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos).getState(); // CraftBukkit - capture BlockState before remove block
+        level.removeBlock(pos, false);
+        BlockPos blockPos = pos.relative(state.getValue(FACING).getOpposite());
+        if (level.getBlockState(blockPos).is(this)) {
+            level.removeBlock(blockPos, false);
+        }
+
+        Vec3 center = pos.getCenter();
+        level.explode(null, level.damageSources().badRespawnPointExplosion(center).causingBlockSnapshot(blockState), null, center, 5.0F, true, Level.ExplosionInteraction.BLOCK); // CraftBukkit - add state
+        return InteractionResult.SUCCESS_SERVER;
+    }
+    // CraftBukkit end
+
     private boolean kickVillagerOutOfBed(Level level, BlockPos pos) {
         List<Villager> entitiesOfClass = level.getEntitiesOfClass(Villager.class, new AABB(pos), LivingEntity::isSleeping);
         if (entitiesOfClass.isEmpty()) {
@@ -296,6 +_,11 @@
         if (!level.isClientSide()) {
             BlockPos blockPos = pos.relative(state.getValue(FACING));
             level.setBlock(blockPos, state.setValue(PART, BedPart.HEAD), Block.UPDATE_ALL);
+            // CraftBukkit start - SPIGOT-7315: Don't updated if we capture block states
+            if (level.captureBlockStates) {
+                return;
+            }
+            // CraftBukkit end
             level.updateNeighborsAt(pos, Blocks.AIR);
             state.updateNeighbourShapes(level, pos, Block.UPDATE_ALL);
         }
