--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -58,18 +_,22 @@
             int age = state.getValue(AGE);
 
             while (level.getBlockState(pos.below(height)).is(this)) {
-                if (++height == 3 && age == 15) {
+                if (++height == level.paperConfig().maxGrowthHeight.cactus && age == 15) { // Paper - Configurable cactus/bamboo/reed growth height
                     return;
                 }
             }
 
             if (age == 8 && this.canSurvive(this.defaultBlockState(), level, pos.above())) {
-                double chanceToGrowFlower = height >= 3 ? 0.25 : 0.1;
+                double chanceToGrowFlower = height >= level.paperConfig().maxGrowthHeight.cactus ? 0.25 : 0.1; // Paper - Configurable cactus/bamboo/reed growth height
                 if (random.nextDouble() <= chanceToGrowFlower) {
                     level.setBlockAndUpdate(above, Blocks.CACTUS_FLOWER.defaultBlockState());
                 }
-            } else if (age == 15 && height < 3) {
-                level.setBlockAndUpdate(above, this.defaultBlockState());
+            } else if (age == 15 && height < level.paperConfig().maxGrowthHeight.cactus) { // Paper - Configurable cactus/bamboo/reed growth height
+                // Paper start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, above, this.defaultBlockState(), Block.UPDATE_ALL)) {
+                    return;
+                }
+                // Paper end
                 BlockState aboveBlock = state.setValue(AGE, 0);
                 level.setBlock(pos, aboveBlock, Block.UPDATE_NONE);
                 level.neighborChanged(aboveBlock, above, this, null, false);
@@ -131,7 +_,8 @@
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
-        entity.hurt(level.damageSources().cactus(), 1.0F);
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
+        entity.hurt(level.damageSources().cactus().eventBlockDamager(level, pos), 1.0F); // CraftBukkit
     }
 
     @Override
