--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -87,6 +_,7 @@
             CopperGolem copperGolem = EntityType.COPPER_GOLEM.create(level, EntitySpawnReason.TRIGGERED);
             if (copperGolem != null) {
                 spawnGolemInWorld(level, copperGolemMatch, copperGolem, copperGolemMatch.getBlock(0, 0, 0).getPos());
+                if (!copperGolem.valid) return; // Paper - entityspawnevent - entity was not added to the world so prevent world mutation
                 this.replaceCopperBlockWithChest(level, copperGolemMatch);
                 copperGolem.spawn(this.getWeatherStateFromPattern(copperGolemMatch));
             }
@@ -105,9 +_,22 @@
     }
 
     private static void spawnGolemInWorld(final Level level, final BlockPattern.BlockPatternMatch match, final Entity golem, final BlockPos spawnPos) {
-        clearPatternBlocks(level, match);
+        // clearPatternBlocks(level, patternMatch); // Paper - moved down
         golem.snapTo(spawnPos.getX() + 0.5, spawnPos.getY() + 0.05, spawnPos.getZ() + 0.5, 0.0F, 0.0F);
-        level.addFreshEntity(golem);
+        // Paper start
+        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason;
+        if (golem.getType() == net.minecraft.world.entity.EntityType.SNOW_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN;
+        } else if (golem.getType() == net.minecraft.world.entity.EntityType.COPPER_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM;
+        } else {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM;
+        }
+        if (!level.addFreshEntity(golem, spawnReason)) {
+            return;
+        }
+        // Paper end
+        clearPatternBlocks(level, match); // Paper - from above
 
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
             CriteriaTriggers.SUMMONED_ENTITY.trigger(player, golem);
