From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/ChangeOverTimeBlock.java b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
index 3aa5be8b7a5bd55ffd92a4af4b6cec2b98c699ee..4f98ae8bd4bfb681883132eddb57cbc5703d7d9e 100644
--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -16,7 +16,7 @@ public interface ChangeOverTimeBlock<T extends Enum<T>> {
     default void changeOverTime(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         float f = 0.05688889F;
         if (random.nextFloat() < 0.05688889F) {
-            this.getNextState(state, level, pos, random).ifPresent(blockState -> level.setBlockAndUpdate(pos, blockState));
+            this.getNextState(state, level, pos, random).ifPresent(blockState -> org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, blockState)); // CraftBukkit
         }
     }
 
