--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -20,7 +20,7 @@
 
         if (random.nextFloat() < 0.05688889F) {
             this.getNextState(state, world, pos, random).ifPresent((iblockdata1) -> {
-                world.setBlockAndUpdate(pos, iblockdata1);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(world, pos, iblockdata1); // CraftBukkit
             });
         }
 
