From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/ChorusFlowerBlock.java b/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 5bcd96cda9de0159041d56ac19437a00a534a631..3af8f150474a8a461c5829850172196b7b9e7bfd 100644
--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -96,8 +96,10 @@ public class ChorusFlowerBlock extends Block {
                 }
 
                 if (flag && allNeighborsEmpty(level, blockPos, null) && level.isEmptyBlock(pos.above(2))) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, ageValue), 2)) { // CraftBukkit - add event
                     level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
                     this.placeGrownFlower(level, blockPos, ageValue);
+                    } // CraftBukkit
                 } else if (ageValue < 4) {
                     int i = random.nextInt(4);
                     if (flag1) {
@@ -112,18 +114,28 @@ public class ChorusFlowerBlock extends Block {
                         if (level.isEmptyBlock(blockPos1)
                             && level.isEmptyBlock(blockPos1.below())
                             && allNeighborsEmpty(level, blockPos1, randomDirection.getOpposite())) {
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos1, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, ageValue + 1), 2)) { // CraftBukkit - add event
                             this.placeGrownFlower(level, blockPos1, ageValue + 1);
                             flag2 = true;
+                            } // CraftBukkit
                         }
                     }
 
                     if (flag2) {
                         level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
                     } else {
+                        // CraftBukkit start - add event
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
                         this.placeDeadFlower(level, pos);
+                        }
+                        // CraftBukkit end
                     }
                 } else {
+                    // CraftBukkit start - add event
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
                     this.placeDeadFlower(level, pos);
+                    }
+                    // CraftBukkit end
                 }
             }
         }
@@ -261,6 +273,11 @@ public class ChorusFlowerBlock extends Block {
     protected void onProjectileHit(Level level, BlockState state, BlockHitResult hit, Projectile projectile) {
         BlockPos blockPos = hit.getBlockPos();
         if (level instanceof ServerLevel serverLevel && projectile.mayInteract(serverLevel, blockPos) && projectile.mayBreak(serverLevel)) {
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(projectile, blockPos, state.getFluidState().createLegacyBlock())) { // Paper - fix wrong block state
+                return;
+            }
+            // CraftBukkit end
             level.destroyBlock(blockPos, true, projectile);
         }
     }
