--- a/net/minecraft/world/level/block/CommandBlock.java
+++ b/net/minecraft/world/level/block/CommandBlock.java
@@ -69,8 +_,17 @@
         }
     }
 
-    private void setPoweredAndUpdate(final Level level, final BlockPos pos, final CommandBlockEntity commandBlock, final boolean isPowered) {
+    private void setPoweredAndUpdate(final Level level, final BlockPos pos, final CommandBlockEntity commandBlock, boolean isPowered) { // Paper - remove final from isPowered
         boolean wasPowered = commandBlock.isPowered();
+        // CraftBukkit start
+        org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
+        int old = wasPowered ? 15 : 0;
+        int current = isPowered ? 15 : 0;
+
+        org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(bukkitBlock, old, current);
+        level.getCraftServer().getPluginManager().callEvent(eventRedstone);
+        isPowered = eventRedstone.getNewCurrent() > 0;
+        // CraftBukkit end
         if (isPowered != wasPowered) {
             commandBlock.setPowered(isPowered);
             if (isPowered) {
@@ -129,7 +_,7 @@
         final BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult
     ) {
         BlockEntity blockEntity = level.getBlockEntity(pos);
-        if (blockEntity instanceof CommandBlockEntity && player.canUseGameMasterBlocks()) {
+        if (blockEntity instanceof CommandBlockEntity && (player.canUseGameMasterBlocks() || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
             player.openCommandBlock((CommandBlockEntity)blockEntity);
             return InteractionResult.SUCCESS;
         } else {
