--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -48,6 +_,11 @@
         boolean hasNeighborSignal = level.hasNeighborSignal(pos);
         if (hasNeighborSignal != state.getValue(POWERED)) {
             BlockState blockState = state;
+            // Paper start - Call BlockRedstoneEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, hasNeighborSignal)) {
+                return;
+            }
+            // Paper end - Call BlockRedstoneEvent
             if (!state.getValue(POWERED)) {
                 blockState = state.cycle(LIT);
                 level.playSound(null, pos, blockState.getValue(LIT) ? SoundEvents.COPPER_BULB_TURN_ON : SoundEvents.COPPER_BULB_TURN_OFF, SoundSource.BLOCKS);
