--- a/net/minecraft/world/level/block/DecoratedPotBlock.java
+++ b/net/minecraft/world/level/block/DecoratedPotBlock.java
@@ -217,6 +_,11 @@
     protected void onProjectileHit(final Level level, final BlockState state, final BlockHitResult blockHit, final Projectile projectile) {
         BlockPos pos = blockHit.getBlockPos();
         if (level instanceof ServerLevel serverLevel && projectile.mayInteract(serverLevel, pos) && projectile.mayBreak(serverLevel)) {
+            // CraftBukkit start - call EntityChangeBlockEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(projectile, pos, this.getFluidState(state).createLegacyBlock())) {
+                return;
+            }
+            // CraftBukkit end
             level.setBlock(pos, state.setValue(CRACKED, true), Block.UPDATE_NONE);
             level.destroyBlock(pos, true, projectile);
         }
