--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -189,6 +_,11 @@
         if (!level.isClientSide()) {
             boolean hasPower = level.hasNeighborSignal(pos);
             if (state.getValue(POWERED) != hasPower) {
+                // Paper start - Call BlockRedstoneEvent
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, hasPower)) {
+                    return;
+                }
+                // Paper end - Call BlockRedstoneEvent
                 level.setBlock(pos, state.setValue(POWERED, hasPower).setValue(OPEN, hasPower), Block.UPDATE_CLIENTS);
                 if (state.getValue(OPEN) != hasPower) {
                     level.playSound(
