--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -176,14 +_,14 @@
                     level.getChunk(spreadPos.pos()).markPosForPostprocessing(spreadPos.pos());
                 }
 
-                return level.setBlock(spreadPos.pos(), spreadState, Block.UPDATE_CLIENTS);
+                return org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, spreadPos.source(), spreadPos.pos(), spreadState, Block.UPDATE_CLIENTS, true); // CraftBukkit
             } else {
                 return false;
             }
         }
     }
 
-    public record SpreadPos(BlockPos pos, Direction face) {
+    public record SpreadPos(BlockPos pos, Direction face, BlockPos source) { // CraftBukkit
     }
 
     @FunctionalInterface
@@ -195,19 +_,19 @@
         SAME_POSITION {
             @Override
             public MultifaceSpreader.SpreadPos getSpreadPos(final BlockPos pos, final Direction spreadDirection, final Direction fromFace) {
-                return new MultifaceSpreader.SpreadPos(pos, spreadDirection);
+                return new MultifaceSpreader.SpreadPos(pos, spreadDirection, pos); // CraftBukkit
             }
         },
         SAME_PLANE {
             @Override
             public MultifaceSpreader.SpreadPos getSpreadPos(final BlockPos pos, final Direction spreadDirection, final Direction fromFace) {
-                return new MultifaceSpreader.SpreadPos(pos.relative(spreadDirection), fromFace);
+                return new MultifaceSpreader.SpreadPos(pos.relative(spreadDirection), fromFace, pos); // CraftBukkit
             }
         },
         WRAP_AROUND {
             @Override
             public MultifaceSpreader.SpreadPos getSpreadPos(final BlockPos pos, final Direction spreadDirection, final Direction fromFace) {
-                return new MultifaceSpreader.SpreadPos(pos.relative(spreadDirection).relative(fromFace), spreadDirection.getOpposite());
+                return new MultifaceSpreader.SpreadPos(pos.relative(spreadDirection).relative(fromFace), spreadDirection.getOpposite(), pos); // CraftBukkit
             }
         };
 
