--- a/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -125,6 +_,7 @@
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
             serverLevel.destroyBlock(pos, true, entity);
         }
@@ -147,7 +_,7 @@
     @Override
     public void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         float growthSpeed = CropBlock.getGrowthSpeed(this, level, pos);
-        boolean shouldProgressGrowth = random.nextInt((int)(25.0F / growthSpeed) + 1) == 0;
+        boolean shouldProgressGrowth = random.nextFloat() < (level.spigotConfig.pitcherPlantModifier / (100.0F * (Math.floor(25.0F / growthSpeed) + 1))); // Paper - Fix Spigot growth modifiers
         if (shouldProgressGrowth) {
             this.grow(level, state, pos, 1);
         }
@@ -157,7 +_,7 @@
         int updatedAge = Math.min(lowerState.getValue(AGE) + increase, 4);
         if (this.canGrow(level, lowerPos, lowerState, updatedAge)) {
             BlockState newLowerState = lowerState.setValue(AGE, updatedAge);
-            level.setBlock(lowerPos, newLowerState, Block.UPDATE_CLIENTS);
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, blockState, Block.UPDATE_CLIENTS)) return; // Paper
             if (isDouble(updatedAge)) {
                 level.setBlock(lowerPos.above(), newLowerState.setValue(HALF, DoubleBlockHalf.UPPER), Block.UPDATE_ALL);
             }
