--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -117,7 +_,7 @@
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         int distance = getDistance(level, pos);
         BlockState blockState = state.setValue(DISTANCE, distance).setValue(BOTTOM, this.isBottom(level, pos, distance));
-        if (blockState.getValue(DISTANCE) == 7) {
+        if (blockState.getValue(DISTANCE) == 7 && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, blockState.getFluidState().createLegacyBlock()).isCancelled()) { // CraftBukkit - BlockFadeEvent // Paper - fix wrong block state
             if (state.getValue(DISTANCE) == 7) {
                 FallingBlockEntity.fall(level, pos, blockState);
             } else {
