--- a/net/minecraft/world/level/block/SculkShriekerBlock.java
+++ b/net/minecraft/world/level/block/SculkShriekerBlock.java
@@ -60,6 +_,7 @@
         if (level instanceof ServerLevel serverLevel) {
             ServerPlayer player = SculkShriekerBlockEntity.tryGetPlayer(entity);
             if (player != null) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null).isCancelled()) return; // CraftBukkit
                 serverLevel.getBlockEntity(pos, BlockEntityType.SCULK_SHRIEKER).ifPresent(shrieker -> shrieker.tryShriek(serverLevel, player));
             }
         }
@@ -126,9 +_,16 @@
     @Override
     protected void spawnAfterBreak(final BlockState state, final ServerLevel level, final BlockPos pos, final ItemStack tool, final boolean dropExperience) {
         super.spawnAfterBreak(state, level, pos, tool, dropExperience);
+        // CraftBukkit start - Delegate to getExpDrop
+    }
+
+    @Override
+    public int getExpDrop(final BlockState state, final ServerLevel level, final BlockPos pos, final ItemStack tool, final boolean dropExperience) {
         if (dropExperience) {
-            this.tryDropExperience(level, pos, tool, ConstantInt.of(5));
+            return this.tryDropExperience(level, pos, tool, ConstantInt.of(5));
         }
+        return 0;
+        // CraftBukkit end
     }
 
     @Override
