--- a/net/minecraft/world/level/block/SculkBlock.java
+++ b/net/minecraft/world/level/block/SculkBlock.java
@@ -42,8 +_,9 @@
                 if (random.nextInt(xpPerGrowthSpawn) < charge) {
                     BlockPos growthPlacement = chargePos.above();
                     BlockState growthState = this.getRandomGrowthState(level, growthPlacement, random, spreader.isWorldGeneration());
-                    level.setBlock(growthPlacement, growthState, Block.UPDATE_ALL);
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, originPos, growthPlacement, growthState, Block.UPDATE_ALL)) { // CraftBukkit - Call BlockSpreadEvent
                     level.playSound(null, chargePos, growthState.getSoundType().getPlaceSound(), SoundSource.BLOCKS, 1.0F, 1.0F);
+                    } // CraftBukkit - Call BlockSpreadEvent
                 }
 
                 return Math.max(0, charge - xpPerGrowthSpawn);
@@ -84,6 +_,7 @@
             int growthCount = 0;
 
             for (BlockPos blockPos : BlockPos.betweenClosed(pos.offset(-4, 0, -4), pos.offset(4, 2, 4))) {
+                if (level instanceof net.minecraft.server.level.WorldGenRegion worldGenRegion && !worldGenRegion.isWithinWriteZone(blockPos)) continue; // Paper - Fix MC-307918
                 BlockState state = level.getBlockState(blockPos);
                 if (state.is(Blocks.SCULK_SENSOR) || state.is(Blocks.SCULK_SHRIEKER)) {
                     growthCount++;
