--- a/net/minecraft/world/level/block/SculkBlock.java
+++ b/net/minecraft/world/level/block/SculkBlock.java
@@ -42,8 +_,9 @@
                 if (random.nextInt(xpPerGrowthSpawn) < charge) {
                     BlockPos growthPlacement = chargePos.above();
                     BlockState growthState = this.getRandomGrowthState(level, growthPlacement, random, spreader.isWorldGeneration());
-                    level.setBlock(growthPlacement, growthState, Block.UPDATE_ALL);
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, originPos, growthPlacement, growthState, Block.UPDATE_ALL)) { // CraftBukkit - Call BlockSpreadEvent
                     level.playSound(null, chargePos, growthState.getSoundType().getPlaceSound(), SoundSource.BLOCKS, 1.0F, 1.0F);
+                    } // CraftBukkit - Call BlockSpreadEvent
                 }
 
                 return Math.max(0, charge - xpPerGrowthSpawn);
