--- a/net/minecraft/world/level/block/WeatheringCopperGolemStatueBlock.java
+++ b/net/minecraft/world/level/block/WeatheringCopperGolemStatueBlock.java
@@ -58,14 +_,18 @@
                     return InteractionResult.PASS;
                 }
 
-                this.updatePose(level, state, pos, player);
-                return InteractionResult.SUCCESS;
+                return this.updatePose(level, state, pos, player); // Paper - call EntityChangeBlockEvent
             }
 
             if (this.getAge().equals(WeatheringCopper.WeatherState.UNAFFECTED)) {
                 CopperGolem copperGolem = copperGolemStatueBlockEntity.removeStatue(state);
-                stack.hurtAndBreak(1, player, hand.asEquipmentSlot());
                 if (copperGolem != null) {
+                    // Paper start - call EntityChangeBlockEvent
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, Blocks.AIR.defaultBlockState())) {
+                        return InteractionResult.PASS;
+                    }
+                    stack.hurtAndBreak(1, player, hand.asEquipmentSlot()); // Paper - moved after event
+                    // Paper end - call EntityChangeBlockEvent
                     level.addFreshEntity(copperGolem);
                     level.removeBlock(pos, false);
                     return InteractionResult.SUCCESS;
