--- a/net/minecraft/world/entity/decoration/Mannequin.java
+++ b/net/minecraft/world/entity/decoration/Mannequin.java
@@ -33,7 +_,7 @@
         Mannequin.class, EntityDataSerializers.OPTIONAL_COMPONENT
     );
     public static final byte ALL_LAYERS = (byte)Arrays.stream(PlayerModelPart.values()).mapToInt(PlayerModelPart::getMask).reduce(0, (i, i1) -> i | i1);
-    private static final Set<Pose> VALID_POSES = Set.of(Pose.STANDING, Pose.CROUCHING, Pose.SWIMMING, Pose.FALL_FLYING, Pose.SLEEPING);
+    public static final Set<Pose> VALID_POSES = Set.of(Pose.STANDING, Pose.CROUCHING, Pose.SWIMMING, Pose.FALL_FLYING, Pose.SLEEPING); // Paper - public
     public static final Codec<Pose> POSE_CODEC = Pose.CODEC
         .validate(pose -> VALID_POSES.contains(pose) ? DataResult.success(pose) : DataResult.error(() -> "Invalid pose: " + pose.getSerializedName()));
     private static final Codec<Byte> LAYERS_CODEC = PlayerModelPart.CODEC
