--- a/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
@@ -2,7 +2,6 @@
 
 import com.google.common.collect.Lists;
 import java.util.List;
-import net.minecraft.Util;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
@@ -16,8 +15,7 @@
     private static final int MAX_PAYLOAD_SIZE = 32767;
     public static final StreamCodec<FriendlyByteBuf, ServerboundCustomPayloadPacket> STREAM_CODEC = CustomPacketPayload.codec((minecraftkey) -> {
         return DiscardedPayload.codec(minecraftkey, 32767);
-    }, (List) Util.make(Lists.newArrayList(new CustomPacketPayload.TypeAndCodec[]{new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC)}), (arraylist) -> {
-    })).map(ServerboundCustomPayloadPacket::new, ServerboundCustomPayloadPacket::payload);
+    }, java.util.Collections.emptyList()).map(ServerboundCustomPayloadPacket::new, ServerboundCustomPayloadPacket::payload); // CraftBukkit - treat all packets the same
 
     @Override
     public PacketType<ServerboundCustomPayloadPacket> type() {
