--- a/net/minecraft/world/item/crafting/BrewingRecipe.java
+++ b/net/minecraft/world/item/crafting/BrewingRecipe.java
@@ -102,4 +_,16 @@
     public interface Factory<T extends BrewingRecipe> {
         T create(final Recipe.CommonInfo commonInfo, final PotionIngredient input, final PotionIngredient reagent, final ItemStackTemplate output);
     }
+
+    // Paper start - Recipe API
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe(final org.bukkit.NamespacedKey id) {
+        return new org.bukkit.inventory.BrewingRecipe(
+            id,
+            org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(this.output),
+            org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.input.ingredient()),
+            org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.reagent.ingredient())
+        );
+    }
+    // Paper end - Recipe API
 }
