From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/crafting/SmokingRecipe.java b/net/minecraft/world/item/crafting/SmokingRecipe.java
index 8d849399052f98186d9b924e92370f38015558ac..fad1338a1832763e5ecc512ed8c8220ab3b7eb35 100644
--- a/net/minecraft/world/item/crafting/SmokingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmokingRecipe.java
@@ -28,4 +28,17 @@ public class SmokingRecipe extends AbstractCookingRecipe {
     public RecipeBookCategory recipeBookCategory() {
         return RecipeBookCategories.SMOKER_FOOD;
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+        org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.result());
+
+        org.bukkit.craftbukkit.inventory.CraftSmokingRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftSmokingRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toBukkit(this.input()), this.experience(), this.cookingTime());
+        recipe.setGroup(this.group());
+        recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+
+        return recipe;
+    }
+    // CraftBukkit end
 }
