--- a/net/minecraft/world/entity/monster/warden/AngerManagement.java
+++ b/net/minecraft/world/entity/monster/warden/AngerManagement.java
@@ -142,7 +_,7 @@
 
     public int increaseAnger(final Entity entity, final int increment) {
         boolean newSuspect = !this.angerBySuspect.containsKey(entity);
-        int currentAnger = this.angerBySuspect.computeInt(entity, (k, anger) -> Math.min(150, (anger == null ? 0 : anger) + increment));
+        int currentAnger = this.angerBySuspect.computeInt(entity, (k, anger) -> Math.min(150, (anger == null ? 0 : anger) + increment)); // Paper - diff on change (Warden#increaseAngerAt WardenAngerChangeEvent)
         if (newSuspect) {
             int serializedAnger = this.angerByUuid.removeInt(entity.getUUID());
             currentAnger += serializedAnger;
