--- a/net/minecraft/world/item/consume_effects/ApplyStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/ApplyStatusEffectsConsumeEffect.java
@@ -12,6 +12,9 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
+// CraftBukkit start
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+// CraftBukkit end
 
 public record ApplyStatusEffectsConsumeEffect(List<MobEffectInstance> effects, float probability) implements ConsumeEffect {
 
@@ -38,8 +41,8 @@
     }
 
     @Override
-    public boolean apply(Level world, ItemStack stack, LivingEntity user) {
-        if (user.getRandom().nextFloat() >= this.probability) {
+    public boolean apply(Level world, ItemStack itemstack, LivingEntity entityliving, EntityPotionEffectEvent.Cause cause) { // CraftBukkit
+        if (entityliving.getRandom().nextFloat() >= this.probability) {
             return false;
         } else {
             boolean flag = false;
@@ -48,7 +51,7 @@
             while (iterator.hasNext()) {
                 MobEffectInstance mobeffect = (MobEffectInstance) iterator.next();
 
-                if (user.addEffect(new MobEffectInstance(mobeffect))) {
+                if (entityliving.addEffect(new MobEffectInstance(mobeffect), cause)) { // CraftBukkit
                     flag = true;
                 }
             }
