From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/core/HolderLookup.java b/net/minecraft/core/HolderLookup.java
index 5812796dd0bbc66fd326e4afe831a9b2733d9224..71d3aa91c5202f782263016062daf01c109afda2 100644
--- a/net/minecraft/core/HolderLookup.java
+++ b/net/minecraft/core/HolderLookup.java
@@ -68,6 +68,9 @@ public interface HolderLookup<T> extends HolderGetter<T> {
     }
 
     public interface RegistryLookup<T> extends HolderLookup<T>, HolderOwner<T> {
+
+        Optional<T> getValueForCopying(ResourceKey<T> resourceKey); // Paper - add method to get the value for pre-filling builders in the reg mod API
+
         ResourceKey<? extends Registry<? extends T>> key();
 
         Lifecycle registryLifecycle();
@@ -80,6 +83,13 @@ public interface HolderLookup<T> extends HolderGetter<T> {
 
         default HolderLookup.RegistryLookup<T> filterElements(final Predicate<T> predicate) {
             return new HolderLookup.RegistryLookup.Delegate<T>() {
+                // Paper start - add getValueForCopying
+                @Override
+                public Optional<T> getValueForCopying(final ResourceKey<T> resourceKey) {
+                    return this.parent().getValueForCopying(resourceKey).filter(predicate);
+                }
+                // Paper end - add getValueForCopying
+
                 @Override
                 public HolderLookup.RegistryLookup<T> parent() {
                     return RegistryLookup.this;
@@ -100,6 +110,13 @@ public interface HolderLookup<T> extends HolderGetter<T> {
         public interface Delegate<T> extends HolderLookup.RegistryLookup<T> {
             HolderLookup.RegistryLookup<T> parent();
 
+            // Paper start - add getValueForCopying
+            @Override
+            default Optional<T> getValueForCopying(ResourceKey<T> resourceKey) {
+                return this.parent().getValueForCopying(resourceKey);
+            }
+            // Paper end - add getValueForCopying
+
             @Override
             default ResourceKey<? extends Registry<? extends T>> key() {
                 return this.parent().key();
