--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -115,7 +_,7 @@
         Entity effectSource = this.getEffectSource();
         PotionContents potionContents = this.getPotionContents();
         float potionDurationScale = this.getPotionDurationScale();
-        potionContents.forEachEffect(mobEffectInstance -> target.addEffect(mobEffectInstance, effectSource), potionDurationScale);
+        potionContents.forEachEffect(mobEffectInstance -> target.addEffect(mobEffectInstance, effectSource, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW), potionDurationScale); // CraftBukkit
     }
 
     @Override
