From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/projectile/EvokerFangs.java b/net/minecraft/world/entity/projectile/EvokerFangs.java
index 76d2d591f83eac51e84edd6393a782f7e5f00507..d97cfa98cbe11564f699ae4bfbb2f20f00b43d6b 100644
--- a/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -107,7 +107,7 @@ public class EvokerFangs extends Entity implements TraceableEntity {
             }
 
             if (--this.lifeTicks < 0) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
@@ -116,7 +116,7 @@ public class EvokerFangs extends Entity implements TraceableEntity {
         LivingEntity owner = this.getOwner();
         if (target.isAlive() && !target.isInvulnerable() && target != owner) {
             if (owner == null) {
-                target.hurt(this.damageSources().magic(), 6.0F);
+                target.hurt(this.damageSources().magic().eventEntityDamager(this), 6.0F); // CraftBukkit // Paper - fix DamageSource API
             } else {
                 if (owner.isAlliedTo(target)) {
                     return;
