--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -31,6 +_,17 @@
 
     BlockState getBlockState(final BlockPos pos);
 
+    // Paper start - if loaded util
+    @Nullable BlockState getBlockStateIfLoaded(BlockPos pos);
+
+    default net.minecraft.world.level.block.@Nullable Block getBlockIfLoaded(BlockPos pos) {
+        BlockState type = this.getBlockStateIfLoaded(pos);
+        return type == null ? null : type.getBlock();
+    }
+
+    @Nullable FluidState getFluidIfLoaded(BlockPos pos);
+    // Paper end
+
     FluidState getFluidState(BlockPos pos);
 
     default int getLightEmission(final BlockPos pos) {
@@ -63,18 +_,42 @@
     }
 
     default BlockHitResult clip(final ClipContext c) {
+        // Paper start - Add predicate for blocks when raytracing
+        return clip(c, (java.util.function.Predicate<org.bukkit.block.Block>) null);
+    }
+
+    default BlockHitResult clip(ClipContext traverseContext, BlockPos traversePos) {
+        return clip(traverseContext, traversePos, null);
+    }
+
+    default BlockHitResult clip(ClipContext c, BlockPos pos, java.util.function.@Nullable Predicate<? super org.bukkit.block.Block> canCollide) {
+        // Prevent raytrace from loading chunks
+        BlockState blockState = this.getBlockStateIfLoaded(pos);
+        if (blockState == null) {
+            // copied the last function parameter (listed below)
+            Vec3 vec3d = c.getFrom().subtract(c.getTo());
+
+            return BlockHitResult.miss(c.getTo(), Direction.getApproximateNearest(vec3d.x, vec3d.y, vec3d.z), BlockPos.containing(c.getTo()));
+        }
+        // Paper end - Prevent raytrace from loading chunks
+        if (blockState.isAir() || (canCollide != null && this instanceof LevelAccessor levelAccessor && !canCollide.test(org.bukkit.craftbukkit.block.CraftBlock.at(levelAccessor, pos)))) return null; // Paper - Perf: optimise air cases & check canCollide predicate
+        FluidState fluidState = blockState.getFluidState(); // Paper - Perf: don't need to go to world state again
+        Vec3 from = c.getFrom();
+        Vec3 to = c.getTo();
+        VoxelShape blockShape = c.getBlockShape(blockState, this, pos);
+        BlockHitResult blockResult = this.clipWithInteractionOverride(from, to, pos, blockShape, blockState);
+        VoxelShape fluidShape = c.getFluidShape(fluidState, this, pos);
+        BlockHitResult liquidResult = fluidShape.clip(from, to, pos);
+        double blockDistanceSquared = blockResult == null ? Double.MAX_VALUE : c.getFrom().distanceToSqr(blockResult.getLocation());
+        double liquidDistanceSquared = liquidResult == null ? Double.MAX_VALUE : c.getFrom().distanceToSqr(liquidResult.getLocation());
+        return blockDistanceSquared <= liquidDistanceSquared ? blockResult : liquidResult;
+    }
+
+
+    default BlockHitResult clip(ClipContext c, java.util.function.@Nullable Predicate<? super org.bukkit.block.Block> canCollide) {
+        // Paper end - Add predicate for blocks when raytracing
         return traverseBlocks(c.getFrom(), c.getTo(), c, (context, pos) -> {
-            BlockState blockState = this.getBlockState(pos);
-            FluidState fluidState = this.getFluidState(pos);
-            Vec3 from = context.getFrom();
-            Vec3 to = context.getTo();
-            VoxelShape blockShape = context.getBlockShape(blockState, this, pos);
-            BlockHitResult blockResult = this.clipWithInteractionOverride(from, to, pos, blockShape, blockState);
-            VoxelShape fluidShape = context.getFluidShape(fluidState, this, pos);
-            BlockHitResult liquidResult = fluidShape.clip(from, to, pos);
-            double blockDistanceSquared = blockResult == null ? Double.MAX_VALUE : context.getFrom().distanceToSqr(blockResult.getLocation());
-            double liquidDistanceSquared = liquidResult == null ? Double.MAX_VALUE : context.getFrom().distanceToSqr(liquidResult.getLocation());
-            return blockDistanceSquared <= liquidDistanceSquared ? blockResult : liquidResult;
+            return this.clip(context, pos, canCollide); // CraftBukkit - moved into separate method // Paper - Add predicate for blocks when raytracing
         }, context -> {
             Vec3 delta = context.getFrom().subtract(context.getTo());
             return BlockHitResult.miss(context.getTo(), Direction.getApproximateNearest(delta.x, delta.y, delta.z), BlockPos.containing(context.getTo()));
