--- a/net/minecraft/world/level/EmptyBlockAndTintGetter.java
+++ b/net/minecraft/world/level/EmptyBlockAndTintGetter.java
@@ -39,6 +_,18 @@
         return Blocks.AIR.defaultBlockState();
     }
 
+    // Paper start
+    @Override
+    public @Nullable BlockState getBlockStateIfLoaded(final BlockPos pos) {
+        return null;
+    }
+
+    @Override
+    public @Nullable FluidState getFluidIfLoaded(final BlockPos pos) {
+        return null;
+    }
+    // Paper end
+
     @Override
     public FluidState getFluidState(BlockPos pos) {
         return Fluids.EMPTY.defaultFluidState();
