From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Tue, 10 Feb 2026 20:21:21 +0100
Subject: [PATCH] fixup! Eigencraft redstone implementation


diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index ff747a1ecdf3c888bca0d69de4f85dcd810b6139..e7d1a5fd79680886af16572f19c5896bd8713f17 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -1,5 +1,7 @@
 package io.papermc.paper.redstone;
 
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ThreadLocalRandom;
@@ -12,11 +14,9 @@ import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.RedStoneWireBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.block.CraftBlockRedstoneEvent;
 import org.bukkit.event.block.BlockRedstoneEvent;
 
-import com.google.common.collect.Lists;
-import com.google.common.collect.Maps;
-
 public final class RedstoneWireTurbo {
     /*
      * This is Helper class for BlockRedstoneWire.  It implements a minimally-invasive
@@ -892,7 +892,7 @@ public final class RedstoneWireTurbo {
         // egg82's amendment
         // Adding Bukkit's BlockRedstoneEvent - er.. event.
         if (i != j) {
-            BlockRedstoneEvent event = new BlockRedstoneEvent(CraftBlock.at(worldIn, upd.self), i, j);
+            BlockRedstoneEvent event = new CraftBlockRedstoneEvent(CraftBlock.at(worldIn, upd.self), i, j);
             worldIn.getCraftServer().getPluginManager().callEvent(event);
             j = event.getNewCurrent();
         }
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 6d98bc37d88459d1e0a171b52bbff5810d775c38..1b601a6339c0d15cb3f5ca9cb5c9fefd2636c0e6 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -300,7 +300,7 @@ public class RedStoneWireBlock extends Block {
         int oldPower = state.getValue(POWER);
         int newPower = ((DefaultRedstoneWireEvaluator) evaluator).calculateTargetStrength(level, pos);
         if (oldPower != newPower) {
-            org.bukkit.event.block.BlockRedstoneEvent event = new org.bukkit.event.block.BlockRedstoneEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), oldPower, newPower);
+            org.bukkit.event.block.BlockRedstoneEvent event = new org.bukkit.craftbukkit.event.block.CraftBlockRedstoneEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), oldPower, newPower);
             level.getCraftServer().getPluginManager().callEvent(event);
 
             newPower = event.getNewCurrent();
