From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bl4stySKID <gamwtimana69@gmail.com>
Date: Sat, 30 Aug 2025 02:50:11 +0300
Subject: [PATCH] Fix MC-118740 Right click resets attack cooldown


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 8f94c963f2c59668d72d162f46f7505d6a6b06a5..18aa4666396fe6565e565cfa3fe44650f4371bde 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2436,7 +2436,12 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     @Override
     public void swing(InteractionHand hand) {
         super.swing(hand);
-        this.resetAttackStrengthTicker();
+
+        // Paper start - Fix MC-118740
+        if (hand == InteractionHand.MAIN_HAND) {
+            this.resetAttackStrengthTicker();
+        }
+        // Paper end
     }
 
     public boolean isChangingDimension() {
