From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Mon, 24 Aug 2026 18:19:23 +0200
Subject: [PATCH] fixup! Call AsyncServerDataFixerRemoveBlockEntityEvent from
 DC


diff --git a/ca/spottedleaf/dataconverter/minecraft/converters/tileentity/ConverterAbstractTileEntityRemove.java b/ca/spottedleaf/dataconverter/minecraft/converters/tileentity/ConverterAbstractTileEntityRemove.java
index e70408061c8a495af319c810ee93d51a2644e465..b3b8e8e98c17da61f7ff615f2f2b41c227ce2e9d 100644
--- a/ca/spottedleaf/dataconverter/minecraft/converters/tileentity/ConverterAbstractTileEntityRemove.java
+++ b/ca/spottedleaf/dataconverter/minecraft/converters/tileentity/ConverterAbstractTileEntityRemove.java
@@ -91,7 +91,7 @@ public final class ConverterAbstractTileEntityRemove {
                                 if (blockEntity.getMap("PublicBukkitValues") instanceof final ca.spottedleaf.dataconverter.types.nbt.NBTMapType nbtMapType) {
                                     container.putAll(nbtMapType.getTag());
                                 }
-                                new io.papermc.paper.event.server.AsyncServerDataFixerRemoveBlockEntityEvent(
+                                new io.papermc.paper.event.server.PaperAsyncServerDataFixerRemoveBlockEntityEvent(
                                     worldKey,
                                     net.kyori.adventure.key.Key.key(type),
                                     io.papermc.paper.math.Position.block(x.intValue(), y.intValue(), z.intValue()),
