From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AGC Developer <developer@agc.com>
Date: Thu, 11 Jun 2026 22:37:31 +0900
Subject: [PATCH] Meteus empty entity tracker broadcast fast path


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index a4c3d48294a046164b79216361947b7977feb706..78068db1c3068f2019cac11468fb2905a8af37ab 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1372,6 +1372,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayers(final Packet<? super ClientGamePacketListener> packet) {
+            // Meteus start - empty entity tracker broadcast fast path
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().meteus.performance.emptyEntityTrackerBroadcastFastPath && this.seenBy.isEmpty()) {
+                return;
+            }
+            // Meteus end - empty entity tracker broadcast fast path
             for (ServerPlayerConnection connection : this.seenBy) {
                 connection.send(packet);
             }
@@ -1387,6 +1392,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayersFiltered(final Packet<? super ClientGamePacketListener> packet, final Predicate<ServerPlayer> targetPredicate) {
+            // Meteus start - empty entity tracker broadcast fast path
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().meteus.performance.emptyEntityTrackerBroadcastFastPath && this.seenBy.isEmpty()) {
+                return;
+            }
+            // Meteus end - empty entity tracker broadcast fast path
             for (ServerPlayerConnection connection : this.seenBy) {
                 if (targetPredicate.test(connection.getPlayer())) {
                     connection.send(packet);
@@ -1395,6 +1405,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         }
 
         public void broadcastRemoved() {
+            // Meteus start - empty entity tracker broadcast fast path
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().meteus.performance.emptyEntityTrackerBroadcastFastPath && this.seenBy.isEmpty()) {
+                return;
+            }
+            // Meteus end - empty entity tracker broadcast fast path
             for (ServerPlayerConnection connection : this.seenBy) {
                 this.serverEntity.removePairing(connection.getPlayer());
             }
