--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -106,9 +_,14 @@
     }
 
     private EntitySelector parse(StringReader reader, boolean allowSelectors) throws CommandSyntaxException {
+        // CraftBukkit start
+        return this.parse(reader, allowSelectors, false);
+    }
+    public EntitySelector parse(StringReader reader, boolean allowSelectors, boolean overridePermissions) throws CommandSyntaxException {
+        // CraftBukkit end
         int i = 0;
         EntitySelectorParser entitySelectorParser = new EntitySelectorParser(reader, allowSelectors);
-        EntitySelector entitySelector = entitySelectorParser.parse();
+        EntitySelector entitySelector = entitySelectorParser.parse(overridePermissions); // CraftBukkit
         if (entitySelector.getMaxResults() > 1 && this.single) {
             if (this.playersOnly) {
                 reader.setCursor(0);
@@ -130,9 +_,15 @@
         if (context.getSource() instanceof SharedSuggestionProvider sharedSuggestionProvider) {
             StringReader stringReader = new StringReader(builder.getInput());
             stringReader.setCursor(builder.getStart());
+            // Paper start - Fix EntityArgument permissions
+            final boolean permission = sharedSuggestionProvider instanceof CommandSourceStack stack
+                ? stack.bypassSelectorPermissions || stack.hasPermission(Permissions.COMMANDS_ENTITY_SELECTORS, "minecraft.command.selector")
+                // Only CommandSourceStack implements SharedSuggestionProvider. If *somehow* anything else ends up here, try to query its permission level, otherwise yield false.
+                : sharedSuggestionProvider.permissions().hasPermission(Permissions.COMMANDS_ENTITY_SELECTORS);
             EntitySelectorParser entitySelectorParser = new EntitySelectorParser(
-                stringReader, sharedSuggestionProvider.permissions().hasPermission(Permissions.COMMANDS_ENTITY_SELECTORS)
+                stringReader, permission
             );
+            // Paper end - Fix EntityArgument permissions
 
             try {
                 entitySelectorParser.parse();
