--- a/net/minecraft/commands/arguments/MessageArgument.java
+++ b/net/minecraft/commands/arguments/MessageArgument.java
@@ -41,6 +_,12 @@
 
     public static void resolveChatMessage(final CommandContext<CommandSourceStack> context, final String name, final Consumer<PlayerChatMessage> task) throws CommandSyntaxException {
         MessageArgument.Message message = context.getArgument(name, MessageArgument.Message.class);
+        // Paper start - brig message argument support
+        resolveChatMessage(message, context, name, task);
+    }
+
+    public static void resolveChatMessage(final MessageArgument.Message message, final CommandContext<CommandSourceStack> context, final String name, final Consumer<PlayerChatMessage> task) throws CommandSyntaxException {
+        // Paper end - brig message argument support
         CommandSourceStack sender = context.getSource();
         Component formatted = message.resolveComponent(sender);
         CommandSigningContext signingContext = sender.getSigningContext();
@@ -55,17 +_,21 @@
     private static void resolveSignedMessage(final Consumer<PlayerChatMessage> task, final CommandSourceStack sender, final PlayerChatMessage signedArgument) {
         MinecraftServer server = sender.getServer();
         CompletableFuture<FilteredText> filteredFuture = filterPlainText(sender, signedArgument);
-        Component decorated = server.getChatDecorator().decorate(sender.getPlayer(), signedArgument.decoratedContent());
-        sender.getChatMessageChainer().append(filteredFuture, filtered -> {
-            PlayerChatMessage filteredMessage = signedArgument.withUnsignedContent(decorated).filter(filtered.mask());
+        // Paper start - support asynchronous chat decoration
+        CompletableFuture<Component> decoratedFuture = server.getChatDecorator().decorate(sender.getPlayer(), sender, signedArgument.decoratedContent());
+        sender.getChatMessageChainer().append(CompletableFuture.allOf(filteredFuture, decoratedFuture), filtered -> {
+            PlayerChatMessage filteredMessage = signedArgument.withUnsignedContent(decoratedFuture.join()).filter(filteredFuture.join().mask());
+            // Paper end - support asynchronous chat decoration
             task.accept(filteredMessage);
         });
     }
 
     private static void resolveDisguisedMessage(final Consumer<PlayerChatMessage> task, final CommandSourceStack sender, final PlayerChatMessage argument) {
         ChatDecorator decorator = sender.getServer().getChatDecorator();
-        Component decorated = decorator.decorate(sender.getPlayer(), argument.decoratedContent());
-        task.accept(argument.withUnsignedContent(decorated));
+        // Paper start - support asynchronous chat decoration
+        CompletableFuture<Component> decoratedFuture = decorator.decorate(sender.getPlayer(), sender, argument.decoratedContent());
+        sender.getChatMessageChainer().append(decoratedFuture, result -> task.accept(argument.withUnsignedContent(result)));
+        // Paper end - support asynchronous chat decoration
     }
 
     private static CompletableFuture<FilteredText> filterPlainText(final CommandSourceStack sender, final PlayerChatMessage message) {
