From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java b/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java
index 6993dfdf55b68ecd82c614010e5f2c272a6c6d4a..66580946785d67e080ca898fe5379bd81491b970 100644
--- a/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java
+++ b/net/minecraft/world/entity/ai/goal/SitWhenOrderedToGoal.java
@@ -20,7 +20,7 @@ public class SitWhenOrderedToGoal extends Goal {
     @Override
     public boolean canUse() {
         if (!this.mob.isTame()) {
-            return false;
+            return this.mob.isOrderedToSit() && this.mob.getTarget() == null; // CraftBukkit - Allow sitting for wild animals
         } else if (this.mob.isInWaterOrBubble()) {
             return false;
         } else if (!this.mob.onGround()) {
