From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
index 3760164dda42be969a84cfe9418aea3c80c1c210..f5e93036b5ecb46e4f81aff27d93d27cda49e1b0 100644
--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -110,7 +110,9 @@ public class HarvestFarmland extends Behavior<Villager> {
                 Block block = blockState.getBlock();
                 Block blockBelow = level.getBlockState(this.aboveFarmlandPos.below()).getBlock();
                 if (block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockState)) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(body, this.aboveFarmlandPos, blockState.getFluidState().createLegacyBlock())) { // CraftBukkit // Paper - fix wrong block state
                     level.destroyBlock(this.aboveFarmlandPos, true, body);
+                    } // CraftBukkit
                 }
 
                 if (blockState.isAir() && blockBelow instanceof FarmlandBlock && body.hasFarmSeeds()) {
@@ -121,9 +123,11 @@ public class HarvestFarmland extends Behavior<Villager> {
                         boolean ok = false;
                         if (!itemStack.isEmpty() && itemStack.is(ItemTags.VILLAGER_PLANTABLE_SEEDS) && itemStack.getItem() instanceof BlockItem blockItem) {
                             BlockState place = blockItem.getBlock().defaultBlockState();
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(body, this.aboveFarmlandPos, place)) { // CraftBukkit
                             level.setBlockAndUpdate(this.aboveFarmlandPos, place);
                             level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(body, place));
                             ok = true;
+                            } // CraftBukkit
                         }
 
                         if (ok) {
