From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/RamTarget.java b/net/minecraft/world/entity/ai/behavior/RamTarget.java
index 4b44284f7b6a59692f03038f93d857a6263f1516..175dc2ce4dd934f22e569dbcb44b1c717020e381 100644
--- a/net/minecraft/world/entity/ai/behavior/RamTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/RamTarget.java
@@ -93,7 +93,7 @@ public class RamTarget extends Behavior<Goat> {
             DamageSource damageSource1 = level.damageSources().mobAttack(owner);
             float f3 = livingEntity.applyItemBlocking(level, damageSource1, f);
             float f4 = f3 > 0.0F ? 0.5F : 1.0F;
-            livingEntity.knockback(f4 * f2 * this.getKnockbackForce.applyAsDouble(owner), this.ramDirection.x(), this.ramDirection.z());
+            livingEntity.knockback(f4 * f2 * this.getKnockbackForce.applyAsDouble(owner), this.ramDirection.x(), this.ramDirection.z(), owner, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.ENTITY_ATTACK); // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
             this.finishRam(level, owner);
             level.playSound(null, owner, this.getImpactSound.apply(owner), SoundSource.NEUTRAL, 1.0F, 1.0F);
         } else if (this.hasRammedHornBreakingBlock(level, owner)) {
