From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java b/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
index ec9761d39419211917a65ec800db0240d093bab1..80c644f27cbe9916feffffd9e1bf61f7dba9604e 100644
--- a/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
+++ b/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
@@ -33,6 +33,12 @@ public class TryLaySpawnOnFluidNearLand {
                                     BlockPos spawnPos = relativePos.above();
                                     if (level.getBlockState(spawnPos).isAir()) {
                                         BlockState newState = spawnBlock.defaultBlockState();
+                                        // CraftBukkit start
+                                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(body, spawnPos, newState)) {
+                                            pregnant.erase();
+                                            return true;
+                                        }
+                                        // CraftBukkit end
                                         level.setBlock(spawnPos, newState, Block.UPDATE_ALL);
                                         level.gameEvent(GameEvent.BLOCK_PLACE, spawnPos, GameEvent.Context.of(body, newState));
                                         level.playSound(null, body, SoundEvents.FROG_LAY_SPAWN, SoundSource.BLOCKS, 1.0F, 1.0F);
