From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
index cbd5d1aabbe2a2741094fb80bb6de28d1fbfa122..37cf6c4cef5a8885fddd9f637cac024892b498a3 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -111,11 +111,17 @@ public class VillagerMakeLove extends Behavior<Villager> {
         if (child == null) {
             return Optional.empty();
         } else {
-            source.setAge(6000);
-            target.setAge(6000);
             child.setAge(-24000);
+            // Paper - Move age setting down
             child.snapTo(source.getX(), source.getY(), source.getZ(), 0.0F, 0.0F);
-            level.addFreshEntityWithPassengers(child);
+            // CraftBukkit start - call EntityBreedEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(child, source, target, null, null, 0).isCancelled()) {
+                return Optional.empty();
+            }
+            source.setAge(6000);
+            target.setAge(6000);
+            level.addFreshEntityWithPassengers(child, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);
+            // CraftBukkit end - call EntityBreedEvent
             level.broadcastEntityEvent(child, EntityEvent.LOVE_HEARTS);
             return Optional.of(child);
         }
