--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -320,7 +_,7 @@
         }
 
         this.makeSound(this.getDeathSound());
-        this.remove(Entity.RemovalReason.DISCARDED);
+        this.remove(Entity.RemovalReason.DISCARDED, null); // CraftBukkit - add Bukkit remove cause
     }
 
     public void creakingDeathEffects(final DamageSource source) {
@@ -448,9 +_,9 @@
     }
 
     @Override
-    public void knockback(final double power, final double xd, final double zd, final DamageSource source, final float damage) {
+    public void knockback(final double power, final double xd, final double zd, final DamageSource source, final float damage, @Nullable final Entity attacker, final io.papermc.paper.event.entity.EntityKnockbackEvent.Cause cause) { // Paper - knockback events
         if (this.canMove()) {
-            super.knockback(power, xd, zd, source, damage);
+            super.knockback(power, xd, zd, source, damage, attacker, cause); // Paper - knockback events
         }
     }
 
