--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -206,7 +206,7 @@
 
     @Override
     public BehaviorController<Creaking> getBrain() {
-        return super.getBrain();
+        return (BehaviorController<Creaking>) super.getBrain(); // CraftBukkit - decompile error
     }
 
     @Override
@@ -329,7 +329,7 @@
         }
 
         this.makeSound(this.getDeathSound());
-        this.remove(Entity.RemovalReason.DISCARDED);
+        this.remove(Entity.RemovalReason.DISCARDED, null); // CraftBukkit - add Bukkit remove cause
     }
 
     public void creakingDeathEffects(DamageSource damagesource) {
@@ -476,7 +476,7 @@
 
     @Override
     protected SoundEffect getHurtSound(DamageSource damagesource) {
-        return this.isHeartBound() ? SoundEffects.CREAKING_SWAY : super.getHurtSound(damagesource);
+        return SoundEffects.CREAKING_SWAY;
     }
 
     @Override
@@ -549,7 +549,7 @@
     }
 
     public void activate(EntityHuman entityhuman) {
-        this.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, (Object) entityhuman);
+        this.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, entityhuman); // CraftBukkit - decompile error
         this.gameEvent(GameEvent.ENTITY_ACTION);
         this.makeSound(SoundEffects.CREAKING_ACTIVATE);
         this.setIsActive(true);
