--- a/net/minecraft/world/entity/ConversionType.java
+++ b/net/minecraft/world/entity/ConversionType.java
@@ -21,7 +_,7 @@
 
                 for (Entity entity : newMob.getPassengers()) {
                     entity.stopRiding();
-                    entity.remove(Entity.RemovalReason.DISCARDED);
+                    entity.remove(Entity.RemovalReason.DISCARDED, org.bukkit.event.entity.EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
                 }
 
                 firstPassenger.startRiding(newMob);
