From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/Display.java b/net/minecraft/world/entity/Display.java
index eb564aab8efd110f74f77e57ba6d73a2b6ffe3f3..ff49fdcddb43ee479a377aa219eb18fcaefeffca 100644
--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -213,7 +213,7 @@ public abstract class Display extends Entity {
         if (tag.contains("transformation")) {
             Transformation.EXTENDED_CODEC
                 .decode(NbtOps.INSTANCE, tag.get("transformation"))
-                .resultOrPartial(Util.prefix("Display entity", LOGGER::error))
+                .result() // Paper - Hide text display error on spawn
                 .ifPresent(pair -> this.setTransformation(pair.getFirst()));
         }
 
