--- a/net/minecraft/world/entity/EntityFluidInteraction.java
+++ b/net/minecraft/world/entity/EntityFluidInteraction.java
@@ -77,6 +_,11 @@
                                             }
 
                                             tracker.accumulateCurrent(flow);
+                                            // CraftBukkit start - store last lava contact location
+                                            if (fluidState.is(net.minecraft.tags.FluidTags.LAVA)) {
+                                                entity.lastLavaContact = mutablePos.immutable();
+                                            }
+                                            // CraftBukkit end - store last lava contact location
                                         }
                                     }
                                 }
