--- a/net/minecraft/world/entity/EntityEquipment.java
+++ b/net/minecraft/world/entity/EntityEquipment.java
@@ -70,4 +_,11 @@
     public void clear() {
         this.items.replaceAll((equipmentSlot, itemStack) -> ItemStack.EMPTY);
     }
+
+    // Paper start - EntityDeathEvent
+    // Needed to not set ItemStack.EMPTY to not existent slot.
+    public boolean has(final EquipmentSlot slot) {
+        return this.items.containsKey(slot);
+    }
+    // Paper end - EntityDeathEvent
 }
