--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -34,6 +_,14 @@
         return true;
     }
 
+    // CraftBukkit start
+    public void boost(final int boostTime) {
+        this.boosting = true;
+        this.boostTime = 0;
+        this.entityData.set(this.boostTimeAccessor, boostTime);
+    }
+    // CraftBukkit end
+
     public void tickBoost() {
         if (this.boosting && this.boostTime++ > this.boostTimeTotal()) {
             this.boosting = false;
