From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ItemBasedSteering.java b/net/minecraft/world/entity/ItemBasedSteering.java
index 6ca4214f4aaca4429fa3d50e3f6f45df206459f5..7c0e3af0cb4f91ec3fed5d8707ff5a4f44c36e11 100644
--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -51,6 +51,14 @@ public class ItemBasedSteering {
         return this.entityData.get(this.boostTimeAccessor);
     }
 
+    // CraftBukkit add setBoostTicks(int)
+    public void setBoostTicks(int ticks) {
+        this.boosting = true;
+        this.boostTime = 0;
+        this.entityData.set(this.boostTimeAccessor, ticks);
+    }
+    // CraftBukkit end
+
     public void addAdditionalSaveData(CompoundTag nbt) {
         nbt.putBoolean("Saddle", this.hasSaddle());
     }
