--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -133,6 +_,8 @@
         boolean verify
     ) {
         super(
+            null, // Paper
+            null, // Paper
             serverThread,
             storageSource,
             packRepository,
@@ -148,8 +_,15 @@
 
     @Override
     public boolean initServer() {
-        this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {});
-        this.loadLevel();
+        // Paper start
+        this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {
+            @Override
+            public void loadAndSaveFiles() {
+                throw new UnsupportedOperationException("Should not be called in a GameTestServer");
+            }
+        });
+        this.loadLevel("blah");
+        // Paper end
         ServerLevel serverLevel = this.overworld();
         this.testBatches = this.evaluateTestsToRun(serverLevel);
         LOGGER.info("Started game test server");
@@ -338,6 +_,13 @@
     public boolean shouldInformAdmins() {
         return false;
     }
+
+    // Paper start
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(final net.minecraft.commands.CommandSourceStack wrapper) {
+        throw new UnsupportedOperationException("Not supported.");
+    }
+    // Paper end
 
     @Override
     public boolean isSingleplayerOwner(GameProfile profile) {
