--- a/net/minecraft/world/level/block/grower/WorldGenMegaTreeProvider.java
+++ b/net/minecraft/world/level/block/grower/WorldGenMegaTreeProvider.java
@@ -39,6 +39,7 @@
             return false;
         } else {
             ((WorldGenFeatureTreeConfiguration) worldgenfeatureconfigured.f).b();
+            setTreeType(worldgenfeatureconfigured); // CraftBukkit
             IBlockData iblockdata1 = Blocks.AIR.getBlockData();
 
             worldserver.setTypeAndData(blockposition.b(i, 0, j), iblockdata1, 4);
