--- a/net/minecraft/world/level/block/grower/WorldGenMegaTreeProvider.java
+++ b/net/minecraft/world/level/block/grower/WorldGenMegaTreeProvider.java
@@ -38,6 +38,7 @@
         if (holder == null) {
             return false;
         } else {
+            setTreeType(holder); // CraftBukkit
             WorldGenFeatureConfigured<?, ?> worldgenfeatureconfigured = (WorldGenFeatureConfigured) holder.value();
             IBlockData iblockdata1 = Blocks.AIR.defaultBlockState();
 
