From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 3 Jan 2021 20:03:40 -0800
Subject: [PATCH] Added missing vanilla tags


diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 69462a0679c255acee2adb84f590523d3fc3c9aa..4beb607c31705e182ac04c48bdc2a1de13eb9b4c 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -767,6 +767,12 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla item tag representing all chest boat items.
      */
     Tag<Material> ITEMS_CHEST_BOATS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("chest_boats"), Material.class);
+    // Paper start
+    /**
+     * Vanilla item tag representing all non-flammable wood items.
+     */
+    Tag<Material> ITEMS_NON_FLAMMABLE_WOOD = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("non_flammable_wood"), Material.class);
+    // Paper end
     /**
      * Vanilla item tag representing all fish items.
      */
