From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 3 Jan 2021 20:03:40 -0800
Subject: [PATCH] Added missing vanilla tags


diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -0,0 +0,0 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla item tag representing all chest boat items.
      */
     Tag<Material> ITEMS_CHEST_BOATS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("chest_boats"), Material.class);
+    // Paper start
+    /**
+     * Vanilla item tag representing all non-flammable wood items.
+     */
+    Tag<Material> ITEMS_NON_FLAMMABLE_WOOD = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("non_flammable_wood"), Material.class);
+    // Paper end
     /**
      * Vanilla item tag representing all fish items.
      */
@@ -0,0 +0,0 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla tag representing entities which can be eaten by frogs.
      */
     Tag<EntityType> ENTITY_TYPES_FROG_FOOD = Bukkit.getTag(REGISTRY_ENTITY_TYPES, NamespacedKey.minecraft("frog_food"), EntityType.class);
+    // Paper start
+    /**
+     * Key for the built-in entity registry
+     * @deprecated use {@link #REGISTRY_ENTITY_TYPES}
+     */
+    @Deprecated(forRemoval = true)
+    String REGISTRY_ENTITIES = REGISTRY_ENTITY_TYPES;
+    /**
+     * Vanilla entity tag representing arrow entities.
+     * @deprecated use {@link #ENTITY_TYPES_ARROWS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> ARROWS = ENTITY_TYPES_ARROWS;
+    /**
+     * Vanilla entity tag representing entities that live in beehives
+     * @deprecated use {@link #ENTITY_TYPES_BEEHIVE_INHABITORS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> BEEHIVE_INHABITORS = ENTITY_TYPES_BEEHIVE_INHABITORS;
+    /**
+     * Vanilla entity tag representing projectiles that impact
+     * @deprecated use {@link #ENTITY_TYPES_IMPACT_PROJECTILES}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> IMPACT_PROJECTILES = ENTITY_TYPES_IMPACT_PROJECTILES;
+    /**
+     * Vanilla entity tag for village raiders
+     * @deprecated use {@link #ENTITY_TYPES_RAIDERS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> RAIDERS = ENTITY_TYPES_RAIDERS;
+    /**
+     * Vanilla entity tag for skeleton types
+     * @deprecated use {@link #ENTITY_TYPES_SKELETONS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> SKELETONS = ENTITY_TYPES_SKELETONS;
+    // Paper end
 
     /**
      * Returns whether or not this tag has an entry for the specified item.
