From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Wed, 13 Jul 2022 15:29:53 +0200
Subject: [PATCH] Allow trident custom damage


diff --git a/src/main/java/org/bukkit/entity/Trident.java b/src/main/java/org/bukkit/entity/Trident.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/org/bukkit/entity/Trident.java
+++ b/src/main/java/org/bukkit/entity/Trident.java
@@ -0,0 +0,0 @@ public interface Trident extends AbstractArrow, ThrowableProjectile {
      * @param hasDealtDamage has dealt damage or hit the floor
      */
     void setHasDealtDamage(boolean hasDealtDamage);
+
+    /**
+     * Sets the base amount of damage this trident will do.
+     *
+     * @param damage new damage amount
+     */
+    void setDamage(double damage);
+
+    /**
+     * Gets the base amount of damage this trident will do.
+     *
+     * Defaults to 8.0 for a normal trident with
+     * <code>0.5 * (1 + power level)</code> added for trident fired from
+     * damage enchanted bows.
+     *
+     * @return base damage amount
+     */
+    double getDamage();
 }
 // Paper end
