--- a/net/minecraft/world/entity/decoration/EntityPainting.java
+++ b/net/minecraft/world/entity/decoration/EntityPainting.java
@@ -26,6 +26,10 @@
 
     public EntityPainting(EntityTypes<? extends EntityPainting> entitytypes, World world) {
         super(entitytypes, world);
+        // CraftBukkit start - generate a non-null painting
+        List<Paintings> list = Lists.newArrayList(Paintings.a);
+        this.art = (Paintings) list.get(this.random.nextInt(list.size()));
+        // CraftBukkit end
     }
 
     public EntityPainting(World world, BlockPosition blockposition, EnumDirection enumdirection) {
