--- a/net/minecraft/world/damagesource/EntityDamageSourceIndirect.java
+++ b/net/minecraft/world/damagesource/EntityDamageSourceIndirect.java
@@ -53,4 +53,10 @@
             return IChatBaseComponent.translatable(s, entityliving.getDisplayName(), ichatbasecomponent);
         }
     }
+
+    // CraftBukkit start
+    public Entity getProximateDamageSource() {
+        return super.getEntity();
+    }
+    // CraftBukkit end
 }
