--- a/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -16,6 +_,11 @@
         super(server, registries, playerIo, server.notificationManager());
         this.setViewDistance(server.viewDistance());
         this.setSimulationDistance(server.simulationDistance());
+        // Paper start - fix converting txt to json file; moved from constructor
+    }
+    @Override
+    public void loadAndSaveFiles() {
+        // Paper end - fix converting txt to json file
         this.loadUserBanList();
         this.saveUserBanList();
         this.loadIpBanList();
