--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -27,9 +_,11 @@
     public static final Codec<FluidState> CODEC = codec(BuiltInRegistries.FLUID.byNameCodec(), Fluid::defaultFluidState).stable();
     public static final int AMOUNT_MAX = 9;
     public static final int AMOUNT_FULL = 8;
+    protected final boolean isEmpty; // Paper - Perf: moved from isEmpty()
 
     public FluidState(final Fluid owner, final Reference2ObjectArrayMap<Property<?>, Comparable<?>> values, final MapCodec<FluidState> propertiesCodec) {
         super(owner, values, propertiesCodec);
+        this.isEmpty = owner.isEmpty(); // Paper - Perf: moved from isEmpty()
     }
 
     public Fluid getType() {
@@ -45,7 +_,7 @@
     }
 
     public boolean isEmpty() {
-        return this.getType().isEmpty();
+        return this.isEmpty; // Paper - Perf: moved into constructor
     }
 
     public float getHeight(final BlockGetter level, final BlockPos pos) {
