--- a/net/minecraft/world/level/redstone/ExperimentalRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/ExperimentalRedstoneWireEvaluator.java
@@ -39,6 +_,11 @@
             int intValue = entry.getIntValue();
             int i = unpackPower(intValue);
             BlockState blockState = level.getBlockState(blockPos);
+            // Paper start - Call BlockRedstoneEvent
+            if (blockState.is(this.wireBlock) && !blockState.getValue(RedStoneWireBlock.POWER).equals(i)) {
+                i = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, blockPos, blockState.getValue(RedStoneWireBlock.POWER), i).getNewCurrent();
+            }
+            // Paper end - Call BlockRedstoneEvent
             if (blockState.is(this.wireBlock) && !blockState.getValue(RedStoneWireBlock.POWER).equals(i)) {
                 int i1 = Block.UPDATE_CLIENTS;
                 if (!updateShape || !flag) {
