--- /dev/null
+++ b/com/mojang/logging/LogUtils.java
@@ -1,0 +_,69 @@
+package com.mojang.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.LifeCycle;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.LoggerConfig;
+import org.apache.logging.log4j.spi.LoggerContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Marker;
+import org.slf4j.MarkerFactory;
+
+import java.util.function.Supplier;
+
+public class LogUtils {
+    public static final String FATAL_MARKER_ID = "FATAL";
+    public static final Marker FATAL_MARKER = MarkerFactory.getMarker(FATAL_MARKER_ID);
+    private static final StackWalker STACK_WALKER = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
+
+    public static boolean isLoggerActive() {
+        final LoggerContext loggerContext = LogManager.getContext();
+        if (loggerContext instanceof LifeCycle lifeCycle) {
+            return !lifeCycle.isStopped();
+        }
+        return true; // Sensible default? In worst case, no logs - so not a huge loss
+    }
+
+    public static void configureRootLoggingLevel(final org.slf4j.event.Level level) {
+        final org.apache.logging.log4j.core.LoggerContext ctx = (org.apache.logging.log4j.core.LoggerContext) LogManager.getContext(false);
+        final Configuration config = ctx.getConfiguration();
+        final LoggerConfig loggerConfig = config.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
+        loggerConfig.setLevel(convertLevel(level));
+        ctx.updateLoggers();
+    }
+
+    private static Level convertLevel(final org.slf4j.event.Level level) {
+        return switch (level) {
+            case INFO -> Level.INFO;
+            case WARN -> Level.WARN;
+            case DEBUG -> Level.DEBUG;
+            case ERROR -> Level.ERROR;
+            case TRACE -> Level.TRACE;
+        };
+    }
+
+    public static Object defer(final Supplier<Object> result) {
+        class ToString {
+            @Override
+            public String toString() {
+                return result.get().toString();
+            }
+        }
+
+        return new ToString();
+    }
+
+    /**
+     * Caller sensitive, DO NOT WRAP
+     */
+    public static Logger getLogger() {
+        return LoggerFactory.getLogger(STACK_WALKER.getCallerClass());
+    }
+    // Paper start
+    public static Logger getClassLogger() {
+        return LoggerFactory.getLogger(STACK_WALKER.getCallerClass().getSimpleName());
+    }
+    // Paper end
+}
