--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -72,7 +_,7 @@
     public void tick() {
         boolean shouldTryTeleportToOwner = this.tamable.shouldTryTeleportToOwner();
         if (!shouldTryTeleportToOwner) {
-            this.tamable.getLookControl().setLookAt(this.owner, 10.0F, this.tamable.getMaxHeadXRot());
+            if (this.tamable.distanceToSqr(this.owner) <= 16 * 16) this.tamable.getLookControl().setLookAt(this.owner, 10.0F, this.tamable.getMaxHeadXRot()); // Paper - Limit pet look distance
         }
 
         if (--this.timeToRecalcPath <= 0) {
