--- a/net/minecraft/world/entity/ai/goal/SwellGoal.java
+++ b/net/minecraft/world/entity/ai/goal/SwellGoal.java
@@ -20,6 +_,7 @@
         return this.creeper.getSwellDir() > 0 || target != null && this.creeper.distanceToSqr(target) < 9.0;
     }
 
+
     @Override
     public void start() {
         this.creeper.getNavigation().stop();
@@ -40,7 +_,7 @@
     public void tick() {
         if (this.target == null) {
             this.creeper.setSwellDir(-1);
-        } else if (this.creeper.distanceToSqr(this.target) > 49.0) {
+        } else if (this.creeper.distanceToSqr(this.target) > 49.0 || !net.minecraft.world.entity.EntitySelector.NO_CREATIVE_OR_SPECTATOR.test(this.target)) { // Paper - Fix MC-179072 - consider creative / spectator players as out of reach
             this.creeper.setSwellDir(-1);
         } else if (!this.creeper.getSensing().hasLineOfSight(this.target)) {
             this.creeper.setSwellDir(-1);
