From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/CrashReportCategory.java b/net/minecraft/CrashReportCategory.java
index 6264463142c5f45bf560401aeb528b416525503f..65387eaecd854715e442faeec3ba441aa959227d 100644
--- a/net/minecraft/CrashReportCategory.java
+++ b/net/minecraft/CrashReportCategory.java
@@ -138,6 +138,7 @@ public class CrashReportCategory {
         } else {
             this.stackTrace = new StackTraceElement[stackTrace.length - 3 - size];
             System.arraycopy(stackTrace, 3 + size, this.stackTrace, 0, this.stackTrace.length);
+            this.stackTrace = io.papermc.paper.util.StacktraceDeobfuscator.INSTANCE.deobfuscateStacktrace(this.stackTrace); // Paper
             return this.stackTrace.length;
         }
     }
