From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/advancements/DisplayInfo.java b/net/minecraft/advancements/DisplayInfo.java
index 689345eb1d12680287b9448ae6e1cb0c36a1612d..fb263d4975f2a0f628a0704a18733b4971b98b19 100644
--- a/net/minecraft/advancements/DisplayInfo.java
+++ b/net/minecraft/advancements/DisplayInfo.java
@@ -37,6 +37,7 @@ public class DisplayInfo {
     private final boolean hidden;
     private float x;
     private float y;
+    public final io.papermc.paper.advancement.AdvancementDisplay paper = new io.papermc.paper.advancement.PaperAdvancementDisplay(this); // Paper - Add more advancement API
 
     public DisplayInfo(
         ItemStack icon,
