From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 11 May 2021 14:54:56 -0700
Subject: [PATCH] Inventory#close


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
index c3fa97ac34e1fc61ae02f224f8afe5a0b486fb4d..f6001047ada8308cfa1d9b26677a7a5d7774de51 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
@@ -449,6 +449,14 @@ public class CraftInventory implements Inventory {
             clear(i);
         }
     }
+    // Paper start
+    @Override
+    public int close() {
+        int count = this.inventory.getViewers().size();
+        com.google.common.collect.Lists.newArrayList(this.inventory.getViewers()).forEach(HumanEntity::closeInventory);
+        return count;
+    }
+    // Paper end
 
     @Override
     public ListIterator<ItemStack> iterator() {
