From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/AxeItem.java b/net/minecraft/world/item/AxeItem.java
index 148ee3f4f87b01f90e2fb47d9e14d76c6e4769b4..333048c688146df9853f76be41b74b6a8912728b 100644
--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -72,6 +72,11 @@ public class AxeItem extends Item {
         }
 
         ItemStack itemInHand = context.getItemInHand();
+        // Paper start - EntityChangeBlockEvent
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, newBlock.get())) {
+            return InteractionResult.PASS;
+        }
+        // Paper end
         if (player instanceof ServerPlayer) {
             CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger((ServerPlayer)player, pos, itemInHand);
         }
