From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index 3024444b2df4dfcc37d260fdbdc12757dd0533e3..1eca60c5a3e1eeb6f44edbb81fe1af011edd8732 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -41,11 +41,17 @@ public class EndCrystalItem extends Item {
                     if (level instanceof ServerLevel) {
                         EndCrystal crystal = new EndCrystal(level, x + 0.5, y, z + 0.5);
                         crystal.setShowBottom(false);
+                        // CraftBukkit start
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(context, crystal).isCancelled()) {
+                            if (context.getPlayer() != null) context.getPlayer().containerMenu.sendAllDataToRemote(); // Paper - Fix inventory desync
+                            return InteractionResult.FAIL;
+                        }
+                        // CraftBukkit end
                         level.addFreshEntity(crystal);
                         level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, above);
                         EnderDragonFight fight = ((ServerLevel)level).getDragonFight();
                         if (fight != null) {
-                            fight.tryRespawn();
+                            fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup
                         }
                     }
 
