From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/FireChargeItem.java b/net/minecraft/world/item/FireChargeItem.java
index fef7f0848eff9b7046f4a61b498f0a9804ea83df..654929d846d46dd22b939bbe2da478b7410464ec 100644
--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -36,12 +36,28 @@ public class FireChargeItem extends Item implements ProjectileItem {
         if (!CampfireBlock.canLight(blockState) && !CandleBlock.canLight(blockState) && !CandleCakeBlock.canLight(blockState)) {
             clickedPos = clickedPos.relative(context.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, clickedPos, context.getHorizontalDirection())) {
+                // CraftBukkit start - fire BlockIgniteEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, clickedPos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, context.getPlayer()).isCancelled()) {
+                    if (!context.getPlayer().getAbilities().instabuild) {
+                        context.getItemInHand().shrink(1);
+                    }
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 this.playSound(level, clickedPos);
                 level.setBlockAndUpdate(clickedPos, BaseFireBlock.getState(level, clickedPos));
                 level.gameEvent(context.getPlayer(), GameEvent.BLOCK_PLACE, clickedPos);
                 flag = true;
             }
         } else {
+            // CraftBukkit start - fire BlockIgniteEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, clickedPos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, context.getPlayer()).isCancelled()) {
+                if (!context.getPlayer().getAbilities().instabuild) {
+                    context.getItemInHand().shrink(1);
+                }
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             this.playSound(level, clickedPos);
             level.setBlockAndUpdate(clickedPos, blockState.setValue(BlockStateProperties.LIT, true));
             level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, clickedPos);
