--- a/net/minecraft/resources/RegistryOps.java
+++ b/net/minecraft/resources/RegistryOps.java
@@ -107,9 +_,18 @@
         public int hashCode() {
             return this.lookupProvider.hashCode();
         }
+
+        // Paper start - add method to get the value for pre-filling builders in the reg mod API
+        @Override
+        public HolderLookup.Provider lookupForValueCopyViaBuilders() {
+            return this.lookupProvider;
+        }
+        // Paper end - add method to get the value for pre-filling builders in the reg mod API
     }
 
     public interface RegistryInfoLookup {
         <T> Optional<HolderGetter<T>> lookup(ResourceKey<? extends Registry<? extends T>> registryKey);
+
+        HolderLookup.Provider lookupForValueCopyViaBuilders(); // Paper - add method to get the value for pre-filling builders in the reg mod API
     }
 }
