From e5b5b1c567121fc928252436476aad3485adf72a Mon Sep 17 00:00:00 2001
From: Iceee <andrew@opticgaming.tv>
Date: Tue, 24 Jun 2014 00:03:34 -0500
Subject: [PATCH] Actually fix TNT movement


diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index a442cb1..742e731 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -19,11 +19,11 @@ public class EntityTNTPrimed extends Entity {
     public EntityTNTPrimed(World world, double d0, double d1, double d2, EntityLiving entityliving) {
         this(world);
         this.setPosition(d0, d1, d2);
-        float f = (float) (Math.random() * 3.1415927410125732D * 2.0D);
+        // float f = (float) (Math.random() * 3.1415927410125732D * 2.0D); // PaperSpigot
 
-        this.motX = (double) (-((float) Math.sin((double) f)) * 0.02F);
+        this.motX = 0; // PaperSpigot - Fix TNT movement // (double) (-((float) Math.sin((double) f)) * 0.02F);
         this.motY = 0.20000000298023224D;
-        this.motZ = (double) (-((float) Math.cos((double) f)) * 0.02F);
+        this.motZ = 0; // PaperSpigot - Fix TNT movement  // (double) (-((float) Math.cos((double) f)) * 0.02F);
         this.fuseTicks = 80;
         this.lastX = d0;
         this.lastY = d1;
-- 
1.9.1

