From 2ad338cf6855c7e88f2c97a01a5f6459044da79e Mon Sep 17 00:00:00 2001
From: "Evan A. Haskell" <eah2119@gmail.com>
Date: Thu, 26 Jun 2014 18:37:29 -0400
Subject: [PATCH] Fix Null Tile Entity Worlds


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1d949e7..ca1a02c 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2021,6 +2021,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
 
+                tileentity.a(this); // PaperSpigot - No null worlds
                 this.a.add(tileentity);
             } else {
                 this.tileEntityList.add(tileentity);
-- 
1.9.1

