From e285e92e541fc9d440331ffd024d0979da6ddc76 Mon Sep 17 00:00:00 2001
From: ninja- <xninja@openmailbox.org>
Date: Tue, 8 Oct 2013 14:34:49 +0200
Subject: [PATCH] Implement respawn API.


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 348b6e2..272f7df 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1331,6 +1331,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             getHandle().collidesWithEntities = collides;
             getHandle().k = collides; // First boolean of Entity
         }
+
+        @Override
+        public void respawn()
+        {
+            if ( getHealth() <= 0 && isOnline() )
+            {
+                server.getServer().getPlayerList().moveToWorld( getHandle(), 0, false );
+            }
+        }
     };
 
     public Player.Spigot spigot()
-- 
2.1.0

