From c4fd50b8f90a07ce9b2d2fff3df0cfa3f1dcc79e Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Tue, 2 Jul 2013 13:51:38 +1000
Subject: [PATCH] Silence End of Stream Errors


diff --git a/src/main/java/net/minecraft/server/Packet.java b/src/main/java/net/minecraft/server/Packet.java
index b11c26b..9389a7d 100644
--- a/src/main/java/net/minecraft/server/Packet.java
+++ b/src/main/java/net/minecraft/server/Packet.java
@@ -112,7 +112,7 @@ public abstract class Packet {
             ++o;
             p += (long) packet.a();
         } catch (EOFException eofexception) {
-            iconsolelogmanager.severe("Reached end of stream for " + socket.getInetAddress());
+            // iconsolelogmanager.severe("Reached end of stream for " + socket.getInetAddress()); // Spigot - too spammy
             return null;
         }
 
-- 
1.8.1.2

