From 314e1a289073e5ffdd6853a686121a1eacaf6d53 Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Tue, 2 Jul 2013 14:30:00 +1000
Subject: [PATCH] Really Disable Netty


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index ccbd6bc..fead3d5 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -101,7 +101,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
 
         try {
             // Spigot start
-            this.s = ( org.spigotmc.SpigotConfig.listeners.get( 0 ).netty )
+            this.s = ( false )
                     ? new org.spigotmc.netty.NettyServerConnection( this, inetaddress, this.I() )
                     : new DedicatedServerConnection( this, inetaddress, this.I() );
             // Spigot end
-- 
1.8.1.2

