From 82c0825e94c0fbd1ef2e8eb8664c51158f7d8265 Mon Sep 17 00:00:00 2001
From: ItsHarry <info@itsjerryandharry.com>
Date: Wed, 10 Jul 2013 13:45:08 +0200
Subject: [PATCH] Fixes StructureGrowEvent.getPlayer(); always returning null
 regardless of whether bonemeal was used or not.


diff --git a/src/main/java/net/minecraft/server/ItemDye.java b/src/main/java/net/minecraft/server/ItemDye.java
index d0ffd84..3a98903 100644
--- a/src/main/java/net/minecraft/server/ItemDye.java
+++ b/src/main/java/net/minecraft/server/ItemDye.java
@@ -29,7 +29,7 @@ public class ItemDye extends Item {
             return false;
         } else {
             if (itemstack.getData() == 15) {
-                if (a(itemstack, world, i, j, k)) {
+                if (a(itemstack, world, i, j, k, entityhuman)) {
                     if (!world.isStatic) {
                         world.triggerEffect(2005, i, j, k, 0);
                     }
-- 
1.8.3.msysgit.0

