From a5c94cb10d115a07f16fd2c4d1af8bd71a1e5746 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Fri, 13 Dec 2013 15:21:02 +1100
Subject: [PATCH] Prevent Mineshaft Saving


diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
index 4d336d8..f835b01 100644
--- a/src/main/java/net/minecraft/server/StructureGenerator.java
+++ b/src/main/java/net/minecraft/server/StructureGenerator.java
@@ -179,7 +179,7 @@ public abstract class StructureGenerator extends WorldGenBase {
     private void a(World world) {
         if (this.e == null) {
             // Spigot Start
-            if ( world.spigotConfig.saveStructureInfo )
+            if ( world.spigotConfig.saveStructureInfo && !this.a().equals( "Mineshaft" ) )
             {
             this.e = (PersistentStructure) world.a(PersistentStructure.class, this.a());
             } else
-- 
1.9.1

