From 7f6ffc377f63db0c5bfe0bfb9189e14f2a670270 Mon Sep 17 00:00:00 2001
From: Thinkofdeath <thethinkofdeath@gmail.com>
Date: Mon, 2 Dec 2013 23:42:09 +0000
Subject: [PATCH] Fix some chunks not being sent to the client


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 99c30f9..d342f8f 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -919,6 +919,12 @@ public class Chunk {
     }
 
     public boolean k() {
+        // Spigot Start
+        if ( !( this.m && this.done && this.lit ) )
+        {
+            this.b( false );
+        }
+        // Spigot End
         return this.m && this.done && this.lit;
     }
 
-- 
1.8.4.msysgit.0

