From aa98163eefaa1479c99e6d527e5c81d79ae146fc Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Fri, 10 Jan 2014 15:15:50 +1100
Subject: [PATCH] Fix ItemStack Unbreakable Code


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index f32322e..fd1b18a 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -130,7 +130,7 @@ public final class ItemStack {
     }
 
     public boolean g() {
-        return this.item.getMaxDurability() <= 0 ? false : !this.hasTag() || !this.getTag().getBoolean("Unbreakable");
+        return this.item.getMaxDurability() <= 0 ? false : (!this.hasTag() || !this.getTag().getBoolean("Unbreakable")); // Spigot
     }
 
     public boolean usesData() {
-- 
1.8.3.2

