From 2317d87940f106c27011d4246d99e6e074d15acd Mon Sep 17 00:00:00 2001
From: Thinkofdeath <thethinkofdeath@gmail.com>
Date: Tue, 14 Jan 2014 20:11:25 +0000
Subject: [PATCH] Fix ConcurrentModificationException while being idle kicked
 in a vehicle


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 703b919..d525845 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1031,6 +1031,7 @@ public abstract class World implements IBlockAccess {
             this.everyoneSleeping();
         }
 
+        if (!guardEntityList) { // Spigot - It will get removed after the tick if we are ticking
         int i = entity.ae;
         int j = entity.ag;
 
@@ -1047,6 +1048,7 @@ public abstract class World implements IBlockAccess {
             this.entityList.remove(index);
         }
         // CraftBukkit end
+        } // Spigot
         this.b(entity);
     }
 
-- 
2.1.0

