From 60220a4c67cb653701d79926a425b6f8160dd010 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sun, 13 Apr 2014 09:00:59 +1000
Subject: [PATCH] Print Stack on InternalException


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 78d5949..65b66cd 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -71,6 +71,7 @@ public class NetworkManager extends SimpleChannelInboundHandler {
     public void exceptionCaught(ChannelHandlerContext channelhandlercontext, Throwable throwable) {
         NetworkManager.f.debug("Disconnecting " + this.getSocketAddress(), throwable);
         this.close(new ChatMessage("disconnect.genericReason", new Object[] { "Internal Exception: " + throwable}));
+        if (MinecraftServer.getServer().isDebugging()) throwable.printStackTrace(); // Spigot
     }
 
     protected void a(ChannelHandlerContext channelhandlercontext, Packet packet) {
-- 
2.1.0

