From 2a33ccd140c05c3df0e88c5d6fd37db64bcd5744 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Mon, 14 Apr 2014 09:46:20 +1000
Subject: [PATCH] Use Provided Case for Non Existent Offline Players


diff --git a/src/main/java/net/minecraft/server/UserCache.java b/src/main/java/net/minecraft/server/UserCache.java
index 78f1f57..021853e 100644
--- a/src/main/java/net/minecraft/server/UserCache.java
+++ b/src/main/java/net/minecraft/server/UserCache.java
@@ -142,7 +142,7 @@ public class UserCache {
             this.e.remove(gameprofile);
             this.e.addFirst(gameprofile);
         } else {
-            gameprofile = a(this.f, s1);
+            gameprofile = a(this.f, s); // Spigot - use correct case for offline players
             if (gameprofile != null) {
                 this.a(gameprofile);
                 usercache_usercacheentry = (UserCache.UserCacheEntry) this.c.get(s1);
-- 
2.1.4

