From d864b9c6274fb735641ffaadf42403c7855543d6 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Wed, 16 Apr 2014 10:09:56 +1000
Subject: [PATCH] Fix ban expire dates.


diff --git a/src/main/java/net/minecraft/server/GameProfileBanEntry.java b/src/main/java/net/minecraft/server/GameProfileBanEntry.java
index a5803cc..444f26b 100644
--- a/src/main/java/net/minecraft/server/GameProfileBanEntry.java
+++ b/src/main/java/net/minecraft/server/GameProfileBanEntry.java
@@ -12,7 +12,7 @@ public class GameProfileBanEntry extends ExpirableListEntry {
     }
 
     public GameProfileBanEntry(GameProfile gameprofile, Date date, String s, Date date1, String s1) {
-        super(gameprofile, date1, s, date1, s1);
+        super(gameprofile, date, s, date1, s1); // Spigot
     }
 
     public GameProfileBanEntry(JsonObject jsonobject) {
-- 
2.1.0

