From b6507b27adeaf9bfdd200dcbe04438f90f6d3776 Mon Sep 17 00:00:00 2001
From: Thinkofdeath <thethinkofdeath@gmail.com>
Date: Sun, 23 Mar 2014 01:12:10 +0000
Subject: [PATCH] Check for manually prefixed commands or commands that don't
 need a prefix for calling vanilla commands with command blocks


diff --git a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
index 0cd512d..278832c 100644
--- a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
+++ b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
@@ -110,6 +110,14 @@ public abstract class CommandBlockListenerAbstract implements ICommandListener {
                 }
             }
 
+            // Spigot start - check for manually prefixed command or commands that don't need a prefix
+            org.bukkit.command.Command commandBlockCommand = commandMap.getCommand(args[0]);
+            if (commandBlockCommand instanceof VanillaCommandWrapper) {
+                this.b = ((VanillaCommandWrapper) commandBlockCommand).dispatchVanillaCommandBlock(this, this.e);
+                return;
+            }
+            // Spigot end
+
             // Make sure this is a valid command
             if (commandMap.getCommand(args[0]) == null) {
                 this.b = 0;
-- 
1.9.1

