From 433062e42e703ec2032fcdef0845b3569c8635a7 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Wed, 26 Mar 2014 21:21:35 +1100
Subject: [PATCH] Fix several occurances of missed diff.


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index e91d53f..22becf7 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -182,7 +182,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         this.c((this.getMaxBuildHeight() + 8) / 16 * 16);
         this.c(MathHelper.a(this.getMaxBuildHeight(), 64, 256));
         this.propertyManager.a("max-build-height", Integer.valueOf(this.getMaxBuildHeight()));
-        h.info("Preparing level \"" + this.M() + "\"");
+        h.info("Preparing level \"" + this.N() + "\"");
         this.a(this.N(), this.N(), k, worldtype, s2);
         long i1 = System.nanoTime() - j;
         String s3 = String.format("%.3fs", new Object[] { Double.valueOf((double) i1 / 1.0E9D)});
diff --git a/src/main/java/net/minecraft/server/EntityEnderDragon.java b/src/main/java/net/minecraft/server/EntityEnderDragon.java
index f53b183..dbf30cb 100644
--- a/src/main/java/net/minecraft/server/EntityEnderDragon.java
+++ b/src/main/java/net/minecraft/server/EntityEnderDragon.java
@@ -135,7 +135,7 @@ public class EntityEnderDragon extends EntityInsentient implements IComplex, IMo
                     d2 = this.locZ + (this.bj - this.locZ) / (double) this.bg;
                     d3 = MathHelper.g(this.bk - (double) this.yaw);
                     this.yaw = (float) ((double) this.yaw + d3 / (double) this.bg);
-                    this.pitch = (float) ((double) this.pitch + (this.bm - (double) this.pitch) / (double) this.bg);
+                    this.pitch = (float) ((double) this.pitch + (this.bl - (double) this.pitch) / (double) this.bg);
                     --this.bg;
                     this.setPosition(d0, d1, d2);
                     this.b(this.yaw, this.pitch);
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 1f0e29a..b0b1462 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -78,7 +78,7 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
         ChunkCoordinates chunkcoordinates = world.getSpawn();
 
         this.setPositionRotation((double) chunkcoordinates.x + 0.5D, (double) (chunkcoordinates.y + 1), (double) chunkcoordinates.z + 0.5D, 0.0F, 0.0F);
-        this.az = 180.0F;
+        this.aZ = 180.0F;
         this.maxFireTicks = 20;
     }
 
@@ -190,7 +190,7 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
         }
 
         if (d0 < -d3) {
-            this.bu = this.by = this.locX;
+            this.bu = this.bx = this.locX;
         }
 
         if (d2 < -d3) {
@@ -198,7 +198,7 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
         }
 
         if (d1 < -d3) {
-            this.bv = this.bz = this.locY;
+            this.bv = this.by = this.locY;
         }
 
         this.bx += d0 * 0.25D;
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 0f1bcc7..5eb893f 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -320,7 +320,7 @@ public class EntityItem extends Entity {
     public ItemStack getItemStack() {
         ItemStack itemstack = this.getDataWatcher().getItemStack(10);
 
-        return itemstack == null ? new ItemStack(Blocks.STONE) : itemstack; 
+        return itemstack == null ? new ItemStack(Blocks.STONE) : itemstack;
     }
 
     public void setItemStack(ItemStack itemstack) {
diff --git a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
index f565028..166206c 100644
--- a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
+++ b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
@@ -199,7 +199,7 @@ public abstract class EntityMinecartAbstract extends Entity {
         }
 
         if (this.locY < -64.0D) {
-            this.G();
+            this.F();
         }
 
         int i;
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index b2cde75..b8d08e2 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -353,7 +353,7 @@ public class PlayerConnection implements PacketPlayInListener {
                 }
 
                 this.player.i();
-                this.player.W = 0.0F;
+                this.player.V = 0.0F;
                 this.player.setLocation(this.y, this.z, this.q, f2, f3);
                 if (!this.checkMovement) {
                     return;
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index c4698d4..25810d3 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -51,7 +51,7 @@ public abstract class PlayerList {
     public IPlayerFileData playerFileData; // CraftBukkit - private -> public
     public boolean hasWhitelist; // CraftBukkit - private -> public
     protected int maxPlayers;
-    protected int m;
+    private int m;
     private EnumGamemode n;
     private boolean o;
     private int p;
diff --git a/src/main/java/net/minecraft/server/StatisticManager.java b/src/main/java/net/minecraft/server/StatisticManager.java
index e86e21b..418ec40 100644
--- a/src/main/java/net/minecraft/server/StatisticManager.java
+++ b/src/main/java/net/minecraft/server/StatisticManager.java
@@ -8,11 +8,10 @@ public class StatisticManager {
 
     protected final Map a = Maps.newConcurrentMap();
 
-    public StatisticManager() {
-    }
+    public StatisticManager() {}
 
     public boolean a(Achievement achievement) {
-        return this.getStatisticValue((Statistic) achievement) > 0;
+        return this.getStatisticValue(achievement) > 0;
     }
 
     public boolean b(Achievement achievement) {
-- 
1.8.5.2.msysgit.0

