From 061a6b2daa24da1fe8e7b138d556d42d40014383 Mon Sep 17 00:00:00 2001
From: Suddenly <suddenly@suddenly.coffee>
Date: Sat, 5 Jul 2014 16:56:44 +0100
Subject: [PATCH] Add More Information to session.lock Errors


diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
index efdcad7..141248e 100644
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -54,7 +54,7 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
             }
         } catch (IOException ioexception) {
             ioexception.printStackTrace();
-            throw new RuntimeException("Failed to check session lock, aborting");
+            throw new RuntimeException("Failed to check session lock for world located at " + this.baseDir + ", aborting. Stop the server and delete the session.lock in this world to prevent further issues."); // Spigot
         }
     }
 
@@ -69,13 +69,13 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
 
             try {
                 if (datainputstream.readLong() != this.sessionId) {
-                    throw new ExceptionWorldConflict("The save is being accessed from another location, aborting");
+                    throw new ExceptionWorldConflict("The save for world located at " + this.baseDir + " is being accessed from another location, aborting");  // Spigot
                 }
             } finally {
                 datainputstream.close();
             }
         } catch (IOException ioexception) {
-            throw new ExceptionWorldConflict("Failed to check session lock, aborting");
+            throw new ExceptionWorldConflict("Failed to check session lock for world located at " + this.baseDir + ", aborting. Stop the server and delete the session.lock in this world to prevent further issues."); // Spigot
         }
     }
 
-- 
1.9.1

