From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <1254957+zachbr@users.noreply.github.com>
Date: Sat, 14 Jun 2014 22:39:58 -0500
Subject: [PATCH] Ability to disable asynccatcher


diff --git a/src/main/java/org/spigotmc/AsyncCatcher.java b/src/main/java/org/spigotmc/AsyncCatcher.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -0,0 +0,0 @@ import net.minecraft.server.MinecraftServer;
 public class AsyncCatcher
 {
 
-    public static boolean enabled = true;
+    public static boolean enabled = SpigotConfig.asyncCatcherFeature; // PaperSpigot - AsyncCatcher override feature
 
     public static void catchOp(String reason)
     {
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -0,0 +0,0 @@ public class SpigotConfig
     {
         intCacheLimit = getInt( "settings.int-cache-limit", 1024 );
     }
+
+    public static boolean asyncCatcherFeature;
+    private static void asyncCatcherFeature()
+    {
+        asyncCatcherFeature = getBoolean( "settings.async-plugin-bad-magic-catcher", true );
+        if ( !asyncCatcherFeature ) {
+            Bukkit.getLogger().log( Level.INFO, "Disabling async plugin bad ju-ju catcher, this might be bad depending on your plugins" );
+        }
+    }
 }
--