--- a/net/minecraft/server/notifications/NotificationService.java
+++ b/net/minecraft/server/notifications/NotificationService.java
@@ -38,7 +_,7 @@
 
     void playerUnbanned(NameAndId nameAndId);
 
-    <T> void onGameRuleChanged(GameRule<T> rule, T value);
+    <T> void onGameRuleChanged(net.minecraft.server.level.ServerLevel level, GameRule<T> rule, T value); // Paper - per-world game rules (add level param)
 
     void statusHeartbeat();
 }
