--- a/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
+++ b/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
@@ -35,6 +_,27 @@
             .required(LootContextParams.ENCHANTMENT_ACTIVE)
             .required(LootContextParams.ENCHANTMENT_LEVEL)
     );
+    // Paper start
+    public static final ContextKeySet ALL_PARAMS_OPTIONAL = register(
+        "generic_optional",
+        builder -> builder.optional(LootContextParams.THIS_ENTITY)
+            .optional(LootContextParams.LAST_DAMAGE_PLAYER)
+            .optional(LootContextParams.DAMAGE_SOURCE)
+            .optional(LootContextParams.ATTACKING_ENTITY)
+            .optional(LootContextParams.DIRECT_ATTACKING_ENTITY)
+            .optional(LootContextParams.ORIGIN)
+            .optional(LootContextParams.BLOCK_STATE)
+            .optional(LootContextParams.BLOCK_ENTITY)
+            .optional(LootContextParams.TOOL)
+            .optional(LootContextParams.EXPLOSION_RADIUS)
+            .optional(LootContextParams.ADDITIONAL_COST_COMPONENT_ALLOWED)
+            .optional(LootContextParams.CONTAINER)
+            .optional(LootContextParams.INTERACTING_ENTITY)
+            .optional(LootContextParams.TARGET_ENTITY)
+            .optional(LootContextParams.ENCHANTMENT_ACTIVE)
+            .optional(LootContextParams.ENCHANTMENT_LEVEL)
+    );
+    // Paper end
     public static final ContextKeySet CHEST = register("chest", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY));
     public static final ContextKeySet COMMAND = register(
         "command", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY)
@@ -194,5 +_,11 @@
         if (!missingFromAllParams.isEmpty()) {
             throw new IllegalStateException("Missing parameters from 'all_params': " + missingFromAllParams);
         }
+        // Paper start
+        Set<ContextKey<?>> missingFromAllParamsOptional = Sets.difference(allParams, ALL_PARAMS_OPTIONAL.allowed());
+        if (!missingFromAllParamsOptional.isEmpty()) {
+            throw new IllegalStateException("Missing parameters from 'all_params_optional': " + missingFromAllParamsOptional);
+        }
+        // Paper end
     }
 }
