--- a/net/minecraft/world/entity/projectile/EntityEnderPearl.java
+++ b/net/minecraft/world/entity/projectile/EntityEnderPearl.java
@@ -15,6 +15,13 @@
 import net.minecraft.world.phys.MovingObjectPosition;
 import net.minecraft.world.phys.MovingObjectPositionEntity;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+// CraftBukkit end
+
 public class EntityEnderPearl extends EntityProjectileThrowable {
 
     public EntityEnderPearl(EntityTypes<? extends EntityEnderPearl> entitytypes, World world) {
@@ -51,23 +58,36 @@
                 EntityPlayer entityplayer = (EntityPlayer) entity;
 
                 if (entityplayer.connection.isAcceptingMessages() && entityplayer.level() == this.level() && !entityplayer.isSleeping()) {
-                    if (this.random.nextFloat() < 0.05F && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
-                        EntityEndermite entityendermite = (EntityEndermite) EntityTypes.ENDERMITE.create(this.level());
+                    // CraftBukkit start - Fire PlayerTeleportEvent
+                    org.bukkit.craftbukkit.entity.CraftPlayer player = entityplayer.getBukkitEntity();
+                    org.bukkit.Location location = getBukkitEntity().getLocation();
+                    location.setPitch(player.getLocation().getPitch());
+                    location.setYaw(player.getLocation().getYaw());
 
-                        if (entityendermite != null) {
-                            entityendermite.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
-                            this.level().addFreshEntity(entityendermite);
+                    PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.ENDER_PEARL);
+                    Bukkit.getPluginManager().callEvent(teleEvent);
+
+                    if (!teleEvent.isCancelled() && entityplayer.connection.isAcceptingMessages()) {
+                        if (this.random.nextFloat() < 0.05F && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+                            EntityEndermite entityendermite = (EntityEndermite) EntityTypes.ENDERMITE.create(this.level());
+
+                            if (entityendermite != null) {
+                                entityendermite.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
+                                this.level().addFreshEntity(entityendermite, CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
+                            }
                         }
-                    }
 
-                    if (entity.isPassenger()) {
-                        entityplayer.dismountTo(this.getX(), this.getY(), this.getZ());
-                    } else {
-                        entity.teleportTo(this.getX(), this.getY(), this.getZ());
-                    }
+                        if (entity.isPassenger()) {
+                            entity.stopRiding();
+                        }
 
-                    entity.resetFallDistance();
-                    entity.hurt(this.damageSources().fall(), 5.0F);
+                        entityplayer.connection.teleport(teleEvent.getTo());
+                        entity.resetFallDistance();
+                        CraftEventFactory.entityDamage = this;
+                        entity.hurt(this.damageSources().fall(), 5.0F);
+                        CraftEventFactory.entityDamage = null;
+                    }
+                    // CraftBukkit end
                 }
             } else if (entity != null) {
                 entity.teleportTo(this.getX(), this.getY(), this.getZ());
@@ -96,7 +116,7 @@
     public Entity changeDimension(WorldServer worldserver) {
         Entity entity = this.getOwner();
 
-        if (entity != null && entity.level().dimension() != worldserver.dimension()) {
+        if (entity != null && worldserver != null && entity.level().dimension() != worldserver.dimension()) { // CraftBukkit - SPIGOT-6113
             this.setOwner((Entity) null);
         }
 
