--- a/net/minecraft/world/entity/projectile/EntitySmallFireball.java
+++ b/net/minecraft/world/entity/projectile/EntitySmallFireball.java
@@ -14,6 +14,8 @@
 import net.minecraft.world.phys.MovingObjectPositionBlock;
 import net.minecraft.world.phys.MovingObjectPositionEntity;
 
+import org.bukkit.event.entity.EntityCombustByEntityEvent; // CraftBukkit
+
 public class EntitySmallFireball extends EntityFireballFireball {
 
     public EntitySmallFireball(EntityTypes<? extends EntitySmallFireball> entitytypes, World world) {
@@ -22,6 +24,11 @@
 
     public EntitySmallFireball(World world, EntityLiving entityliving, double d0, double d1, double d2) {
         super(EntityTypes.SMALL_FIREBALL, entityliving, d0, d1, d2, world);
+        // CraftBukkit start
+        if (this.getShooter() != null && this.getShooter() instanceof EntityInsentient) {
+            isIncendiary = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+        }
+        // CraftBukkit end
     }
 
     public EntitySmallFireball(World world, double d0, double d1, double d2, double d3, double d4, double d5) {
@@ -38,7 +45,14 @@
                 Entity entity1 = this.getShooter();
                 int i = entity.getFireTicks();
 
-                entity.setOnFire(5);
+                // CraftBukkit start - Entity damage by entity event + combust event
+                EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+                entity.level.getCraftServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    entity.setOnFire(event.getDuration(), false);
+                }
+                // CraftBukkit end
                 boolean flag = entity.damageEntity(DamageSource.fireball(this, entity1), 5.0F);
 
                 if (!flag) {
@@ -57,10 +71,10 @@
         if (!this.level.isClientSide) {
             Entity entity = this.getShooter();
 
-            if (!(entity instanceof EntityInsentient) || this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (isIncendiary) { // CraftBukkit
                 BlockPosition blockposition = movingobjectpositionblock.getBlockPosition().shift(movingobjectpositionblock.getDirection());
 
-                if (this.level.isEmpty(blockposition)) {
+                if (this.level.isEmpty(blockposition) && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, blockposition, this).isCancelled()) { // CraftBukkit
                     this.level.setTypeUpdate(blockposition, BlockFireAbstract.a((IBlockAccess) this.level, blockposition));
                 }
             }
