--- a/net/minecraft/world/entity/projectile/EntityLlamaSpit.java
+++ b/net/minecraft/world/entity/projectile/EntityLlamaSpit.java
@@ -33,7 +33,7 @@
         Vec3D vec3d = this.getMot();
         MovingObjectPosition movingobjectposition = ProjectileHelper.a((Entity) this, this::a);
 
-        this.a(movingobjectposition);
+        this.preOnHit(movingobjectposition); // CraftBukkit - projectile hit event
         double d0 = this.locX() + vec3d.x;
         double d1 = this.locY() + vec3d.y;
         double d2 = this.locZ() + vec3d.z;
