--- a/net/minecraft/world/entity/projectile/EntitySpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/EntitySpectralArrow.java
@@ -45,7 +45,7 @@
         super.a(entityliving);
         MobEffect mobeffect = new MobEffect(MobEffects.GLOWING, this.duration, 0);
 
-        entityliving.addEffect(mobeffect, this.x());
+        entityliving.addEffect(mobeffect, this.x(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
     }
 
     @Override
