--- a/net/minecraft/world/entity/projectile/EntityEvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EntityEvokerFangs.java
@@ -15,6 +15,10 @@
 import net.minecraft.world.entity.TraceableEntity;
 import net.minecraft.world.level.World;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityEvokerFangs extends Entity implements TraceableEntity {
 
     public static final int ATTACK_DURATION = 20;
@@ -119,7 +123,7 @@
             }
 
             if (--this.lifeTicks < 0) {
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
         }
 
@@ -130,7 +134,7 @@
 
         if (entityliving.isAlive() && !entityliving.isInvulnerable() && entityliving != entityliving1) {
             if (entityliving1 == null) {
-                entityliving.hurt(this.damageSources().magic(), 6.0F);
+                entityliving.hurt(this.damageSources().magic().customCausingEntity(this), 6.0F); // CraftBukkit
             } else {
                 if (entityliving1.isAlliedTo((Entity) entityliving)) {
                     return;
