--- a/net/minecraft/world/level/portal/TeleportTransition.java
+++ b/net/minecraft/world/level/portal/TeleportTransition.java
@@ -20,11 +_,18 @@
     boolean asPassenger,
     Set<Relative> relatives,
     TeleportTransition.PostTeleportTransition postTeleportTransition
+    , org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause // CraftBukkit
 ) {
     public static final TeleportTransition.PostTeleportTransition DO_NOTHING = entity -> {};
     public static final TeleportTransition.PostTeleportTransition PLAY_PORTAL_SOUND = TeleportTransition::playPortalSound;
     public static final TeleportTransition.PostTeleportTransition PLACE_PORTAL_TICKET = TeleportTransition::placePortalTicket;
 
+    // CraftBukkit start
+    public TeleportTransition(ServerLevel newLevel, Vec3 position, Vec3 deltaMovement, float yRot, float xRot, boolean missingRespawnBlock, boolean asPassenger, Set<Relative> relatives, TeleportTransition.PostTeleportTransition postTeleportTransition) {
+        this(newLevel, position, deltaMovement, yRot, xRot, missingRespawnBlock, asPassenger, relatives, postTeleportTransition, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.UNKNOWN);
+    }
+    // CraftBukkit end
+
     public TeleportTransition(
         ServerLevel newLevel, Vec3 position, Vec3 deltaMovement, float yRot, float xRot, TeleportTransition.PostTeleportTransition postTeleportTransition
     ) {
@@ -40,7 +_,21 @@
         Set<Relative> relatives,
         TeleportTransition.PostTeleportTransition postTeleportTransition
     ) {
-        this(newLevel, position, deltaMovement, yRot, xRot, false, false, relatives, postTeleportTransition);
+        // CraftBukkit start
+        this(newLevel, position, deltaMovement, yRot, xRot, relatives, postTeleportTransition, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.UNKNOWN);
+    }
+    public TeleportTransition(
+        ServerLevel newLevel,
+        Vec3 position,
+        Vec3 deltaMovement,
+        float yRot,
+        float xRot,
+        Set<Relative> relatives,
+        TeleportTransition.PostTeleportTransition postTeleportTransition,
+        org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause
+    ) {
+        this(newLevel, position, deltaMovement, yRot, xRot, false, false, relatives, postTeleportTransition, cause);
+        // CraftBukkit end
     }
 
     private static void playPortalSound(Entity entity) {
@@ -99,7 +_,8 @@
             this.missingRespawnBlock(),
             this.asPassenger(),
             this.relatives(),
-            this.postTeleportTransition()
+            this.postTeleportTransition(),
+            this.cause() // Paper - keep cause
         );
     }
 
@@ -113,7 +_,8 @@
             this.missingRespawnBlock(),
             this.asPassenger(),
             this.relatives(),
-            this.postTeleportTransition()
+            this.postTeleportTransition(),
+            this.cause() // Paper - keep cause
         );
     }
 
@@ -127,7 +_,8 @@
             this.missingRespawnBlock(),
             true,
             this.relatives(),
-            this.postTeleportTransition()
+            this.postTeleportTransition(),
+            this.cause() // Paper - keep cause
         );
     }
 
