--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -72,6 +_,11 @@
         }
 
         ItemStack itemInHand = context.getItemInHand();
+        // Paper start - EntityChangeBlockEvent
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, newBlock.get())) {
+            return InteractionResult.PASS;
+        }
+        // Paper end
         if (player instanceof ServerPlayer) {
             CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger((ServerPlayer)player, pos, itemInHand);
         }
