From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/BoneMealItem.java b/net/minecraft/world/item/BoneMealItem.java
index 356252e85a39de9dfec8e680fea0cede607d4da7..b76760c2d3f06a7058418b6276f3362171d45645 100644
--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -33,12 +33,17 @@ public class BoneMealItem extends Item {
 
     @Override
     public InteractionResult useOn(UseOnContext context) {
+        // CraftBukkit start - extract bonemeal application logic to separate, static method
+        return BoneMealItem.applyBonemeal(context);
+    }
+    public static InteractionResult applyBonemeal(UseOnContext context) {
+        // CraftBukkit end
         Level level = context.getLevel();
         BlockPos clickedPos = context.getClickedPos();
         BlockPos blockPos = clickedPos.relative(context.getClickedFace());
         if (growCrop(context.getItemInHand(), level, clickedPos)) {
             if (!level.isClientSide) {
-                context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH);
+                if (context.getPlayer() != null) context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
                 level.levelEvent(1505, clickedPos, 15);
             }
 
@@ -48,7 +53,7 @@ public class BoneMealItem extends Item {
             boolean isFaceSturdy = blockState.isFaceSturdy(level, clickedPos, context.getClickedFace());
             if (isFaceSturdy && growWaterPlant(context.getItemInHand(), level, blockPos, context.getClickedFace())) {
                 if (!level.isClientSide) {
-                    context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH);
+                    if (context.getPlayer() != null) context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
                     level.levelEvent(1505, blockPos, 15);
                 }
 
