--- a/net/minecraft/world/item/DebugStickItem.java
+++ b/net/minecraft/world/item/DebugStickItem.java
@@ -49,7 +_,7 @@
     }
 
     public boolean handleInteraction(Player player, BlockState stateClicked, LevelAccessor level, BlockPos pos, boolean shouldCycleState, ItemStack debugStack) {
-        if (!player.canUseGameMasterBlocks()) {
+        if (!player.canUseGameMasterBlocks() && !(player.getAbilities().instabuild && player.getBukkitEntity().hasPermission("minecraft.debugstick")) && !player.getBukkitEntity().hasPermission("minecraft.debugstick.always")) { // Spigot
             return false;
         } else {
             Holder<Block> blockHolder = stateClicked.getBlockHolder();
