--- a/net/minecraft/world/item/EmptyMapItem.java
+++ b/net/minecraft/world/item/EmptyMapItem.java
@@ -17,10 +_,16 @@
     public InteractionResult use(final Level level, final Player player, final InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
         if (level instanceof ServerLevel serverLevel) {
+            org.bukkit.inventory.ItemStack emptyMap = itemStack.asBukkitCopy(); // Paper - PlayerMapFilledEvent
             itemStack.consume(1, player);
             player.awardStat(Stats.ITEM_USED.get(this));
             serverLevel.playSound(null, player, SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT, player.getSoundSource(), 1.0F, 1.0F);
             ItemStack map = MapItem.create(serverLevel, player.getBlockX(), player.getBlockZ(), (byte)0, true, false);
+            // Paper start - PlayerMapFilledEvent
+            io.papermc.paper.event.player.PlayerMapFilledEvent event = new io.papermc.paper.event.player.PlayerMapFilledEvent((org.bukkit.entity.Player) player.getBukkitEntity(), emptyMap, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(map));
+            event.callEvent();
+            map = org.bukkit.craftbukkit.inventory.CraftItemStack.unwrap(event.getCreatedMap());
+            // Paper end - PlayerMapFilledEvent
             if (itemStack.isEmpty()) {
                 return InteractionResult.SUCCESS.heldItemTransformedTo(map);
             } else {
