--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -99,7 +99,11 @@
                 entityendersignal.setItem(itemstack);
                 entityendersignal.signalTo(blockposition);
                 world.gameEvent((Holder) GameEvent.PROJECTILE_SHOOT, entityendersignal.position(), GameEvent.Context.of((Entity) user));
-                world.addFreshEntity(entityendersignal);
+                // CraftBukkit start
+                if (!world.addFreshEntity(entityendersignal)) {
+                    return InteractionResult.FAIL;
+                }
+                // CraftBukkit end
                 if (user instanceof ServerPlayer) {
                     ServerPlayer entityplayer = (ServerPlayer) user;
 
