--- a/net/minecraft/world/item/FlintAndSteelItem.java
+++ b/net/minecraft/world/item/FlintAndSteelItem.java
@@ -32,6 +_,12 @@
         if (!CampfireBlock.canLight(state) && !CandleBlock.canLight(state) && !CandleCakeBlock.canLight(state)) {
             BlockPos relativePos = pos.relative(context.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, relativePos, context.getHorizontalDirection())) {
+                // CraftBukkit start - Store the clicked block
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, relativePos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+                    context.getItemInHand().hurtAndBreak(1, player, context.getHand().asEquipmentSlot());
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 level.playSound(player, relativePos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
                 BlockState fireState = BaseFireBlock.getState(level, relativePos);
                 level.setBlock(relativePos, fireState, Block.UPDATE_ALL_IMMEDIATE);
@@ -47,6 +_,12 @@
                 return InteractionResult.FAIL;
             }
         } else {
+            // CraftBukkit start - Store the clicked block
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, pos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+                context.getItemInHand().hurtAndBreak(1, player, context.getHand().asEquipmentSlot());
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             level.playSound(player, pos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
             level.setBlock(pos, state.setValue(BlockStateProperties.LIT, true), Block.UPDATE_ALL_IMMEDIATE);
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
