--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -54,10 +_,13 @@
                 Player player = context.getPlayer();
                 level.playSound(player, pos, SoundEvents.HOE_TILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 if (!level.isClientSide()) {
-                    action.accept(context);
+                    // Paper start - Fire EntityBlockChangeEvent when dirt is tilled
                     if (player != null) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, net.minecraft.world.level.block.Blocks.FARMLAND.defaultBlockState());
                         context.getItemInHand().hurtAndBreak(1, player, context.getHand().asEquipmentSlot());
                     }
+                    action.accept(context);
+                    // Paper end - Fire EntityBlockChangeEvent when dirt is tilled
                 }
 
                 return InteractionResult.SUCCESS;
