--- a/net/minecraft/world/item/ItemCooldowns.java
+++ b/net/minecraft/world/item/ItemCooldowns.java
@@ -56,6 +_,13 @@
     }
 
     public void addCooldown(final Identifier cooldownGroup, final int time) {
+        // Paper start - Item cooldown events
+        this.addCooldown(cooldownGroup, time, true);
+    }
+
+    public void addCooldown(Identifier cooldownGroup, int time, boolean callEvent) {
+        // Event called in server override
+        // Paper end - Item cooldown events
         this.cooldowns.put(cooldownGroup, new ItemCooldowns.CooldownInstance(this.tickCount, this.tickCount + time));
         this.onCooldownStarted(cooldownGroup, time);
     }
