From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/NameTagItem.java b/net/minecraft/world/item/NameTagItem.java
index c104e28be26c5b8760daf9ad1fb79e84261aba31..438d98b11275d792f18301c643254dfb733c0dd6 100644
--- a/net/minecraft/world/item/NameTagItem.java
+++ b/net/minecraft/world/item/NameTagItem.java
@@ -18,8 +18,14 @@ public class NameTagItem extends Item {
         Component component = stack.get(DataComponents.CUSTOM_NAME);
         if (component != null && target.getType().canSerialize() && target.canBeNameTagged()) {
             if (!player.level().isClientSide && target.isAlive()) {
-                target.setCustomName(component);
-                if (target instanceof Mob mob) {
+                // Paper start - Add PlayerNameEntityEvent
+                io.papermc.paper.event.player.PlayerNameEntityEvent event = new io.papermc.paper.event.player.PlayerNameEntityEvent(((net.minecraft.server.level.ServerPlayer) player).getBukkitEntity(), target.getBukkitLivingEntity(), io.papermc.paper.adventure.PaperAdventure.asAdventure(stack.getHoverName()), true);
+                if (!event.callEvent()) return InteractionResult.PASS;
+
+                LivingEntity newEntity = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getEntity()).getHandle();
+                newEntity.setCustomName(event.getName() != null ? io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getName()) : null);
+                if (event.isPersistent() && newEntity instanceof Mob mob) {
+                    // Paper end - Add PlayerNameEntityEvent
                     mob.setPersistenceRequired();
                 }
 
