--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -60,6 +_,7 @@
 
                     return InteractionResult.FAIL;
                 } else {
+                    if (level.paperConfig().entities.spawning.disableMobSpawnerSpawnEggTransformation) return InteractionResult.FAIL; // Paper - Allow disabling mob spawner spawn egg transformation
                     spawnerHolder.setEntityId(type, level.getRandom());
                     level.sendBlockUpdated(pos, blockState, blockState, Block.UPDATE_ALL);
                     level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, pos);
@@ -171,7 +_,7 @@
 
         offspring.snapTo(pos.x(), pos.y(), pos.z(), 0.0F, 0.0F);
         offspring.applyComponentsFromItemStack(spawnEggStack);
-        level.addFreshEntityWithPassengers(offspring);
+        level.addFreshEntityWithPassengers(offspring, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // CraftBukkit
         spawnEggStack.consume(1, player);
         return Optional.of(offspring);
     }
