--- a/net/minecraft/world/level/block/BlockWitherRose.java
+++ b/net/minecraft/world/level/block/BlockWitherRose.java
@@ -31,7 +31,7 @@
                 EntityLiving entityliving = (EntityLiving) entity;
 
                 if (!entityliving.isInvulnerable(DamageSource.WITHER)) {
-                    entityliving.addEffect(new MobEffect(MobEffects.WITHER, 40));
+                    entityliving.addEffect(new MobEffect(MobEffects.WITHER, 40), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WITHER_ROSE); // CraftBukkit
                 }
             }
 
