--- a/net/minecraft/world/level/block/BlockTrapdoor.java
+++ b/net/minecraft/world/level/block/BlockTrapdoor.java
@@ -29,6 +29,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.phys.shapes.VoxelShapeCollision;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public class BlockTrapdoor extends BlockFacingHorizontal implements IBlockWaterlogged {
 
     public static final BlockStateBoolean OPEN = BlockProperties.OPEN;
@@ -117,6 +119,19 @@
             boolean flag1 = world.isBlockIndirectlyPowered(blockposition);
 
             if (flag1 != (Boolean) iblockdata.get(BlockTrapdoor.POWERED)) {
+                // CraftBukkit start
+                org.bukkit.World bworld = world.getWorld();
+                org.bukkit.block.Block bblock = bworld.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+
+                int power = bblock.getBlockPower();
+                int oldPower = (Boolean) iblockdata.get(OPEN) ? 15 : 0;
+
+                if (oldPower == 0 ^ power == 0 || block.getBlockData().isPowerSource()) {
+                    BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bblock, oldPower, power);
+                    world.getServer().getPluginManager().callEvent(eventRedstone);
+                    flag1 = eventRedstone.getNewCurrent() > 0;
+                }
+                // CraftBukkit end
                 if ((Boolean) iblockdata.get(BlockTrapdoor.OPEN) != flag1) {
                     iblockdata = (IBlockData) iblockdata.set(BlockTrapdoor.OPEN, flag1);
                     this.a((EntityHuman) null, world, blockposition, flag1);
