--- a/net/minecraft/world/level/block/BlockDropper.java
+++ b/net/minecraft/world/level/block/BlockDropper.java
@@ -16,6 +16,12 @@
 import net.minecraft.world.level.block.entity.TileEntityHopper;
 import net.minecraft.world.level.block.state.BlockBase;
 
+// CraftBukkit start
+import net.minecraft.world.InventoryLargeChest;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.inventory.InventoryMoveItemEvent;
+// CraftBukkit end
+
 public class BlockDropper extends BlockDispenser {
 
     private static final IDispenseBehavior c = new DispenseBehaviorItem();
@@ -53,8 +59,25 @@
                 if (iinventory == null) {
                     itemstack1 = BlockDropper.c.dispense(sourceblock, itemstack);
                 } else {
-                    itemstack1 = TileEntityHopper.addItem(tileentitydispenser, iinventory, itemstack.cloneItemStack().cloneAndSubtract(1), enumdirection.opposite());
-                    if (itemstack1.isEmpty()) {
+                    // CraftBukkit start - Fire event when pushing items into other inventories
+                    CraftItemStack oitemstack = CraftItemStack.asCraftMirror(itemstack.cloneItemStack().cloneAndSubtract(1));
+
+                    org.bukkit.inventory.Inventory destinationInventory;
+                    // Have to special case large chests as they work oddly
+                    if (iinventory instanceof InventoryLargeChest) {
+                        destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest((InventoryLargeChest) iinventory);
+                    } else {
+                        destinationInventory = iinventory.getOwner().getInventory();
+                    }
+
+                    InventoryMoveItemEvent event = new InventoryMoveItemEvent(tileentitydispenser.getOwner().getInventory(), oitemstack.clone(), destinationInventory, true);
+                    worldserver.getServer().getPluginManager().callEvent(event);
+                    if (event.isCancelled()) {
+                        return;
+                    }
+                    itemstack1 = TileEntityHopper.addItem(tileentitydispenser, iinventory, CraftItemStack.asNMSCopy(event.getItem()), enumdirection.opposite());
+                    if (event.getItem().equals(oitemstack) && itemstack1.isEmpty()) {
+                        // CraftBukkit end
                         itemstack1 = itemstack.cloneItemStack();
                         itemstack1.subtract(1);
                     } else {
