--- a/net/minecraft/world/level/block/BlockChorusFlower.java
+++ b/net/minecraft/world/level/block/BlockChorusFlower.java
@@ -20,6 +20,8 @@
 import net.minecraft.world.level.block.state.properties.BlockStateInteger;
 import net.minecraft.world.phys.MovingObjectPositionBlock;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class BlockChorusFlower extends Block {
 
     public static final BlockStateInteger AGE = BlockProperties.ah;
@@ -84,8 +86,12 @@
                 }
 
                 if (flag && b((IWorldReader) worldserver, blockposition1, (EnumDirection) null) && worldserver.isEmpty(blockposition.up(2))) {
-                    worldserver.setTypeAndData(blockposition, this.b.a((IBlockAccess) worldserver, blockposition), 2);
-                    this.a(worldserver, blockposition1, i);
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockSpreadEvent(worldserver, blockposition, blockposition1, this.getBlockData().set(BlockChorusFlower.AGE, Integer.valueOf(i)), 2)) {
+                        worldserver.setTypeAndData(blockposition, this.b.a((IBlockAccess) worldserver, blockposition), 2);
+                        this.a(worldserver, blockposition1, i);
+                    }
+                    // CraftBukkit end
                 } else if (i < 4) {
                     j = random.nextInt(4);
                     if (flag1) {
@@ -99,18 +105,30 @@
                         BlockPosition blockposition2 = blockposition.shift(enumdirection);
 
                         if (worldserver.isEmpty(blockposition2) && worldserver.isEmpty(blockposition2.down()) && b((IWorldReader) worldserver, blockposition2, enumdirection.opposite())) {
-                            this.a(worldserver, blockposition2, i + 1);
-                            flag2 = true;
+                            // CraftBukkit start - add event
+                            if (CraftEventFactory.handleBlockSpreadEvent(worldserver, blockposition, blockposition2, this.getBlockData().set(BlockChorusFlower.AGE, Integer.valueOf(i + 1)), 2)) {
+                                this.a(worldserver, blockposition2, i + 1);
+                                flag2 = true;
+                            }
+                            // CraftBukkit end
                         }
                     }
 
                     if (flag2) {
                         worldserver.setTypeAndData(blockposition, this.b.a((IBlockAccess) worldserver, blockposition), 2);
                     } else {
-                        this.a((World) worldserver, blockposition);
+                        // CraftBukkit - add event
+                        if (CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition, this.getBlockData().set(BlockChorusFlower.AGE, Integer.valueOf(5)), 2)) {
+                            this.a((World) worldserver, blockposition);
+                        }
+                        // CraftBukkit end
                     }
                 } else {
-                    this.a((World) worldserver, blockposition);
+                    // CraftBukkit - add event
+                    if (CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition, this.getBlockData().set(BlockChorusFlower.AGE, Integer.valueOf(5)), 2)) {
+                        this.a((World) worldserver, blockposition);
+                    }
+                    // CraftBukkit end
                 }
 
             }
