--- a/net/minecraft/world/level/block/BlockGrassPath.java
+++ b/net/minecraft/world/level/block/BlockGrassPath.java
@@ -43,6 +43,11 @@
 
     @Override
     public void tick(IBlockData iblockdata, WorldServer worldserver, BlockPosition blockposition, Random random) {
+        // CraftBukkit start - do not fade if the block is valid here
+        if (iblockdata.canSurvive(worldserver, blockposition)) {
+            return;
+        }
+        // CraftBukkit end
         BlockSoil.turnToDirt(iblockdata, worldserver, blockposition);
     }
 
