--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -14,11 +_,11 @@
     public static final int SECTION_HEIGHT = 16;
     public static final int SECTION_SIZE = 4096;
     public static final int BIOME_CONTAINER_BITS = 2;
-    private short nonEmptyBlockCount;
+    short nonEmptyBlockCount; // Paper - package private
     private short tickingBlockCount;
     private short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
-    private PalettedContainerRO<Holder<Biome>> biomes;
+    private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
 
     private LevelChunkSection(LevelChunkSection section) {
         this.nonEmptyBlockCount = section.nonEmptyBlockCount;
@@ -28,7 +_,7 @@
         this.biomes = section.biomes.copy();
     }
 
-    public LevelChunkSection(PalettedContainer<BlockState> states, PalettedContainerRO<Holder<Biome>> biomes) {
+    public LevelChunkSection(PalettedContainer<BlockState> states, PalettedContainer<Holder<Biome>> biomes) { // CraftBukkit - read/write
         this.states = states;
         this.biomes = biomes;
         this.recalcBlockCounts();
@@ -44,7 +_,7 @@
     }
 
     public FluidState getFluidState(int x, int y, int z) {
-        return this.states.get(x, y, z).getFluidState();
+        return this.states.get(x, y, z).getFluidState(); // Paper - Perf: Optimise LevelChunk#getFluidState; diff on change - we expect this to be effectively just get(x, y, z).getFluidState(). If this changes we need to check other patches that use BlockBehaviour#getFluidState.
     }
 
     public void acquire() {
@@ -181,6 +_,11 @@
     public Holder<Biome> getNoiseBiome(int x, int y, int z) {
         return this.biomes.get(x, y, z);
     }
+    // CraftBukkit start
+    public void setBiome(int x, int y, int z, Holder<Biome> biome) {
+        this.biomes.set(x, y, z, biome);
+    }
+    // CraftBukkit end
 
     public void fillBiomesFromNoise(BiomeResolver biomeResolver, Climate.Sampler climateSampler, int x, int y, int z) {
         PalettedContainer<Holder<Biome>> palettedContainer = this.biomes.recreate();
