--- a/net/minecraft/world/level/chunk/PalettedContainerFactory.java
+++ b/net/minecraft/world/level/chunk/PalettedContainerFactory.java
@@ -18,6 +_,7 @@
     Strategy<Holder<Biome>> biomeStrategy,
     Holder<Biome> defaultBiome,
     Codec<PalettedContainerRO<Holder<Biome>>> biomeContainerCodec
+    , Codec<PalettedContainer<Holder<Biome>>> biomeContainerRWCodec // Paper
 ) {
     public static PalettedContainerFactory create(final RegistryAccess registries) {
         Strategy<BlockState> blockStateStrategy = Strategy.createForBlockStates(Block.BLOCK_STATE_REGISTRY);
@@ -32,6 +_,7 @@
             biomeStrategy,
             defaultBiome,
             PalettedContainer.codecRO(biomes.holderByNameCodec(), biomeStrategy, defaultBiome)
+            , PalettedContainer.codecRW(biomes.holderByNameCodec(), biomeStrategy, defaultBiome) // Paper
         );
     }
 
