--- a/net/minecraft/world/entity/animal/armadillo/ArmadilloAi.java
+++ b/net/minecraft/world/entity/animal/armadillo/ArmadilloAi.java
@@ -181,6 +_,7 @@
             }
 
             if (owner.shouldSwitchToScaredState()) {
+                if (owner.isFrozen()) return; // Paper - checks if the armadillo is frozen.
                 owner.switchToState(Armadillo.ArmadilloState.SCARED);
                 if (owner.onGround()) {
                     owner.playSound(SoundEvents.ARMADILLO_LAND);
@@ -201,10 +_,12 @@
                     }
 
                     if (timeUntilExpiry < Armadillo.ArmadilloState.UNROLLING.animationDuration()) {
+                        if (owner.isFrozen()) return; // Paper - checks if the armadillo is frozen.
                         owner.playSound(SoundEvents.ARMADILLO_UNROLL_START);
                         owner.switchToState(Armadillo.ArmadilloState.UNROLLING);
                     }
                 } else if (state == Armadillo.ArmadilloState.UNROLLING && timeUntilExpiry > Armadillo.ArmadilloState.UNROLLING.animationDuration()) {
+                    if (owner.isFrozen()) return; // Paper - checks if the armadillo is frozen.
                     owner.switchToState(Armadillo.ArmadilloState.SCARED);
                 }
             }
