--- a/net/minecraft/world/level/RayTrace.java
+++ b/net/minecraft/world/level/RayTrace.java
@@ -24,7 +24,7 @@
         this.b = vec3d1;
         this.c = raytrace_blockcollisionoption;
         this.d = raytrace_fluidcollisionoption;
-        this.e = VoxelShapeCollision.a(entity);
+        this.e = (entity == null) ? VoxelShapeCollision.a() : VoxelShapeCollision.a(entity); // CraftBukkit
     }
 
     public Vec3D a() {
@@ -53,7 +53,7 @@
 
         private final Predicate<Fluid> predicate;
 
-        private FluidCollisionOption(Predicate predicate) {
+        private FluidCollisionOption(Predicate<Fluid> predicate) { // CraftBukkit - decompile error
             this.predicate = predicate;
         }
 
