From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/monster/skeleton/Skeleton.java b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
index dd9017161c0efee43dbccb7dfab8c96580d136bc..f2759462a4423d2c5d31d2c7fbfe96dc61081c77 100644
--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -93,11 +93,17 @@ public class Skeleton extends AbstractSkeleton {
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityType.STRAY, ConversionParams.single(this, true, true), stray -> {
+        final Stray entity = this.convertTo(EntityType.STRAY, ConversionParams.single(this, true, true), stray -> { // Paper - Fix issues with mob conversion; reset conversion time to prevent event spam
             if (!this.isSilent()) {
                 this.level().levelEvent(null, LevelEvent.SOUND_SKELETON_TO_STRAY, this.blockPosition(), 0);
             }
-        });
+        // Paper start - add spawn and transform reasons
+        }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN);
+        if (entity == null) {
+            // Reset conversion time to prevent event spam
+            this.conversionTime = 300;
+        }
+        // Paper end - add spawn and transform reasons
     }
 
     @Override
