From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/monster/skeleton/Skeleton.java b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
index a3a33954efee47d1a1c5738823c4e472bf8b22f5..f23ffcbc12867a5ed068c95f5c570497a0d53f13 100644
--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -94,11 +94,17 @@ public class Skeleton extends AbstractSkeleton {
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityTypes.STRAY, ConversionParams.single(this, true, true), stray -> {
+        final Stray entity = this.convertTo(EntityTypes.STRAY, ConversionParams.single(this, true, true), stray -> { // Paper - Fix issues with mob conversion; reset conversion time to prevent event spam
             if (!this.isSilent()) {
                 this.level().levelEvent(null, LevelEvent.SOUND_SKELETON_TO_STRAY, this.blockPosition(), 0);
             }
-        });
+        // Paper start - add spawn and transform reasons
+        }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN);
+        if (entity == null) {
+            // Reset conversion time to prevent event spam
+            this.conversionTime = 300;
+        }
+        // Paper end - add spawn and transform reasons
     }
 
     @Override
