From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvent.java b/net/minecraft/server/bossevents/CustomBossEvent.java
index b1a6fbee0757a3758e5ac57df58565ae0b80171a..877b90eaaf5574a43e2a4f6aa9027e5126f49491 100644
--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -23,6 +23,16 @@ public class CustomBossEvent extends ServerBossEvent {
     private final Set<UUID> players = Sets.newHashSet();
     private int value;
     private int max = 100;
+    // CraftBukkit start
+    private @javax.annotation.Nullable org.bukkit.boss.KeyedBossBar bossBar;
+
+    public org.bukkit.boss.KeyedBossBar getBukkitEntity() {
+        if (this.bossBar == null) {
+            this.bossBar = new org.bukkit.craftbukkit.boss.CraftKeyedBossbar(this);
+        }
+        return this.bossBar;
+    }
+    // CraftBukkit end
 
     public CustomBossEvent(ResourceLocation id, Component name) {
         super(name, BossEvent.BossBarColor.WHITE, BossEvent.BossBarOverlay.PROGRESS);
