--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
@@ -34,7 +_,7 @@
     public void doServerTick(ServerLevel level) {
         this.time++;
         if (this.targetLocation == null) {
-            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.dragon.getFightOrigin()));
+            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
             this.targetLocation = Vec3.atBottomCenterOf(heightmapPos);
         }
 
