--- a/net/minecraft/world/entity/monster/EntitySpider.java
+++ b/net/minecraft/world/entity/monster/EntitySpider.java
@@ -172,7 +172,7 @@
             Holder<MobEffectList> holder = entityspider_groupdataspider.effect;
 
             if (holder != null) {
-                this.addEffect(new MobEffect(holder, -1));
+                this.addEffect(new MobEffect(holder, -1), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN); // CraftBukkit
             }
         }
 
