--- a/net/minecraft/world/entity/monster/EntityEndermite.java
+++ b/net/minecraft/world/entity/monster/EntityEndermite.java
@@ -25,6 +25,10 @@
 import net.minecraft.world.level.World;
 import net.minecraft.world.level.block.state.IBlockData;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityEndermite extends EntityMonster {
 
     private static final int MAX_LIFE = 2400;
@@ -113,7 +117,7 @@
             }
 
             if (this.life >= 2400) {
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
         }
 
