--- a/net/minecraft/world/entity/monster/EntityCreeper.java
+++ b/net/minecraft/world/entity/monster/EntityCreeper.java
@@ -43,6 +43,12 @@
 import net.minecraft.world.level.World;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
+
 public class EntityCreeper extends EntityMonster implements PowerableMob {
 
     private static final DataWatcherObject<Integer> DATA_SWELL_DIR = DataWatcher.defineId(EntityCreeper.class, DataWatcherRegistry.INT);
@@ -218,9 +224,19 @@
     @Override
     public void thunderHit(WorldServer worldserver, EntityLightning entitylightning) {
         super.thunderHit(worldserver, entitylightning);
-        this.entityData.set(EntityCreeper.DATA_IS_POWERED, true);
+        // CraftBukkit start
+        if (CraftEventFactory.callCreeperPowerEvent(this, entitylightning, org.bukkit.event.entity.CreeperPowerEvent.PowerCause.LIGHTNING).isCancelled()) {
+            return;
+        }
+
+        this.setPowered(true);
     }
 
+    public void setPowered(boolean powered) {
+        this.entityData.set(EntityCreeper.DATA_IS_POWERED, powered);
+    }
+    // CraftBukkit end
+
     @Override
     protected EnumInteractionResult mobInteract(EntityHuman entityhuman, EnumHand enumhand) {
         ItemStack itemstack = entityhuman.getItemInHand(enumhand);
@@ -250,10 +266,18 @@
         if (!this.level.isClientSide) {
             float f = this.isPowered() ? 2.0F : 1.0F;
 
-            this.dead = true;
-            this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float) this.explosionRadius * f, World.a.MOB);
-            this.discard();
-            this.spawnLingeringCloud();
+            // CraftBukkit start
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+            this.level.getCraftServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                this.dead = true;
+                this.level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), World.a.MOB);
+                this.discard();
+                this.spawnLingeringCloud();
+            } else {
+                swell = 0;
+            }
+            // CraftBukkit end
         }
 
     }
@@ -264,6 +288,7 @@
         if (!collection.isEmpty()) {
             EntityAreaEffectCloud entityareaeffectcloud = new EntityAreaEffectCloud(this.level, this.getX(), this.getY(), this.getZ());
 
+            entityareaeffectcloud.setOwner(this); // CraftBukkit
             entityareaeffectcloud.setRadius(2.5F);
             entityareaeffectcloud.setRadiusOnUse(-0.5F);
             entityareaeffectcloud.setWaitTime(10);
@@ -277,7 +302,7 @@
                 entityareaeffectcloud.addEffect(new MobEffect(mobeffect));
             }
 
-            this.level.addFreshEntity(entityareaeffectcloud);
+            this.level.addFreshEntity(entityareaeffectcloud, CreatureSpawnEvent.SpawnReason.EXPLOSION); // CraftBukkit
         }
 
     }
