From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/common/custom/DiscardedPayload.java b/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
index 56e1ad9038392019b3fd80a99eeb4dd020cb862e..c712d7c83394e77452f897da15c2a6589a7eebf2 100644
--- a/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
+++ b/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
@@ -4,13 +4,19 @@ import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.Identifier;
 
-public record DiscardedPayload(Identifier id) implements CustomPacketPayload {
+public record DiscardedPayload(Identifier id, byte[] data) implements CustomPacketPayload { // Paper - store data
     public static <T extends FriendlyByteBuf> StreamCodec<T, DiscardedPayload> codec(Identifier id, int maxSize) {
-        return CustomPacketPayload.codec((value, output) -> {}, buffer -> {
+        return CustomPacketPayload.codec((value, output) -> {
+            // Paper start
+            // Always write data
+            output.writeBytes(value.data);
+        }, buffer -> {
             int i = buffer.readableBytes();
             if (i >= 0 && i <= maxSize) {
-                buffer.skipBytes(i);
-                return new DiscardedPayload(id);
+                final byte[] data = new byte[i];
+                buffer.readBytes(data);
+                return new DiscardedPayload(id, data);
+                // Paper end
             } else {
                 throw new IllegalArgumentException("Payload may not be larger than " + maxSize + " bytes");
             }
