--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -74,6 +_,7 @@
                                 return false;
                             } else {
                                 nextScheduledStart.setValue(timestamp + 20L + random.nextInt(20));
+                                if (level.paperConfig().entities.behavior.stuckEntityPoiRetryDelay.enabled() && body.getNavigation().isStuck()) nextScheduledStart.add(level.paperConfig().entities.behavior.stuckEntityPoiRetryDelay.intValue()); // Paper - Next stuck check delay config
                                 PoiManager poiManager = level.getPoiManager();
                                 batchCache.long2ObjectEntrySet().removeIf(entry -> !entry.getValue().isStillValid(timestamp));
                                 Predicate<BlockPos> cacheTest = pos -> {
