From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
index 689f78b962fbb6584a126fc07b0abca0ca5a25cc..4106549bd4dec1cc47d8765be8f5d119fe33bf56 100644
--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -110,7 +110,9 @@ public class HarvestFarmland extends Behavior<Villager> {
                 Block block = blockState.getBlock();
                 Block block1 = level.getBlockState(this.aboveFarmlandPos.below()).getBlock();
                 if (block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockState)) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(owner, this.aboveFarmlandPos, blockState.getFluidState().createLegacyBlock())) { // CraftBukkit // Paper - fix wrong block state
                     level.destroyBlock(this.aboveFarmlandPos, true, owner);
+                    } // CraftBukkit
                 }
 
                 if (blockState.isAir() && block1 instanceof FarmBlock && owner.hasFarmSeeds()) {
@@ -121,9 +123,11 @@ public class HarvestFarmland extends Behavior<Villager> {
                         boolean flag = false;
                         if (!item.isEmpty() && item.is(ItemTags.VILLAGER_PLANTABLE_SEEDS) && item.getItem() instanceof BlockItem blockItem) {
                             BlockState blockState1 = blockItem.getBlock().defaultBlockState();
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(owner, this.aboveFarmlandPos, blockState1)) { // CraftBukkit
                             level.setBlockAndUpdate(this.aboveFarmlandPos, blockState1);
                             level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(owner, blockState1));
                             flag = true;
+                            } // CraftBukkit
                         }
 
                         if (flag) {
