From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
index b8365065cfcdfb580497d2722dd93ae07f2878ca..dfd39ff8f4a17b42bda60f2bcec5ebc163c25954 100644
--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -75,6 +75,16 @@ public class PrepareRamNearestTarget<E extends PathfinderMob> extends Behavior<E
             .flatMap(
                 nearestVisibleLivingEntities -> nearestVisibleLivingEntities.findClosest(livingEntity -> this.ramTargeting.test(level, entity, livingEntity))
             )
+            // CraftBukkit start
+            .map((livingEntity) -> {
+                org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(entity, livingEntity, (livingEntity instanceof net.minecraft.server.level.ServerPlayer) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+                if (event.isCancelled() || event.getTarget() == null) {
+                    return null;
+                }
+                livingEntity = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
+                return livingEntity;
+            })
+            // CraftBukkit end
             .ifPresent(entity1 -> this.chooseRamPosition(entity, entity1));
     }
 
