--- a/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
+++ b/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
@@ -323,7 +_,10 @@
             } else {
                 boolean flag1 = this.isWantedBlock(mob, transportItemTarget.state)
                     && !this.isPositionAlreadyVisited(visited, unreachable, transportItemTarget, level)
-                    && !this.isContainerLocked(transportItemTarget);
+                    // Paper start - CopperGolemFindTargetEvent
+                    && !this.isContainerLocked(transportItemTarget)
+                    && new io.papermc.paper.event.entity.CopperGolemValidateTargetEvent((org.bukkit.entity.CopperGolem) mob.getBukkitEntity(), new org.bukkit.craftbukkit.block.CraftBlock(level, transportItemTarget.pos)).callEvent();
+                    // Paper end - CopperGolemFindTargetEvent
                 return flag1 ? transportItemTarget : null;
             }
         }
