From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
index 0a2f1487fc1222913af2e34dce4d4fe0350eaece..c65f80aadeb27685653f6a4731888c612d4a59d6 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -45,7 +45,7 @@ public class VillagerGoalPackages {
             Pair.of(1, new MoveToTargetSink()),
             Pair.of(2, PoiCompetitorScan.create()),
             Pair.of(3, new LookAndFollowTradingPlayerSink(speedModifier)),
-            Pair.of(5, GoToWantedItem.create(speedModifier, false, 4)),
+            Pair.of(5, GoToWantedItem.create(villager -> !villager.isSleeping(), speedModifier, false, 4)), // Paper - Fix MC-157464
             Pair.of(
                 6,
                 AcquirePoi.create(
