From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java b/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
index 9d82dde190201982bb5bcaf291f17dc31811b113..83ea42defa7d2fcaab8ed11c3cc49e35e133cd71 100644
--- a/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
+++ b/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
@@ -86,7 +86,9 @@ public class WorkAtComposter extends WorkAtPoi {
                 inventory.removeItemType(Items.WHEAT, i3);
                 ItemStack itemStack = inventory.addItem(new ItemStack(Items.BREAD, min));
                 if (!itemStack.isEmpty()) {
+                    villager.forceDrops = true; // Paper - Add missing forceDrop toggles
                     villager.spawnAtLocation(level, itemStack, 0.5F);
+                    villager.forceDrops = false; // Paper - Add missing forceDrop toggles
                 }
             }
         }
