From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java b/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
index f03f3c50990083492628cb28ae30a794f9110479..63ba05d7e0411bb6c9fde0e7c4dafa540eb1890d 100644
--- a/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
@@ -423,7 +423,7 @@ public class ItemStackTheFlatteningFix extends DataFix {
                 if (DAMAGE_IDS.contains(id.get().getSecond())) {
                     Typed<?> tag = input.getOrCreateTyped(tagF);
                     Dynamic<?> tagRest = tag.get(DSL.remainderFinder());
-                    tagRest = tagRest.set("Damage", tagRest.createInt(data));
+                    if (data != 0) tagRest = tagRest.set("Damage", tagRest.createInt(data)); // CraftBukkit
                     output = output.set(tagF, tag.set(DSL.remainderFinder(), tagRest));
                 }
 
