--- a/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
@@ -19,6 +_,6 @@
 
     @Override
     public boolean apply(final Level level, final ItemStack stack, final LivingEntity user) {
-        return user.removeAllEffects();
+        return user.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SELF_CONSUME); // CraftBukkit
     }
 }
