From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java b/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
index 2eea12f3390360dd02ae2ec994e64d91ca48148f..41e1e076a4567d3d3202cf8e426a1ebb391d85e8 100644
--- a/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
@@ -18,7 +18,9 @@ public record ClearAllStatusEffectsConsumeEffect() implements ConsumeEffect {
     }
 
     @Override
-    public boolean apply(Level level, ItemStack stack, LivingEntity entity) {
-        return entity.removeAllEffects();
+    // CraftBukkit start
+    public boolean apply(Level level, ItemStack stack, LivingEntity entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause) {
+        return entity.removeAllEffects(cause);
+        // CraftBukkit end
     }
 }
