--- a/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
@@ -14,6 +14,9 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
+// CraftBukkit start
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+// CraftBukkit end
 
 public record RemoveStatusEffectsConsumeEffect(HolderSet<MobEffect> effects) implements ConsumeEffect {
 
@@ -32,14 +35,14 @@
     }
 
     @Override
-    public boolean apply(Level world, ItemStack stack, LivingEntity user) {
+    public boolean apply(Level world, ItemStack itemstack, LivingEntity entityliving, EntityPotionEffectEvent.Cause cause) { // CraftBukkit
         boolean flag = false;
         Iterator iterator = this.effects.iterator();
 
         while (iterator.hasNext()) {
             Holder<MobEffect> holder = (Holder) iterator.next();
 
-            if (user.removeEffect(holder)) {
+            if (entityliving.removeEffect(holder, cause)) { // CraftBukkit
                 flag = true;
             }
         }
