--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -64,7 +_,13 @@
             }
 
             Vec3 oldPos = user.position();
-            if (user.randomTeleport(xx, yy, zz, true, BlockTags.CONSUMABLE_DOES_NOT_TELEPORT_TO)) {
+            // CraftBukkit start - handle canceled status of teleport event
+            net.minecraft.world.entity.LivingEntity.RandomTeleportResult status = user.randomTeleport(xx, yy, zz, true, net.minecraft.tags.BlockTags.CONSUMABLE_DOES_NOT_TELEPORT_TO, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.CONSUMABLE_EFFECT);
+
+            // teleport event was canceled, no more tries
+            if (status == net.minecraft.world.entity.LivingEntity.RandomTeleportResult.EVENT_CANCELLED) break;
+            if (status == net.minecraft.world.entity.LivingEntity.RandomTeleportResult.SUCCESS) {
+                // CraftBukkit end
                 level.gameEvent(GameEvent.TELEPORT, oldPos, GameEvent.Context.of(user));
                 SoundSource soundSource;
                 SoundEvent soundEvent;
