From 5e6b33d90f104b017d79526da9538bfd0934fc41 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
index 47daa7ef1e86a303e416c1155bd379b55d11a243..6b969005efa9243aeb46b64670091c33c76e4c33 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -115,10 +115,13 @@ public class IglooPieces {
         ) {
             if ("chest".equals(markerId)) {
                 level.setBlock(position, Blocks.AIR.defaultBlockState(), Block.UPDATE_ALL);
-                BlockEntity chest = level.getBlockEntity(position.below());
-                if (chest instanceof ChestBlockEntity) {
-                    ((ChestBlockEntity)chest).setLootTable(BuiltInLootTables.IGLOO_CHEST, random.nextLong());
-                }
+                // CraftBukkit start - ensure block transformation
+                // BlockEntity blockEntity = level.getBlockEntity(position.below());
+                // if (blockEntity instanceof ChestBlockEntity) {
+                //     ((ChestBlockEntity)blockEntity).setLootTable(BuiltInLootTables.IGLOO_CHEST, random.nextLong());
+                // }
+                this.setCraftLootTable(level, position.below(), random, BuiltInLootTables.IGLOO_CHEST);
+                // CraftBukkit end
             }
         }
 
