From c8366d72279d30331daff185f7b263159a86ff4d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
index 1bcf140aaeca616680cbdc6b999e5b6461cc2eef..fa05dc41e4368659d204b12c2896de456cefae60 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -1084,10 +1084,13 @@ public class NetherFortressPieces {
                 BlockPos worldPos = this.getWorldPos(3, 5, 5);
                 if (box.isInside(worldPos)) {
                     this.hasPlacedSpawner = true;
-                    level.setBlock(worldPos, Blocks.SPAWNER.defaultBlockState(), Block.UPDATE_CLIENTS);
-                    if (level.getBlockEntity(worldPos) instanceof SpawnerBlockEntity spawnerBlockEntity) {
-                        spawnerBlockEntity.setEntityId(EntityType.BLAZE, random);
-                    }
+                    // CraftBukkit start
+                    // level.setBlock(worldPos, Blocks.SPAWNER.defaultBlockState(), Block.UPDATE_CLIENTS);
+                    // if (level.getBlockEntity(worldPos) instanceof SpawnerBlockEntity spawnerBlockEntity) {
+                    //     spawnerBlockEntity.setEntityId(EntityType.BLAZE, random);
+                    // }
+                    this.placeCraftSpawner(level, worldPos, org.bukkit.entity.EntityType.BLAZE, Block.UPDATE_CLIENTS);
+                    // CraftBukkit end
                 }
             }
 
