--- a/net/minecraft/world/item/ItemRecord.java
+++ b/net/minecraft/world/item/ItemRecord.java
@@ -46,6 +46,7 @@
             ItemStack itemstack = itemactioncontext.getItemInHand();
 
             if (!world.isClientSide) {
+                if (true) return EnumInteractionResult.SUCCESS; // CraftBukkit - handled in ItemStack
                 EntityHuman entityhuman = itemactioncontext.getPlayer();
                 TileEntity tileentity = world.getBlockEntity(blockposition);
 
