--- a/net/minecraft/world/item/ItemRecord.java
+++ b/net/minecraft/world/item/ItemRecord.java
@@ -41,6 +41,7 @@
             ItemStack itemstack = itemactioncontext.getItemInHand();
 
             if (!world.isClientSide) {
+                if (true) return EnumInteractionResult.SUCCESS; // CraftBukkit - handled in ItemStack
                 ((BlockJukeBox) Blocks.JUKEBOX).setRecord(itemactioncontext.getPlayer(), world, blockposition, iblockdata, itemstack);
                 world.levelEvent((EntityHuman) null, 1010, blockposition, Item.getId(this));
                 itemstack.shrink(1);
