From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/com/mojang/brigadier/suggestion/IntegerSuggestion.java b/com/mojang/brigadier/suggestion/IntegerSuggestion.java
index 025bb337093e6a2a827d4520287949184d15f374..180ac38241d9c199dfb339f4c7585a66c625df7d 100644
--- a/com/mojang/brigadier/suggestion/IntegerSuggestion.java
+++ b/com/mojang/brigadier/suggestion/IntegerSuggestion.java
@@ -53,7 +53,7 @@ public class IntegerSuggestion extends Suggestion {
 
     @Override
     public int compareTo(final Suggestion o) {
-        if (o instanceof IntegerSuggestion) {
+        if (false && o instanceof IntegerSuggestion) { // Paper - fix unstable Suggestion comparison
             return Integer.compare(value, ((IntegerSuggestion) o).value);
         }
         return super.compareTo(o);
@@ -61,6 +61,6 @@ public class IntegerSuggestion extends Suggestion {
 
     @Override
     public int compareToIgnoreCase(final Suggestion b) {
-        return compareTo(b);
+        return super.compareToIgnoreCase(b); // Paper - fix unstable Suggestion comparison
     }
 }
