--- a/net/minecraft/world/entity/animal/EntityFish.java
+++ b/net/minecraft/world/entity/animal/EntityFish.java
@@ -68,7 +68,7 @@
 
     @Override
     public boolean isTypeNotPersistent(double d0) {
-        return !this.isFromBucket() && !this.hasCustomName();
+        return true; // CraftBukkit
     }
 
     @Override
@@ -90,6 +90,7 @@
     @Override
     public void setFromBucket(boolean flag) {
         this.entityData.set(EntityFish.FROM_BUCKET, flag);
+        this.setPersistenceRequired(this.isPersistent()); // CraftBukkit - SPIGOT-4106 update persistence
     }
 
     @Override
