--- a/net/minecraft/world/entity/animal/EntityParrot.java
+++ b/net/minecraft/world/entity/animal/EntityParrot.java
@@ -258,7 +258,7 @@
             }
 
             if (!this.level.isClientSide) {
-                if (this.random.nextInt(10) == 0) {
+                if (this.random.nextInt(10) == 0 && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTameEvent(this, entityhuman).isCancelled()) { // CraftBukkit
                     this.tame(entityhuman);
                     this.level.broadcastEntityEvent(this, (byte) 7);
                 } else {
@@ -272,7 +272,7 @@
                 itemstack.shrink(1);
             }
 
-            this.addEffect(new MobEffect(MobEffects.POISON, 900));
+            this.addEffect(new MobEffect(MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
             if (entityhuman.isCreative() || !this.isInvulnerable()) {
                 this.hurt(DamageSource.playerAttack(entityhuman), Float.MAX_VALUE);
             }
@@ -384,7 +384,7 @@
 
     @Override
     public boolean isPushable() {
-        return true;
+        return super.isPushable(); // CraftBukkit - collidable API
     }
 
     @Override
@@ -399,7 +399,7 @@
         if (this.isInvulnerableTo(damagesource)) {
             return false;
         } else {
-            this.setOrderedToSit(false);
+            // this.setOrderedToSit(false); // CraftBukkit - moved into EntityLiving.damageEntity(DamageSource, float)
             return super.hurt(damagesource, f);
         }
     }
