--- a/net/minecraft/world/entity/animal/EntityTurtle.java
+++ b/net/minecraft/world/entity/animal/EntityTurtle.java
@@ -314,7 +314,9 @@
     protected void n() {
         super.n();
         if (!this.isBaby() && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+            this.forceDrops = true; // CraftBukkit
             this.a((IMaterial) Items.SCUTE, 1);
+            this.forceDrops = false; // CraftBukkit
         }
 
     }
@@ -341,7 +343,9 @@
 
     @Override
     public void onLightningStrike(WorldServer worldserver, EntityLightning entitylightning) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.entityDamage = entitylightning; // CraftBukkit
         this.damageEntity(DamageSource.LIGHTNING_BOLT, Float.MAX_VALUE);
+        org.bukkit.craftbukkit.event.CraftEventFactory.entityDamage = null; // CraftBukkit
     }
 
     private static class e extends ControllerMove {
@@ -487,8 +491,12 @@
                 } else if (this.turtle.layEggCounter > 200) {
                     World world = this.turtle.level;
 
+                    // CraftBukkit start
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.turtle, this.blockPos.up(), (IBlockData) Blocks.TURTLE_EGG.getBlockData().set(BlockTurtleEgg.EGGS, this.turtle.random.nextInt(4) + 1)).isCancelled()) {
                     world.playSound((EntityHuman) null, blockposition, SoundEffects.TURTLE_LAY_EGG, SoundCategory.BLOCKS, 0.3F, 0.9F + world.random.nextFloat() * 0.2F);
                     world.setTypeAndData(this.blockPos.up(), (IBlockData) Blocks.TURTLE_EGG.getBlockData().set(BlockTurtleEgg.EGGS, this.turtle.random.nextInt(4) + 1), 3);
+                    }
+                    // CraftBukkit end
                     this.turtle.setHasEgg(false);
                     this.turtle.w(false);
                     this.turtle.setLoveTicks(600);
