--- a/net/minecraft/world/entity/animal/EntityChicken.java
+++ b/net/minecraft/world/entity/animal/EntityChicken.java
@@ -99,10 +99,12 @@
 
         if (world instanceof WorldServer worldserver) {
             if (this.isAlive() && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) {
+                this.forceDrops = true; // CraftBukkit
                 if (this.dropFromGiftLootTable(worldserver, LootTables.CHICKEN_LAY, this::spawnAtLocation)) {
                     this.playSound(SoundEffects.CHICKEN_EGG, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
                     this.gameEvent(GameEvent.ENTITY_PLACE);
                 }
+                this.forceDrops = false; // CraftBukkit
 
                 this.eggTime = this.random.nextInt(6000) + 6000;
             }
