--- a/net/minecraft/world/scores/Objective.java
+++ b/net/minecraft/world/scores/Objective.java
@@ -18,7 +_,7 @@
     private final String name;
     private final ObjectiveCriteria criteria;
     private Component displayName;
-    private Component formattedDisplayName;
+    private @Nullable Component formattedDisplayName; // Paper - Lazy initialization because only used internally by commands
     private ObjectiveCriteria.RenderType renderType;
     private boolean displayAutoUpdate;
     private @Nullable NumberFormat numberFormat;
@@ -36,7 +_,6 @@
         this.name = name;
         this.criteria = criteria;
         this.displayName = displayName;
-        this.formattedDisplayName = this.createFormattedDisplayName();
         this.renderType = renderType;
         this.displayAutoUpdate = displayAutoUpdate;
         this.numberFormat = numberFormat;
@@ -81,12 +_,21 @@
     }
 
     public Component getFormattedDisplayName() {
-        return this.formattedDisplayName;
+        // Paper start - Lazy initialization because only used internally by commands
+        Component formatted = this.formattedDisplayName;
+        if (formatted != null) {
+            return formatted;
+        }
+
+        formatted = this.createFormattedDisplayName();
+        this.formattedDisplayName = formatted;
+        return formatted;
+        // Paper end - Lazy initialization because only used internally by commands
     }
 
     public void setDisplayName(final Component name) {
         this.displayName = name;
-        this.formattedDisplayName = this.createFormattedDisplayName();
+        this.formattedDisplayName = null; // Paper - Lazy initialization because only used internally by commands
         this.scoreboard.onObjectiveChanged(this);
     }
 
