--- a/net/minecraft/world/scores/Scoreboard.java
+++ b/net/minecraft/world/scores/Scoreboard.java
@@ -375,7 +_,7 @@
     }
 
     protected List<PlayerTeam.Packed> packPlayerTeams() {
-        return this.getPlayerTeams().stream().map(PlayerTeam::pack).toList();
+        return this.getPlayerTeams().stream().filter(p -> io.papermc.paper.configuration.GlobalConfiguration.get().scoreboards.saveEmptyScoreboardTeams || !p.getPlayers().isEmpty()).map(PlayerTeam::pack).toList(); // Paper - Don't save empty scoreboard teams to scoreboard.dat
     }
 
     protected void loadPlayerTeam(PlayerTeam.Packed packed) {
