--- a/net/minecraft/world/item/component/ChargedProjectiles.java
+++ b/net/minecraft/world/item/component/ChargedProjectiles.java
@@ -22,6 +_,7 @@
         .xmap(ChargedProjectiles::new, chargedProjectiles -> chargedProjectiles.items);
     public static final StreamCodec<RegistryFriendlyByteBuf, ChargedProjectiles> STREAM_CODEC = ItemStack.STREAM_CODEC
         .apply(ByteBufCodecs.list())
+        .apply(ByteBufCodecs::increaseDepth) // Paper - Track codec depth
         .map(ChargedProjectiles::new, chargedProjectiles -> chargedProjectiles.items);
     private final List<ItemStack> items;
 
