--- a/net/minecraft/world/item/component/DeathProtection.java
+++ b/net/minecraft/world/item/component/DeathProtection.java
@@ -37,7 +_,7 @@
 
     public void applyEffects(final ItemStack itemStack, final LivingEntity entity) {
         for (ConsumeEffect effect : this.deathEffects) {
-            effect.apply(entity.level(), itemStack, entity);
+            effect.apply(entity.level(), itemStack, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.TOTEM); // CraftBukkit
         }
     }
 }
