--- a/net/minecraft/world/item/component/WrittenBookContent.java
+++ b/net/minecraft/world/item/component/WrittenBookContent.java
@@ -92,7 +_,7 @@
 
     public static boolean resolveForItem(final ItemStack itemStack, final CommandSourceStack source, final @Nullable Player player) {
         WrittenBookContent content = itemStack.get(DataComponents.WRITTEN_BOOK_CONTENT);
-        if (content != null && !content.resolved()) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.resolveSelectorsInBooks && content != null && !content.resolved()) { // Paper - Disable component selector resolving in books by default
             WrittenBookContent resolvedContent = content.resolve(source, player);
             if (resolvedContent != null) {
                 itemStack.set(DataComponents.WRITTEN_BOOK_CONTENT, resolvedContent);
