From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
index 419ef83d609ae2d5346d2a76fffb268ef03d083c..9bdd5f276ca78b623c11eb6773bc3c8085bd9c6c 100644
--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -31,8 +31,10 @@ public class ClientboundInitializeBorderPacket implements Packet<ClientGamePacke
     }
 
     public ClientboundInitializeBorderPacket(WorldBorder worldBorder) {
-        this.newCenterX = worldBorder.getCenterX();
-        this.newCenterZ = worldBorder.getCenterZ();
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldBorder.getCenterX() * worldBorder.world.dimensionType().coordinateScale();
+        this.newCenterZ = worldBorder.getCenterZ() * worldBorder.world.dimensionType().coordinateScale();
+        // CraftBukkit end
         this.oldSize = worldBorder.getSize();
         this.newSize = worldBorder.getLerpTarget();
         this.lerpTime = worldBorder.getLerpRemainingTime();
