--- a/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
@@ -2,6 +_,7 @@
 
 import java.util.ArrayList;
 import java.util.List;
+
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
@@ -19,9 +_,11 @@
     }
 
     private static void pack(List<SynchedEntityData.DataValue<?>> dataValues, RegistryFriendlyByteBuf buffer) {
+        try (io.papermc.paper.util.ItemObfuscationSession ignored = io.papermc.paper.util.ItemObfuscationSession.start(io.papermc.paper.util.ItemObfuscationBinding.LEVEL)) { // Paper - data sanitization
         for (SynchedEntityData.DataValue<?> dataValue : dataValues) {
             dataValue.write(buffer);
         }
+        } // Paper - data sanitization
 
         buffer.writeByte(255);
     }
