From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
index c317b80c6c537537f4e71b4626c988b56ab89695..a6297dd18f58c83aff8048cd20e5b7a2e4b37ec4 100644
--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -16,6 +16,16 @@ public record ClientboundSystemChatPacket(Component content, boolean overlay) im
         ClientboundSystemChatPacket::overlay,
         ClientboundSystemChatPacket::new
     );
+    // Spigot start
+    public ClientboundSystemChatPacket(net.md_5.bungee.api.chat.BaseComponent[] content, boolean overlay) {
+        this(org.bukkit.craftbukkit.util.CraftChatMessage.fromJSON(net.md_5.bungee.chat.ComponentSerializer.toString(content)), overlay);
+    }
+    // Spigot end
+    // Paper start
+    public ClientboundSystemChatPacket(net.kyori.adventure.text.Component content, boolean overlay) {
+        this(io.papermc.paper.adventure.PaperAdventure.asVanilla(content), overlay);
+    }
+    // Paper end
 
     @Override
     public PacketType<ClientboundSystemChatPacket> type() {
