From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java b/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
index f5edeb4fccd94bd84fc2fa46de41e958ab1ecdba..a5ffd95aba30dc0c4c663d44a4a97e10b87b3a9e 100644
--- a/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
@@ -14,6 +14,7 @@ public class ServerboundUseItemOnPacket implements Packet<ServerGamePacketListen
     private final BlockHitResult blockHit;
     private final InteractionHand hand;
     private final int sequence;
+    public long timestamp; // Spigot
 
     public ServerboundUseItemOnPacket(InteractionHand hand, BlockHitResult blockHit, int sequence) {
         this.hand = hand;
@@ -22,6 +23,7 @@ public class ServerboundUseItemOnPacket implements Packet<ServerGamePacketListen
     }
 
     private ServerboundUseItemOnPacket(FriendlyByteBuf buffer) {
+        this.timestamp = System.currentTimeMillis(); // Spigot
         this.hand = buffer.readEnum(InteractionHand.class);
         this.blockHit = buffer.readBlockHitResult();
         this.sequence = buffer.readVarInt();
