From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
index b1df0135c74be06739f764dc789ccec4d6d43b0a..4dacbf362c33b0770977d1dcb8afd58d26bd8333 100644
--- a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -14,6 +14,7 @@ public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener
     private final int sequence;
     private final float yRot;
     private final float xRot;
+    public long timestamp; // Spigot
 
     public ServerboundUseItemPacket(InteractionHand hand, int sequence, float yRot, float xRot) {
         this.hand = hand;
@@ -23,6 +24,7 @@ public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener
     }
 
     private ServerboundUseItemPacket(FriendlyByteBuf buffer) {
+        this.timestamp = System.currentTimeMillis(); // Spigot
         this.hand = buffer.readEnum(InteractionHand.class);
         this.sequence = buffer.readVarInt();
         this.yRot = buffer.readFloat();
