--- a/net/minecraft/server/packs/PathPackResources.java
+++ b/net/minecraft/server/packs/PathPackResources.java
@@ -112,6 +_,12 @@
         try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
             for (Path path1 : directoryStream) {
                 String string = path1.getFileName().toString();
+                // Paper start - Improve logging and errors
+                if (!Files.isDirectory(path1)) {
+                    LOGGER.error("Invalid directory entry: {} in {}.", string, this.root, new java.nio.file.NotDirectoryException(string));
+                    continue;
+                }
+                // Paper end - Improve logging and errors
                 if (ResourceLocation.isValidNamespace(string)) {
                     set.add(string);
                 } else {
