--- a/net/minecraft/server/commands/BanPlayerCommands.java
+++ b/net/minecraft/server/commands/BanPlayerCommands.java
@@ -57,7 +_,7 @@
                 );
                 ServerPlayer player = source.getServer().getPlayerList().getPlayer(gameProfile.getId());
                 if (player != null) {
-                    player.connection.disconnect(Component.translatable("multiplayer.disconnect.banned"));
+                    player.connection.disconnect(Component.translatable("multiplayer.disconnect.banned"), org.bukkit.event.player.PlayerKickEvent.Cause.BANNED); // Paper - kick event cause
                 }
             }
         }
