--- a/net/minecraft/server/commands/ChaseCommand.java
+++ b/net/minecraft/server/commands/ChaseCommand.java
@@ -30,6 +30,7 @@ public class ChaseCommand {
     public static void register(final CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(
             Commands.literal("chase")
+                .requires(Commands.hasPermission(Commands.LEVEL_ADMINS))
                 .then(
                     Commands.literal("follow")
                         .then(
