--- a/net/minecraft/server/commands/CommandEffect.java
+++ b/net/minecraft/server/commands/CommandEffect.java
@@ -84,7 +84,7 @@
             if (entity instanceof EntityLiving) {
                 MobEffect mobeffect = new MobEffect(holder, k, i, false, flag);
 
-                if (((EntityLiving) entity).addEffect(mobeffect, commandlistenerwrapper.getEntity())) {
+                if (((EntityLiving) entity).addEffect(mobeffect, commandlistenerwrapper.getEntity(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                     ++j;
                 }
             }
@@ -114,7 +114,7 @@
         while (iterator.hasNext()) {
             Entity entity = (Entity) iterator.next();
 
-            if (entity instanceof EntityLiving && ((EntityLiving) entity).removeAllEffects()) {
+            if (entity instanceof EntityLiving && ((EntityLiving) entity).removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                 ++i;
             }
         }
@@ -144,7 +144,7 @@
         while (iterator.hasNext()) {
             Entity entity = (Entity) iterator.next();
 
-            if (entity instanceof EntityLiving && ((EntityLiving) entity).removeEffect(holder)) {
+            if (entity instanceof EntityLiving && ((EntityLiving) entity).removeEffect(holder, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                 ++i;
             }
         }
