--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -40,14 +_,14 @@
 
     private static <T> int setRule(final CommandContext<CommandSourceStack> context, final GameRule<T> gameRule) throws CommandSyntaxException {
         CommandSourceStack source = context.getSource();
-        T value = context.getArgument("value", gameRule.valueClass());
+        T value = org.bukkit.craftbukkit.event.CraftEventFactory.handleGameRuleSet(gameRule, context.getArgument("value", gameRule.valueClass()), source.getLevel(), source.getBukkitSender()).value(); // Paper - per-world game rules and event
         GameRules gameRules = source.getLevel().getGameRules();
         String seralizedValue = gameRule.serialize(value);
         if (gameRules.get(gameRule).equals(value)) {
             throw ERROR_GAME_RULE_NOT_SET.create(gameRule, seralizedValue);
         }
 
-        gameRules.set(gameRule, value, context.getSource().getServer());
+        // gameRules.set(gameRule, value, context.getSource().getServer()); // Paper - per-world game rules and event - handle in CraftEventFactory.handleGameRuleSet
         source.sendSuccess(() -> Component.translatable("commands.gamerule.set", gameRule.id(), seralizedValue), true);
         return gameRule.getCommandResult(value);
     }
