--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -34,8 +_,7 @@
 
     private static <T> int setRule(final CommandContext<CommandSourceStack> context, final GameRule<T> gameRule) {
         CommandSourceStack source = context.getSource();
-        T value = context.getArgument("value", gameRule.valueClass());
-        source.getLevel().getGameRules().set(gameRule, value, context.getSource().getServer());
+        T value = org.bukkit.craftbukkit.event.CraftEventFactory.handleGameRuleSet(gameRule, context.getArgument("value", gameRule.valueClass()), source.getLevel(), context.getSource().getBukkitSender()).value(); // Paper - per-world game rules and event
         source.sendSuccess(() -> Component.translatable("commands.gamerule.set", gameRule.id(), gameRule.serialize(value)), true);
         return gameRule.getCommandResult(value);
     }
