From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/server/commands/LocateCommand.java b/net/minecraft/server/commands/LocateCommand.java
index b41d247006f9b7f1725a6490980aa5eee3f8a368..13bcd8653d766cd0b754a22e9aab261fbc62b0a5 100644
--- a/net/minecraft/server/commands/LocateCommand.java
+++ b/net/minecraft/server/commands/LocateCommand.java
@@ -202,6 +202,6 @@ public class LocateCommand {
     private static float dist(int x1, int z1, int x2, int z2) {
         int i = x2 - x1;
         int i1 = z2 - z1;
-        return Mth.sqrt(i * i + i1 * i1);
+        return (float) Math.hypot(i, i1); // Paper - Fix MC-177381
     }
 }
