From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/server/commands/ReloadCommand.java b/net/minecraft/server/commands/ReloadCommand.java
index 2d33a2f2bddda03baa37db9e2337a846b93af448..42ffdbeb21324cec477a665cf455afcaae82e041 100644
--- a/net/minecraft/server/commands/ReloadCommand.java
+++ b/net/minecraft/server/commands/ReloadCommand.java
@@ -16,7 +16,7 @@ public class ReloadCommand {
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public static void reloadPacks(Collection<String> selectedIds, CommandSourceStack source) {
-        source.getServer().reloadResources(selectedIds).exceptionally(throwable -> {
+        source.getServer().reloadResources(selectedIds, io.papermc.paper.event.server.ServerResourcesReloadedEvent.Cause.COMMAND).exceptionally(throwable -> { // Paper - Add ServerResourcesReloadedEvent
             LOGGER.warn("Failed to execute reload", throwable);
             source.sendFailure(Component.translatable("commands.reload.failure"));
             return null;
@@ -24,7 +24,7 @@ public class ReloadCommand {
     }
 
     private static Collection<String> discoverNewPacks(PackRepository packRepository, WorldData worldData, Collection<String> selectedIds) {
-        packRepository.reload();
+        packRepository.reload(true); // Paper - will perform a full reload
         Collection<String> list = Lists.newArrayList(selectedIds);
         Collection<String> disabled = worldData.getDataConfiguration().dataPacks().getDisabled();
 
@@ -37,6 +37,16 @@ public class ReloadCommand {
         return list;
     }
 
+    // CraftBukkit start
+    public static void reload(MinecraftServer server) {
+        PackRepository packRepository = server.getPackRepository();
+        WorldData worldData = server.getWorldData();
+        Collection<String> selectedIds = packRepository.getSelectedIds();
+        Collection<String> collection = discoverNewPacks(packRepository, worldData, selectedIds);
+        server.reloadResources(collection, io.papermc.paper.event.server.ServerResourcesReloadedEvent.Cause.PLUGIN); // Paper - Add ServerResourcesReloadedEvent
+    }
+    // CraftBukkit end
+
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(Commands.literal("reload").requires(source -> source.hasPermission(2)).executes(context -> {
             CommandSourceStack commandSourceStack = context.getSource();
