--- a/net/minecraft/server/commands/SetSpawnCommand.java
+++ b/net/minecraft/server/commands/SetSpawnCommand.java
@@ -102,8 +_,11 @@
         ServerPlayer.RespawnConfig respawnConfig = new ServerPlayer.RespawnConfig(respawnData, true);
 
         for (ServerPlayer target : targets) {
-            target.setRespawnPosition(respawnConfig, false);
-            tracker.track(target);
+            // Paper start - Add PlayerSetSpawnEvent
+            if (target.setRespawnPosition(respawnConfig, false, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.COMMAND)) {
+                tracker.track(target);
+            }
+            // Paper end - Add PlayerSetSpawnEvent
         }
 
         return tracker.sendFeedback(source, true, RESPONSE_SET, respawnData);
