--- a/net/minecraft/server/commands/SpreadPlayersCommand.java
+++ b/net/minecraft/server/commands/SpreadPlayersCommand.java
@@ -93,7 +93,7 @@
             if (entity instanceof Player) {
                 set.add(entity.getTeam());
             } else {
-                set.add((Object) null);
+                set.add((Team) null); // CraftBukkit - decompile error
             }
         }
 
@@ -203,7 +203,7 @@
                 commandspreadplayers_a = piles[j++];
             }
 
-            entity.teleportTo(world, (double) Mth.floor(commandspreadplayers_a.x) + 0.5D, (double) commandspreadplayers_a.getSpawnY(world, maxY), (double) Mth.floor(commandspreadplayers_a.z) + 0.5D, Set.of(), entity.getYRot(), entity.getXRot(), true);
+            entity.teleportTo(world, (double) Mth.floor(commandspreadplayers_a.x) + 0.5D, (double) commandspreadplayers_a.getSpawnY(world, maxY), (double) Mth.floor(commandspreadplayers_a.z) + 0.5D, Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND); // CraftBukkit - handle teleport reason
             d1 = Double.MAX_VALUE;
             SpreadPlayersCommand.Position[] acommandspreadplayers_a1 = piles;
             int k = piles.length;
