--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -47,7 +_,7 @@
     protected final RegionBitmap usedSectors = new RegionBitmap();
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final boolean sync) throws IOException {
-        this(info, path, externalFileDir, RegionFileVersion.getSelected(), sync);
+        this(info, path, externalFileDir, RegionFileVersion.getCompressionFormat(), sync); // Paper - Configurable region compression format
     }
 
     public RegionFile(final RegionStorageInfo info, final Path path, final Path externalFileDir, final RegionFileVersion version, final boolean sync) throws IOException {
@@ -83,6 +_,14 @@
                     if (offset != 0) {
                         int sectorNumber = getSectorNumber(offset);
                         int numSectors = getNumSectors(offset);
+                        // Spigot start
+                        if (numSectors == 255) {
+                            // We're maxed out, so we need to read the proper length from the section
+                            ByteBuffer realLen = ByteBuffer.allocate(4);
+                            this.file.read(realLen, sectorNumber * 4096);
+                            numSectors = (realLen.getInt(0) + 4) / 4096 + 1;
+                        }
+                        // Spigot end
                         if (sectorNumber < 2) {
                             LOGGER.warn("Region file {} has invalid sector at index: {}; sector {} overlaps with header", path, i, sectorNumber);
                             this.offsets.put(i, 0);
@@ -117,6 +_,13 @@
         } else {
             int sectorNumber = getSectorNumber(offset);
             int numSectors = getNumSectors(offset);
+            // Spigot start
+            if (numSectors == 255) {
+                ByteBuffer realLen = ByteBuffer.allocate(4);
+                this.file.read(realLen, sectorNumber * 4096);
+                numSectors = (realLen.getInt(0) + 4) / 4096 + 1;
+            }
+            // Spigot end
             int sectorsLength = numSectors * 4096;
             ByteBuffer buffer = ByteBuffer.allocate(sectorsLength);
             this.file.read(buffer, sectorNumber * 4096);
@@ -258,6 +_,7 @@
                     return true;
                 }
             } catch (IOException var9) {
+                com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(var9); // Paper - ServerExceptionEvent
                 return false;
             }
         }
@@ -329,6 +_,11 @@
         try (FileChannel extFile = FileChannel.open(tmpPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
             data.position(5);
             extFile.write(data);
+            // Paper start - ServerExceptionEvent
+        } catch (Throwable throwable) {
+            com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(throwable);
+            throw throwable;
+            // Paper end - ServerExceptionEvent
         }
 
         return () -> Files.move(tmpPath, path, StandardCopyOption.REPLACE_EXISTING);
