--- a/net/minecraft/world/effect/PoisonMobEffect.java
+++ b/net/minecraft/world/effect/PoisonMobEffect.java
@@ -11,7 +11,7 @@
     @Override
     public boolean applyEffectTick(EntityLiving entityliving, int i) {
         if (entityliving.getHealth() > 1.0F) {
-            entityliving.hurt(entityliving.damageSources().magic(), 1.0F);
+            entityliving.hurt(entityliving.damageSources().poison(), 1.0F);  // CraftBukkit - DamageSource.MAGIC -> CraftEventFactory.POISON
         }
 
         return true;
