--- a/net/minecraft/world/level/material/FluidTypeFlowing.java
+++ b/net/minecraft/world/level/material/FluidTypeFlowing.java
@@ -30,6 +30,15 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.phys.shapes.VoxelShapes;
 
+// CraftBukkit start
+import org.bukkit.block.BlockFace;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.data.CraftBlockData;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockFromToEvent;
+import org.bukkit.event.block.FluidLevelChangeEvent;
+// CraftBukkit end
+
 public abstract class FluidTypeFlowing extends FluidType {
 
     public static final BlockStateBoolean FALLING = BlockProperties.FALLING;
@@ -130,6 +139,15 @@
             Fluid fluid1 = this.getNewLiquid(world, blockposition1, iblockdata1);
 
             if (this.canSpreadTo(world, blockposition, iblockdata, EnumDirection.DOWN, blockposition1, iblockdata1, world.getFluidState(blockposition1), fluid1.getType())) {
+                // CraftBukkit start
+                org.bukkit.block.Block source = CraftBlock.at(world, blockposition);
+                BlockFromToEvent event = new BlockFromToEvent(source, BlockFace.DOWN);
+                world.getCraftServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+                // CraftBukkit end
                 this.spreadTo(world, blockposition1, iblockdata1, EnumDirection.DOWN, fluid1);
                 if (this.sourceNeighborCount(world, blockposition) >= 3) {
                     this.spreadToSides(world, blockposition, fluid, iblockdata);
@@ -160,6 +178,15 @@
                 IBlockData iblockdata1 = world.getBlockState(blockposition1);
 
                 if (this.canSpreadTo(world, blockposition, iblockdata, enumdirection, blockposition1, iblockdata1, world.getFluidState(blockposition1), fluid1.getType())) {
+                    // CraftBukkit start
+                    org.bukkit.block.Block source = CraftBlock.at(world, blockposition);
+                    BlockFromToEvent event = new BlockFromToEvent(source, org.bukkit.craftbukkit.block.CraftBlock.notchToBlockFace(enumdirection));
+                    world.getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        continue;
+                    }
+                    // CraftBukkit end
                     this.spreadTo(world, blockposition1, iblockdata1, enumdirection, fluid1);
                 }
             }
@@ -438,12 +465,23 @@
 
             if (fluid1.isEmpty()) {
                 fluid = fluid1;
-                world.setBlock(blockposition, Blocks.AIR.defaultBlockState(), 3);
+                // CraftBukkit start
+                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, blockposition, Blocks.AIR.defaultBlockState());
+                if (event.isCancelled()) {
+                    return;
+                }
+                world.setBlock(blockposition, ((CraftBlockData) event.getNewData()).getState(), 3);
+                // CraftBukkit end
             } else if (!fluid1.equals(fluid)) {
                 fluid = fluid1;
                 IBlockData iblockdata = fluid1.createLegacyBlock();
-
-                world.setBlock(blockposition, iblockdata, 2);
+                // CraftBukkit start
+                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, blockposition, iblockdata);
+                if (event.isCancelled()) {
+                    return;
+                }
+                world.setBlock(blockposition, ((CraftBlockData) event.getNewData()).getState(), 2);
+                // CraftBukkit end
                 world.scheduleTick(blockposition, fluid1.getType(), i);
                 world.updateNeighborsAt(blockposition, iblockdata.getBlock());
             }
