--- a/net/minecraft/world/level/storage/loot/LootDataType.java
+++ b/net/minecraft/world/level/storage/loot/LootDataType.java
@@ -26,6 +_,11 @@
         ContextKeySet contextKeys = this.contextGetter.context(value);
         ValidationContext rootContext = contextSource.context(contextKeys).enterElement(new ProblemReporter.RootElementPathElement(key), key);
         value.validate(rootContext);
+        // CraftBukkit start
+        if (value instanceof LootTable lootTable) {
+            lootTable.craftLootTable = new org.bukkit.craftbukkit.CraftLootTable(org.bukkit.craftbukkit.util.CraftNamespacedKey.fromMinecraft(key.identifier()), lootTable);
+        }
+        // CraftBukkit end
     }
 
     public void runValidation(final ValidationContextSource contextSource, final HolderLookup<T> lookup) {
