--- a/net/minecraft/tags/TagNetworkSerialization.java
+++ b/net/minecraft/tags/TagNetworkSerialization.java
@@ -60,7 +_,8 @@
         public static final TagNetworkSerialization.NetworkPayload EMPTY = new TagNetworkSerialization.NetworkPayload(Map.of());
         private final Map<Identifier, IntList> tags;
 
-        NetworkPayload(final Map<Identifier, IntList> tags) {
+        // Paper public for Per Player Quick Action & Pause Menu Dialogs
+        public NetworkPayload(final Map<Identifier, IntList> tags) {
             this.tags = tags;
         }
 
