--- a/net/minecraft/world/level/block/entity/TileEntityCommand.java
+++ b/net/minecraft/world/level/block/entity/TileEntityCommand.java
@@ -24,6 +24,13 @@
     private boolean auto;
     private boolean conditionMet;
     private final CommandBlockListenerAbstract commandBlock = new CommandBlockListenerAbstract() {
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandListenerWrapper wrapper) {
+            return new org.bukkit.craftbukkit.command.CraftBlockCommandSender(wrapper, TileEntityCommand.this);
+        }
+        // CraftBukkit end
+
         @Override
         public void setCommand(String s) {
             super.setCommand(s);
