From 96e4ace8a69c515c1d75f731dcd6f01e57d19245 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
index 1af5ff0e83d76d10b7e369acc4d97d12d845c385..2070a7b0a9c65c291f3bcbebfa9e30ed74b164cd 100644
--- a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
@@ -42,7 +42,7 @@ public class GroundPathNavigation extends PathNavigation {
     }
 
     @Override
-    public Path createPath(BlockPos pos, final int reachRange) {
+    public Path createPath(BlockPos pos, @org.jspecify.annotations.Nullable Entity entity, final int reachRange) { // Paper - EntityPathfindEvent
         LevelChunk chunk = this.level.getChunkSource().getChunkNow(SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getZ()));
         if (chunk == null) {
             return null;
@@ -51,7 +51,7 @@ public class GroundPathNavigation extends PathNavigation {
                 pos = this.findSurfacePosition(chunk, pos, reachRange);
             }
 
-            return super.createPath(pos, reachRange);
+            return super.createPath(pos, entity, reachRange); // Paper - EntityPathfindEvent
         }
     }
 
@@ -91,7 +91,7 @@ public class GroundPathNavigation extends PathNavigation {
 
     @Override
     public Path createPath(final Entity target, final int reachRange) {
-        return this.createPath(target.blockPosition(), reachRange);
+        return this.createPath(target.blockPosition(), target, reachRange); // Paper - EntityPathfindEvent
     }
 
     private int getSurfaceY() {
