From 09671551669244ef4f259d8b27547e463d6795d4 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/level/block/entity/TestBlockEntity.java b/net/minecraft/world/level/block/entity/TestBlockEntity.java
index 09ba8ecc0af8da4ec95080d6502bf1734616eebc..b97db8a9883a5680eb55238536da0a15e382253a 100644
--- a/net/minecraft/world/level/block/entity/TestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestBlockEntity.java
@@ -36,6 +36,7 @@ public class TestBlockEntity extends BlockEntity {
 
     @Override
     public void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
+        super.loadAdditional(tag, registries); // Paper - load the PDC
         this.mode = tag.read("mode", TestBlockMode.CODEC).orElse(TestBlockMode.FAIL);
         this.message = tag.getStringOr("message", "");
         this.powered = tag.getBooleanOr("powered", false);
