--- a/net/minecraft/world/entity/ai/goal/PathfinderGoalSit.java
+++ b/net/minecraft/world/entity/ai/goal/PathfinderGoalSit.java
@@ -21,7 +21,7 @@
 
     @Override
     public boolean canUse() {
-        if (!this.mob.isTame()) {
+        if (false && !this.mob.isTame()) { // CraftBukkit - Allow sitting for wild animals
             return false;
         } else if (this.mob.isInWaterOrBubble()) {
             return false;
