--- a/net/minecraft/server/PathfinderGoalSit.java
+++ b/net/minecraft/server/PathfinderGoalSit.java
@@ -19,7 +19,7 @@
     @Override
     public boolean a() {
         if (!this.entity.isTamed()) {
-            return false;
+            return this.entity.isWillSit() && this.entity.getGoalTarget() == null; // CraftBukkit - Allow sitting for wild animals
         } else if (this.entity.aH()) {
             return false;
         } else if (!this.entity.isOnGround()) {
