--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -35,6 +_,8 @@
     private final PrimaryLevelData.SpecialWorldProperty specialWorldProperty;
     private final Lifecycle worldGenSettingsLifecycle;
     private LevelData.RespawnData respawnData;
+    private static final String PAPER_RESPAWN_DIMENSION = "paperSpawnDimension"; // Paper
+    public net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = net.minecraft.world.level.Level.OVERWORLD; // Paper
     private long gameTime;
     private final @Nullable UUID singlePlayerUUID;
     private final int version;
@@ -43,6 +_,20 @@
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
 
+    // CraftBukkit start - Add world and pdc
+    private net.minecraft.server.level.ServerLevel world;
+    protected net.minecraft.nbt.Tag pdc;
+
+    public void setWorld(net.minecraft.server.level.ServerLevel world) {
+        if (this.world != null) {
+            return;
+        }
+        this.world = world;
+        world.getWorld().readBukkitValues(this.pdc);
+        this.pdc = null;
+    }
+    // CraftBukkit end
+
     private PrimaryLevelData(
         final @Nullable UUID singlePlayerUUID,
         final boolean wasModded,
@@ -93,7 +_,7 @@
     ) {
         long gameTime = input.get("Time").asLong(0L);
         LevelVersion levelVersion = LevelVersion.parse(input);
-        return new PrimaryLevelData(
+        PrimaryLevelData data = new PrimaryLevelData( // Paper
             input.get("singleplayer_uuid").flatMap(UUIDUtil.CODEC::parse).result().orElse(null),
             input.get("WasModded").asBoolean(false),
             input.get("spawn").read(LevelData.RespawnData.CODEC).result().orElse(LevelData.RespawnData.DEFAULT),
@@ -106,6 +_,13 @@
             specialWorldProperty,
             worldGenSettingsLifecycle
         );
+        // Paper start
+        data.respawnDimension = input.get(PAPER_RESPAWN_DIMENSION)
+            .read(net.minecraft.world.level.Level.RESOURCE_KEY_CODEC)
+            .result()
+            .orElse(data.respawnData.dimension());
+        return data;
+        // Paper end
     }
 
     @Override
@@ -130,6 +_,8 @@
         NbtUtils.addCurrentDataVersion(tag);
         tag.putInt("GameType", this.settings.gameType().getId());
         tag.store("spawn", LevelData.RespawnData.CODEC, this.respawnData);
+        // TODO - snapshot - Missing WorldGenSettings.encode diff
+        tag.store(PAPER_RESPAWN_DIMENSION, net.minecraft.world.level.Level.RESOURCE_KEY_CODEC, this.respawnDimension); // Paper
         tag.putLong("Time", this.gameTime);
         writeLastPlayed(tag);
         tag.putString("LevelName", this.settings.levelName());
@@ -142,6 +_,8 @@
         }
 
         tag.store(WorldDataConfiguration.MAP_CODEC, this.settings.dataConfiguration());
+        tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
+        this.world.getWorld().storeBukkitValues(tag); // CraftBukkit - add pdc
     }
 
     public static void writeLastPlayed(final CompoundTag tag) {
@@ -249,6 +_,12 @@
     @Override
     public void setDifficulty(final Difficulty difficulty) {
         this.settings = this.settings.withDifficulty(difficulty);
+        // CraftBukkit start
+        net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket packet = new net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+        for (net.minecraft.server.level.ServerPlayer player : (java.util.List<net.minecraft.server.level.ServerPlayer>) (java.util.List) this.world.players()) {
+            player.connection.send(packet);
+        }
+        // CraftBukkit end
     }
 
     @Override
@@ -322,6 +_,14 @@
     public LevelSettings getLevelSettings() {
         return this.settings.copy();
     }
+
+    // CraftBukkit start - Check if the name stored in NBT is the correct one
+    public void checkName(String name) {
+        if (!this.settings.levelName().equals(name)) {
+            this.settings = this.settings.withLevelName(name);
+        }
+    }
+    // CraftBukkit end
 
     @Deprecated
     public static enum SpecialWorldProperty {
