From 782378e964c0df175a8bb46055b1d0e63fbf3dfb Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 8 Aug 2014 22:51:26 -0500
Subject: [PATCH] Player affects spawning API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index b2ad5b7..b2a8e2e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1102,6 +1102,28 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
         {
             throw new UnsupportedOperationException( "Not supported yet." );
         }
+
+        /**
+         * Get whether the player affects mob spawning
+         *
+         * @return whether or not the player affects
+         * mob spawning.
+         */
+        public boolean getAffectsSpawning()
+        {
+            throw new  UnsupportedOperationException( "Not supported yet." );
+        }
+
+        /**
+         * Set whether or not the player affects mob spawning
+         *
+         * @param affects whether or not the player should affect
+         * spawning or not.
+         */
+        public void setAffectsSpawning(boolean affects)
+        {
+            throw new UnsupportedOperationException( "Not supported yet" );
+        }
     }
 
     Spigot spigot();
-- 
1.9.1

