From 777ac30514f3a7c0fa43ab7d3a64732580a0236b Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 29 Feb 2016 17:22:34 -0600
Subject: [PATCH] Player affects spawning API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d24ef805..01c61e8e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1396,6 +1396,20 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public AdvancementProgress getAdvancementProgress(Advancement advancement);
 
+    /**
+     * Get whether the player can affect mob spawning
+     *
+     * @return if the player can affect mob spawning
+     */
+    public boolean getAffectsSpawning();
+
+    /**
+     * Set whether the player can affect mob spawning
+     *
+     * @param affects Whether the player can affect mob spawning
+     */
+    public void setAffectsSpawning(boolean affects);
+
     // Spigot start
     public class Spigot extends Entity.Spigot
     {
-- 
2.11.0

