From 7c0c42257e29ed95c6de268ff1c754d7b233c844 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Mon, 29 Feb 2016 18:05:37 -0600
Subject: [PATCH] Add player view distance API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 313dde80..5a02b2e8 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1452,6 +1452,20 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public void setAffectsSpawning(boolean affects);
 
+    /**
+     * Gets the view distance for this player
+     *
+     * @return the player's view distance
+     */
+    public int getViewDistance();
+
+    /**
+     * Sets the view distance for this player
+     *
+     * @param viewDistance the player's view distance
+     */
+    public void setViewDistance(int viewDistance);
+
     // Spigot start
     public class Spigot extends Entity.Spigot
     {
-- 
2.15.0

