From 4a940f6fb654c6297cb3aa11127d0ed03027c578 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 25 Nov 2012 13:47:27 -0600
Subject: [PATCH] Add methods for working with arrows stuck in living entities


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 8827c6c..9d25133 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -394,4 +394,20 @@ public interface LivingEntity extends Attributable, Entity, Damageable, Projecti
      * @return true if the entity has AI, otherwise false.
      */
     boolean hasAI();
+
+    // Paper start
+
+    /**
+     * Get the number of arrows stuck in this entity
+     * * @return Number of arrows stuck
+     */
+    int getArrowsStuck();
+
+    /**
+     * Set the number of arrows stuck in this entity
+     *
+     * @param arrows Number of arrows to stick in this entity
+     */
+    void setArrowsStuck(int arrows);
+    // Paper end
 }
-- 
2.7.4

