From ad5444f104e8943b5cd8fe2110dfa6e74c47ab0c Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 20 Mar 2016 06:44:49 -0400
Subject: [PATCH] Access items by EquipmentSlot


diff --git a/src/main/java/org/bukkit/inventory/PlayerInventory.java b/src/main/java/org/bukkit/inventory/PlayerInventory.java
index 5d75bd49..ddded787 100644
--- a/src/main/java/org/bukkit/inventory/PlayerInventory.java
+++ b/src/main/java/org/bukkit/inventory/PlayerInventory.java
@@ -198,4 +198,22 @@ public interface PlayerInventory extends Inventory {
     public void setHeldItemSlot(int slot);
 
     public HumanEntity getHolder();
+
+    // Paper start
+    /**
+     * Gets the {@link ItemStack} found in the slot.
+     *
+     * @param slot The slot
+     * @return The item stack in the slot
+     */
+    ItemStack getItem(EquipmentSlot slot);
+
+    /**
+     * Sets the {@link ItemStack} at the given {@link EquipmentSlot}.
+     *
+     * @param slot The slot for the stack
+     * @param stack The item stack to set
+     */
+    void setItem(EquipmentSlot slot, ItemStack stack);
+    // Paper end
 }
-- 
2.18.0

