From 11c3bdd9e6dbef27b415b8ca0c46cb6cdfd454f7 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Wed, 14 Dec 2016 08:27:57 -0800
Subject: [PATCH] Add API methods to control if armour stands can move and
 perform entity lookups


diff --git a/src/main/java/org/bukkit/entity/ArmorStand.java b/src/main/java/org/bukkit/entity/ArmorStand.java
index b4f0451..cdf4f66 100644
--- a/src/main/java/org/bukkit/entity/ArmorStand.java
+++ b/src/main/java/org/bukkit/entity/ArmorStand.java
@@ -258,4 +258,38 @@ public interface ArmorStand extends LivingEntity {
      * @param marker whether this is a marker
      */
     void setMarker(boolean marker);
+
+    // Paper start
+    /**
+     * Tests if this armour stand can move.
+     *
+     * <p>The default value is {@code true}.</p>
+     *
+     * @return {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    boolean canMove();
+
+    /**
+     * Sets if this armour stand can move.
+     *
+     * @param flag {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    void setCanMove(boolean flag);
+
+    /**
+     * Tests if this armour stand can perform entity lookups.
+     *
+     * <p>The default value is {@code true}.</p>
+     *
+     * @return {@code true} if this armour stand can perform entity lookups, {@code false} otherwise
+     */
+    boolean canPerformEntityLookups();
+
+    /**
+     * Sets if this armour stand can perform entity lookups.
+     *
+     * @param flag {@code true} if this armour stand can perform entity lookups, {@code false} otherwise
+     */
+    void setCanPerformEntityLookups(boolean flag);
+    // Paper end
 }
-- 
2.10.2

