From e4a91b4babe80ccab09955b6069d1498676c5e53 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 14 Jan 2017 16:15:20 -0600
Subject: [PATCH] Add fromBottle flag to Experience Orbs


diff --git a/src/main/java/org/bukkit/entity/ExperienceOrb.java b/src/main/java/org/bukkit/entity/ExperienceOrb.java
index c286edfd..e8a83c06 100644
--- a/src/main/java/org/bukkit/entity/ExperienceOrb.java
+++ b/src/main/java/org/bukkit/entity/ExperienceOrb.java
@@ -18,4 +18,11 @@ public interface ExperienceOrb extends Entity {
      * @param value Amount of experience
      */
     public void setExperience(int value);
+
+    /**
+     * Check if this orb was spawned from a {@link ThrownExpBottle}
+     *
+     * @return if orb was spawned from a bottle
+     */
+    public boolean isFromBottle();
 }
-- 
2.13.0.windows.1

