From e0ca64d7522079f570a571193c5bdd574a63ff7d Mon Sep 17 00:00:00 2001
From: AlphaBlend <whizkid3000@hotmail.com>
Date: Tue, 21 Feb 2017 16:44:15 -0800
Subject: [PATCH] Add PlayerArmorStandCreateEvent


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerArmorStandCreateEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerArmorStandCreateEvent.java
new file mode 100644
index 0000000..76e25c9
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerArmorStandCreateEvent.java
@@ -0,0 +1,48 @@
+package com.destroystokyo.paper.event.player;
+
+import javax.annotation.Nonnull;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+
+/**
+ * Called when a player is about to create an armor stand
+ */
+public class PlayerArmorStandCreateEvent extends PlayerEvent implements Cancellable {
+    private static HandlerList handlers = new HandlerList();
+    private ArmorStand stand;
+    private boolean cancelled;
+
+    public PlayerArmorStandCreateEvent(Player player, @Nonnull final ArmorStand stand) {
+        super(player);
+        this.stand = stand;
+    }
+
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    public void setCancelled(boolean cancelled) {
+        this.cancelled = cancelled;
+    }
+
+    /**
+     * Gets the armor stand that will be created
+     * @return the armor stand that will be created
+     */
+    @Nonnull
+    public ArmorStand getArmorStand() {
+        return stand;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
\ No newline at end of file
-- 
2.10.0.windows.1

