From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Mon, 31 Jul 2017 01:49:43 -0500
Subject: [PATCH] LivingEntity#setKiller


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 3154b1cb18b1cf97705ccd2e52bd87bd6cf5d8cf..a2bdecc8b27652460064126f8e5bb8d038f5d11a 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -279,6 +279,15 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     @Nullable
     public Player getKiller();
 
+    // Paper start
+    /**
+     * Sets the player identified as the killer of the living entity.
+     *
+     * @param killer player
+     */
+    public void setKiller(@Nullable Player killer);
+    // Paper end
+
     /**
      * Adds the given {@link PotionEffect} to the living entity.
      *
