From b1935b33977f5215db28c156d7f75428b14f6ff3 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 15:29:21 -0400
Subject: [PATCH] Vex#getSummoner API

Get's the Mob that summoned this Vex

diff --git a/src/main/java/org/bukkit/entity/Vex.java b/src/main/java/org/bukkit/entity/Vex.java
index a2f2fcae..d395e405 100644
--- a/src/main/java/org/bukkit/entity/Vex.java
+++ b/src/main/java/org/bukkit/entity/Vex.java
@@ -3,4 +3,10 @@ package org.bukkit.entity;
 /**
  * Represents a Vex.
  */
-public interface Vex extends Monster { }
+public interface Vex extends Monster {
+    /**
+     * @return What Entity (most likely an Evoker, but not guaranteed) summoned this Vex
+     */
+    Mob getSummoner(); // Paper
+
+}
-- 
2.20.1

