From bbb9cf97e9b7033826d511f51950b79c2741fe10 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Thu, 19 Jul 2018 17:32:47 -0500
Subject: [PATCH] Use specific 1.13 versioning branch/build name


diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index c45faf4cb..7041085c3 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -266,7 +266,7 @@ public class VersionCommand extends BukkitCommand {
             return getFromJenkins(currentVer);
         } catch (NumberFormatException ex) {
             verInfo = verInfo.replace("\"", "");
-            return getFromRepo("PaperMC/Paper", verInfo);
+            return getFromRepo("PaperMC/Paper", "pre/1.13", verInfo);
         }
             /*
             BufferedReader reader = Resources.asCharSource(
@@ -288,7 +288,7 @@ public class VersionCommand extends BukkitCommand {
     private static int getFromJenkins(int currentVer) {
         try {
             BufferedReader reader = Resources.asCharSource(
-                    new URL("https://ci.destroystokyo.com/job/Paper/lastSuccessfulBuild/buildNumber"), // Paper
+                    new URL("https://ci.destroystokyo.com/job/Paper-1.13/lastSuccessfulBuild/buildNumber"), // Paper
                     Charsets.UTF_8
             ).openBufferedStream();
             try {
@@ -307,10 +307,9 @@ public class VersionCommand extends BukkitCommand {
     }
 
     // Contributed by Techcable <Techcable@outlook.com> in GH PR #65
-    private static final String BRANCH = "master";
-    private static int getFromRepo(String repo, String hash) {
+    private static int getFromRepo(String repo, String branch, String hash) {
         try {
-            HttpURLConnection connection = (HttpURLConnection) new URL("https://api.github.com/repos/" + repo + "/compare/" + BRANCH + "..." + hash).openConnection();
+            HttpURLConnection connection = (HttpURLConnection) new URL("https://api.github.com/repos/" + repo + "/compare/" + branch + "..." + hash).openConnection();
             connection.connect();
             if (connection.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) return -2; // Unknown commit
             try (
-- 
2.18.0

