From ed8a733093c4111805af091a2c46bb57aa5c07aa Mon Sep 17 00:00:00 2001
From: miclebrick <miclebrick@outlook.com>
Date: Sat, 1 Sep 2018 12:28:16 -0400
Subject: [PATCH] Custom plugin advancement API


diff --git a/src/main/java/com/destroystokyo/paper/advancement/AdvancementData.java b/src/main/java/com/destroystokyo/paper/advancement/AdvancementData.java
new file mode 100644
index 00000000..fac732dd
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/advancement/AdvancementData.java
@@ -0,0 +1,213 @@
+package com.destroystokyo.paper.advancement;
+
+import org.bukkit.Material;
+import org.bukkit.inventory.ItemStack;
+
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Set;
+
+public class AdvancementData {
+    private final Material parent;
+    private final AdvancementDisplay display;
+    private final AdvancementRewards rewards;
+    private final Material key;
+
+    public AdvancementData(@Nullable Material parent, AdvancementDisplay display, AdvancementRewards rewards, Material key) {
+        this.parent = parent;
+        this.display = display;
+        this.rewards = rewards;
+        this.key = key;
+    }
+
+    public Material getKey() {
+        return key;
+    }
+
+    /**
+     * The optional parent advancement of this advancement (does not apply for the root advancement).
+     * @return The parent advancement if present otherwise null
+     */
+    @Nullable
+    public Material getParent() {
+        return parent;
+    }
+
+    /**
+     * The optional display data. Controls how the advancement looks for the client.
+     * @return Display data if present otherwise null
+     */
+    @Nullable
+    public AdvancementDisplay getDisplay() {
+        return display;
+    }
+
+    /**
+     * Required rewards object which gives rewards when a player completes the advancement.
+     * Can be used to define a custom function for executing commands (which can then execute code..)
+     * @return The rewards
+     */
+    public AdvancementRewards getRewards() {
+        return rewards;
+    }
+
+    /**
+     * Rewards class is used for giving rewards to players upon completing an advancement.
+     */
+    public static class AdvancementRewards {
+        private int experience;
+        private Set<Material> loot;
+        private Set<Material> recipes;
+        @Nullable
+        private Material customFunction;
+
+        public AdvancementRewards(int experience) {
+            this(experience, new HashSet<>(), new HashSet<>(), null);
+        }
+
+        public AdvancementRewards(int experience, Material customFunction) {
+            this(experience, new HashSet<>(), new HashSet<>(), customFunction);
+        }
+
+        public AdvancementRewards(Material customFunction) {
+            this(0, new HashSet<>(), new HashSet<>(), customFunction);
+        }
+
+        public AdvancementRewards(int experience, Set<Material> loot, Set<Material> recipes, @Nullable Material customFunction) {
+            this.experience = experience;
+            this.loot = loot;
+            this.recipes = recipes;
+            this.customFunction = customFunction;
+        }
+
+        /**
+         * Amount of experience to give
+         * @return The defined amount of XP
+         */
+        public int getExperience() {
+            return experience;
+        }
+
+        /**
+         * Set of Minecraft IDs for item loot
+         * @return The defined loot items
+         */
+        public Set<Material> getLoot() {
+            return loot;
+        }
+
+        /**
+         * Set of minecraft IDs for recipes
+         * @return The defined recipes
+         */
+        public Set<Material> getRecipes() {
+            return recipes;
+        }
+
+        /**
+         * Functions are text files with the extension .mcfunction in the datapacks folder of your main world
+         * @return The defined custom function
+         */
+        @Nullable
+        public Material getCustomFunction() {
+            return customFunction;
+        }
+    }
+
+    /**
+     * Display class is used for controlling how the advancement appears to the client
+     */
+    public static class AdvancementDisplay {
+        private String title;
+        private String description;
+        private ItemStack icon;
+        private Material background;
+        private DisplayFrameType frameType;
+        private boolean showToast;
+        private boolean announceToChat;
+        private boolean hidden;
+
+        public AdvancementDisplay(String title, String description, ItemStack icon, Material background, DisplayFrameType frameType, boolean showToast, boolean announceToChat, boolean hidden) {
+            this.title = title;
+            this.description = description;
+            this.icon = icon;
+            this.background = background;
+            this.frameType = frameType;
+            this.showToast = showToast;
+            this.announceToChat = announceToChat;
+            this.hidden = hidden;
+        }
+
+        /**
+         * The title of the advancement
+         * @return The defined title
+         */
+        public String getTitle() {
+            return title;
+        }
+
+        /**
+         * The description of the advancement
+         * @return The defined description
+         */
+        public String getDescription() {
+            return description;
+        }
+
+        /**
+         * The item to show as the icon
+         * @return The defined item
+         */
+        public ItemStack getIcon() {
+            return icon;
+        }
+
+        /**
+         * The item type to use as the background
+         * @return The material to be used for the background
+         */
+        @Nullable
+        public Material getBackground() {
+            return background;
+        }
+
+        /**
+         * The shape of the display frame
+         * @return The defined type
+         */
+        public DisplayFrameType getFrameType() {
+            return frameType;
+        }
+
+        /**
+         * Whether to show a toast notification upon completion
+         * @return Boolean, true if should show
+         */
+        public boolean showToast() {
+            return showToast;
+        }
+
+        /**
+         * Whether to announce to chat when someone completes this advancement
+         * @return Boolean, true if should announce
+         */
+        public boolean announceToChat() {
+            return announceToChat;
+        }
+
+        /**
+         * Whether to hide this in the advancement screen
+         * @return Boolean, true if should hide
+         */
+        public boolean isHidden() {
+            return hidden;
+        }
+    }
+
+    /**
+     * Enum used for frame type in advancement menu
+     */
+    public enum DisplayFrameType {
+        TASK, CHALLENGE, GOAL
+    }
+}
diff --git a/src/main/java/com/destroystokyo/paper/advancement/AdvancementManager.java b/src/main/java/com/destroystokyo/paper/advancement/AdvancementManager.java
new file mode 100644
index 00000000..f2ecbb2e
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/advancement/AdvancementManager.java
@@ -0,0 +1,22 @@
+package com.destroystokyo.paper.advancement;
+
+import org.bukkit.plugin.Plugin;
+
+/**
+ * Manager of Advancements
+ */
+public interface AdvancementManager {
+    /**
+     * Clear a plugin's custom advancements
+     * @param plugin the plugin to clear the advancements of
+     */
+    void clearAdvancements(Plugin plugin);
+
+    /**
+     * Registers an advancement to the plugin.
+     * Note: Changes do not take effect until the next tick or Bukkit.reloadData() is called
+     * @param plugin the plugin to register the advancement to
+     * @param advancementData the advancement data to register
+     */
+    void registerAdvancement(Plugin plugin, AdvancementData advancementData);
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2912bdae..1d53dc4e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -863,6 +863,16 @@ public interface Server extends PluginMessageRecipient {
      */
     ScoreboardManager getScoreboardManager();
 
+    // Paper start - custom advancement API
+
+    /**
+     * Gets the instance of the advancement manager.
+     *
+     * @return the advancement manager
+     */
+    com.destroystokyo.paper.advancement.AdvancementManager getAdvancementManager();
+    // Paper end
+
     /**
      * Gets an instance of the server's default server-icon.
      *
diff --git a/src/main/java/org/bukkit/advancement/AdvancementProgress.java b/src/main/java/org/bukkit/advancement/AdvancementProgress.java
index 46b6637d..866508f2 100644
--- a/src/main/java/org/bukkit/advancement/AdvancementProgress.java
+++ b/src/main/java/org/bukkit/advancement/AdvancementProgress.java
@@ -1,5 +1,7 @@
 package org.bukkit.advancement;
 
+import org.bukkit.advancement.Advancement;
+
 import java.util.Collection;
 import java.util.Date;
 
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index cb2b0b9c..37565bb3 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -468,6 +468,7 @@ public final class SimplePluginManager implements PluginManager {
                 handlePluginException("Error occurred (in the plugin loader) while unregistering plugin channels for "
                         + plugin.getDescription().getFullName() + " (Is it up to date?)", ex, plugin); // Paper
             }
+            server.getAdvancementManager().clearAdvancements(plugin); // Paper - custom advancements API
         }
     }
 
-- 
2.17.1

