From 0c385de7f4c7f5f079a56b7ee658c65c026eed9a Mon Sep 17 00:00:00 2001
From: Mafkees92 <unconfigured@null.spigotmc.org>
Date: Wed, 12 Aug 2020 23:29:19 +0200
Subject: [PATCH] test change player


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fac5ded7..be11d315 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -37,6 +37,8 @@ import org.bukkit.scoreboard.Scoreboard;
  */
 public interface Player extends HumanEntity, Conversable, CommandSender, OfflinePlayer, PluginMessageRecipient, com.destroystokyo.paper.network.NetworkClient { // Paper - Extend NetworkClient
 
+	
+	public void sayWelcome();
     /**
      * Gets the "friendly" name to display of this player. This may include
      * color.
-- 
2.27.0.windows.1

