From 9e401d6a50f9c84e1ee21dbe52e61e9559d8a276 Mon Sep 17 00:00:00 2001
From: Mafkees92 <unconfigured@null.spigotmc.org>
Date: Wed, 12 Aug 2020 23:56:21 +0200
Subject: [PATCH] test


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index be11d315..de3d6bc7 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -39,6 +39,8 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
 
 	
 	public void sayWelcome();
+	public void sayWelcomeTest();
+	
     /**
      * Gets the "friendly" name to display of this player. This may include
      * color.
-- 
2.27.0.windows.1

