From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 15 Aug 2018 01:04:58 -0400
Subject: [PATCH] Ability to get Tile Entities from a chunk without snapshots


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index fa576096e908f8fbdbef53e1bd91215ac9e73ed6..b4ef6297f78d1f0c216e718024a21e6aa07cd1c6 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -103,13 +103,26 @@ public interface Chunk extends PersistentDataHolder {
     @NotNull
     Entity[] getEntities();
 
+    // Paper start
+    /**
+     * Get a list of all tile entities in the chunk.
+     *
+     * @return The tile entities.
+     */
+    @NotNull
+    default BlockState[] getTileEntities() {
+        return getTileEntities(true);
+    }
+
     /**
      * Get a list of all tile entities in the chunk.
      *
+     * @param useSnapshot Take snapshots or direct references
      * @return The tile entities.
      */
     @NotNull
-    BlockState[] getTileEntities();
+    BlockState[] getTileEntities(boolean useSnapshot);
+    // Paper end
 
     /**
      * Checks if the chunk is loaded.
