From a2a0bb90cb0867e746c4ccb41031f58319a15d0e Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Tue, 4 Sep 2018 15:01:54 -0500
Subject: [PATCH] Expose attack cooldown methods for Player


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ddf226a4..db6bcae5 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1883,6 +1883,26 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param profile The new profile to use
      */
     void setPlayerProfile(PlayerProfile profile);
+
+    /**
+     * Returns the amount of ticks the current cooldown lasts
+     *
+     * @return Amount of ticks cooldown will last
+     */
+    float getCooldownPeriod();
+
+    /**
+     * Returns the percentage of attack power available based on the cooldown (zero to one).
+     *
+     * @param adjustTicks Amount of ticks to add to cooldown counter for this calculation
+     * @return Percentage of attack power available
+     */
+    float getCooledAttackStrength(float adjustTicks);
+
+    /**
+     * Reset the cooldown counter to 0, effectively starting the cooldown period.
+     */
+    void resetCooldown();
     // Paper end
 
     // Spigot start
-- 
2.19.0.windows.1

