From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 22 Sep 2018 18:41:01 -0400
Subject: [PATCH] Remove Precondition on name for AttributeModifier

Vanilla allows empty names

diff --git a/src/main/java/org/bukkit/attribute/AttributeModifier.java b/src/main/java/org/bukkit/attribute/AttributeModifier.java
index a97631474545bb51dfcbc769401f4f1ffede580d..f0dff145a8c5ae729de5e5f547fff60333c6501e 100644
--- a/src/main/java/org/bukkit/attribute/AttributeModifier.java
+++ b/src/main/java/org/bukkit/attribute/AttributeModifier.java
@@ -32,10 +32,10 @@ public class AttributeModifier implements ConfigurationSerializable {
 
     public AttributeModifier(@NotNull UUID uuid, @NotNull String name, double amount, @NotNull Operation operation, @Nullable EquipmentSlot slot) {
         Validate.notNull(uuid, "UUID cannot be null");
-        Validate.notEmpty(name, "Name cannot be empty");
+        //Validate.notEmpty(name, "Name cannot be empty"); // Paper
         Validate.notNull(operation, "Operation cannot be null");
         this.uuid = uuid;
-        this.name = name;
+        this.name = name != null ? name : ""; // Paper
         this.amount = amount;
         this.operation = operation;
         this.slot = slot;
