From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 6 Oct 2018 21:14:29 -0400
Subject: [PATCH] Material API additions


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index c7316bd2ef89e63ced9f8d37f0936f1551c1157d..769c65e7928be5e0a15e592c874fc68cc6d30148 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -85,6 +85,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * An enum of all material IDs accepted by the official server and client
  */
+@SuppressWarnings({"DeprecatedIsStillUsed", "deprecation"}) // Paper
 public enum Material implements Keyed {
     //<editor-fold desc="Materials" defaultstate="collapsed">
     ACACIA_BOAT(27326, 1),
@@ -3563,6 +3564,22 @@ public enum Material implements Keyed {
         }
     }
 
+    // Paper start
+
+    /**
+     * @return If the type is either AIR, CAVE_AIR or VOID_AIR
+     */
+    public boolean isEmpty() {
+        switch (this) {
+            case AIR:
+            case CAVE_AIR:
+            case VOID_AIR:
+                return true;
+        }
+        return false;
+    }
+    // Paper end
+
     /**
      * Do not use for any reason.
      *
