From 57018093cd49e64641888299c3c0645fe7d157e3 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 7 Oct 2018 00:54:15 -0500
Subject: [PATCH] Add sun related API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 668c9b7a..67e062b8 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1446,6 +1446,16 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public void setFullTime(long time);
 
+    // Paper start
+
+    /**
+     * Check if it is currently daytime in this world
+     *
+     * @return True if it is daytime
+     */
+    public boolean isDayTime();
+    // Paper end
+
     /**
      * Returns whether the world has an ongoing storm.
      *
diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index 48eddcd3..1b108a4a 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -13,6 +13,13 @@ public interface Mob extends LivingEntity, Lootable {
      * @return Pathfinding Manager for this entity
      */
     com.destroystokyo.paper.entity.Pathfinder getPathfinder();
+
+    /**
+     * Check if this mob is exposed to daylight
+     *
+     * @return True if mob is exposed to daylight
+     */
+    boolean isInDaylight();
     // Paper end
 
     /**
-- 
2.20.1

