From 7401b4dc7ba58b3d5d00428a19bc05f18a3e74c5 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 7 Oct 2018 00:54:15 -0500
Subject: [PATCH] Add sun related API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 38dae649..107f4173 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1517,6 +1517,16 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public void setFullTime(long time);
 
+    // Paper start
+
+    /**
+     * Check if it is currently daytime in this world
+     *
+     * @return True if it is daytime
+     */
+    public boolean isDayTime();
+    // Paper end
+
     /**
      * Returns whether the world has an ongoing storm.
      *
diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index afdc103f..784db447 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -16,6 +16,13 @@ public interface Mob extends LivingEntity, Lootable {
      */
     @NotNull
     com.destroystokyo.paper.entity.Pathfinder getPathfinder();
+
+    /**
+     * Check if this mob is exposed to daylight
+     *
+     * @return True if mob is exposed to daylight
+     */
+    boolean isInDaylight();
     // Paper end
 
     /**
-- 
2.21.0

