From 51801f043e701cadcef0689f4d17747eea5684dd Mon Sep 17 00:00:00 2001
From: MTM123 <matiss234@gmail.com>
Date: Mon, 4 Feb 2019 23:06:34 +0200
Subject: [PATCH] Add damaged slot to PlayerItemDamageEvent


diff --git a/src/main/java/org/bukkit/event/player/PlayerItemDamageEvent.java b/src/main/java/org/bukkit/event/player/PlayerItemDamageEvent.java
index 0ece784a..4f054dd5 100644
--- a/src/main/java/org/bukkit/event/player/PlayerItemDamageEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerItemDamageEvent.java
@@ -14,14 +14,25 @@ public class PlayerItemDamageEvent extends PlayerEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
     private final ItemStack item;
     private int damage;
+    private final int slot; //Paper - add damaged slot
     private boolean cancelled = false;
 
     public PlayerItemDamageEvent(Player player, ItemStack what, int damage) {
         super(player);
         this.item = what;
         this.damage = damage;
+        this.slot = -1;
     }
 
+    //Paper start
+    public PlayerItemDamageEvent(Player player, ItemStack what, int damage, int slot){
+        super(player);
+        this.item = what;
+        this.damage = damage;
+        this.slot = slot;
+    }
+    //Paper end
+
     /**
      * Gets the item being damaged.
      *
@@ -44,6 +55,17 @@ public class PlayerItemDamageEvent extends PlayerEvent implements Cancellable {
         this.damage = damage;
     }
 
+    //Paper start
+    /**
+     * Gets the slot where the item took durability damage
+     *
+     * @return inventory slot or -1 if the slot is unknown
+     */
+    public int getSlot() {
+        return slot;
+    }
+    //Paper end
+
     @Override
     public boolean isCancelled() {
         return cancelled;
-- 
2.19.0.windows.1

