From 02cffbac63eaa817faab80430412fe9c61de135a Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Fri, 5 Apr 2019 02:23:38 -0700
Subject: [PATCH] Add API for retreiving the itemstack used to launch a
 projectile


diff --git a/src/main/java/org/bukkit/entity/Projectile.java b/src/main/java/org/bukkit/entity/Projectile.java
index c854860c..a6e28ebc 100644
--- a/src/main/java/org/bukkit/entity/Projectile.java
+++ b/src/main/java/org/bukkit/entity/Projectile.java
@@ -1,5 +1,6 @@
 package org.bukkit.entity;
 
+import org.bukkit.inventory.ItemStack;
 import org.bukkit.projectiles.ProjectileSource;
 import org.jetbrains.annotations.Nullable;
 
@@ -39,4 +40,10 @@ public interface Projectile extends Entity {
      * @param doesBounce whether or not it should bounce.
      */
     public void setBounce(boolean doesBounce);
+
+    /**
+     * @return A copy of the item used to launch this projectile, or {@code null} if there was no item
+     */
+    @Nullable
+    public ItemStack getLaunchedItem();
 }
-- 
2.21.0

