From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: simpleauthority <jacob@algorithmjunkie.com>
Date: Thu, 11 Jul 2019 10:35:56 -0700
Subject: [PATCH] Amend PlayerInteractAtEntityEvent javadoc for ArmorStands


diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
index 1075dbb8135d2fd2fd69d1e821dd1ba7f90824d5..3f24d302e28170f7f6e5885b5b9abb22cbbb7d66 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
@@ -13,6 +13,9 @@ import org.jetbrains.annotations.NotNull;
  * <br>
  * Note that the client may sometimes spuriously send this packet in addition to {@link PlayerInteractEntityEvent}.
  * Users are advised to listen to this (parent) class unless specifically required.
+ * <br>
+ * Note that interacting with Armor Stands fires this event only and not its parent and as such users are expressly required
+ * to listen to this event for that scenario.
  */
 public class PlayerInteractAtEntityEvent extends PlayerInteractEntityEvent {
     private static final HandlerList handlers = new HandlerList();
