From d5f83fc5905e609069e33d12b25bd147d2ca15a0 Mon Sep 17 00:00:00 2001
From: slicklibro <slicklibro@gmail.com>
Date: Sat, 28 Sep 2019 00:01:23 +1200
Subject: [PATCH] Ticket-level per player view distance API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8d8d8957..efcd3c35 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1861,6 +1861,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     /**
      * Gets the view distance for this player
+     * <br>
+     * This determines how far the player can add tickets to chunks.
      *
      * @return the player's view distance
      */
@@ -1868,10 +1870,30 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     /**
      * Sets the view distance for this player
+     * <br>
+     * This determines how far the player can add tickets to chunks.
      *
      * @param viewDistance the player's view distance
      */
     public void setViewDistance(int viewDistance);
+
+    /**
+     * Gets the tracking view distance for this player
+     * <br>
+     * This determines how far the player can see/load all chunks with tickets, allowing shared view distance.
+     *
+     * @return the player's tracking view distance
+     */
+    public int getTrackViewDistance();
+
+    /**
+     * Sets the tracking view distance for this player
+     * <br>
+     * This determines how far the player can see/load all chunks with tickets, allowing shared view distance.
+     *
+     * @param trackViewDistance the player's tracking view distance
+     */
+    public void setTrackViewDistance(int trackViewDistance);
     // Paper end
 
     /**
-- 
2.20.1.windows.1

