From 1b3b055f0c4aabdcc8d4edde81898d16a9cc02c6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jakub=20Zag=C3=B3rski?= <jaqobbdev@gmail.com>
Date: Mon, 7 Oct 2019 19:23:48 +0200
Subject: [PATCH] Thrown projectiles display item methods


diff --git a/src/main/java/com/destroystokyo/paper/entity/projectile/ThrownProjectile.java b/src/main/java/com/destroystokyo/paper/entity/projectile/ThrownProjectile.java
new file mode 100644
index 00000000..f1d862c5
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/entity/projectile/ThrownProjectile.java
@@ -0,0 +1,31 @@
+package com.destroystokyo.paper.entity.projectile;
+
+import org.bukkit.Material;
+import org.bukkit.entity.Projectile;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a thrown projectile.
+ */
+public interface ThrownProjectile extends Projectile {
+
+    /**
+     * Get this thrown projectile's current display item.
+     *
+     * @return This thrown projectile's current display item.
+     */
+    @NotNull
+    ItemStack getDisplayItem();
+
+    /**
+     * Set this thrown projectile's display item. If the new display item
+     * is null or its type is {@link Material#isEmpty}, the display item
+     * will be set to the default item (snowball if the projectile was
+     * originally a snowball, ender pearl if it was an ender pearl, etc.).
+     *
+     * @param displayItem The new display item.
+     */
+    void setDisplayItem(@Nullable ItemStack displayItem);
+}
diff --git a/src/main/java/org/bukkit/entity/Egg.java b/src/main/java/org/bukkit/entity/Egg.java
index 2dcc00b9..6df40f5d 100644
--- a/src/main/java/org/bukkit/entity/Egg.java
+++ b/src/main/java/org/bukkit/entity/Egg.java
@@ -3,4 +3,4 @@ package org.bukkit.entity;
 /**
  * Represents a thrown egg.
  */
-public interface Egg extends Projectile {}
+public interface Egg extends com.destroystokyo.paper.entity.projectile.ThrownProjectile {} // Paper
diff --git a/src/main/java/org/bukkit/entity/EnderPearl.java b/src/main/java/org/bukkit/entity/EnderPearl.java
index db18a90b..1b17eb2c 100644
--- a/src/main/java/org/bukkit/entity/EnderPearl.java
+++ b/src/main/java/org/bukkit/entity/EnderPearl.java
@@ -3,6 +3,6 @@ package org.bukkit.entity;
 /**
  * Represents a thrown Ender Pearl entity
  */
-public interface EnderPearl extends Projectile {
+public interface EnderPearl extends com.destroystokyo.paper.entity.projectile.ThrownProjectile { // Paper
 
 }
diff --git a/src/main/java/org/bukkit/entity/Snowball.java b/src/main/java/org/bukkit/entity/Snowball.java
index 8c6b4333..c7f80805 100644
--- a/src/main/java/org/bukkit/entity/Snowball.java
+++ b/src/main/java/org/bukkit/entity/Snowball.java
@@ -3,4 +3,4 @@ package org.bukkit.entity;
 /**
  * Represents a snowball.
  */
-public interface Snowball extends Projectile {}
+public interface Snowball extends com.destroystokyo.paper.entity.projectile.ThrownProjectile {} // Paper
diff --git a/src/main/java/org/bukkit/entity/ThrownExpBottle.java b/src/main/java/org/bukkit/entity/ThrownExpBottle.java
index 671282ed..2a6e0c7b 100644
--- a/src/main/java/org/bukkit/entity/ThrownExpBottle.java
+++ b/src/main/java/org/bukkit/entity/ThrownExpBottle.java
@@ -3,6 +3,6 @@ package org.bukkit.entity;
 /**
  * Represents a thrown Experience bottle.
  */
-public interface ThrownExpBottle extends Projectile {
+public interface ThrownExpBottle extends Projectile { // Paper
 
 }
-- 
2.23.0.windows.1

