From 08e9770903dc7026a0c1cb1ec9df1a180d37e244 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sun, 2 Feb 2020 03:28:56 -0600
Subject: [PATCH] Implement tick times API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 388c2bbf..ec59e4c0 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1603,6 +1603,25 @@ public final class Bukkit {
     public static int getCurrentTick() {
         return server.getCurrentTick();
     }
+
+    /**
+     * Get a sample of the servers last tick times (in nanos)
+     *
+     * @return A sample of the servers last tick times (in nanos)
+     */
+    @NotNull
+    public static long @NotNull[] getTickTimes() {
+        return server.getTickTimes();
+    }
+
+    /**
+     * Get the average tick time (in millis) of the last 100 ticks
+     *
+     * @return Average tick time (in millis)
+     */
+    public static double getAverageTickTime() {
+        return server.getAverageTickTime();
+    }
     // Paper end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a1371730..1d8f1d19 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1403,5 +1403,20 @@ public interface Server extends PluginMessageRecipient {
      * @return Current tick
      */
     int getCurrentTick();
+
+    /**
+     * Get a sample of the servers last tick times (in nanos)
+     *
+     * @return A sample of the servers last tick times (in nanos)
+     */
+    @NotNull
+    long @NotNull[] getTickTimes();
+
+    /**
+     * Get the average tick time (in millis) of the last 100 ticks
+     *
+     * @return Average tick time (in millis)
+     */
+    double getAverageTickTime();
     // Paper end
 }
-- 
2.24.0

