From d11f9d96a0d801a97709ace16030dc233eb58fb9 Mon Sep 17 00:00:00 2001
From: TheVIperShow <thevipershowita@gmail.com>
Date: Sun, 29 Mar 2020 21:04:03 +0200
Subject: [PATCH] Increased TPS averages number and improved tps command


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index b9973406..e0833f56 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1412,6 +1412,18 @@ public final class Bukkit {
     public static double[] getTPS() {
         return server.getTPS();
     }
+
+    /**
+     * Returns the current server TPS from the specified time interval
+     * the server only keeps track of the last 30 minutes
+     *
+     * @param indexes any integer from 0 to 29
+     * @return current server TPS.
+     */
+    @NotNull
+    public static double[] getTPS(@NotNull int...indexes) {
+        return server.getTPS(indexes);
+    }
     // Paper end
 
     /**
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 80f9abdc..ab30e9ef 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1188,6 +1188,20 @@ public interface Server extends PluginMessageRecipient {
      */
     @NotNull
     public double[] getTPS();
+
+    /**
+     * Provide the average tps at from the current indexes
+     * the index represent the minutes, for example:
+     * 2,5,10 , will provide the the TPS averages from
+     * two , five and ten minutes ago.
+     * this method only saves 30 minutes so the biggest that
+     * can be used is 29
+     *
+     * @param indexes any value from 0 to 29
+     * @return a double array that represents the TPS averages at the specified indexes.
+     */
+    @NotNull
+    public double[] getTPS(@NotNull int... indexes);
     // Paper end
 
     // Paper start
-- 
2.21.1 (Apple Git-122.3)

