From 74cde8a498b882475b236e27fc34adeee51c7973 Mon Sep 17 00:00:00 2001
From: TheViperShow <29604693+TheViperShow@users.noreply.github.com>
Date: Wed, 22 Apr 2020 09:40:23 +0200
Subject: [PATCH] Implemented BlockFailedDispenseEvent


diff --git a/src/main/java/com/destroystokyo/paper/event/block/BlockFailedDispenseEvent.java b/src/main/java/com/destroystokyo/paper/event/block/BlockFailedDispenseEvent.java
new file mode 100644
index 00000000..99ab7b65
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/block/BlockFailedDispenseEvent.java
@@ -0,0 +1,26 @@
+package com.destroystokyo.paper.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when a block tries to dispense an item, but its inventory is empty.
+ */
+public class BlockFailedDispenseEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+
+    public BlockFailedDispenseEvent(@NotNull Block theBlock) {
+        super(theBlock);
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
-- 
2.26.0.windows.1
