From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jan Tuck <jan@tuck.dk>
Date: Mon, 22 Jun 2020 00:29:10 +0200
Subject: [PATCH] Expose the minecraftkey associated with the biome that is
 associated with the chunk.


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 11b691cf59b4a72493c91055fd8c46d7681b4527..32d7d01e8d7846b4d4b66f9b45e51fc7fc4a3051 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -7,6 +7,7 @@ import org.bukkit.block.data.BlockData;
 import org.bukkit.entity.Entity;
 import org.bukkit.plugin.Plugin;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable; // Paper
 
 /**
  * Represents a chunk of blocks
@@ -262,4 +263,14 @@ public interface Chunk {
      * @return if the block is contained within
      */
     boolean contains(@NotNull BlockData block);
+
+    // Paper - Start
+    /**
+     * Get the namespacedkey of the biome the chunk is in.
+     * May be slightly inconsistent to what the client shows in the f3 menu.
+     * @return the namespaced key associated with this chunk, e.g. minecraft:taiga_mountains
+     */
+    @Nullable
+    NamespacedKey getNamespacedKey();
+    // Paper - End
 }
