From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ysl3000 <yannicklamprecht@live.de>
Date: Sun, 28 Jun 2020 21:34:45 +0200
Subject: [PATCH] Create HoverEvent from ItemStack Entity


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 329e2dd1c057788d6bb5a1adc7ee9282820e927d..56b38068df513888a829cec1b5e98380b0857134 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -13,6 +13,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+import net.md_5.bungee.api.chat.BaseComponent; // Paper
+import net.md_5.bungee.api.chat.HoverEvent; // Paper
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -1743,6 +1745,45 @@ public final class Bukkit {
     public static com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
         return server.getMobGoals();
     }
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that ItemStack for displaying.
+     *
+     * @param itemStack ItemStack to create the HoverEvent for
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that ItemStack
+     */
+
+    @NotNull
+    public static HoverEvent hoverEventOf(@NotNull ItemStack itemStack) {
+        return server.hoverEventOf(itemStack);
+    }
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity} for displaying.
+     *
+     * @param entity Entity to create the HoverEvent for
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     *
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity}
+     */
+
+    @NotNull
+    public static HoverEvent hoverEventOf(@NotNull Entity entity, @Nullable String customName) {
+        return server.hoverEventOf(entity, customName);
+    }
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity} for displaying.
+     *
+     * @param entity Entity to create the HoverEvent for
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     *
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity}
+     */
+    @NotNull
+    public static HoverEvent hoverEventOf(@NotNull Entity entity, @Nullable BaseComponent customName) {
+        return server.hoverEventOf(entity, customName);
+    }
     // Paper end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b464cc1a70aca3450a5de84a657a1e069a1a8ab6..e3cf8f64630aad90ac6cb8e0fe07d5e1441d5fa5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -13,6 +13,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+import net.md_5.bungee.api.chat.BaseComponent; // Paper
+import net.md_5.bungee.api.chat.HoverEvent; // Paper
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -1523,5 +1525,36 @@ public interface Server extends PluginMessageRecipient {
      */
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that ItemStack for displaying.
+     *
+     * @param itemStack
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that ItemStack
+     */
+    @NotNull
+    HoverEvent hoverEventOf(@NotNull ItemStack itemStack);
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity} for displaying.
+     *
+     * @param entity Entity to create the HoverEvent for
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     *
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity}
+     */
+    @NotNull
+    HoverEvent hoverEventOf(@NotNull Entity entity, @Nullable String customName);
+
+    /**
+     * Creates a {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity} for displaying.
+     *
+     * @param entity Entity to create the HoverEvent for
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     *
+     * @return the {@link net.md_5.bungee.api.chat.HoverEvent} of that {@link Entity}
+     */
+    @NotNull
+    HoverEvent hoverEventOf(@NotNull Entity entity, @Nullable BaseComponent customName);
     // Paper end
 }
