From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Wed, 5 Aug 2020 00:10:45 +0200
Subject: [PATCH] Add Switch.Face <-> FaceAttachable.AttachedFace converters


diff --git a/src/main/java/org/bukkit/block/data/FaceAttachable.java b/src/main/java/org/bukkit/block/data/FaceAttachable.java
index 9599e1237b9717ddbf84c3738bf6c1293e8b3c54..8c2341b03dc9f5fe62ae67d5828ed51d698de73a 100644
--- a/src/main/java/org/bukkit/block/data/FaceAttachable.java
+++ b/src/main/java/org/bukkit/block/data/FaceAttachable.java
@@ -1,5 +1,6 @@
 package org.bukkit.block.data;
 
+import org.bukkit.block.data.type.Switch;
 import org.jetbrains.annotations.NotNull;
 
 /**
@@ -41,5 +42,25 @@ public interface FaceAttachable extends BlockData {
          * The switch is mounted to the ceiling and pointing dowanrds.
          */
         CEILING;
+
+        // Paper start
+        /**
+         * @return The {@link Switch.Face} equivalent to this.
+         * @deprecated Prefer {@link AttachedFace} over {@link Switch.Face}.
+         */
+        @Deprecated
+        public Switch.Face toSwitchFace() {
+            switch (this) {
+                case FLOOR:
+                    return Switch.Face.FLOOR;
+                case WALL:
+                    return Switch.Face.WALL;
+                case CEILING:
+                    return Switch.Face.CEILING;
+            }
+
+            throw new IllegalStateException("Unknown face: " + this);
+        }
+        // Paper end
     }
 }
diff --git a/src/main/java/org/bukkit/block/data/type/Switch.java b/src/main/java/org/bukkit/block/data/type/Switch.java
index be06f8db02ca41d5cc3a5dc02951ad27e3cc8f9d..ddfa0edb8e5c8718b10d5040d5f619134d98e45a 100644
--- a/src/main/java/org/bukkit/block/data/type/Switch.java
+++ b/src/main/java/org/bukkit/block/data/type/Switch.java
@@ -45,5 +45,25 @@ public interface Switch extends Directional, FaceAttachable, Powerable {
          * The switch is mounted to the ceiling and pointing dowanrds.
          */
         CEILING;
+
+        // Paper start
+        /**
+         * @return The {@link AttachedFace} equivalent to this.
+         * @deprecated Prefer {@link AttachedFace} over {@link Switch.Face}.
+         */
+        @Deprecated
+        public AttachedFace toAttachedFace() {
+            switch (this) {
+                case FLOOR:
+                    return AttachedFace.FLOOR;
+                case WALL:
+                    return AttachedFace.WALL;
+                case CEILING:
+                    return AttachedFace.CEILING;
+            }
+
+            throw new IllegalStateException("Unknown face: " + this);
+        }
+        // Paper end
     }
 }
