From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DigitalRegent <misterwener@gmail.com>
Date: Mon, 6 Apr 2020 20:30:09 +0200
Subject: [PATCH] Brand support


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index b584f8b15f7ee36b42bba0e0bae721aae8f6f14b..991f757a89d2e917186bb152a2f92bad8b6c2897 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1873,6 +1873,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         }
     }
 
+    // Paper start - brand support
+    /**
+     * Returns player's client brand name. If the client didn't send this information, the brand name will be null.<br>
+     * For the Notchian client this name defaults to <code>vanilla</code>. Some modified clients report other names such as <code>forge</code>.<br>
+     * @return client brand name
+     */
+    @Nullable
+    String getClientBrandName();
+    // Paper end
+
     @NotNull
     @Override
     Spigot spigot();
