From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Josh Roy <10731363+JRoy@users.noreply.github.com>
Date: Sat, 12 Sep 2020 19:09:49 +0000
Subject: [PATCH] Add Server#isStarting & Bukkit#isStarting API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index bfe842364ee0a4bf39dacdbb6972477d57a4ef8a..7b382f5ba33ed52a27cc08b08de76de463b2ad36 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1772,6 +1772,15 @@ public final class Bukkit {
         return server.isStopping();
     }
 
+    /**
+     * Checks if the server is in the process of being started.
+     * 
+     * @return true if server is in the process of being started
+     */
+    public static boolean isStarting() {
+        return server.isStarting();
+    }
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7c0a788900c93c29d14d8c45ac5ae3317cf4a94e..4f5b37d489a7690de5641d9c3fcab1f9875d7b84 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1548,6 +1548,13 @@ public interface Server extends PluginMessageRecipient {
      */
     boolean isStopping();
 
+    /**
+     * Checks if the server is in the process of being started.
+     * 
+     * @return true if server is in the process of being started
+     */
+    boolean isStarting();
+
     /**
      * Returns the {@link com.destroystokyo.paper.entity.ai.MobGoals} manager
      *
