From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ineusia <ineusia@yahoo.com>
Date: Mon, 26 Oct 2020 11:37:48 -0500
Subject: [PATCH] Add Destroy Speed API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 7616c5601adee3cbe0e5f722646a2458b535ab77..6933fd6ad353a2d008c4a64c52a64bf36bd8035c 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -581,5 +581,16 @@ public interface Block extends Metadatable {
      */
     @NotNull
     String getTranslationKey();
+
+    /**
+     * Gets the speed at which this block will be destroyed by a given {@link ItemStack}
+     *
+     * <p>Default value is 1.0</p>
+     *
+     * @param itemStack {@link ItemStack} used to mine this Block
+     * @return the speed that this Block will be mined by the given {@link ItemStack}
+     */
+    @NotNull
+    float getDestroySpeed(@NotNull ItemStack itemStack);
     // Paper end
 }
