From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheFruxz <cedricspitzer@outlook.de>
Date: Wed, 6 Jan 2021 01:50:34 +0100
Subject: [PATCH] Removed redundant initializer and made final


diff --git a/src/main/java/org/bukkit/event/entity/EntityCreatePortalEvent.java b/src/main/java/org/bukkit/event/entity/EntityCreatePortalEvent.java
index 397baaefd2fb45ae9d50b7b9bd5276e6ad980f20..dc320d01b41f5d0ae219f686b4eb4e7f952068cb 100644
--- a/src/main/java/org/bukkit/event/entity/EntityCreatePortalEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityCreatePortalEvent.java
@@ -19,7 +19,7 @@ public class EntityCreatePortalEvent extends EntityEvent implements Cancellable
     private static final HandlerList handlers = new HandlerList();
     private final List<BlockState> blocks;
     private boolean cancelled = false;
-    private PortalType type = PortalType.CUSTOM;
+    private final PortalType type; // Paper | Removed redundant initializer (PortalType.CUSTOM) and became final
 
     public EntityCreatePortalEvent(@NotNull final LivingEntity what, @NotNull final List<BlockState> blocks, @NotNull final PortalType type) {
         super(what);
