From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CDFN <codefun@protonmail.com>
Date: Tue, 7 Jul 2020 17:53:23 +0200
Subject: [PATCH] Return chat component with empty text instead of throwing
 exception


diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index a4e3d526db2d17dc923cbe82e53d3c902d61e1f3..2448e70d75ae7a678c6befac4506c103edb78875 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -450,7 +450,7 @@ public abstract class InventoryView {
     /**
      * Get the title of this inventory window.
      *
-     * @return The title.
+     * @return The title or empty string when title is {@code null}. <!-- Paper -->
      */
     @NotNull
     public /*abstract*/ net.kyori.adventure.text.Component title() {
