From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Tue, 16 Mar 2021 02:24:34 -0500
Subject: [PATCH] Chunk.getChunkData API


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 98263d896f316983609432c45b85401a2692432d..3d1eb063de18c49c1e6358f043a86016b11cd29a 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -75,6 +75,29 @@ public interface Chunk extends PersistentDataHolder {
     @NotNull
     Block getBlock(int x, int y, int z);
 
+    // Paper start
+    /**
+     * Returns an array with blocks in the chunk.
+     * The order filled goes x, z, y.
+     *
+     * @return an array containing the blocks in the chunk
+     */
+    @NotNull
+    default Material[] getBlockArray() {
+        return this.getBlockArray(new Material[0]);
+    }
+
+    /**
+     * Fills the supplied array with blocks in the chunk.
+     * The order filled goes x, z, y.
+     *
+     * @param array the array to reuse
+     * @return a new array if the passed array is too small
+     */
+    @NotNull
+    Material[] getBlockArray(@NotNull Material[] array);
+    // Paper end
+
     /**
      * Capture thread-safe read-only snapshot of chunk data
      *
