From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Swedz <me@swedz.net>
Date: Sat, 29 May 2021 00:12:27 -0400
Subject: [PATCH] Expose Modified Enchantment Level API


diff --git a/src/main/java/org/bukkit/enchantments/Enchantment.java b/src/main/java/org/bukkit/enchantments/Enchantment.java
index 799813410f40589ba1d7d530a47d87f609764705..287653e9e2911fd8a0fbb18dbfaa84f9cdedf98d 100644
--- a/src/main/java/org/bukkit/enchantments/Enchantment.java
+++ b/src/main/java/org/bukkit/enchantments/Enchantment.java
@@ -329,7 +329,23 @@ public abstract class Enchantment implements Keyed {
      */
     @NotNull
     public abstract io.papermc.paper.enchantments.EnchantmentRarity getRarity();
-
+    
+    /**
+     * Gets the modified enchantment level minimum for the enchantment given the level
+     *
+     * @param level the enchantment level
+     * @return the minimum modified enchantment level
+     */
+    public abstract int getModifiedEnchantmentLevelMinimum(int level);
+    
+    /**
+     * Gets the modified enchantment level maximum for the enchantment given the level
+     *
+     * @param level the enchantment level
+     * @return the maximum modified enchantment level
+     */
+    public abstract int getModifiedEnchantmentLevelMaximum(int level);
+    
     /**
      * Gets the damage increase as a result of the level and entity category specified
      *
diff --git a/src/main/java/org/bukkit/enchantments/EnchantmentWrapper.java b/src/main/java/org/bukkit/enchantments/EnchantmentWrapper.java
index ca097db904db96e81c0ea4b06b445a13e9333cfe..f0763d4ddc54531ca1387a8b23338bee7c6b4f66 100644
--- a/src/main/java/org/bukkit/enchantments/EnchantmentWrapper.java
+++ b/src/main/java/org/bukkit/enchantments/EnchantmentWrapper.java
@@ -85,7 +85,19 @@ public class EnchantmentWrapper extends Enchantment {
     public io.papermc.paper.enchantments.EnchantmentRarity getRarity() {
         return getEnchantment().getRarity();
     }
-
+    
+    @Override
+    public int getModifiedEnchantmentLevelMinimum(int level)
+    {
+        return getEnchantment().getModifiedEnchantmentLevelMinimum(level);
+    }
+    
+    @Override
+    public int getModifiedEnchantmentLevelMaximum(int level)
+    {
+        return getEnchantment().getModifiedEnchantmentLevelMaximum(level);
+    }
+    
     @Override
     public float getDamageIncrease(int level, @NotNull org.bukkit.entity.EntityCategory entityCategory) {
         return getEnchantment().getDamageIncrease(level, entityCategory);
