--- a/net/minecraft/server/BiomeStorage.java
+++ b/net/minecraft/server/BiomeStorage.java
@@ -12,7 +12,7 @@
     public static final int a = 1 << BiomeStorage.e + BiomeStorage.e + BiomeStorage.f;
     public static final int b = (1 << BiomeStorage.e) - 1;
     public static final int c = (1 << BiomeStorage.f) - 1;
-    private final Registry<BiomeBase> g;
+    public final Registry<BiomeBase> g; // PAIL
     private final BiomeBase[] h;
 
     public BiomeStorage(Registry<BiomeBase> registry, BiomeBase[] abiomebase) {
@@ -87,4 +87,14 @@
 
         return this.h[i1 << BiomeStorage.e + BiomeStorage.e | j1 << BiomeStorage.e | l];
     }
+
+    // CraftBukkit start
+    public void setBiome(int i, int j, int k, BiomeBase biome) {
+        int l = i & BiomeStorage.b;
+        int i1 = MathHelper.clamp(j, 0, BiomeStorage.c);
+        int j1 = k & BiomeStorage.b;
+
+        this.h[i1 << BiomeStorage.e + BiomeStorage.e | j1 << BiomeStorage.e | l] = biome;
+    }
+    // CraftBukkit end
 }
