--- a/net/minecraft/server/BiomeStorage.java
+++ b/net/minecraft/server/BiomeStorage.java
@@ -77,7 +77,7 @@
         int[] aint = new int[this.f.length];
 
         for (int i = 0; i < this.f.length; ++i) {
-            aint[i] = IRegistry.BIOME.a((Object) this.f[i]);
+            aint[i] = IRegistry.BIOME.a(this.f[i]); // CraftBukkit - decompile error
         }
 
         return aint;
@@ -90,7 +90,7 @@
         for (int j = 0; j < i; ++j) {
             BiomeBase biomebase = abiomebase[j];
 
-            packetdataserializer.writeInt(IRegistry.BIOME.a((Object) biomebase));
+            packetdataserializer.writeInt(IRegistry.BIOME.a(biomebase)); // CraftBukkit - decompile error
         }
 
     }
@@ -107,4 +107,14 @@
 
         return this.f[i1 << BiomeStorage.d + BiomeStorage.d | j1 << BiomeStorage.d | l];
     }
+
+    // CraftBukkit start
+    public void setBiome(int i, int j, int k, BiomeBase biome) {
+        int l = i & BiomeStorage.b;
+        int i1 = MathHelper.clamp(j, 0, BiomeStorage.c);
+        int j1 = k & BiomeStorage.b;
+
+        this.f[i1 << BiomeStorage.d + BiomeStorage.d | j1 << BiomeStorage.d | l] = biome;
+    }
+    // CraftBukkit end
 }
