--- a/net/minecraft/server/BlockFluids.java
+++ b/net/minecraft/server/BlockFluids.java
@@ -111,14 +111,20 @@
                 Fluid fluid = world.getFluid(blockposition);
 
                 if (fluid.isSource()) {
-                    world.setTypeUpdate(blockposition, Blocks.OBSIDIAN.getBlockData());
-                    this.fizz(world, blockposition);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(world, blockposition, Blocks.OBSIDIAN.getBlockData())) {
+                        this.fizz(world, blockposition);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
 
                 if (fluid.getHeight(world, blockposition) >= 0.44444445F) {
-                    world.setTypeUpdate(blockposition, Blocks.COBBLESTONE.getBlockData());
-                    this.fizz(world, blockposition);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(world, blockposition, Blocks.COBBLESTONE.getBlockData())) {
+                        this.fizz(world, blockposition);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
             }
