--- a/net/minecraft/server/CustomFunctionData.java
+++ b/net/minecraft/server/CustomFunctionData.java
@@ -32,7 +32,7 @@
     private final Map<MinecraftKey, CustomFunction> g = Maps.newHashMap();
     private final ArrayDeque<CustomFunctionData.a> h = new ArrayDeque();
     private boolean i;
-    private final Tags<CustomFunction> j = new Tags((minecraftkey) -> {
+    private final Tags<CustomFunction> j = new Tags<>((minecraftkey) -> { // CraftBukkit - decompile error
         return this.a(minecraftkey) != null;
     }, this::a, "tags/functions", true, "function");
     private final List<CustomFunction> k = Lists.newArrayList();
@@ -60,7 +60,7 @@
     }
 
     public com.mojang.brigadier.CommandDispatcher<CommandListenerWrapper> d() {
-        return this.server.getCommandDispatcher().a();
+        return this.server.vanillaCommandDispatcher.a(); // CraftBukkit
     }
 
     public void tick() {
@@ -164,7 +164,7 @@
             arraylist.add(CompletableFuture.supplyAsync(() -> {
                 return a(iresourcemanager, minecraftkey);
             }, Resource.a).thenApplyAsync((list) -> {
-                return CustomFunction.a(minecraftkey, this, list);
+                return CustomFunction.a(minecraftkey1, this, list); // CraftBukkit - decompile error
             }).handle((customfunction, throwable) -> {
                 return this.a(customfunction, throwable, minecraftkey);
             }));
