--- a/net/minecraft/server/EntityPainting.java
+++ b/net/minecraft/server/EntityPainting.java
@@ -11,6 +11,10 @@
 
     public EntityPainting(World world) {
         super(EntityTypes.PAINTING, world);
+        // CraftBukkit start - generate a non-null painting
+        ArrayList arraylist = Lists.newArrayList(Paintings.a);
+        this.art = (Paintings) arraylist.get(this.random.nextInt(arraylist.size()));
+        // CraftBukkit end
     }
 
     public EntityPainting(World world, BlockPosition blockposition, EnumDirection enumdirection) {
