--- a/net/minecraft/server/EntityPigZombie.java
+++ b/net/minecraft/server/EntityPigZombie.java
@@ -6,7 +6,7 @@
 
     private static final UUID b = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
     private static final AttributeModifier c = (new AttributeModifier(EntityPigZombie.b, "Attacking speed boost", 0.05D, 0)).a(false);
-    private int angerLevel;
+    public int angerLevel; // CraftBukkit - public
     private int soundDelay;
     private UUID hurtBy;
 
