--- ../work/decompile-8eb82bde//net/minecraft/server/EntityChicken.java	2014-11-28 11:22:55.882823164 +0000
+++ src/main/java/net/minecraft/server/EntityChicken.java	2014-11-28 11:01:54.000000000 +0000
@@ -35,6 +35,11 @@
     }
 
     public void m() {
+        // CraftBukkit start
+        if (this.isChickenJockey()) {
+            this.persistent = !this.isTypeNotPersistent();
+        }
+        // CraftBukkit end
         super.m();
         this.bo = this.bk;
         this.bn = this.bm;
