--- ../work/decompile-8eb82bde//net/minecraft/server/EntityIronGolem.java	2014-11-28 11:22:55.926823163 +0000
+++ src/main/java/net/minecraft/server/EntityIronGolem.java	2014-11-28 11:01:54.000000000 +0000
@@ -57,7 +57,7 @@
 
     protected void s(Entity entity) {
         if (entity instanceof IMonster && this.bb().nextInt(20) == 0) {
-            this.setGoalTarget((EntityLiving) entity);
+            this.setGoalTarget((EntityLiving) entity, org.bukkit.event.entity.EntityTargetLivingEntityEvent.TargetReason.COLLISION, true); // CraftBukkit - set reason
         }
 
         super.s(entity);
