--- ../work/decompile-8eb82bde//net/minecraft/server/EntityLargeFireball.java	2014-11-28 11:22:55.938823163 +0000
+++ src/main/java/net/minecraft/server/EntityLargeFireball.java	2014-11-28 11:01:54.000000000 +0000
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
+
 public class EntityLargeFireball extends EntityFireball {
 
     public int yield = 1;
@@ -21,7 +23,16 @@
 
             boolean flag = this.world.getGameRules().getBoolean("mobGriefing");
 
-            this.world.createExplosion((Entity) null, this.locX, this.locY, this.locZ, (float) this.yield, flag, flag);
+            // CraftBukkit start - fire ExplosionPrimeEvent
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.entity.CraftEntity.getEntity(this.world.getServer(), this));
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                // give 'this' instead of (Entity) null so we know what causes the damage
+                this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), flag);
+            }
+            // CraftBukkit end
+        
             this.die();
         }
 
@@ -35,7 +46,8 @@
     public void a(NBTTagCompound nbttagcompound) {
         super.a(nbttagcompound);
         if (nbttagcompound.hasKeyOfType("ExplosionPower", 99)) {
-            this.yield = nbttagcompound.getInt("ExplosionPower");
+            // CraftBukkit - set bukkitYield when setting explosionpower
+            bukkitYield = this.yield = nbttagcompound.getInt("ExplosionPower");
         }
 
     }
