--- a/net/minecraft/server/EntityMushroomCow.java
+++ b/net/minecraft/server/EntityMushroomCow.java
@@ -2,6 +2,11 @@
 
 import java.util.UUID;
 import org.apache.commons.lang3.tuple.Pair;
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.player.PlayerShearEntityEvent;
+// CraftBukkit end
 
 public class EntityMushroomCow extends EntityCow {
 
@@ -72,9 +77,17 @@
             int i;
 
             if (itemstack.getItem() == Items.SHEARS && this.getAge() >= 0) {
+                // CraftBukkit start
+                PlayerShearEntityEvent event = new PlayerShearEntityEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), this.getBukkitEntity());
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return false;
+                }
+                // CraftBukkit end
                 this.world.addParticle(Particles.EXPLOSION, this.locX, this.locY + (double) (this.getHeight() / 2.0F), this.locZ, 0.0D, 0.0D, 0.0D);
                 if (!this.world.isClientSide) {
-                    this.die();
+                    // this.die(); // CraftBukkit - moved down
                     EntityCow entitycow = (EntityCow) EntityTypes.COW.a(this.world);
 
                     entitycow.setPositionRotation(this.locX, this.locY, this.locZ, this.yaw, this.pitch);
@@ -84,7 +97,14 @@
                         entitycow.setCustomName(this.getCustomName());
                     }
 
-                    this.world.addEntity(entitycow);
+                    // CraftBukkit start
+                    if (CraftEventFactory.callEntityTransformEvent(this, entitycow, EntityTransformEvent.TransformReason.SHEARED).isCancelled()) {
+                        return false;
+                    }
+                    this.world.addEntity(entitycow, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SHEARED);
+
+                    this.die(); // CraftBukkit - from above
+                    // CraftBukkit end
 
                     for (i = 0; i < 5; ++i) {
                         this.world.addEntity(new EntityItem(this.world, this.locX, this.locY + (double) this.getHeight(), this.locZ, new ItemStack(this.dV().d.getBlock())));
