--- ../work/decompile-8eb82bde//net/minecraft/server/EntityGuardian.java	Thu Dec 11 13:57:14 2014
+++ src/main/java/net/minecraft/server/EntityGuardian.java	Thu Dec 11 13:55:14 2014
@@ -14,7 +14,7 @@
     private EntityLiving bn;
     private int bo;
     private boolean bp;
-    private PathfinderGoalRandomStroll bq;
+    public PathfinderGoalRandomStroll bq; // CraftBukkit - private to public
 
     public EntityGuardian(World world) {
         super(world);
@@ -35,7 +35,8 @@
         this.c = this.b = this.random.nextFloat();
     }
 
-    protected void aW() {
+    // PAIL: Access + rename please
+    public void aW() { // CraftBukkit - protected to public
         super.aW();
         this.getAttributeInstance(GenericAttributes.e).setValue(6.0D);
         this.getAttributeInstance(GenericAttributes.d).setValue(0.5D);
