--- a/net/minecraft/server/ItemFishingRod.java
+++ b/net/minecraft/server/ItemFishingRod.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.player.PlayerFishEvent; // CraftBukkit
+
 public class ItemFishingRod extends Item {
 
     public ItemFishingRod() {
@@ -19,7 +21,6 @@
             itemstack.damage(i, entityhuman);
             entityhuman.a(enumhand);
         } else {
-            world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.I, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemFishingRod.j.nextFloat() * 0.4F + 0.8F));
             if (!world.isClientSide) {
                 EntityFishingHook entityfishinghook = new EntityFishingHook(world, entityhuman);
                 int j = EnchantmentManager.c(itemstack);
@@ -34,6 +35,17 @@
                     entityfishinghook.b(k);
                 }
 
+                // CraftBukkit start
+                PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), null, (org.bukkit.entity.Fish) entityfishinghook.getBukkitEntity(), PlayerFishEvent.State.FISHING);
+                world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+                if (playerFishEvent.isCancelled()) {
+                    entityhuman.hookedFish = null;
+                    return new InteractionResultWrapper(EnumInteractionResult.PASS, itemstack);
+                }
+                world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.I, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemFishingRod.j.nextFloat() * 0.4F + 0.8F));
+                // CraftBukkit end
+
                 world.addEntity(entityfishinghook);
             }
 
