--- /home/matt/mc-dev-private//net/minecraft/server/NBTTagList.java	2015-04-16 22:20:21.171362490 +0100
+++ src/main/java/net/minecraft/server/NBTTagList.java	2015-04-16 22:20:21.191362490 +0100
@@ -40,6 +40,7 @@
             nbtreadlimiter.a(8L);
             this.type = datainput.readByte();
             int j = datainput.readInt();
+            nbtreadlimiter.a(j * 8); // CraftBukkit
 
             this.list = Lists.newArrayList();
 
