--- a/net/minecraft/server/NetworkManager.java
+++ b/net/minecraft/server/NetworkManager.java
@@ -127,7 +127,7 @@
     }
 
     private static <T extends PacketListener> void a(Packet<T> packet, PacketListener packetlistener) {
-        packet.a(packetlistener);
+        packet.a((T) packetlistener); // CraftBukkit - decompile error
     }
 
     public void setPacketListener(PacketListener packetlistener) {
@@ -237,7 +237,7 @@
 
     public void close(IChatBaseComponent ichatbasecomponent) {
         if (this.channel.isOpen()) {
-            this.channel.close().awaitUninterruptibly();
+            this.channel.close(); // We can't wait as this may be called from an event loop.
             this.m = ichatbasecomponent;
         }
 
