--- a/net/minecraft/server/PacketPlayOutBlockChange.java
+++ b/net/minecraft/server/PacketPlayOutBlockChange.java
@@ -5,7 +5,7 @@
 public class PacketPlayOutBlockChange implements Packet<PacketListenerPlayOut> {
 
     private BlockPosition a;
-    private IBlockData block;
+    public IBlockData block; // CraftBukkit - public
 
     public PacketPlayOutBlockChange() {}
 
@@ -27,8 +27,4 @@
     public void a(PacketListenerPlayOut packetlistenerplayout) {
         packetlistenerplayout.a(this);
     }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayOut) packetlistener);
-    }
 }
