--- ../work/decompile-8eb82bde//net/minecraft/server/PacketPlayInCloseWindow.java	2014-11-28 17:43:43.273707431 +0000
+++ src/main/java/net/minecraft/server/PacketPlayInCloseWindow.java	2014-11-28 17:38:22.000000000 +0000
@@ -6,6 +6,17 @@
 
     public PacketPlayInCloseWindow() {}
 
+    // CraftBukkit start
+    @Override
+    public void a(PacketListener pl) {
+        a((PacketListenerPlayIn) pl);
+    }    
+    
+    public PacketPlayInCloseWindow(int id) {
+        this.id = id;
+    }
+    // CraftBukkit end
+
     public void a(PacketListenerPlayIn packetlistenerplayin) {
         packetlistenerplayin.a(this);
     }
@@ -17,8 +28,4 @@
     public void b(PacketDataSerializer packetdataserializer) {
         packetdataserializer.writeByte(this.id);
     }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayIn) packetlistener);
-    }
 }
