--- ../work/decompile-bb26c12b/net/minecraft/server/Path.java	2014-11-27 08:59:46.849421195 +1100
+++ src/main/java/net/minecraft/server/Path.java	2014-11-27 08:42:10.088851036 +1100
@@ -2,7 +2,7 @@
 
 public class Path {
 
-    private PathPoint[] a = new PathPoint[1024];
+    private PathPoint[] a = new PathPoint[128]; // CraftBukkit - reduce default size
     private int b;
 
     public Path() {}
