--- ../work/decompile-bb26c12b/net/minecraft/server/PathfinderGoalSit.java	2014-11-27 08:59:46.845421213 +1100
+++ src/main/java/net/minecraft/server/PathfinderGoalSit.java	2014-11-27 08:42:10.096851020 +1100
@@ -12,7 +12,7 @@
 
     public boolean a() {
         if (!this.entity.isTamed()) {
-            return false;
+            return this.willSit && this.entity.getGoalTarget() == null; // CraftBukkit - Allow sitting for wild animals
         } else if (this.entity.V()) {
             return false;
         } else if (!this.entity.onGround) {
