--- /home/matt/mc-dev-private//net/minecraft/server/PathfinderGoalMakeLove.java	2015-02-26 22:40:22.951608135 +0000
+++ src/main/java/net/minecraft/server/PathfinderGoalMakeLove.java	2015-02-26 22:40:22.951608135 +0000
@@ -87,7 +87,7 @@
         this.b.o(false);
         entityvillager.setAgeRaw(-24000);
         entityvillager.setPositionRotation(this.b.locX, this.b.locY, this.b.locZ, 0.0F, 0.0F);
-        this.d.addEntity(entityvillager);
+        this.d.addEntity(entityvillager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
         this.d.broadcastEntityEffect(entityvillager, (byte) 12);
     }
 }
