--- ../work/decompile-bb26c12b/net/minecraft/server/PathfinderGoalNearestAttackableTargetInsentient.java	2014-11-27 08:59:46.833421266 +1100
+++ src/main/java/net/minecraft/server/PathfinderGoalNearestAttackableTargetInsentient.java	2014-11-27 08:42:10.084851043 +1100
@@ -54,7 +54,7 @@
     }
 
     public void c() {
-        this.b.setGoalTarget(this.e);
+        this.b.setGoalTarget(this.e, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit - reason
         super.c();
     }
 
