--- /home/matt/mc-dev-private//net/minecraft/server/PathfinderGoalOwnerHurtTarget.java	2015-02-26 22:40:22.963608135 +0000
+++ src/main/java/net/minecraft/server/PathfinderGoalOwnerHurtTarget.java	2015-02-26 22:40:22.963608135 +0000
@@ -30,7 +30,7 @@
     }
 
     public void c() {
-        this.e.setGoalTarget(this.b);
+        this.e.setGoalTarget(this.b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit - reason
         EntityLiving entityliving = this.a.getOwner();
 
         if (entityliving != null) {
