--- a/net/minecraft/server/RecipeFireworks.java
+++ b/net/minecraft/server/RecipeFireworks.java
@@ -4,11 +4,15 @@
 import java.util.ArrayList;
 import javax.annotation.Nullable;
 
-public class RecipeFireworks implements IRecipe {
+public class RecipeFireworks extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
 
     private ItemStack a;
 
-    public RecipeFireworks() {}
+    // CraftBukkit start - Delegate to new parent class with bogus info
+    public RecipeFireworks() {
+        super(new ItemStack(Items.FIREWORKS, 0, 0), java.util.Arrays.asList(new ItemStack(Items.GUNPOWDER, 0, 5)));
+    }
+    // CraftBukkit end
 
     public boolean a(InventoryCrafting inventorycrafting, World world) {
         this.a = null;
