--- a/net/minecraft/server/RecipeBookServer.java
+++ b/net/minecraft/server/RecipeBookServer.java
@@ -9,6 +9,7 @@
 import java.util.function.Consumer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public class RecipeBookServer extends RecipeBook {
 
@@ -25,7 +26,7 @@
             IRecipe<?> irecipe = (IRecipe) iterator.next();
             MinecraftKey minecraftkey = irecipe.getKey();
 
-            if (!this.recipes.contains(minecraftkey) && !irecipe.isComplex()) {
+            if (!this.recipes.contains(minecraftkey) && !irecipe.isComplex() && CraftEventFactory.handlePlayerRecipeListUpdateEvent(entityplayer, minecraftkey)) { // CraftBukkit
                 this.a(minecraftkey);
                 this.d(minecraftkey);
                 list.add(minecraftkey);
@@ -59,6 +60,7 @@
     }
 
     private void a(PacketPlayOutRecipes.Action packetplayoutrecipes_action, EntityPlayer entityplayer, List<MinecraftKey> list) {
+        if (entityplayer.playerConnection == null) return; // SPIGOT-4478 during PlayerLoginEvent
         entityplayer.playerConnection.sendPacket(new PacketPlayOutRecipes(packetplayoutrecipes_action, list, Collections.emptyList(), this.a()));
     }
 
