--- a/net/minecraft/server/RecipesBanner.java
+++ b/net/minecraft/server/RecipesBanner.java
@@ -2,9 +2,16 @@
 
 import javax.annotation.Nullable;
 
-public class RecipesBanner {    public static class AddRecipe implements IRecipe {
+// CraftBukkit - decompile weirdness
+public class RecipesBanner {
 
-        public AddRecipe() {}
+    public static class AddRecipe extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
+
+        // CraftBukkit start - Delegate to new parent class with bogus info
+        public AddRecipe() {
+            super("", new ItemStack(Items.BANNER, 0, 0), NonNullList.a(RecipeItemStack.a, RecipeItemStack.a(Items.BANNER)));
+        }
+        // CraftBukkit end
 
         public boolean a(InventoryCrafting inventorycrafting, World world) {
             boolean flag = false;
@@ -186,9 +193,13 @@
         }
     }
 
-    public static class DuplicateRecipe implements IRecipe {
+    public static class DuplicateRecipe extends ShapelessRecipes implements IRecipe { // CraftBukkit - added extends
 
-        public DuplicateRecipe() {}
+        // CraftBukkit start - Delegate to new parent class with bogus info
+        public DuplicateRecipe() {
+            super("", new ItemStack(Items.BANNER, 0, 0), NonNullList.a(RecipeItemStack.a, RecipeItemStack.a(Items.DYE)));
+        }
+        // CraftBukkit end
 
         public boolean a(InventoryCrafting inventorycrafting, World world) {
             ItemStack itemstack = ItemStack.a;
