--- a/net/minecraft/server/RemoteControlListener.java
+++ b/net/minecraft/server/RemoteControlListener.java
@@ -109,7 +109,7 @@
         RemoteControlListener.LOGGER.debug("closeSocket: {}", serversocket);
 
         try {
-            serversocket.close();
+            if (serversocket != null) serversocket.close(); // CraftBukkit - SPIGOT-5867, MC-193339: NPE during shutdown if no password
         } catch (IOException ioexception) {
             RemoteControlListener.LOGGER.warn("Failed to close socket", ioexception);
         }
