--- a/net/minecraft/server/StructureGenerator.java
+++ b/net/minecraft/server/StructureGenerator.java
@@ -43,7 +43,7 @@
                     StructureStart structurestart = this.a(generatoraccess, chunkgenerator, (SeededRandom) random, i2);
 
                     if (structurestart != StructureGenerator.a && structurestart.c().a(l, i1, l + 15, i1 + 15)) {
-                        ((LongSet) chunkgenerator.getStructureCache(this).computeIfAbsent(j1, (i) -> {
+                        ((LongSet) chunkgenerator.getStructureCache(this).computeIfAbsent(j1, (x) -> { // CraftBukkit - fix decompile error
                             return new LongOpenHashSet();
                         })).add(i2);
                         generatoraccess.getChunkProvider().a(j, k, true).a(this.a(), i2);
@@ -204,7 +204,7 @@
                 return structurestart;
             } else {
                 ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(i);
-                IChunkAccess ichunkaccess = generatoraccess.getChunkProvider().a(chunkcoordintpair.x, chunkcoordintpair.z, false);
+                IChunkAccess ichunkaccess = generatoraccess.getChunkProvider().getChunkAt(chunkcoordintpair.x, chunkcoordintpair.z, false, false); // CraftBukkit - don't load chunks
 
                 if (ichunkaccess != null) {
                     structurestart = ichunkaccess.a(this.a());
