--- ../work/decompile-bb26c12b/net/minecraft/server/Village.java	2014-11-27 08:59:46.921420877 +1100
+++ src/main/java/net/minecraft/server/Village.java	2014-11-27 08:42:10.104851005 +1100
@@ -60,7 +60,7 @@
                 EntityIronGolem entityirongolem = new EntityIronGolem(this.a);
 
                 entityirongolem.setPosition(vec3d.a, vec3d.b, vec3d.c);
-                this.a.addEntity(entityirongolem);
+                this.a.addEntity(entityirongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE); // CraftBukkit
                 ++this.l;
             }
         }
