--- a/net/minecraft/server/WorldProviderNormal.java
+++ b/net/minecraft/server/WorldProviderNormal.java
@@ -13,10 +13,14 @@
         super(world, dimensionmanager);
     }
 
+    // CraftBukkit start
+    /*
     @Override
     public DimensionManager getDimensionManager() {
         return DimensionManager.OVERWORLD;
     }
+    */
+    // CraftBukkit end
 
     @Override
     public ChunkGenerator<? extends GeneratorSettingsDefault> getChunkGenerator() {
