--- ../work/decompile-8eb82bde//net/minecraft/server/WorldGenGroundBush.java	2014-11-28 17:43:43.421707428 +0000
+++ src/main/java/net/minecraft/server/WorldGenGroundBush.java	2014-11-28 17:38:18.000000000 +0000
@@ -46,7 +46,11 @@
                     }
                 }
             }
+        // CraftBukkit start - Return false if gen was unsuccessful
+        } else {
+            return false;
         }
+        // CraftBukkit end
 
         return true;
     }
