--- a/net/minecraft/server/WorldBorder.java
+++ b/net/minecraft/server/WorldBorder.java
@@ -18,6 +18,7 @@
     private double j;
     private int k;
     private int l;
+    public WorldServer world; // CraftBukkit
 
     public WorldBorder() {
         this.e = this.d;
@@ -177,6 +178,7 @@
     }
 
     public void a(IWorldBorderListener iworldborderlistener) {
+        if (a.contains(iworldborderlistener)) return; // CraftBukkit
         this.a.add(iworldborderlistener);
     }
 
