--- ../work/decompile-8eb82bde/net/minecraft/server/WorldBorderListener.java	2014-12-10 19:16:20.876465384 +0000
+++ src/main/java/net/minecraft/server/WorldBorderListener.java	2014-12-10 18:59:49.024476151 +0000
@@ -9,23 +9,23 @@
     }
 
     public void a(WorldBorder worldborder, double d0) {
-        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_SIZE));
+        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_SIZE), worldborder.world); // CraftBukkit
     }
 
     public void a(WorldBorder worldborder, double d0, double d1, long i) {
-        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.LERP_SIZE));
+        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.LERP_SIZE), worldborder.world); // CraftBukkit
     }
 
     public void a(WorldBorder worldborder, double d0, double d1) {
-        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_CENTER));
+        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_CENTER), worldborder.world); // CraftBukkit
     }
 
     public void a(WorldBorder worldborder, int i) {
-        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_WARNING_TIME));
+        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_WARNING_TIME), worldborder.world); // CraftBukkit
     }
 
     public void b(WorldBorder worldborder, int i) {
-        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_WARNING_BLOCKS));
+        this.a.sendAll(new PacketPlayOutWorldBorder(worldborder, EnumWorldBorderAction.SET_WARNING_BLOCKS), worldborder.world); // CraftBukkit
     }
 
     public void b(WorldBorder worldborder, double d0) {}
