From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Wed, 21 Dec 2016 11:47:25 -0600
Subject: [PATCH] Add API methods to control if armour stands can move


diff --git a/src/main/java/org/bukkit/entity/ArmorStand.java b/src/main/java/org/bukkit/entity/ArmorStand.java
index d08c5eb82eab8ad0ca7c8602b4d3d92fb0df02b3..6a1d563d1b7ea7e1060dfacfba98e4e1d4295790 100644
--- a/src/main/java/org/bukkit/entity/ArmorStand.java
+++ b/src/main/java/org/bukkit/entity/ArmorStand.java
@@ -344,4 +344,21 @@ public interface ArmorStand extends LivingEntity {
          */
         ADDING;
     }
+    // Paper start
+    /**
+     * Tests if this armor stand can move.
+     *
+     * <p>The default value is {@code true}.</p>
+     *
+     * @return {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    boolean canMove();
+
+    /**
+     * Sets if this armor stand can move.
+     *
+     * @param move {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    void setCanMove(boolean move);
+    // Paper end
 }
