From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Mon, 31 Jul 2017 02:08:55 -0500
Subject: [PATCH] Make /plugins list alphabetical


diff --git a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
index bcb576a4271b1ec7b1cfe6f83cf161b7d89ed2e5..337200fb94774ff4938e155518521c9928f77091 100644
--- a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
@@ -34,15 +34,22 @@ public class PluginsCommand extends BukkitCommand {
 
     @NotNull
     private String getPluginList() {
-        StringBuilder pluginList = new StringBuilder();
-        Plugin[] plugins = Bukkit.getPluginManager().getPlugins();
+        // Paper start
+        java.util.TreeMap<String, Plugin> plugins = new java.util.TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+
+        for (Plugin plugin : Bukkit.getPluginManager().getPlugins()) {
+            plugins.put(plugin.getDescription().getName(), plugin);
+        }
 
-        for (Plugin plugin : plugins) {
+        StringBuilder pluginList = new StringBuilder();
+        for (java.util.Map.Entry<String, Plugin> entry : plugins.entrySet()) {
             if (pluginList.length() > 0) {
                 pluginList.append(ChatColor.WHITE);
                 pluginList.append(", ");
             }
 
+            Plugin plugin = entry.getValue();
+            
             pluginList.append(plugin.isEnabled() ? ChatColor.GREEN : ChatColor.RED);
             pluginList.append(plugin.getDescription().getName());
 
@@ -51,6 +58,8 @@ public class PluginsCommand extends BukkitCommand {
             }
         }
 
-        return "(" + plugins.length + "): " + pluginList.toString();
+        return "(" + plugins.size() + "): " + pluginList.toString();
+        // Paper end
     }
+
 }
