From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 1 Apr 2018 02:28:43 +0300
Subject: [PATCH] Add openSign method to HumanEntity


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index b4452d18a530c4c4a9bc10f368aeced9430b42a2..35d72d8ab78ca4095545fab54d6b440c040223eb 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -538,6 +538,26 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     @Deprecated(since = "1.12")
     public void setShoulderEntityRight(@Nullable Entity entity);
+    // Paper start - Add method to open already placed sign
+    /**
+     * Opens an editor window for the specified sign
+     *
+     * @param sign The sign to open
+     * @deprecated use {@link #openSign(org.bukkit.block.Sign, org.bukkit.block.sign.Side)}
+     */
+    @Deprecated
+    default void openSign(@NotNull org.bukkit.block.Sign sign) {
+        this.openSign(sign, org.bukkit.block.sign.Side.FRONT);
+    }
+
+    /**
+     * Opens an editor window for the specified sign
+     *
+     * @param sign The sign to open
+     * @param side The side of the sign to open
+     */
+    void openSign(org.bukkit.block.@NotNull Sign sign, org.bukkit.block.sign.@NotNull Side side);
+    // Paper end
 
     /**
      * Make the entity drop the item in their hand.
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3611ef03abe9c81cfb956c9957c57e7df05faf32..949b34e9c9f8b4be48c09f80d78d9453f0ff0ee3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3115,10 +3115,12 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     /**
      * Open a Sign for editing by the Player.
      *
-     * The Sign must be placed in the same world as the player.
+     * The Sign must be in the same world as the player.
      *
      * @param sign The sign to edit
+     * @deprecated use {@link #openSign(Sign, Side)}
      */
+    @Deprecated
     public void openSign(@NotNull Sign sign);
 
     /**
