From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 16 May 2018 20:39:09 -0400
Subject: [PATCH] WitchThrowPotionEvent

Fired when a witch throws a potion at a player

diff --git a/src/main/java/com/destroystokyo/paper/event/entity/WitchThrowPotionEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/WitchThrowPotionEvent.java
index 6ef6367b67261c2b653a97322b9703a9409b3499..011cae7064cb9488db3cf4d261b8a784a4cc667e 100644
--- a/src/main/java/com/destroystokyo/paper/event/entity/WitchThrowPotionEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/entity/WitchThrowPotionEvent.java
@@ -4,26 +4,69 @@ import org.bukkit.event.Cancellable;
 import org.bukkit.event.Event;
 import org.bukkit.event.HandlerList;
 
-public class WitchThrowPotionEvent extends Event implements Cancellable {
-    public WitchThrowPotionEvent() {
+/**
+ * Fired when a witch throws a potion at a player
+ */
+public class WitchThrowPotionEvent extends org.bukkit.event.entity.EntityEvent implements Cancellable {
+    @org.jetbrains.annotations.NotNull private final org.bukkit.entity.LivingEntity target;
+    @org.jetbrains.annotations.Nullable private org.bukkit.inventory.ItemStack potion;
+
+    public WitchThrowPotionEvent(@org.jetbrains.annotations.NotNull org.bukkit.entity.Witch witch, @org.jetbrains.annotations.NotNull org.bukkit.entity.LivingEntity target, @org.jetbrains.annotations.Nullable org.bukkit.inventory.ItemStack potion) {
+        super(witch);
+        this.target = target;
+        this.potion = potion;
     }
 
+    @org.jetbrains.annotations.NotNull
+    @Override
+    public org.bukkit.entity.Witch getEntity() {
+        return (org.bukkit.entity.Witch) super.getEntity();
+    }
+
+    /**
+     * @return The target of the potion
+     */
+    @org.jetbrains.annotations.NotNull
+    public org.bukkit.entity.LivingEntity getTarget() {
+        return target;
+    }
+
+    /**
+     * @return The potion the witch will throw at a player
+     */
+    @org.jetbrains.annotations.Nullable
+    public org.bukkit.inventory.ItemStack getPotion() {
+        return potion;
+    }
+
+    /**
+     * Sets the potion to be thrown at a player
+     * @param potion The potion
+     */
+    public void setPotion(@org.jetbrains.annotations.Nullable org.bukkit.inventory.ItemStack potion) {
+        this.potion = potion != null ? potion.clone() : null;
+    }
 
     private static final HandlerList handlers = new HandlerList();
 
+    @org.jetbrains.annotations.NotNull
     public HandlerList getHandlers() {
         return handlers;
     }
 
+    @org.jetbrains.annotations.NotNull
     public static HandlerList getHandlerList() {
         return handlers;
     }
 
     private boolean cancelled = false;
 
+    /**
+     * @return Event was cancelled or potion was null
+     */
     @Override
     public boolean isCancelled() {
-        return cancelled;
+        return cancelled || potion == null;
     }
 
     @Override
