From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Aug 2020 15:22:44 +0200
Subject: [PATCH] Add more Evoker API


diff --git a/src/main/java/org/bukkit/entity/Evoker.java b/src/main/java/org/bukkit/entity/Evoker.java
index e9fffd8967e95d370c3e64b4558e41e5a9d580a8..a0a46e6e1114f9f5e2843347f835f3cbf5a75d1b 100644
--- a/src/main/java/org/bukkit/entity/Evoker.java
+++ b/src/main/java/org/bukkit/entity/Evoker.java
@@ -64,4 +64,19 @@ public interface Evoker extends Spellcaster {
      */
     @Deprecated(since = "1.11.2")
     void setCurrentSpell(@Nullable Spell spell);
+
+    // Paper start
+    /**
+     * @return the sheep being targeted by the {@link Spell#WOLOLO wololo spell}, or {@code null} if none
+     */
+    @Nullable
+    Sheep getWololoTarget();
+
+    /**
+     * Set the sheep to be the target of the {@link Spell#WOLOLO wololo spell}, or {@code null} to clear.
+     *
+     * @param sheep new wololo target
+     */
+    void setWololoTarget(@Nullable Sheep sheep);
+    // Paper end
 }
