From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Aug 2020 16:32:03 +0200
Subject: [PATCH] Add moon phase API


diff --git a/src/main/java/io/papermc/paper/world/MoonPhase.java b/src/main/java/io/papermc/paper/world/MoonPhase.java
new file mode 100644
index 0000000000000000000000000000000000000000..df05153397b42930cd53d37b30824c7e5f008f7e
--- /dev/null
+++ b/src/main/java/io/papermc/paper/world/MoonPhase.java
@@ -0,0 +1,36 @@
+package io.papermc.paper.world;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public enum MoonPhase {
+    FULL_MOON(0L),
+    WANING_GIBBOUS(1L),
+    LAST_QUARTER(2L),
+    WANING_CRESCENT(3L),
+    NEW_MOON(4L),
+    WAXING_CRESCENT(5L),
+    FIRST_QUARTER(6L),
+    WAXING_GIBBOUS(7L);
+
+    private final long day;
+
+    MoonPhase(long day) {
+        this.day = day;
+    }
+
+    private static final Map<Long, MoonPhase> BY_DAY = new HashMap<>();
+
+    static {
+        for (MoonPhase phase : values()) {
+            BY_DAY.put(phase.day, phase);
+        }
+    }
+
+    @NotNull
+    public static MoonPhase getPhase(long day) {
+        return BY_DAY.get(day % 8L);
+    }
+}
diff --git a/src/main/java/org/bukkit/RegionAccessor.java b/src/main/java/org/bukkit/RegionAccessor.java
index bc70f4f5b82ce5b92ab0b2612c8fb8126ed1a8c2..f587a529e4d7b097b3f204a34c636da0bbac6747 100644
--- a/src/main/java/org/bukkit/RegionAccessor.java
+++ b/src/main/java/org/bukkit/RegionAccessor.java
@@ -421,4 +421,12 @@ public interface RegionAccessor {
      * {@link HeightMap}
      */
     public int getHighestBlockYAt(@NotNull Location location, @NotNull HeightMap heightMap);
+
+    // Paper start
+    /**
+     * @return the current moon phase at the current time in the world
+     */
+    @NotNull
+    io.papermc.paper.world.MoonPhase getMoonPhase();
+    // Paper end
 }
