From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Trigary <trigary0@gmail.com>
Date: Tue, 14 Apr 2020 12:06:14 +0200
Subject: [PATCH] Player elytra boost API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8f7667562187a141ec037eaafa2e3e8d1abd6bff..8d4a010d826fb84feb3d6b927275c49d8eec9f71 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2516,6 +2516,19 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @NotNull
     <T> T getClientOption(@NotNull ClientOption<T> option);
+
+    /**
+     * Boost a Player that's {@link #isGliding()} using a {@link Firework}.
+     * If the creation of the entity is cancelled, no boosting is done.
+     * This method does not fire {@link com.destroystokyo.paper.event.player.PlayerElytraBoostEvent}.
+     *
+     * @param firework The {@link Material#FIREWORK_ROCKET} to boost the player with
+     * @return The {@link Firework} boosting the Player or null if the spawning of the entity was cancelled
+     * @throws IllegalArgumentException if {@link #isGliding()} is false
+     * or if the {@code firework} isn't a {@link Material#FIREWORK_ROCKET}
+     */
+    @Nullable
+    Firework boostElytra(@NotNull ItemStack firework);
     // Paper end
 
     // Spigot start
