From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 13 Dec 2020 05:32:12 +0200
Subject: [PATCH] Expose LivingEntity hurt direction


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index a11c94c2ba0c8a1b74aca7cf540853f08911aae6..85740f98460565c5f5f28d278771c9791e042edf 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -350,6 +350,16 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     public void setCooldown(@NotNull Material material, int ticks);
 
+    // Paper start
+    /**
+     * Sets player hurt direction
+     *
+     * @param hurtDirection hurt direction
+     */
+    @Override
+    void setHurtDirection(float hurtDirection);
+    // Paper end
+
     /**
      * Get the sleep ticks of the player. This value may be capped.
      *
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index ed9375a667450669f20bf6302073d482d1d52461..fd5a3ba96fb18e044d52893a5cf28560d1aa97ce 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1101,5 +1101,21 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @param quantity quantity of item
      */
     void playPickupItemAnimation(@NotNull Item item, int quantity);
+
+    /**
+     * Gets player hurt direction
+     *
+     * @return hurt direction
+     */
+    float getHurtDirection();
+
+    /**
+     * Sets player hurt direction
+     *
+     * @param hurtDirection hurt direction
+     * @deprecated use {@link Player#setHurtDirection(float)}
+     */
+    @Deprecated
+    void setHurtDirection(float hurtDirection);
     // Paper end
 }
