From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: oxygencraft <21054297+oxygencraft@users.noreply.github.com>
Date: Sun, 25 Oct 2020 18:35:58 +1100
Subject: [PATCH] Add getOfflinePlayerIfCached(String)


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d5ae00773f37db728907c9deb66ff614d96b83b2..dddd83d74a3f94784dd510813b9a7d5fd7c12abe 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -962,6 +962,27 @@ public final class Bukkit {
         return server.getOfflinePlayer(name);
     }
 
+    // Paper start
+    /**
+     * Gets the player by the given name, regardless if they are offline or
+     * online.
+     * <p>
+     * This will not make a web request to get the UUID for the given name,
+     * thus this method will not block. However this method will return
+     * {@code null} if the player is not cached.
+     * </p>
+     *
+     * @param name the name of the player to retrieve
+     * @return an offline player if cached, {@code null} otherwise
+     * @see #getOfflinePlayer(String)
+     * @see #getOfflinePlayer(java.util.UUID)
+     */
+    @Nullable
+    public static OfflinePlayer getOfflinePlayerIfCached(@NotNull String name) {
+        return server.getOfflinePlayerIfCached(name);
+    }
+    // Paper end
+
     /**
      * Gets the player by the given UUID, regardless if they are offline or
      * online.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9e54765c94d58448becdf0df1bad10d41173e3a7..b632e64a7dd4f08e69573c2fff7ab52031b342d5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -808,6 +808,25 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public OfflinePlayer getOfflinePlayer(@NotNull String name);
 
+    // Paper start
+    /**
+     * Gets the player by the given name, regardless if they are offline or
+     * online.
+     * <p>
+     * This will not make a web request to get the UUID for the given name,
+     * thus this method will not block. However this method will return
+     * {@code null} if the player is not cached.
+     * </p>
+     *
+     * @param name the name of the player to retrieve
+     * @return an offline player if cached, {@code null} otherwise
+     * @see #getOfflinePlayer(String)
+     * @see #getOfflinePlayer(java.util.UUID)
+     */
+    @Nullable
+    public OfflinePlayer getOfflinePlayerIfCached(@NotNull String name);
+    // Paper end
+
     /**
      * Gets the player by the given UUID, regardless if they are offline or
      * online.
