From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Madeline Miller <mnmiller1@me.com>
Date: Mon, 4 Jan 2021 16:40:55 +1000
Subject: [PATCH] Add API to get exact interaction point in PlayerInteractEvent


diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
index 35e0df9fc0338c7c6b8e178ee533cd7833ee66f1..69c800d367d36c8730781b89e3530a75487d000a 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
@@ -1,5 +1,6 @@
 package org.bukkit.event.player;
 
+import org.bukkit.Location;
 import org.bukkit.Material;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockFace;
@@ -234,13 +235,30 @@ public class PlayerInteractEvent extends PlayerEvent implements Cancellable {
      * <p>
      * All vector components are between 0.0 and 1.0 inclusive.
      *
+     * @deprecated misleading, use {@link #getInteractionPoint()}
      * @return the clicked position. May be null.
      */
     @Nullable
+    @Deprecated // Paper
     public Vector getClickedPosition() {
         return clickedPosistion;
     }
 
+    // Paper start
+    /**
+     * The exact point at which the interaction occurred. May be null.
+     *
+     * @return the exact interaction point. May be null.
+     */
+    @Nullable
+    public Location getInteractionPoint() {
+        if (this.blockClicked == null || this.clickedPosistion == null) {
+            return null;
+        }
+        return this.blockClicked.getLocation().add(this.clickedPosistion);
+    }
+    // Paper end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
