From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Thu, 24 Dec 2020 12:43:30 -0800
Subject: [PATCH] Add OBSTRUCTED reason to BedEnterResult


diff --git a/src/main/java/org/bukkit/event/player/PlayerBedEnterEvent.java b/src/main/java/org/bukkit/event/player/PlayerBedEnterEvent.java
index fd6b83f1f9bb2e25613e3f88e497b9d57d393937..ced77de350fa53889439e945336343a1cca5b9ba 100644
--- a/src/main/java/org/bukkit/event/player/PlayerBedEnterEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerBedEnterEvent.java
@@ -42,6 +42,12 @@ public class PlayerBedEnterEvent extends PlayerEvent implements Cancellable {
          * Entering the bed is prevented due to the player being too far away.
          */
         TOO_FAR_AWAY,
+        // Paper start
+        /**
+         * Bed was obstructed.
+         */
+        OBSTRUCTED,
+        // Paper end
         /**
          * Entering the bed is prevented due to there being monsters nearby.
          */
