From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksander Jagiello <themolkapl@gmail.com>
Date: Sun, 24 Jan 2021 22:17:29 +0100
Subject: [PATCH] Add getMainThreadExecutor to BukkitScheduler


diff --git a/src/main/java/org/bukkit/scheduler/BukkitScheduler.java b/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
index 0a0e9461909c6c0bc4a8519483ec05d9c59e67cc..5c277ac7a61df8106f7c13a1ad8ccb1509338699 100644
--- a/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
+++ b/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
@@ -457,4 +457,15 @@ public interface BukkitScheduler {
     @Deprecated(since = "1.7.10")
     @NotNull
     public BukkitTask runTaskTimerAsynchronously(@NotNull Plugin plugin, @NotNull BukkitRunnable task, long delay, long period) throws IllegalArgumentException;
+
+    // Paper start - add getMainThreadExecutor
+    /**
+     * Returns an executor that will run tasks on the next server tick.
+     *
+     * @param plugin the reference to the plugin scheduling tasks
+     * @return an executor associated with the given plugin
+     */
+    @NotNull
+    public java.util.concurrent.Executor getMainThreadExecutor(@NotNull Plugin plugin);
+    // Paper end
 }
