From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 3 Jan 2021 20:03:40 -0800
Subject: [PATCH] Added missing vanilla tags


diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 15699ee58e06880a508689f761ecfdb77d44d182..0bb2f106b93db546e8f740d0dec8ddb02e76de62 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -99,6 +99,12 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla block tag representing all log and bark variants that burn.
      */
     Tag<Material> LOGS_THAT_BURN = Bukkit.getTag(REGISTRY_BLOCKS, NamespacedKey.minecraft("logs_that_burn"), Material.class);
+    // Paper start
+    /**
+     * Vanilla block tag representing naturally occurring overworld logs.
+     */
+    Tag<Material> OVERWORLD_NATURAL_LOGS = Bukkit.getTag(REGISTRY_BLOCKS, NamespacedKey.minecraft("overworld_natural_logs"), Material.class);
+    // Paper end
     /**
      * Vanilla block tag representing all log and bark variants.
      */
@@ -712,6 +718,12 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla item tag representing all chest boat items.
      */
     Tag<Material> ITEMS_CHEST_BOATS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("chest_boats"), Material.class);
+    // Paper start
+    /**
+     * Vanilla item tag representing all non-flammable wood items.
+     */
+    Tag<Material> ITEMS_NON_FLAMMABLE_WOOD = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("non_flammable_wood"), Material.class);
+    // Paper end
     /**
      * Vanilla item tag representing all fish items.
      */
@@ -748,10 +760,7 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla item tag representing all stone tool materials.
      */
     Tag<Material> ITEMS_STONE_TOOL_MATERIALS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("stone_tool_materials"), Material.class);
-    /**
-     * Vanilla item tag representing all furnace materials.
-     */
-    Tag<Material> ITEMS_FURNACE_MATERIALS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("furnace_materials"), Material.class);
+    // Paper - remove ITEMS_FURNACE_MATERIALS (tag doesn't exist anymore)
     /**
      * Vanilla item tag representing all compasses.
      */
@@ -779,6 +788,237 @@ public interface Tag<T extends Keyed> extends Keyed {
      * harvesting clusters (unused).
      */
     Tag<Material> CLUSTER_MAX_HARVESTABLES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("cluster_max_harvestables"), Material.class);
+    // Paper start - add missing item tags
+    /**
+     * @see #SOUL_FIRE_BASE_BLOCKS
+     */
+    Tag<Material> ITEMS_SOUL_FIRE_BASE_BLOCKS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("soul_fire_base_blocks"), Material.class);
+    Tag<Material> ITEMS_STONE_CRAFTING_MATERIALS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("stone_crafting_materials"), Material.class);
+    /**
+     * @see #PIGLIN_REPELLENTS
+     */
+    Tag<Material> ITEMS_PIGLIN_REPELLENTS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("piglin_repellents"), Material.class);
+    /**
+     * @see #WOODEN_FENCES
+     */
+    Tag<Material> ITEMS_WOODEN_FENCES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_fences"), Material.class);
+    /**
+     * @see #WOODEN_SLABS
+     */
+    Tag<Material> ITEMS_WOODEN_SLABS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_slabs"), Material.class);
+    /**
+     * @see #COAL_ORES
+     */
+    Tag<Material> ITEMS_COAL_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("coal_ores"), Material.class);
+    /**
+     * @see #SMALL_FLOWERS
+     */
+    Tag<Material> ITEMS_SMALL_FLOWERS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("small_flowers"), Material.class);
+    /**
+     * @see #WOODEN_TRAPDOORS
+     */
+    Tag<Material> ITEMS_WOODEN_TRAPDOORS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_trapdoors"), Material.class);
+    /**
+     * @see #DAMPENS_VIBRATIONS
+     */
+    Tag<Material> ITEMS_DAMPENS_VIBRATIONS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("dampens_vibrations"), Material.class);
+    /**
+     * @see #MANGROVE_LOGS
+     */
+    Tag<Material> ITEMS_MANGROVE_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("mangrove_logs"), Material.class);
+    /**
+     * @see #JUNGLE_LOGS
+     */
+    Tag<Material> ITEMS_JUNGLE_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("jungle_logs"), Material.class);
+    /**
+     * @see #WOODEN_STAIRS
+     */
+    Tag<Material> ITEMS_WOODEN_STAIRS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_stairs"), Material.class);
+    /**
+     * @see #SIGNS
+     */
+    Tag<Material> ITEMS_SIGNS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("signs"), Material.class);
+    /**
+     * @see #SPRUCE_LOGS
+     */
+    Tag<Material> ITEMS_SPRUCE_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("spruce_logs"), Material.class);
+    /**
+     * @see #WOOL
+     */
+    Tag<Material> ITEMS_WOOL = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wool"), Material.class);
+    /**
+     * @see #WOODEN_BUTTONS
+     */
+    Tag<Material> ITEMS_WOODEN_BUTTONS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_buttons"), Material.class);
+    /**
+     * @see #STAIRS
+     */
+    Tag<Material> ITEMS_STAIRS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("stairs"), Material.class);
+    /**
+     * @see #LOGS
+     */
+    Tag<Material> ITEMS_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("logs"), Material.class);
+    /**
+     * @see #STONE_BRICKS
+     */
+    Tag<Material> ITEMS_STONE_BRICKS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("stone_bricks"), Material.class);
+    /**
+     * @see #WOOL_CARPETS
+     */
+    Tag<Material> ITEMS_WOOL_CARPETS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wool_carpets"), Material.class);
+    /**
+     * @see #SLABS
+     */
+    Tag<Material> ITEMS_SLABS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("slabs"), Material.class);
+    /**
+     * @see #WOODEN_DOORS
+     */
+    Tag<Material> ITEMS_WOODEN_DOORS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_doors"), Material.class);
+    /**
+     * @see #WARPED_STEMS
+     */
+    Tag<Material> ITEMS_WARPED_STEMS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("warped_stems"), Material.class);
+    /**
+     * @see #EMERALD_ORES
+     */
+    Tag<Material> ITEMS_EMERALD_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("emerald_ores"), Material.class);
+    /**
+     * @see #REDSTONE_ORES
+     */
+    Tag<Material> ITEMS_REDSTONE_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("redstone_ores"), Material.class);
+    /**
+     * @see #TRAPDOORS
+     */
+    Tag<Material> ITEMS_TRAPDOORS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("trapdoors"), Material.class);
+    /**
+     * @see #CRIMSON_STEMS
+     */
+    Tag<Material> ITEMS_CRIMSON_STEMS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("crimson_stems"), Material.class);
+    /**
+     * @see #BAMBOO_BLOCKS
+     */
+    Tag<Material> ITEMS_BAMBOO_BLOCKS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("bamboo_blocks"), Material.class);
+    /**
+     * @see #BUTTONS
+     */
+    Tag<Material> ITEMS_BUTTONS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("buttons"), Material.class);
+    /**
+     * @see #FLOWERS
+     */
+    Tag<Material> ITEMS_FLOWERS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("flowers"), Material.class);
+    /**
+     * @see #WART_BLOCKS
+     */
+    Tag<Material> ITEMS_WART_BLOCKS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wart_blocks"), Material.class);
+    /**
+     * @see #TERRACOTTA
+     */
+    Tag<Material> ITEMS_TERRACOTTA = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("terracotta"), Material.class);
+    /**
+     * @see #PLANKS
+     */
+    Tag<Material> ITEMS_PLANKS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("planks"), Material.class);
+    /**
+     * @see #DARK_OAK_LOGS
+     */
+    Tag<Material> ITEMS_DARK_OAK_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("dark_oak_logs"), Material.class);
+    /**
+     * @see #RAILS
+     */
+    Tag<Material> ITEMS_RAILS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("rails"), Material.class);
+    /**
+     * @see #DIAMOND_ORES
+     */
+    Tag<Material> ITEMS_DIAMOND_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("diamond_ores"), Material.class);
+    /**
+     * @see #LEAVES
+     */
+    Tag<Material> ITEMS_LEAVES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("leaves"), Material.class);
+    /**
+     * @see #WALLS
+     */
+    Tag<Material> ITEMS_WALLS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("walls"), Material.class);
+    /**
+     * @see #FENCE_GATES
+     */
+    Tag<Material> ITEMS_FENCE_GATES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("fence_gates"), Material.class);
+    /**
+     * @see #WOODEN_PRESSURE_PLATES
+     */
+    Tag<Material> ITEMS_WOODEN_PRESSURE_PLATES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("wooden_pressure_plates"), Material.class);
+    /**
+     * @see #ACACIA_LOGS
+     */
+    Tag<Material> ITEMS_ACACIA_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("acacia_logs"), Material.class);
+    /**
+     * @see #ANVIL
+     */
+    Tag<Material> ITEMS_ANVIL = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("anvil"), Material.class);
+    /**
+     * @see #CANDLES
+     */
+    Tag<Material> ITEMS_CANDLES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("candles"), Material.class);
+    /**
+     * @see #BIRCH_LOGS
+     */
+    Tag<Material> ITEMS_BIRCH_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("birch_logs"), Material.class);
+    /**
+     * @see #TALL_FLOWERS
+     */
+    Tag<Material> ITEMS_TALL_FLOWERS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("tall_flowers"), Material.class);
+    /**
+     * @see #LAPIS_ORES
+     */
+    Tag<Material> ITEMS_LAPIS_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("lapis_ores"), Material.class);
+    /**
+     * @see #SAND
+     */
+    Tag<Material> ITEMS_SAND = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("sand"), Material.class);
+    /**
+     * @see #COPPER_ORES
+     */
+    Tag<Material> ITEMS_COPPER_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("copper_ores"), Material.class);
+    /**
+     * @see #GOLD_ORES
+     */
+    Tag<Material> ITEMS_GOLD_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("gold_ores"), Material.class);
+    /**
+     * @see #FENCES
+     */
+    Tag<Material> ITEMS_FENCES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("fences"), Material.class);
+    /**
+     * @see #LOGS_THAT_BURN
+     */
+    Tag<Material> ITEMS_LOGS_THAT_BURN = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("logs_that_burn"), Material.class);
+    /**
+     * @see #COMPLETES_FIND_TREE_TUTORIAL
+     */
+    Tag<Material> ITEMS_COMPLETES_FIND_TREE_TUTORIAL = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("completes_find_tree_tutorial"), Material.class);
+    /**
+     * @see #SAPLINGS
+     */
+    Tag<Material> ITEMS_SAPLINGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("saplings"), Material.class);
+    /**
+     * @see #DIRT
+     */
+    Tag<Material> ITEMS_DIRT = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("dirt"), Material.class);
+    /**
+     * @see #BEDS
+     */
+    Tag<Material> ITEMS_BEDS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("beds"), Material.class);
+    /**
+     * @see #IRON_ORES
+     */
+    Tag<Material> ITEMS_IRON_ORES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("iron_ores"), Material.class);
+    /**
+     * @see #OAK_LOGS
+     */
+    Tag<Material> ITEMS_OAK_LOGS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("oak_logs"), Material.class);
+    /**
+     * @see #DOORS
+     */
+    Tag<Material> ITEMS_DOORS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("doors"), Material.class);
+    // Paper end
     /**
      * Key for the built in fluid registry.
      */
@@ -839,6 +1079,44 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla tag representing entities which can be eaten by frogs.
      */
     Tag<EntityType> ENTITY_TYPES_FROG_FOOD = Bukkit.getTag(REGISTRY_ENTITY_TYPES, NamespacedKey.minecraft("frog_food"), EntityType.class);
+    // Paper start
+    /**
+     * Key for the built-in entity registry
+     * @deprecated use {@link #REGISTRY_ENTITY_TYPES}
+     */
+    @Deprecated(forRemoval = true)
+    String REGISTRY_ENTITIES = REGISTRY_ENTITY_TYPES;
+    /**
+     * Vanilla entity tag representing arrow entities.
+     * @deprecated use {@link #ENTITY_TYPES_ARROWS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> ARROWS = ENTITY_TYPES_ARROWS;
+    /**
+     * Vanilla entity tag representing entities that live in beehives
+     * @deprecated use {@link #ENTITY_TYPES_BEEHIVE_INHABITORS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> BEEHIVE_INHABITORS = ENTITY_TYPES_BEEHIVE_INHABITORS;
+    /**
+     * Vanilla entity tag representing projectiles that impact
+     * @deprecated use {@link #ENTITY_TYPES_IMPACT_PROJECTILES}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> IMPACT_PROJECTILES = ENTITY_TYPES_IMPACT_PROJECTILES;
+    /**
+     * Vanilla entity tag for village raiders
+     * @deprecated use {@link #ENTITY_TYPES_RAIDERS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> RAIDERS = ENTITY_TYPES_RAIDERS;
+    /**
+     * Vanilla entity tag for skeleton types
+     * @deprecated use {@link #ENTITY_TYPES_SKELETONS}
+     */
+    @Deprecated(forRemoval = true)
+    Tag<EntityType> SKELETONS = ENTITY_TYPES_SKELETONS;
+    // Paper end
 
     /**
      * Returns whether or not this tag has an entry for the specified item.
