From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thonk <30448663+ExcessiveAmountsOfZombies@users.noreply.github.com>
Date: Wed, 6 Jan 2021 12:05:29 -0800
Subject: [PATCH] Add recipe to cook events


diff --git a/src/main/java/org/bukkit/event/block/BlockCookEvent.java b/src/main/java/org/bukkit/event/block/BlockCookEvent.java
index be7af5440bf9923f0c9c84efa4d70a89337a2f96..a3f1c9cb36c9069ed622985a525bfc2a7a27ab91 100644
--- a/src/main/java/org/bukkit/event/block/BlockCookEvent.java
+++ b/src/main/java/org/bukkit/event/block/BlockCookEvent.java
@@ -14,12 +14,21 @@ public class BlockCookEvent extends BlockEvent implements Cancellable {
     private final ItemStack source;
     private ItemStack result;
     private boolean cancelled;
+    private final org.bukkit.inventory.CookingRecipe<?> recipe; // Paper
 
+    @Deprecated // Paper
     public BlockCookEvent(@NotNull final Block block, @NotNull final ItemStack source, @NotNull final ItemStack result) {
+        // Paper start
+        this(block, source, result, null);
+    }
+
+    public BlockCookEvent(@NotNull final Block block, @NotNull final ItemStack source, @NotNull final ItemStack result, @org.jetbrains.annotations.Nullable org.bukkit.inventory.CookingRecipe<?> recipe) {
+        // Paper end
         super(block);
         this.source = source;
         this.result = result;
         this.cancelled = false;
+        this.recipe = recipe; // Paper
     }
 
     /**
@@ -61,6 +70,18 @@ public class BlockCookEvent extends BlockEvent implements Cancellable {
         this.cancelled = cancel;
     }
 
+    // Paper start
+    /**
+     * Gets the cooking recipe associated with this event.
+     *
+     * @return the recipe
+     */
+    @org.jetbrains.annotations.Nullable
+    public org.bukkit.inventory.CookingRecipe<?> getRecipe() {
+        return recipe;
+    }
+    // Paper end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
diff --git a/src/main/java/org/bukkit/event/inventory/FurnaceSmeltEvent.java b/src/main/java/org/bukkit/event/inventory/FurnaceSmeltEvent.java
index f8f9b08a0bd82a2667ae4e0c99dae9103f0db3f0..25478725bef34153bca204c815a167913b9cd389 100644
--- a/src/main/java/org/bukkit/event/inventory/FurnaceSmeltEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/FurnaceSmeltEvent.java
@@ -12,7 +12,13 @@ import org.jetbrains.annotations.NotNull;
  */
 public class FurnaceSmeltEvent extends BlockCookEvent {
 
+    @Deprecated // Paper
     public FurnaceSmeltEvent(@NotNull final Block furnace, @NotNull final ItemStack source, @NotNull final ItemStack result) {
         super(furnace, source, result);
     }
+    // Paper start
+    public FurnaceSmeltEvent(@NotNull final Block furnace, @NotNull final ItemStack source, @NotNull final ItemStack result, @org.jetbrains.annotations.Nullable org.bukkit.inventory.CookingRecipe<?> recipe) {
+        super(furnace, source, result, recipe);
+    }
+    // Paper end
 }
