From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Thu, 22 Apr 2021 16:45:15 -0700
Subject: [PATCH] add consumeFuel to FurnaceBurnEvent


diff --git a/src/main/java/org/bukkit/event/inventory/FurnaceBurnEvent.java b/src/main/java/org/bukkit/event/inventory/FurnaceBurnEvent.java
index 24077da8e6a7937f66eafc6779206055cf82e8d2..ba4dc8aed85169e55cac276bdd51116919305019 100644
--- a/src/main/java/org/bukkit/event/inventory/FurnaceBurnEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/FurnaceBurnEvent.java
@@ -18,6 +18,7 @@ public class FurnaceBurnEvent extends BlockEvent implements Cancellable {
     private int burnTime;
     private boolean cancelled;
     private boolean burning;
+    private boolean consumeFuel = true; // Paper
 
     public FurnaceBurnEvent(@NotNull final Block furnace, @NotNull final ItemStack fuel, final int burnTime) {
         super(furnace);
@@ -72,6 +73,25 @@ public class FurnaceBurnEvent extends BlockEvent implements Cancellable {
     public void setBurning(boolean burning) {
         this.burning = burning;
     }
+    // Paper start
+    /**
+     * Gets whether the furnace's fuel will be consumed or not.
+     *
+     * @return whether the furnace's fuel will be consumed
+     */
+    public boolean willConsumeFuel() {
+        return consumeFuel;
+    }
+
+    /**
+     * Sets whether the furnace's fuel will be consumed or not.
+     *
+     * @param consumeFuel true to consume the fuel
+     */
+    public void setConsumeFuel(boolean consumeFuel) {
+        this.consumeFuel = consumeFuel;
+    }
+    // Paper end
 
     @Override
     public boolean isCancelled() {
