From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Connor Linfoot <connorlinfoot@me.com>
Date: Wed, 12 May 2021 08:09:19 +0100
Subject: [PATCH] Add raw address to AsyncPlayerPreLoginEvent


diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index e4afc428e9a8625bf6f5967cb659987c14f65ba6..35176b16d208c9756bd00ac74b2a1dcc3fc9febb 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -27,6 +27,7 @@ public class AsyncPlayerPreLoginEvent extends Event {
     private net.kyori.adventure.text.Component message; // Paper
     //private String name; // Paper - Not used anymore
     private final InetAddress ipAddress;
+    private final InetAddress rawAddress; // Paper
     //private UUID uniqueId; // Paper - Not used anymore
 
     @Deprecated
@@ -57,7 +58,23 @@ public class AsyncPlayerPreLoginEvent extends Event {
         this.profile = profile;
     }
 
+    // Paper Start
+    /**
+     * Gets the raw address of the player logging in
+     * @return The address
+     */
+    @NotNull
+    public InetAddress getRawAddress() {
+        return rawAddress;
+    }
+    // Paper end
+
+    @Deprecated
     public AsyncPlayerPreLoginEvent(@NotNull final String name, @NotNull final InetAddress ipAddress, @NotNull final UUID uniqueId, @NotNull PlayerProfile profile) {
+        this(name, ipAddress, ipAddress, uniqueId, profile);
+    }
+
+    public AsyncPlayerPreLoginEvent(@NotNull final String name, @NotNull final InetAddress ipAddress, @NotNull final InetAddress rawAddress, @NotNull final UUID uniqueId, @NotNull PlayerProfile profile) {
         super(true);
         this.profile = profile;
         // Paper end
@@ -65,6 +82,7 @@ public class AsyncPlayerPreLoginEvent extends Event {
         this.message = net.kyori.adventure.text.Component.empty(); // Paper
         //this.name = name; // Paper - Not used anymore
         this.ipAddress = ipAddress;
+        this.rawAddress = rawAddress; // Paper
         //this.uniqueId = uniqueId; // Paper - Not used anymore
     }
 
