From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Tue, 22 Jun 2021 23:16:11 -0400
Subject: [PATCH] Stinger API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index ca74b215d226db16259e9ca9c930a49b3e444a7a..ef13c72963af2356f37a724066fe5b42416dd3eb 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -419,6 +419,52 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     int nextArrowRemoval();
     // Paper end - Add methods for working with arrows stuck in living entities
 
+    // Paper start - Bee Stinger API
+    /**
+     * Gets the time in ticks until the next bee stinger leaves the entity's body.
+     *
+     * @return ticks until bee stinger leaves
+     */
+    public int getBeeStingerCooldown();
+
+    /**
+     * Sets the time in ticks until the next stinger leaves the entity's body.
+     *
+     * @param ticks time until bee stinger leaves
+     */
+    public void setBeeStingerCooldown(int ticks);
+
+    /**
+     * Gets the amount of bee stingers in an entity's body.
+     *
+     * @return amount of bee stingers in body
+     */
+    public int getBeeStingersInBody();
+
+    /**
+     * Set the amount of bee stingers in the entity's body.
+     *
+     * @param count amount of bee stingers in entity's body
+     */
+    public void setBeeStingersInBody(int count);
+
+    /**
+     * Sets the amount of ticks before the next bee stinger gets removed from the entities body.
+     * <p>
+     * A value of 0 will cause the server to re-calculate the amount of ticks on the next tick.
+     *
+     * @param ticks Amount of ticks
+     */
+    void removeNextBeeStingerIn(@org.jetbrains.annotations.Range(from = 0, to = Integer.MAX_VALUE) int ticks);
+
+    /**
+     * Gets the amount of ticks before the next bee stinger gets removed from the entities body.
+     *
+     * @return ticks Amount of ticks
+     */
+    int nextBeeStingerRemoval();
+    // Paper end - Stinger API
+
     /**
      * Returns the living entity's current maximum no damage ticks.
      * <p>
