From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Sun, 24 Oct 2021 16:19:26 -0400
Subject: [PATCH] Add Raw Byte Entity Serialization


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 4cd43815c317cfa1dd0ac15fb469d601aee42e60..5be89089da4e8230dc7aa078712428189f38d9f9 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -169,6 +169,14 @@ public interface UnsafeValues {
 
     ItemStack deserializeItem(byte[] data);
 
+    byte[] serializeEntity(org.bukkit.entity.Entity entity);
+
+    default org.bukkit.entity.Entity deserializeEntity(byte[] data, World world) {
+        return deserializeEntity(data, world, false);
+    }
+
+    org.bukkit.entity.Entity deserializeEntity(byte[] data, World world, boolean preserveUUID);
+
     /**
      * Creates and returns the next EntityId available.
      * <p>
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 3b40c63a380e519ecae2e272754a53aff5aebd9a..464668496f5a611ae78bd5f6392915ec384862ac 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -946,5 +946,32 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @Deprecated
     @NotNull Set<Player> getTrackedPlayers();
+
+    /**
+     * Spawns the entity in the world at the given {@link Location} with the default spawn reason.
+     * <p>
+     * This will not spawn the entity if the entity is already spawned or has previously been despawned.
+     * <p>
+     * Also, this method will fire the same events as a normal entity spawn.
+     *
+     * @param location The location to spawn the entity at.
+     * @return Whether the entity was successfully spawned.
+     */
+    public default boolean spawnAt(@NotNull Location location) {
+        return spawnAt(location, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    /**
+     * Spawns the entity in the world at the given {@link Location} with the reason given.
+     * <p>
+     * This will not spawn the entity if the entity is already spawned or has previously been despawned.
+     * <p>
+     * Also, this method will fire the same events as a normal entity spawn.
+     *
+     * @param location The location to spawn the entity at.
+     * @param reason   The reason for the entity being spawned.
+     * @return Whether the entity was successfully spawned.
+     */
+    public boolean spawnAt(@NotNull Location location, @NotNull org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason);
     // Paper end
 }
