From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 24 Oct 2021 20:58:52 -0700
Subject: [PATCH] Entity powdered snow API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 464668496f5a611ae78bd5f6392915ec384862ac..97b8ac9ae8ebe0f074f5f1f806d9415b8122400c 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -973,5 +973,12 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return Whether the entity was successfully spawned.
      */
     public boolean spawnAt(@NotNull Location location, @NotNull org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason);
+
+    /**
+     * Check if entity is inside powdered snow.
+     *
+     * @return true if in powdered snow.
+     */
+    boolean isInPowderedSnow();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/entity/Skeleton.java b/src/main/java/org/bukkit/entity/Skeleton.java
index 0944d920301d8cd5ec3960454129f7d1c15ee6d4..ce04c03ba5f67e02faf10ed8aaef4d51fbf2b190 100644
--- a/src/main/java/org/bukkit/entity/Skeleton.java
+++ b/src/main/java/org/bukkit/entity/Skeleton.java
@@ -41,6 +41,16 @@ public interface Skeleton extends AbstractSkeleton {
      */
     void setConversionTime(int time);
 
+    // Paper start
+    /**
+     * Gets the time the skeleton
+     * has been inside powdered snow.
+     *
+     * @return time in ticks
+     */
+    int inPowderedSnowTime();
+    // Paper end
+
     /**
      * A legacy enum that defines the different variances of skeleton-like
      * entities on the server.
