From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kieran Wallbanks <kieran.wallbanks@gmail.com>
Date: Mon, 21 Jun 2021 12:33:45 +0100
Subject: [PATCH] Fix NotePlayEvent


diff --git a/src/main/java/org/bukkit/event/block/NotePlayEvent.java b/src/main/java/org/bukkit/event/block/NotePlayEvent.java
index c9ecbebbc27f27734d35f496783706f702dccb29..ea7cb75ba3ecbe40e0390fd5568db77de56c4f16 100644
--- a/src/main/java/org/bukkit/event/block/NotePlayEvent.java
+++ b/src/main/java/org/bukkit/event/block/NotePlayEvent.java
@@ -56,11 +56,13 @@ public class NotePlayEvent extends BlockEvent implements Cancellable {
 
     /**
      * Overrides the {@link Instrument} to be used.
+     * <p>
+     * Only works when the note block isn't under a player head.
+     * For this specific case the 'note_block_sound' property of the
+     * player head state takes the priority.
      *
      * @param instrument the Instrument. Has no effect if null.
-     * @deprecated no effect on newer Minecraft versions
      */
-    @Deprecated(since = "1.13")
     public void setInstrument(@NotNull Instrument instrument) {
         if (instrument != null) {
             this.instrument = instrument;
@@ -71,9 +73,7 @@ public class NotePlayEvent extends BlockEvent implements Cancellable {
      * Overrides the {@link Note} to be played.
      *
      * @param note the Note. Has no effect if null.
-     * @deprecated no effect on newer Minecraft versions
      */
-    @Deprecated(since = "1.13")
     public void setNote(@NotNull Note note) {
         if (note != null) {
             this.note = note;
