From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 14 Oct 2021 12:36:46 -0500
Subject: [PATCH] Left handed API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index c6d8622fba3397cf3434726f1d7c49e95b887e46..12ebceec3eb6f4c3325e6c6a592676795a988136 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -162,4 +162,20 @@ public interface Mob extends LivingEntity, Lootable {
      */
     @Nullable
     public Sound getAmbientSound();
+
+    // Paper start
+    /**
+     * Check if Mob is left-handed
+     *
+     * @return True if left-handed
+     */
+    public boolean isLeftHanded();
+
+    /**
+      * Set if Mob is left-handed
+      *
+      * @param leftHanded True if left-handed
+      */
+    public void setLeftHanded(boolean leftHanded);
+    // Paper end
 }
