From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom <cryptite@gmail.com>
Date: Mon, 20 Sep 2021 08:52:20 -0500
Subject: [PATCH] Add PlayerShieldDisableEvent


diff --git a/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java b/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..93f55f13c0fdd100dc87b2acf0127eb518205f29
--- /dev/null
+++ b/src/main/java/org/bukkit/event/player/PlayerShieldDisableEvent.java
@@ -0,0 +1,44 @@
+package org.bukkit.event.player;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Thrown whenever a {@link Player} dies
+ */
+public class PlayerShieldDisableEvent extends Event {
+    private static final HandlerList handlers = new HandlerList();
+
+    private final Player player;
+    private final Entity damager;
+
+    public PlayerShieldDisableEvent(@NotNull final Player player, @Nullable final Entity damager) {
+        this.player = player;
+        this.damager = damager;
+    }
+
+    @Nullable
+    public Entity getDamager() {
+        return damager;
+    }
+
+    @NotNull
+    public Player getPlayer() {
+        return player;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
