From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 14 Oct 2021 14:12:56 -0500
Subject: [PATCH] Add LivingEntity#safeFallDistance API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 31353bd20404a8c2acf6bf0df524dc3cae324272..055f9febf1d2a02a6209fb92f67e922e15999c3b 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -909,5 +909,22 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @param hurtDirection hurt direction
      */
     void setHurtDirection(float hurtDirection);
+
+    /**
+     * Gets the distance this entity can safely fall without taking damage
+     * <p>
+     * This won't affect mobs that take no fall damage already. Like Bats, bees, basically
+     * all the mobs that override that checkFallDamage method in LivingEntity
+     *
+     * @return Safe fall distance
+     */
+    float getSafeFallDistance();
+
+    /**
+      * Set the distance this entity can safely fall without taking damage
+      *
+      * @param safeFallDistance Safe fall distance
+      */
+    void setSafeFallDistance(float safeFallDistance);
     // Paper end
 }
