From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <admin@yive.me>
Date: Mon, 6 Dec 2021 01:28:13 -0800
Subject: [PATCH] Expose entity type ids


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 014c9984018ad5e51a26228a137e1ba4eb3e80c8..551719c8bca5314160de3b4dbb4713c1e7cf43ce 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -217,5 +217,13 @@ public interface UnsafeValues {
      * @throws IllegalArgumentException if {@link Material#isBlock()} is false
      */
     boolean isCollidable(@org.jetbrains.annotations.NotNull Material material);
+
+    /**
+     * Gets the id of entity types used in packets
+     *
+     * @param type the entity type
+     * @return entity type id
+     */
+    int getEntityProtocolId(org.bukkit.entity.EntityType type);
     // Paper end
 }
