From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <nassim@njahnke.dev>
Date: Sat, 5 Feb 2022 20:25:28 +0100
Subject: [PATCH] Implement regenerateChunk


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 8d5b73b92b78676467c24a73332a2a8c14cd0506..72f1576b8ce5b55b50f053f346ce42c52db4b568 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -465,8 +465,8 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      * @return Whether the chunk was actually regenerated
      *
      * @deprecated regenerating a single chunk is not likely to produce the same
-     * chunk as before as terrain decoration may be spread across chunks. Use of
-     * this method should be avoided as it is known to produce buggy results.
+     * chunk as before as terrain decoration may be spread across chunks. It may
+     * or may not change blocks in the adjacent chunks as well.
      */
     @Deprecated
     public boolean regenerateChunk(int x, int z);
