From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: u9g <git@u9g.dev>
Date: Tue, 14 Jun 2022 19:35:21 -0400
Subject: [PATCH] Add Player#getFishHook


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 10576d39e019c0ca6bbd07b889a4580b2883d875..188c8e27d724a1593dc586b58c7dcb74a1b3d926 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -425,6 +425,13 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Nullable
     public Location getPotentialBedLocation();
     // Paper end
+    // Paper start
+    /**
+     * @return the player's fishing hook if they are fishing
+     */
+    @Nullable
+    FishHook getFishHook();
+    // Paper end
 
     /**
      * Attempts to make the entity sleep at the given location.
