From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: u9g <winworkswow@gmail.com>
Date: Mon, 3 Jan 2022 23:58:30 -0500
Subject: [PATCH] Add new overload to PersistentDataContainer#has


diff --git a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
index 071cf222508ee15d95d7aa76f7979d83c4a6e421..eebb3da156e2d95efbe22d4afa470b977ce19f10 100644
--- a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+++ b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
@@ -157,7 +157,7 @@ public interface PersistentDataContainer {
      * Returns if the persistent metadata provider has metadata registered
      * matching the provided key.
      * 
-     * @param key the key to look up in the custom tag map
+     * @param key the key for which existence should be checked.
      * 
      * @return whether the key exists
      * 
