From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sat, 18 Dec 2021 10:34:21 -0800
Subject: [PATCH] Add GameEvent tags


diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 9694b862d62afbaa831735a6f6d095315bcdf37c..4749d4c01069f0e0cbe948ede1dd043baceaa97b 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -848,6 +848,18 @@ public interface Tag<T extends Keyed> extends Keyed {
      */
     @Deprecated(forRemoval = true)
     Tag<EntityType> SKELETONS = ENTITY_TYPES_SKELETONS;
+
+    String REGISTRY_GAME_EVENTS = "game_events";
+
+    /**
+     * Tag for game events that trigger sculk sensors
+     */
+    Tag<GameEvent> GAME_EVENT_VIBRATIONS = Bukkit.getTag(REGISTRY_GAME_EVENTS, NamespacedKey.minecraft("vibrations"), GameEvent.class);
+
+    /**
+     * Tag for game events that are ignored if the entity is sneaking
+     */
+    Tag<GameEvent> GAME_EVENT_IGNORE_VIBRATIONS_SNEAKING = Bukkit.getTag(REGISTRY_GAME_EVENTS, NamespacedKey.minecraft("ignore_vibrations_sneaking"), GameEvent.class);
     // Paper end
 
     /**
