From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Mon, 14 Mar 2022 22:45:32 -0700
Subject: [PATCH] Add getComputedBiome API


diff --git a/src/main/java/org/bukkit/RegionAccessor.java b/src/main/java/org/bukkit/RegionAccessor.java
index e55f6e2baf35dbd91c433ab9e62713eaac85435b..1d7560a3b2de23e16486608f3785a5dccc296158 100644
--- a/src/main/java/org/bukkit/RegionAccessor.java
+++ b/src/main/java/org/bukkit/RegionAccessor.java
@@ -26,6 +26,7 @@ public interface RegionAccessor {
      *
      * @param location the location of the biome
      * @return Biome at the given location
+     * @see #getComputedBiome(int, int, int)
      */
     @NotNull
     Biome getBiome(@NotNull Location location);
@@ -37,10 +38,33 @@ public interface RegionAccessor {
      * @param y Y-coordinate of the block
      * @param z Z-coordinate of the block
      * @return Biome at the given coordinates
+     * @see #getComputedBiome(int, int, int)
      */
     @NotNull
     Biome getBiome(int x, int y, int z);
 
+    // Paper start
+    /**
+     * Gets the computed {@link Biome} at the given coordinates.
+     *
+     * <p>The computed Biome is the Biome as seen by clients for rendering
+     * purposes and in the "F3" debug menu. This is computed by looking at the noise biome
+     * at this and surrounding quarts and applying complex math operations.</p>
+     *
+     * <p>Most other Biome-related methods named getBiome, setBiome, and similar
+     * operate on the "noise biome", which is stored per-quart, or in other words,
+     * 1 Biome per 4x4x4 block region. This is how Biomes are currently generated and
+     * stored on disk.</p>
+     *
+     * @param x X-coordinate of the block
+     * @param y Y-coordinate of the block
+     * @param z Z-coordinate of the block
+     * @return Biome at the given coordinates
+     */
+    @NotNull
+    Biome getComputedBiome(int x, int y, int z);
+    // Paper end
+
     /**
      * Sets the {@link Biome} at the given {@link Location}.
      *
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 3e980c630452c8ea72227bc4cd92c605253cd41b..f95b7a60dd6f3ed25f84296f249eb0776855d255 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -370,10 +370,22 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
      * Returns the biome that this block resides in
      *
      * @return Biome type containing this block
+     * @see #getComputedBiome()
      */
     @NotNull
     Biome getBiome();
 
+    // Paper start
+    /**
+     * Gets the computed biome at the location of this Block.
+     *
+     * @return computed biome at the location of this Block.
+     * @see org.bukkit.RegionAccessor#getComputedBiome(int, int, int)
+     */
+    @NotNull
+    Biome getComputedBiome();
+    // Paper end
+
     /**
      * Sets the biome that this block resides in
      *
