From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Thu, 2 Jun 2022 02:26:44 -0400
Subject: [PATCH] Move invisible methods to Entity


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 8bc6876c82935988436597161fa0ec94c032174b..e150636a85e694674780df143089c9565abae1db 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -829,5 +829,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return true if in powdered snow.
      */
     boolean isInPowderedSnow();
+
+    /**
+     * Sets whether the entity is invisible or not.
+     *
+     * @param invisible If the entity is invisible
+     */
+    void setInvisible(boolean invisible);
+
+    /**
+     * Gets whether the entity is invisible or not.
+     *
+     * @return Whether the entity is invisible
+     */
+    boolean isInvisible();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 31353bd20404a8c2acf6bf0df524dc3cae324272..174a368295d3df11d38ef570c4a93e5a44c21889 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -767,20 +767,6 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     @NotNull
     public EntityCategory getCategory();
 
-    /**
-     * Sets whether the entity is invisible or not.
-     *
-     * @param invisible If the entity is invisible
-     */
-    public void setInvisible(boolean invisible);
-
-    /**
-     * Gets whether the entity is invisible or not.
-     *
-     * @return Whether the entity is invisible
-     */
-    public boolean isInvisible();
-
     // Paper start
     /**
      * Get the number of arrows stuck in this entity
