From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 7 Feb 2023 08:20:27 -0800
Subject: [PATCH] Fix HandlerList for InventoryBlockStartEvent subclasses


diff --git a/src/main/java/org/bukkit/event/block/BrewingStartEvent.java b/src/main/java/org/bukkit/event/block/BrewingStartEvent.java
index 37be83184cae203d5e99518b0ff5c708fafb0331..43eac972f45d1cbb6278b048f8e6d7882c0aabeb 100644
--- a/src/main/java/org/bukkit/event/block/BrewingStartEvent.java
+++ b/src/main/java/org/bukkit/event/block/BrewingStartEvent.java
@@ -11,7 +11,7 @@ import org.jetbrains.annotations.NotNull;
 @org.jetbrains.annotations.ApiStatus.Experimental // Paper
 public class BrewingStartEvent extends InventoryBlockStartEvent {
 
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - remove HandlerList
     private int brewingTime;
 
     public BrewingStartEvent(@NotNull final Block furnace, @NotNull ItemStack source, int brewingTime) {
@@ -37,14 +37,5 @@ public class BrewingStartEvent extends InventoryBlockStartEvent {
         this.brewingTime = brewTime;
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - remove HandlerList
 }
diff --git a/src/main/java/org/bukkit/event/block/CampfireStartEvent.java b/src/main/java/org/bukkit/event/block/CampfireStartEvent.java
index 2d084214e991fecc51f8e18e3d733e43b1dca248..4b12575107b3f1fa6d0ed7f667bf0d0ae40acae0 100644
--- a/src/main/java/org/bukkit/event/block/CampfireStartEvent.java
+++ b/src/main/java/org/bukkit/event/block/CampfireStartEvent.java
@@ -12,7 +12,7 @@ import org.jetbrains.annotations.NotNull;
 @org.jetbrains.annotations.ApiStatus.Experimental // Paper
 public class CampfireStartEvent extends InventoryBlockStartEvent {
 
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - remove HandlerList
     private int cookingTime;
     private CampfireRecipe campfireRecipe;
 
@@ -50,14 +50,5 @@ public class CampfireStartEvent extends InventoryBlockStartEvent {
         this.cookingTime = cookTime;
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - remove HandlerList
 }
diff --git a/src/main/java/org/bukkit/event/inventory/FurnaceStartSmeltEvent.java b/src/main/java/org/bukkit/event/inventory/FurnaceStartSmeltEvent.java
index d386ab5dd46cc6706ace61fe6b646713ffd50cb7..abfd0441f1dd485e59ce5a9f7fca88ffa32e87f7 100644
--- a/src/main/java/org/bukkit/event/inventory/FurnaceStartSmeltEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/FurnaceStartSmeltEvent.java
@@ -14,7 +14,7 @@ import org.jetbrains.annotations.NotNull;
  * {@link org.bukkit.block.Smoker}, and {@link org.bukkit.block.BlastFurnace}.
  */
 public class FurnaceStartSmeltEvent extends InventoryBlockStartEvent {
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - remove HandlerList
     private final CookingRecipe<?> recipe;
     private int totalCookTime;
 
@@ -59,14 +59,5 @@ public class FurnaceStartSmeltEvent extends InventoryBlockStartEvent {
         this.totalCookTime = cookTime;
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - remove HandlerList
 }
