From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 23 May 2021 18:32:26 -0700
Subject: [PATCH] Add ender dragon boss spawn reason


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 6d5c81824c7cb36f0c76baf532434fb7bf233da6..38d6f20714403b3fc9ee1c3d3bfc8aff9cecac75 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -177,6 +177,12 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a piglin is converted to a zombified piglin.
          */
         PIGLIN_ZOMBIFIED,
+        // Paper start
+        /**
+         * When an ender dragon is spawned as part of a boss battle.
+         */
+        ENDER_DRAGON_BOSS,
+        // Paper end
         /**
          * When an entity is created by a cast spell.
          */
