From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zJanny <jan.trummer@hotmail.com>
Date: Fri, 29 Jul 2022 17:28:06 +0200
Subject: [PATCH] Add new override for setChunckForceLoaded


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index e8c0c853eb52d1473c20231660355f77b1f7e016..17903d37f5b2bccdea23c319108756bde61ca3cb 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -539,6 +539,19 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      */
     public void setChunkForceLoaded(int x, int z, boolean forced);
 
+    // Paper start
+    /**
+     * Sets whether the chunk at the specified chunk coordinates is force
+     * loaded.
+     * <p>
+     * A force loaded chunk will not be unloaded due to lack of player activity.
+     *
+     * @param chunk Chunk to be forceloaded
+     * @param forced force load status
+     */
+    public void setChunkForceLoaded(Chunk chunk, boolean forced);
+    // Paper end
+
     /**
      * Returns all force loaded chunks in this world.
      * <p>
