From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zJanny <jan.trummer@hotmail.com>
Date: Sat, 30 Jul 2022 19:37:37 +0200
Subject: [PATCH] Add Entity#hideEntityByDefault


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 03b35d3ba8ba00c0fa0272450f19355244a014ea..ab1f5fd33a9536eea0abfa44804f02bfa262e128 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -804,6 +804,20 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean fromMobSpawner();
 
+    // Paper start
+    /**
+     * Can be used to set the Entity hidden by default
+     * @param hide wether the Entity should be hidden or not
+     */
+    void setHideEntityByDefault(boolean hide);
+
+    /**
+     *
+     * @return True if the Entity is hidden by default
+     */
+    boolean getIsHiddenByDefault();
+    // Paper end
+
     /**
      * Gets the latest chunk an entity is currently or was in.
      *
