From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Doc <nachito94@msn.com>
Date: Wed, 3 Aug 2022 22:15:16 -0400
Subject: [PATCH] Add CreatureSpawnEvent.SpawnReason.SPAWNER_BUCKET


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 6d5c81824c7cb36f0c76baf532434fb7bf233da6..7d2990b33e2404c3c5942d852e8a948c052efabf 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -71,6 +71,12 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a creature spawns from a Spawner Egg
          */
         SPAWNER_EGG,
+        // Paper start
+        /**
+         * When a creature spawns from a Bucket
+         */
+        SPAWNER_BUCKET,
+        // Paper end
         /**
          * When a creature spawns because of a lightning strike
          */
