From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 12 Oct 2022 00:36:55 +0300
Subject: [PATCH] Allow changing bed's 'occupied' property


diff --git a/src/main/java/org/bukkit/block/data/type/Bed.java b/src/main/java/org/bukkit/block/data/type/Bed.java
index ed519bfebe5b921f60867a3900edfce9859058b6..6e7a456dc5e9bfc28a19029a3381e53fa6453d30 100644
--- a/src/main/java/org/bukkit/block/data/type/Bed.java
+++ b/src/main/java/org/bukkit/block/data/type/Bed.java
@@ -35,6 +35,15 @@ public interface Bed extends Directional {
      */
     boolean isOccupied();
 
+    // Paper start
+    /**
+     * Sets the value of the 'occupied' property.
+     *
+     * @param occupied the new 'occupied' value
+     */
+    void setOccupied(boolean occupied);
+    // Paper end
+
     /**
      * Horizontal half of a bed.
      */
