From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GodOfPro <1387ilia@gmail.com>
Date: Tue, 11 Apr 2023 16:30:58 +0430
Subject: [PATCH] Add Mob Experience reward API


diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index 0c511b64672f9bafb92be3aea6bab8c1c12a7041..30ef0c20df8ff2ed56adb40fea103265dd3e886c 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -212,4 +212,13 @@ public interface Mob extends LivingEntity, Lootable {
       */
     public void setLeftHanded(boolean leftHanded);
     // Paper end
+
+    // Paper start
+    /**
+     * Gets the amount of experience the mob will possibly drop. This value is randomized and it can give different results
+     *
+     * @return the amount of experience the mob will possibly drop
+     */
+    public int getPossibleExperienceReward();
+    // Paper end
 }
