From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TreyRuffy <TreyRuffy@users.noreply.github.com>
Date: Fri, 27 May 2022 02:25:38 -0600
Subject: [PATCH] Flying Fall Damage API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c19e72ec5fcba00948ad039bec7068c4545cc51d..0a75196050beff92948290b3e8d4b7d230c5700b 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1454,6 +1454,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void setAllowFlight(boolean flight);
 
+    // Paper start - flying fall damage
+    /**
+     * Allows you to enable fall damage while {@link #getAllowFlight()} is {@code true}
+     *
+     * @param flyingFallDamage Enables fall damage when {@link #getAllowFlight()} is {@code true}
+     */
+    public void setFlyingFallDamage(@NotNull net.kyori.adventure.util.TriState flyingFallDamage);
+
+    /**
+     * Allows you to get if fall damage is enabled while {@link #getAllowFlight()} is {@code true}
+     *
+     * @return A tristate of whether fall damage is enabled, not set, or disabled when {@link #getAllowFlight()} is {@code true}
+     */
+    @NotNull
+    public net.kyori.adventure.util.TriState hasFlyingFallDamage();
+    // Paper end
+
     /**
      * Hides a player from this player
      *
