From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 4 Mar 2022 09:46:40 -0800
Subject: [PATCH] More Raid API


diff --git a/src/main/java/org/bukkit/Raid.java b/src/main/java/org/bukkit/Raid.java
index 983a8c20a06d2b509602b27f49c090598b8ecc42..8518f0998cc45a822a6e9e40ac62f9ae47145aec 100644
--- a/src/main/java/org/bukkit/Raid.java
+++ b/src/main/java/org/bukkit/Raid.java
@@ -9,7 +9,7 @@ import org.jetbrains.annotations.NotNull;
 /**
  * Represents a raid event.
  */
-public interface Raid {
+public interface Raid extends org.bukkit.persistence.PersistentDataHolder { // Paper
 
     /**
      * Get whether this raid started.
@@ -131,4 +131,20 @@ public interface Raid {
          */
         STOPPED;
     }
+
+    // Paper start
+    /**
+     * Gets the id of this raid.
+     *
+     * @return the raid id
+     */
+    int getId();
+
+    /**
+     * Get the boss bar to be displayed for this raid.
+     *
+     * @return the boss bar
+     */
+    org.bukkit.boss.BossBar getBossBar();
+    // Paper end
 }
