From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 14 Mar 2023 19:37:29 -0700
Subject: [PATCH] Add new HIDE_UPGRADES ItemFlag


diff --git a/src/main/java/org/bukkit/inventory/ItemFlag.java b/src/main/java/org/bukkit/inventory/ItemFlag.java
index 73ed2d3c4aded81a24489381db16184383e625bb..ee99ef92d0afffe90f4e77da1f04e5724f868546 100644
--- a/src/main/java/org/bukkit/inventory/ItemFlag.java
+++ b/src/main/java/org/bukkit/inventory/ItemFlag.java
@@ -35,7 +35,13 @@ public enum ItemFlag {
     /**
      * Setting to show/hide dyes from coloured leather armour
      */
-    HIDE_DYE;
+    HIDE_DYE,
+    // Paper start
+    /**
+     * Setting to show/hide armor trim information.
+     */
+    @org.jetbrains.annotations.ApiStatus.Experimental // Paper - 1.20
+    HIDE_UPGRADES;
     // Paper start
     /**
      * Setting to show/hide item-specific information, including, but not limited to:
