From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LeonidM <wildhs.botm@gmail.com>
Date: Fri, 4 Aug 2023 16:05:09 +0300
Subject: [PATCH] Add Adventure components title and originalTitle in
 InventoryView


diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index 77a0cd901038405ed0e267f0432f13deccae3ab5..d18d507b59551dc7df23afba061e3dc174805064 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -457,6 +457,31 @@ public abstract class InventoryView {
     public /*abstract*/ net.kyori.adventure.text.Component title() {
         return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(this.getTitle());
     }
+
+    /**
+     * Sets the title of this inventory window to the specified title if the
+     * inventory window supports it.
+     * <p>
+     * Note if the inventory does not support titles that can be changed (ie, it
+     * is not creatable or viewed by a player), then this method will throw an
+     * exception.
+     *
+     * @param title The new title.
+     */
+    public /*abstract*/ void title(@NotNull net.kyori.adventure.text.Component title) {
+        this.setTitle(net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(title));
+    }
+
+    /**
+     * Get the original title of this inventory window, before any changes were
+     * made using {@link #setTitle(String)} or {@link #title(net.kyori.adventure.text.Component)}.
+     *
+     * @return the original title
+     */
+    @NotNull
+    public /*abstract*/ net.kyori.adventure.text.Component originalTitle() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(this.getOriginalTitle());
+    }
     // Paper end
 
     /**
@@ -471,10 +496,12 @@ public abstract class InventoryView {
 
     /**
      * Get the original title of this inventory window, before any changes were
-     * made using {@link #setTitle(String)}.
+     * made using {@link #setTitle(String)} or {@link #title(net.kyori.adventure.text.Component)}.
      *
      * @return the original title
+     * @deprecated in favour of {@link #originalTitle()}
      */
+    @Deprecated // Paper
     @NotNull
     public abstract String getOriginalTitle();
 
@@ -487,6 +514,8 @@ public abstract class InventoryView {
      * exception.
      *
      * @param title The new title.
+     * @deprecated in favour of {@link #title()}
      */
+    @Deprecated // Paper
     public abstract void setTitle(@NotNull String title);
 }
