From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Corey Shupe <coreyshupe101@gmail.com>
Date: Wed, 11 Jan 2023 16:40:31 -0500
Subject: [PATCH] Add Listing API for Player


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f5ea1e4709a3d72ea9fe3045f8bced7aa77bc3f4..bd1beef9615eb87297abab6abb55fae34c579626 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1943,6 +1943,32 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @ApiStatus.Experimental
     public boolean canSee(@NotNull Entity entity);
 
+    // Paper start
+    /**
+     * Returns whether the {@code other} player is listed for {@code this}.
+     *
+     * @param other The other {@link Player} to check for listing.
+     * @return True if the {@code other} player is listed for {@code this}.
+     */
+    boolean isListed(@NotNull Player other);
+
+    /**
+     * Unlists the {@code other} player from the tablist.
+     *
+     * @param other The other {@link Player} to de-list.
+     * @return True if the {@code other} player was listed.
+     */
+    boolean unlistPlayer(@NotNull Player other);
+
+    /**
+     * Lists the {@code other} player.
+     *
+     * @param other The other {@link Player} to list.
+     * @return True if the {@code other} player was not listed.
+     */
+    boolean listPlayer(@NotNull Player other);
+    // Paper end
+
     /**
      * Checks to see if this player is currently flying or not.
      *
