From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aurora <aurora@relanet.eu>
Date: Sat, 20 Mar 2021 18:14:24 +0100
Subject: [PATCH] Add EntityExtinguishEvent


diff --git a/src/main/java/io/papermc/paper/event/entity/EntityExtinguishEvent.java b/src/main/java/io/papermc/paper/event/entity/EntityExtinguishEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..cbeeb9f52e7886e4561b6e575a944ed0a78898c7
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/entity/EntityExtinguishEvent.java
@@ -0,0 +1,84 @@
+package io.papermc.paper.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when a burning {@link Entity} is extinguished.
+ */
+public class EntityExtinguishEvent extends EntityEvent implements Cancellable {
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final Cause cause;
+    private boolean cancelled;
+
+    @ApiStatus.Internal
+    public EntityExtinguishEvent(final @NotNull Entity entity, final @NotNull Cause cause) {
+        super(entity);
+        this.cause = cause;
+    }
+
+    @NotNull
+    public Cause getCause() {
+        return cause;
+    }
+
+    @Override
+    public void setCancelled(boolean cancelled) {
+        this.cancelled = cancelled;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    public enum Cause {
+        /**
+         * Extinguish caused by entity standing in water.
+         */
+        WATER,
+        /**
+         * Extinguish caused by entity standing in the rain.
+         */
+        RAIN,
+        /**
+         * Extinguish caused by entity standing in a bubble column.
+         */
+        BUBBLE_COLUMN,
+        /**
+         * Extinguish caused by entity standing in a cauldron with water.
+         */
+        CAULDRON,
+        /**
+         * Extinguish caused by entity dying.
+         */
+        DEATH,
+        /**
+         * Extinguish caused by fire ticks running out.
+         */
+        TIME,
+        /**
+         * Extinguish caused by standing in powdered snow.
+         */
+        POWDER_SNOW,
+        /**
+         * Extinguish caused by being splashed with a water bottle.
+         */
+        WATER_BOTTLE
+    }
+
+    @Override
+    @NotNull
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
