From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Wed, 6 Jan 2021 23:38:19 -0800
Subject: [PATCH] CommandFunction API


diff --git a/src/main/java/io/papermc/paper/datapack/CommandFunction.java b/src/main/java/io/papermc/paper/datapack/CommandFunction.java
new file mode 100644
index 0000000000000000000000000000000000000000..5e7e15af3e656cad42939002e3c5db4a3b2b74fa
--- /dev/null
+++ b/src/main/java/io/papermc/paper/datapack/CommandFunction.java
@@ -0,0 +1,31 @@
+package io.papermc.paper.datapack;
+
+import net.kyori.adventure.key.Keyed;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public interface CommandFunction extends Keyed {
+    /**
+     * Executes the function.
+     *
+     * @return the number of commands run
+     */
+    int executeAsServer();
+
+    /**
+     * Executes the function as the specified sender.
+     *
+     * @param sender the sender to execute the function as
+     * @return the number of commands run
+     */
+    int executeAs(@NotNull CommandSender sender);
+
+    /**
+     * Gets the commands run by this function.
+     *
+     * @return the list of commands
+     */
+    @NotNull List<String> getCommands();
+}
diff --git a/src/main/java/io/papermc/paper/datapack/DatapackManager.java b/src/main/java/io/papermc/paper/datapack/DatapackManager.java
index 58f78d5e91beacaf710f62461cf869f70d08b2a2..b22d3e8f26369af500a46ea599e0595d2360820d 100644
--- a/src/main/java/io/papermc/paper/datapack/DatapackManager.java
+++ b/src/main/java/io/papermc/paper/datapack/DatapackManager.java
@@ -1,6 +1,8 @@
 package io.papermc.paper.datapack;
 
+import net.kyori.adventure.key.Key;
 import org.checkerframework.checker.nullness.qual.NonNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.Collection;
 
@@ -18,4 +20,20 @@ public interface DatapackManager {
     @NonNull
     Collection<Datapack> getEnabledPacks();
 
+    /**
+     * Gets all command functions registered with the server.
+     *
+     * @return the list of functions
+     */
+    @NonNull
+    Collection<CommandFunction> getCommandFunctions();
+
+    /**
+     * Get the command function with this namespaced key.
+     *
+     * @param functionKey the namespaced key
+     * @return the command function if found, null if none
+     */
+    @Nullable
+    CommandFunction getCommandFunction(@Nullable Key functionKey);
 }
