From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TrollyLoki <trollyloki@gmail.com>
Date: Wed, 11 Oct 2023 00:45:54 -0400
Subject: [PATCH] Add API to get the collision shape of a block before it's
 placed


diff --git a/src/main/java/org/bukkit/block/data/BlockData.java b/src/main/java/org/bukkit/block/data/BlockData.java
index 31111cea5ffd018c3c011c1f3b8befbbd33db5e5..66903e0fa65052ede789a163dad723ba05a2cc8c 100644
--- a/src/main/java/org/bukkit/block/data/BlockData.java
+++ b/src/main/java/org/bukkit/block/data/BlockData.java
@@ -204,6 +204,19 @@ public interface BlockData extends Cloneable {
      */
     boolean isFaceSturdy(@NotNull BlockFace face, @NotNull BlockSupport support);
 
+    // Paper start
+    /**
+     * Calculates the collision shape this block data would have at a particular location.
+     * <p>
+     * This does not take into account any block updates that may occur if the block was to be actually placed in the world.
+     *
+     * @param location the location to calculate the collision shape at
+     *
+     * @return a {@link org.bukkit.util.VoxelShape} representing the collision shape of this block data.
+     */
+    @NotNull org.bukkit.util.VoxelShape getCollisionShape(@NotNull Location location);
+    // Paper end
+
     /**
      * Gets the material that a player would use to place this block.
      * <p>
