From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Warrior <50800980+Warriorrrr@users.noreply.github.com>
Date: Sat, 10 Feb 2024 10:05:59 +0100
Subject: [PATCH] Add getChunkSnapshot includeLightData parameter


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 546888898d9d6827079fe041c7bc6eb4e1e4605c..d547ae2b20c58bc703de4532b3b591dd34ddb1c6 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -102,6 +102,23 @@ public interface Chunk extends PersistentDataHolder {
     @NotNull
     ChunkSnapshot getChunkSnapshot(boolean includeMaxblocky, boolean includeBiome, boolean includeBiomeTempRain);
 
+    // Paper start - Add getChunkSnapshot includeLightData parameter
+    /**
+     * Capture thread-safe read-only snapshot of chunk data
+     *
+     * @param includeMaxblocky if true, snapshot includes per-coordinate
+     *     maximum Y values
+     * @param includeBiome if true, snapshot includes per-coordinate biome
+     *     type
+     * @param includeBiomeTempRain if true, snapshot includes per-coordinate
+     *     raw biome temperature and rainfall
+     * @param includeLightData Whether to include per-coordinate light emitted by blocks and sky light data
+     * @return ChunkSnapshot
+     */
+    @NotNull
+    ChunkSnapshot getChunkSnapshot(boolean includeMaxblocky, boolean includeBiome, boolean includeBiomeTempRain, boolean includeLightData);
+    // Paper end - Add getChunkSnapshot includeLightData parameter
+
     /**
      * Checks if entities in this chunk are loaded.
      *
