From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: leguan <longboard.noah@gmail.com>
Date: Sun, 14 Jan 2024 21:26:35 +0100
Subject: [PATCH] CropBlock API


diff --git a/src/main/java/io/papermc/paper/block/CropBlock.java b/src/main/java/io/papermc/paper/block/CropBlock.java
new file mode 100644
index 0000000000000000000000000000000000000000..9bbf45d4ef2c4cad81a092af7296e31f8b33a139
--- /dev/null
+++ b/src/main/java/io/papermc/paper/block/CropBlock.java
@@ -0,0 +1,18 @@
+package io.papermc.paper.block;
+
+import org.bukkit.Material;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents a CropBlock.
+ */
+public interface CropBlock {
+
+    /**
+     * Gets the seed of the planted block. E.g. WHEAT_SEEDS for WHEAT.
+     *
+     * @return The seed of the planted block.
+     */
+    @NotNull Material getBaseSeedId();
+
+}
