From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 1 May 2024 08:22:13 +0300
Subject: [PATCH] More Chest Block API


diff --git a/src/main/java/org/bukkit/block/Chest.java b/src/main/java/org/bukkit/block/Chest.java
index db6affbc78106b2d93b41953b624a0bca0ca1d72..5d02f9c938d0d7d0f4e491ccfaf6beb0a7a61aa4 100644
--- a/src/main/java/org/bukkit/block/Chest.java
+++ b/src/main/java/org/bukkit/block/Chest.java
@@ -27,4 +27,14 @@ public interface Chest extends Container, LootableBlockInventory, Lidded { // Pa
      */
     @NotNull
     Inventory getBlockInventory();
+
+    // Paper start - More Chest Block API
+    /**
+     * Checks whether this chest is blocked
+     * by either a block above or a sitting cat
+     *
+     * @return whether this chest is blocked
+     */
+    boolean isBlocked();
+    // Paper end - More Chest Block API
 }
diff --git a/src/main/java/org/bukkit/block/EnderChest.java b/src/main/java/org/bukkit/block/EnderChest.java
index 1150b449a3f5c40fe10136779c2ccc65ab4d884c..6b66f38e5509f90aad5ee1fffca01003dcbe9896 100644
--- a/src/main/java/org/bukkit/block/EnderChest.java
+++ b/src/main/java/org/bukkit/block/EnderChest.java
@@ -3,4 +3,13 @@ package org.bukkit.block;
 /**
  * Represents a captured state of an ender chest.
  */
-public interface EnderChest extends Lidded, TileState { }
+public interface EnderChest extends Lidded, TileState {
+    // Paper start - More Chest Block API
+    /**
+     * Checks whether this ender chest is blocked by a block above
+     *
+     * @return whether this ender chest is blocked
+     */
+    boolean isBlocked();
+    // Paper end - More Chest Block API
+}
