From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vicisacat <victor.branchu@gmail.com>
Date: Fri, 15 Mar 2024 17:35:18 +0100
Subject: [PATCH] Expanded Hopper API


diff --git a/src/main/java/org/bukkit/block/Hopper.java b/src/main/java/org/bukkit/block/Hopper.java
index 7ade312f180b7e30871d3a3240c76325cc369c26..afaed48ee2cad07d9bab4e3bccd833a8ec4b4a22 100644
--- a/src/main/java/org/bukkit/block/Hopper.java
+++ b/src/main/java/org/bukkit/block/Hopper.java
@@ -6,4 +6,19 @@ import org.bukkit.loot.Lootable;
 /**
  * Represents a captured state of a hopper.
  */
-public interface Hopper extends Container, LootableBlockInventory { } // Paper
+public interface Hopper extends Container, LootableBlockInventory { // Paper
+    // Paper start - Expanded Hopper API
+    /**
+     * Sets the cooldown before the hopper transfers or sucks in another item
+     * @param cooldown the cooldown in ticks
+     */
+    void setTransferCooldown(int cooldown);
+
+    /**
+     * Returns the cooldown before the hopper transfers or sucks in another item
+     * @return the cooldown in ticks
+     */
+    int getTransferCooldown();
+    // Paper end - Expanded Hopper API
+}
+
