From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Badbird5907 <50347938+Badbird5907@users.noreply.github.com>
Date: Mon, 4 Mar 2024 22:18:33 -0500
Subject: [PATCH] Add BlockBreakProgressUpdateEvent


diff --git a/src/main/java/io/papermc/paper/event/block/BlockBreakProgressUpdateEvent.java b/src/main/java/io/papermc/paper/event/block/BlockBreakProgressUpdateEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..0cd1ab085222eae1a8df8ad85b26b49b6dd93a09
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/block/BlockBreakProgressUpdateEvent.java
@@ -0,0 +1,56 @@
+package io.papermc.paper.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Range;
+
+/**
+ * Called when the progress of a block break is updated.
+ */
+public class BlockBreakProgressUpdateEvent extends BlockEvent {
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private final float progress;
+    private final Entity entity;
+
+    @ApiStatus.Internal
+    public BlockBreakProgressUpdateEvent(@NotNull final Block block, final float progress, @NotNull final Entity entity) {
+        super(block);
+        this.progress = progress;
+        this.entity = entity;
+    }
+
+    /**
+     * The progress of the block break
+     * <p>
+     * The progress ranges from 0.0 - 1.0, where 0 is no damage and
+     * 1.0 is the most damaged
+     *
+     * @return The progress of the block break
+     */
+    public float getProgress() {
+        return progress;
+    }
+
+    /**
+     * The entity breaking the block.
+     *
+     * @return The entity breaking the block
+     */
+    @NotNull
+    public Entity getEntity() {
+        return entity;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+    public static @NotNull HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
