From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 1 May 2024 07:44:50 +0300
Subject: [PATCH] Add missing fishing event state


diff --git a/src/main/java/org/bukkit/event/player/PlayerFishEvent.java b/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
index d4001f64a7ee9d5173e9bafd9c45860cbda1fc85..8b4ad421700f859396291508b178af9b51f23b0a 100644
--- a/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
@@ -165,5 +165,13 @@ public class PlayerFishEvent extends PlayerEvent implements Cancellable {
          * in.
          */
         BITE
+        // Paper start - Add missing fishing event state
+        ,
+        /**
+         * Called when a bobber was lured, and is now waiting to be hooked
+         * (when a "fish" starts to swim toward the bobber to bite it).
+         */
+        LURED,
+        // Paper end - Add missing fishing event state
     }
 }
