From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Flo0 <flo.roma@web.de>
Date: Sat, 6 Apr 2024 23:32:45 +0200
Subject: [PATCH] Fixed imports


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 2c2802caea79f853ec14f0cfe382471d32417b19..63d7947cd33b76d49acb199dd2a22d3a3bf933c5 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -54,7 +54,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Represents a player, connected or not
  */
-public interface Player extends org.bukkit.entity.HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
 
     // Paper start
     @Override
@@ -3466,7 +3466,7 @@ public interface Player extends org.bukkit.entity.HumanEntity, Conversable, Offl
 
     // Paper start - elytra boost API
     /**
-     * Boost a Player that's {@link #isGliding()} using a {@link org.bukkit.entity.Firework}.
+     * Boost a Player that's {@link #isGliding()} using a {@link Firework}.
      * If the creation of the entity is cancelled, no boosting is done.
      * This method does not fire {@link com.destroystokyo.paper.event.player.PlayerElytraBoostEvent}.
      *
@@ -3474,10 +3474,10 @@ public interface Player extends org.bukkit.entity.HumanEntity, Conversable, Offl
      * @return The {@link org.bukkit.entity.Firework} boosting the Player or null if the spawning of the entity was cancelled
      * @throws IllegalArgumentException if {@link #isGliding()} is false
      * or if the {@code firework} isn't a {@link Material#FIREWORK_ROCKET}
-     * @deprecated use {@link org.bukkit.entity.HumanEntity#fireworkBoost(ItemStack)} instead. Note that this method <b>does not</b>
+     * @deprecated use {@link HumanEntity#fireworkBoost(ItemStack)} instead. Note that this method <b>does not</b>
      * check if the player is gliding or not.
      */
-    default @Nullable org.bukkit.entity.Firework boostElytra(final @NotNull ItemStack firework) {
+    default @Nullable Firework boostElytra(final @NotNull ItemStack firework) {
         com.google.common.base.Preconditions.checkState(this.isGliding(), "Player must be gliding");
         return this.fireworkBoost(firework);
     }
@@ -3522,7 +3522,7 @@ public interface Player extends org.bukkit.entity.HumanEntity, Conversable, Offl
     // Paper end - custom chat completions API
 
     // Spigot start
-    public class Spigot extends org.bukkit.entity.Entity.Spigot {
+    public class Spigot extends Entity.Spigot {
 
         /**
          * Gets the connection address of this player, regardless of whether it
