From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Wed, 20 Mar 2024 20:23:34 -0400
Subject: [PATCH] fixup! Add BlockPreDispenseEvent


diff --git a/src/main/java/io/papermc/paper/event/block/BlockPreDispenseEvent.java b/src/main/java/io/papermc/paper/event/block/BlockPreDispenseEvent.java
index 62547d5853e7355a805fa1e7f1e5f56d06fca4ba..90374579dafb6a296de84ac1b4c3ef43b62c49fb 100644
--- a/src/main/java/io/papermc/paper/event/block/BlockPreDispenseEvent.java
+++ b/src/main/java/io/papermc/paper/event/block/BlockPreDispenseEvent.java
@@ -12,7 +12,7 @@ public class BlockPreDispenseEvent extends BlockEvent implements Cancellable {
 
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
-    private final ItemStack itemStack;
+    private ItemStack itemStack;
     private final int slot;
 
     private boolean cancelled;
@@ -34,6 +34,17 @@ public class BlockPreDispenseEvent extends BlockEvent implements Cancellable {
         return this.itemStack;
     }
 
+    /**
+     * Sets the itemstack to be used in dispensing this object.
+     * <p>
+     * This does not modify the underlying slot that this itemstack was fetched from.
+     * In which case, see {@link BlockPreDispenseEvent#getSlot()}
+     * @param itemStack new item stack
+     */
+    public void setItemStack(@NotNull final ItemStack itemStack) {
+        this.itemStack = itemStack;
+    }
+
     /**
      * Gets the inventory slot of the dispenser to dispense from.
      *
