From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Flo0 <flo.roma@web.de>
Date: Mon, 8 Apr 2024 16:22:07 +0200
Subject: [PATCH] API for checking sent chunks


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c6cb4f17469a8f2e60dd3e28d41402851ce5fb21..a26afd2c821a2dfdcbe7c922aee3b6b6ba8db539 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3707,6 +3707,41 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     void resetIdleDuration();
     // Paper end
 
+    // Paper start - Add chunk view API
+    /**
+     * Gets the a set of chunk keys for all chunks that have been sent to the player.
+     *
+     * @return an immutable set of chunk keys
+     */
+    java.util.@NotNull @org.jetbrains.annotations.Unmodifiable Set<Long> getSentChunkKeys();
+
+    /**
+     * Gets the set of chunks that have been sent to the player.
+     *
+     * @return an immutable set of chunks
+     */
+    java.util.@NotNull @org.jetbrains.annotations.Unmodifiable Set<org.bukkit.Chunk> getSentChunks();
+
+    /**
+     * Checks if the player has been sent a specific chunk.
+     *
+     * @param chunk the chunk to check
+     * @return true if the player has been sent the chunk, false otherwise
+     */
+    default boolean isChunkSent(@NotNull org.bukkit.Chunk chunk) {
+        return this.isChunkSent(chunk.getChunkKey());
+    }
+
+    /**
+     * Checks if the player has been sent a specific chunk.
+     *
+     * @param chunkKey the chunk key to check
+     * @return true if the player has been sent the chunk, false otherwise
+     * @see org.bukkit.Chunk#getChunkKey()
+     */
+    boolean isChunkSent(long chunkKey);
+    // Paper end
+
     @NotNull
     @Override
     Spigot spigot();
