From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kokiriglade <60290002+celerry@users.noreply.github.com>
Date: Sat, 23 Nov 2024 18:08:13 +0000
Subject: [PATCH] Expanded Art API


diff --git a/src/main/java/org/bukkit/Art.java b/src/main/java/org/bukkit/Art.java
index 00a290f1bf58cdc2238c13fd5a6048a26b7871da..f0ff1fa7f0a2723def50bc1402733bc9dbb33d35 100644
--- a/src/main/java/org/bukkit/Art.java
+++ b/src/main/java/org/bukkit/Art.java
@@ -107,6 +107,32 @@ public interface Art extends OldEnum<Art>, Keyed {
     @NotNull NamespacedKey getKey();
     // Paper end - deprecate getKey
 
+    // Paper start - name and author components, assetId key
+    /**
+     * Get the painting's title.
+     *
+     * @return the title
+     * @since 1.21.3
+     */
+    java.util.@NotNull Optional<net.kyori.adventure.text.Component> title();
+
+    /**
+     * Get the painting's author.
+     *
+     * @return the author
+     * @since 1.21.3
+     */
+    java.util.@NotNull Optional<net.kyori.adventure.text.Component> author();
+
+    /**
+     * Get the painting's asset id
+     *
+     * @return the asset id
+     * @since 1.21.3
+     */
+    net.kyori.adventure.key.@NotNull Key assetId();
+    // Paper end - name and author components, assetId key
+
     /**
      * Get a painting by its numeric ID
      *
