From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tamion <70228790+notTamion@users.noreply.github.com>
Date: Sat, 13 Jul 2024 18:00:37 +0200
Subject: [PATCH] Fix HelpCommand throwing exception when search is empty


diff --git a/src/main/java/org/bukkit/command/defaults/HelpCommand.java b/src/main/java/org/bukkit/command/defaults/HelpCommand.java
index a1699aa93c62f70238822c446326ed46ea71db14..348e8c98382492c7505f6fabfef8acd1beb4e5a6 100644
--- a/src/main/java/org/bukkit/command/defaults/HelpCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/HelpCommand.java
@@ -146,7 +146,7 @@ public class HelpCommand extends BukkitCommand {
         for (HelpTopic topic : Bukkit.getServer().getHelpMap().getHelpTopics()) {
             String trimmedTopic = topic.getName().startsWith("/") ? topic.getName().substring(1) : topic.getName();
 
-            if (trimmedTopic.length() < searchString.length()) {
+            if (searchString.isEmpty() || trimmedTopic.length() < searchString.length()) { // Paper - Fix HelpCommand throwing exception when search is empty
                 continue;
             }
 
