From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tamion <70228790+notTamion@users.noreply.github.com>
Date: Sat, 13 Jul 2024 18:00:37 +0200
Subject: [PATCH] Fix HelpCommand searching


diff --git a/src/main/java/org/bukkit/command/defaults/HelpCommand.java b/src/main/java/org/bukkit/command/defaults/HelpCommand.java
index a1699aa93c62f70238822c446326ed46ea71db14..599078b066ed760b287cb529c37380bd11326bb2 100644
--- a/src/main/java/org/bukkit/command/defaults/HelpCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/HelpCommand.java
@@ -82,7 +82,15 @@ public class HelpCommand extends BukkitCommand {
             return true;
         }
 
-        ChatPaginator.ChatPage page = ChatPaginator.paginate(topic.getFullText(sender), pageNumber, pageWidth, pageHeight);
+        // Paper start - Fix HelpCommand searching
+        String topicText = topic.getFullText(sender);
+        if (topicText.split("\n").length == 1) {
+            sender.sendMessage(ChatColor.RED + "No help for " + command);
+            return true;
+        }
+
+        ChatPaginator.ChatPage page = ChatPaginator.paginate(topicText, pageNumber, pageWidth, pageHeight);
+        // Paper end - Fix HelpCommand searching
 
         StringBuilder header = new StringBuilder();
         header.append(ChatColor.YELLOW);
@@ -146,7 +154,7 @@ public class HelpCommand extends BukkitCommand {
         for (HelpTopic topic : Bukkit.getServer().getHelpMap().getHelpTopics()) {
             String trimmedTopic = topic.getName().startsWith("/") ? topic.getName().substring(1) : topic.getName();
 
-            if (trimmedTopic.length() < searchString.length()) {
+            if (searchString.isEmpty() || trimmedTopic.length() < searchString.length()) { // Paper - Fix HelpCommand searching
                 continue;
             }
 
