From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fabrizio La Rosa <lr.fabrizio@gmail.com>
Date: Fri, 6 Sep 2024 23:59:25 +0200
Subject: [PATCH] Added Decorated Pot cracked API


diff --git a/src/main/java/org/bukkit/block/data/type/DecoratedPot.java b/src/main/java/org/bukkit/block/data/type/DecoratedPot.java
index eb0ffa977450ef0c79caa78302cfe75ee35b34b7..5b6b9a46b0be807c33311e0518ac45d05987b7eb 100644
--- a/src/main/java/org/bukkit/block/data/type/DecoratedPot.java
+++ b/src/main/java/org/bukkit/block/data/type/DecoratedPot.java
@@ -4,4 +4,15 @@ import org.bukkit.block.data.Directional;
 import org.bukkit.block.data.Waterlogged;
 
 public interface DecoratedPot extends Directional, Waterlogged {
+    /**
+     * @return whether the pot is cracked
+     */
+    public boolean isCracked();
+
+    /**
+     * Set whether the pot is cracked.
+     *
+     * @param cracked whether the pot is cracked
+     */
+    public void setCracked(boolean cracked);
 }
