From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Sat, 29 Aug 2020 21:55:41 +0200
Subject: [PATCH] Add PlayerInsertLecternBookEvent


diff --git a/src/main/java/io/papermc/paper/event/player/PlayerInsertLecternBookEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerInsertLecternBookEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..fa26303480d1daec03f1d14e644895d68cf1ce3a
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/player/PlayerInsertLecternBookEvent.java
@@ -0,0 +1,83 @@
+package io.papermc.paper.event.player;
+
+import com.google.common.base.Preconditions;
+import org.bukkit.block.Block;
+import org.bukkit.block.Lectern;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This event is called when a player clicks on a lectern to insert a book.
+ * If this event is cancelled the player will keep the book and the lectern will remain empty.
+ */
+public class PlayerInsertLecternBookEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private boolean cancelled = false;
+    private final Block lectern;
+    private ItemStack book;
+
+    public PlayerInsertLecternBookEvent(@NotNull Player who, @NotNull Block lectern, @NotNull ItemStack book) {
+        super(who);
+        this.lectern = lectern;
+        this.book = book;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    /**
+     * Gets the block of the lectern involved in this event.
+     *
+     * @return the block of the lectern.
+     */
+    @NotNull
+    public Block getLectern() {
+        return lectern;
+    }
+
+    /**
+     * Returns the itemstack the player tried to insert. This is a copy of the item,
+     * changes made to this itemstack will not affect the book that is being placed in the lectern.
+     * If you want to mutate the item stack that ends up in the lectern, use {@link #setBook(ItemStack)}.
+     *
+     * @return the book that is being placed
+     */
+    @NotNull
+    public ItemStack getBook() {
+        return book.clone();
+    }
+
+    @Override
+    public void setCancelled(boolean cancelled) {
+        this.cancelled = cancelled;
+    }
+
+    /**
+     * Sets the itemstack to insert into the lectern.
+     *
+     * @param book the book to insert or {@link ItemStack#empty()} if no book should be inserted.
+     */
+    public void setBook(@NotNull ItemStack book) {
+        Preconditions.checkArgument(book != null, "Cannot set book to null");
+        this.book = book.clone();
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
