From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Abel <abelvanhulst@gmail.com>
Date: Tue, 12 Nov 2024 22:25:35 +0100
Subject: [PATCH] API to allow/disallow tick sleeping


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ba366576a571214e67bcc529dc1bca19e1d59ef8..f55638eb8b315864052f9fe17ab4846e5e9d8dbb 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2578,5 +2578,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * Returns whether the server is sleeping/paused.
      */
     boolean isPaused();
+
+    /**
+     * Allows or disallows the server to sleep/pause.
+     * If any plugin disallows pausing, the server will never pause.
+     *
+     * @param plugin The {@link org.bukkit.plugin.Plugin} that's allowing or disallowing pausing.
+     * @param value Whether to allow sleeping of the server (defaults to true).
+     */
+    void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
 }
