From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Strokkur24 <strokkur24@gmail.com>
Date: Sat, 30 Nov 2024 11:54:38 +0100
Subject: [PATCH] Extend HumanEntity#dropItem API


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 488604ba1a516b477693877c74712e4a45624a8b..75493dabe98490a74fc0ce20c47445fde4d0c171 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -706,6 +706,97 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     public boolean dropItem(boolean dropAll);
 
+    // Paper start - Extend HumanEntity#dropItem API
+    /**
+     * Makes the entity drop an item from their inventory.
+     * <br>
+     * This method calls {@link HumanEntity#dropItem(int slot)}
+     * with the first {@link ItemStack} occurrence in the inventory
+     *
+     * @param itemStack The ItemStack to drop
+     * @return The dropped item, or null if the action was unsuccessful
+     */
+    public Item dropItem(final ItemStack itemStack);
+
+    /**
+     * Makes the entity drop an item from their inventory.
+     * <br>
+     * This method calls {@link HumanEntity#dropItem(int slot, boolean persistThrower)}
+     * with the first {@link ItemStack} occurrence in the inventory
+     *
+     * @param itemStack The ItemStack to drop from their inventory
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @return The dropped item, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final ItemStack itemStack, final boolean persistThrower);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the specified ItemStack.
+     * <br>
+     * This method calls {@link HumanEntity#dropItem(int slot, boolean persistThrower, boolean throwRandomly)}
+     * with the first {@link ItemStack} occurrence in the inventory
+     *
+     * @param itemStack The ItemStack to drop
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @param throwRandomly Whether the item should disperse randomly
+     * @return The dropped item, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final ItemStack itemStack, final boolean persistThrower, final boolean throwRandomly);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the slot.
+     *
+     * @param slot The slot to drop
+     * @return The dropped item, or null if the action was unsuccessful
+     */
+    public Item dropItem(final int slot);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the slot.
+     *
+     * @param slot The slot to drop
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @return The dropped item, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final int slot, final boolean persistThrower);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the slot.
+     *
+     * @param slot The slot to drop
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @param throwRandomly Whether the item should disperse randomly
+     * @return The dropped item entity, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final int slot, final boolean persistThrower, final boolean throwRandomly);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the {@link org.bukkit.inventory.EquipmentSlot}
+     *
+     * @param slot The equipment slot to drop
+     * @return The dropped item entity, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final org.bukkit.inventory.EquipmentSlot slot);
+    /**
+     * Makes the entity drop an item from their inventory based on the {@link org.bukkit.inventory.EquipmentSlot}
+     *
+     * @param slot The equipment slot to drop
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @return The dropped item entity, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final org.bukkit.inventory.EquipmentSlot slot, final boolean persistThrower);
+
+    /**
+     * Makes the entity drop an item from their inventory based on the equipment slot.
+     *
+     * @param slot The equipment slot to drop
+     * @param persistThrower Whether the resulting {@link Item} should have its thrower set to this {@link HumanEntity}
+     * @param throwRandomly Whether the item should disperse randomly
+     * @return The dropped item entity, or null if the action was unsuccessful
+     */
+    public @Nullable Item dropItem(final org.bukkit.inventory.EquipmentSlot slot, final boolean persistThrower, final boolean throwRandomly);
+    // Paper end
+
     /**
      * Gets the players current exhaustion level.
      * <p>
