From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: masmc05 <masmc05@gmail.com>
Date: Sat, 5 Oct 2024 04:23:14 +0300
Subject: [PATCH] Add ItemStack#effectiveName


diff --git a/src/main/java/org/bukkit/inventory/ItemFactory.java b/src/main/java/org/bukkit/inventory/ItemFactory.java
index 198a2a6b5dac78e1cc65e6929ec36f8d61f9f336..b5c167c93290cc244f9bb299028b0da341b87390 100644
--- a/src/main/java/org/bukkit/inventory/ItemFactory.java
+++ b/src/main/java/org/bukkit/inventory/ItemFactory.java
@@ -219,6 +219,19 @@ public interface ItemFactory {
      */
     @NotNull
     net.kyori.adventure.text.event.HoverEvent<net.kyori.adventure.text.event.HoverEvent.ShowItem> asHoverEvent(final @NotNull ItemStack item, final @NotNull java.util.function.UnaryOperator<net.kyori.adventure.text.event.HoverEvent.ShowItem> op);
+
+    /**
+     * Get the formatted display name of the {@link ItemStack}.
+     *
+     * @param itemStack the {@link ItemStack}
+     * @return display name of the {@link ItemStack}
+     * @deprecated Misleading name, represents the chat display name used by vanilla chat messages.
+     * @see ItemStack#chatName()
+     * @see ItemStack#effectiveName()
+     */
+    @Deprecated
+    @NotNull
+    net.kyori.adventure.text.Component displayName(@NotNull ItemStack itemStack);
     // Paper end - Adventure
 
     // Paper start - add getI18NDisplayName
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 80103b070ed42d3ce74866cbd7594f407cdcb58a..b14a812aca05298b0838c16941eb920bad661956 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -701,6 +701,39 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
         return org.bukkit.Bukkit.getServer().getItemFactory().asHoverEvent(this, op);
     }
 
+    /**
+     * Get the formatted display name of the {@link ItemStack}.
+     *
+     * @return display name of the {@link ItemStack}
+     * @deprecated Misleading name, represents the chat display name used by vanilla chat messages.
+     * @see #chatName()
+     * @see #effectiveName()
+     */
+    @Deprecated
+    public net.kyori.adventure.text.@NotNull Component displayName() {
+        return this.chatName();
+    }
+
+    /**
+     * Get the formatted chat display name of the {@link ItemStack}.
+     *
+     * @return chat display name of the {@link ItemStack}
+     */
+    public net.kyori.adventure.text.@NotNull Component chatName() {
+        return this.craftDelegate.chatName();
+    }
+
+    /**
+     * Gets the effective name of this item stack shown to player in inventory.
+     * It takes into account the display name (with italics) from the item meta,
+     * the potion effect, translatable name, rarity etc.
+     *
+     * @return the effective name of this item stack
+     */
+    public net.kyori.adventure.text.@NotNull Component effectiveName() {
+        return this.craftDelegate.effectiveName();
+    }
+
     /**
      * Minecraft updates are converting simple item stacks into more complex NBT oriented Item Stacks.
      *
