--- a/net/minecraft/world/inventory/ContainerBrewingStand.java
+++ b/net/minecraft/world/inventory/ContainerBrewingStand.java
@@ -15,6 +15,11 @@
 import net.minecraft.world.item.alchemy.PotionContents;
 import net.minecraft.world.item.alchemy.PotionRegistry;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventoryBrewer;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+// CraftBukkit end
+
 public class ContainerBrewingStand extends Container {
 
     private static final int BOTTLE_SLOT_START = 0;
@@ -31,12 +36,18 @@
     private final IContainerProperties brewingStandData;
     private final Slot ingredientSlot;
 
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private PlayerInventory player;
+    // CraftBukkit end
+
     public ContainerBrewingStand(int i, PlayerInventory playerinventory) {
         this(i, playerinventory, new InventorySubcontainer(5), new ContainerProperties(2));
     }
 
     public ContainerBrewingStand(int i, PlayerInventory playerinventory, IInventory iinventory, IContainerProperties icontainerproperties) {
         super(Containers.BREWING_STAND, i);
+        player = playerinventory; // CraftBukkit
         checkContainerSize(iinventory, 5);
         checkContainerDataCount(icontainerproperties, 2);
         this.brewingStand = iinventory;
@@ -66,6 +77,7 @@
 
     @Override
     public boolean stillValid(EntityHuman entityhuman) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.brewingStand.stillValid(entityhuman);
     }
 
@@ -196,4 +208,17 @@
             return itemstack.is(Items.BLAZE_POWDER);
         }
     }
+
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryBrewer inventory = new CraftInventoryBrewer(this.brewingStand);
+        bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
