--- a/net/minecraft/world/inventory/InventoryCrafting.java
+++ b/net/minecraft/world/inventory/InventoryCrafting.java
@@ -5,6 +5,10 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.CraftingInput;
 
+// CraftBukkit start
+import net.minecraft.world.item.crafting.RecipeHolder;
+// CraftBukkit end
+
 public interface InventoryCrafting extends IInventory, AutoRecipeOutput {
 
     int getWidth();
@@ -13,6 +17,15 @@
 
     List<ItemStack> getItems();
 
+    // CraftBukkit start
+    default RecipeHolder<?> getCurrentRecipe() {
+        return null;
+    }
+
+    default void setCurrentRecipe(RecipeHolder<?> recipe) {
+    }
+    // CraftBukkit end
+
     default CraftingInput asCraftInput() {
         return this.asPositionedCraftInput().input();
     }
