--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -68,6 +68,11 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.phys.Vec3;
+// CraftBukkit start;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
 
 public class Panda extends Animal {
 
@@ -129,6 +134,7 @@
     }
 
     public void sit(boolean sitting) {
+        if (!new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), sitting).callEvent()) return; // Paper - Add EntityToggleSitEvent
         this.setFlag(8, sitting);
     }
 
@@ -525,7 +531,9 @@
             Panda entitypanda = (Panda) iterator.next();
 
             if (!entitypanda.isBaby() && entitypanda.onGround() && !entitypanda.isInWater() && entitypanda.canPerformAction()) {
+                if (new com.destroystokyo.paper.event.entity.EntityJumpEvent(getBukkitLivingEntity()).callEvent()) { // Paper - Entity Jump API
                 entitypanda.jumpFromGround();
+                } else { this.setJumping(false); } // Paper - Entity Jump API; setJumping(false) stops a potential loop
             }
         }
 
@@ -533,7 +541,9 @@
 
         if (world1 instanceof ServerLevel worldserver) {
             if (worldserver.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+                this.forceDrops = true; // Paper - Add missing forceDrop toggles
                 this.dropFromGiftLootTable(worldserver, BuiltInLootTables.PANDA_SNEEZE, this::spawnAtLocation);
+                this.forceDrops = false; // Paper - Add missing forceDrop toggles
             }
         }
 
@@ -541,14 +551,14 @@
 
     @Override
     protected void pickUpItem(ServerLevel world, ItemEntity itemEntity) {
-        if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && Panda.canPickUpAndEat(itemEntity)) {
+        if (!CraftEventFactory.callEntityPickupItemEvent(this, itemEntity, 0, !(this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && Panda.canPickUpAndEat(itemEntity))).isCancelled()) { // CraftBukkit
             this.onItemPickup(itemEntity);
             ItemStack itemstack = itemEntity.getItem();
 
             this.setItemSlot(EquipmentSlot.MAINHAND, itemstack);
             this.setGuaranteedDrop(EquipmentSlot.MAINHAND);
             this.take(itemEntity, itemstack.getCount());
-            itemEntity.discard();
+            itemEntity.discard(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
         }
 
     }
@@ -657,7 +667,9 @@
                         ItemStack itemstack1 = this.getItemBySlot(EquipmentSlot.MAINHAND);
 
                         if (!itemstack1.isEmpty() && !player.hasInfiniteMaterials()) {
+                            this.forceDrops = true; // Paper - Add missing forceDrop toggles
                             this.spawnAtLocation(worldserver, itemstack1);
+                            this.forceDrops = false; // Paper - Add missing forceDrop toggles
                         }
 
                         this.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(itemstack.getItem(), 1));
@@ -772,7 +784,7 @@
         }
 
         public static Panda.Gene byName(String name) {
-            return (Panda.Gene) Panda.Gene.CODEC.byName(name, (Enum) Panda.Gene.NORMAL);
+            return (Panda.Gene) Panda.Gene.CODEC.byName(name, Panda.Gene.NORMAL); // CraftBukkit - decompile error
         }
 
         public static Panda.Gene getRandom(RandomSource random) {
@@ -876,10 +888,10 @@
         private final Panda panda;
 
         public PandaAvoidGoal(Panda panda, Class<T> fleeFromType, float distance, double slowSpeed, double fastSpeed) {
-            Predicate predicate = EntitySelector.NO_SPECTATORS;
+            // Predicate predicate = IEntitySelector.NO_SPECTATORS;
 
-            Objects.requireNonNull(predicate);
-            super(panda, fleeFromType, distance, slowSpeed, fastSpeed, predicate::test);
+            // Objects.requireNonNull(predicate);
+            super(panda, fleeFromType, distance, slowSpeed, fastSpeed, EntitySelector.NO_SPECTATORS::test);
             this.panda = panda;
         }
 
@@ -935,7 +947,9 @@
             ItemStack itemstack = Panda.this.getItemBySlot(EquipmentSlot.MAINHAND);
 
             if (!itemstack.isEmpty()) {
+                Panda.this.forceDrops = true; // Paper - Add missing forceDrop toggles
                 Panda.this.spawnAtLocation(getServerLevel(Panda.this.level()), itemstack);
+                Panda.this.forceDrops = false; // Paper - Add missing forceDrop toggles
                 Panda.this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
                 int i = Panda.this.isLazy() ? Panda.this.random.nextInt(50) + 10 : Panda.this.random.nextInt(150) + 10;
 
@@ -1116,7 +1130,7 @@
         @Override
         protected void alertOther(Mob mob, LivingEntity target) {
             if (mob instanceof Panda && mob.isAggressive()) {
-                mob.setTarget(target);
+                mob.setTarget(target, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit
             }
 
         }
