--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -215,7 +_,14 @@
                 }
 
                 mob.setPersistenceRequired();
-            });
+                // CraftBukkit start
+            }, null, null);
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callPigZapEvent(this, lightning, zombifiedPiglin).isCancelled()) {
+                return;
+            }
+            level.addFreshEntity(zombifiedPiglin, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
+            // CraftBukkit end
             if (zombifiedPiglin == null) {
                 super.thunderHit(level, lightning);
             }
