--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -155,7 +155,7 @@
         int i = this.getPuffState();
 
         if (target.hurtServer(world, this.damageSources().mobAttack(this), (float) (1 + i))) {
-            target.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this);
+            target.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             this.playSound(SoundEvents.PUFFER_FISH_STING, 1.0F, 1.0F);
         }
 
@@ -171,7 +171,7 @@
                     entityplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
                 }
 
-                player.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this);
+                player.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
         }
 
