From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/animal/ShoulderRidingEntity.java b/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
index d63ff7886984bbaa7107d76bf5714017ff3ab2ab..1489d5704cc1b9d5670ad728e87b6a660571a52a 100644
--- a/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
+++ b/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
@@ -19,7 +19,7 @@ public abstract class ShoulderRidingEntity extends TamableAnimal {
         compoundTag.putString("id", this.getEncodeId());
         this.saveWithoutId(compoundTag);
         if (player.setEntityOnShoulder(compoundTag)) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             return true;
         } else {
             return false;
