From 3aeeaf31b7a7f888f54e47826e8ec0ddbe65c4bf Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/world/entity/animal/WaterAnimal.java b/net/minecraft/world/entity/animal/WaterAnimal.java
index e31d4eda22ea4e9c85b8d3411d6c66ed3d359f1d..ed4f4c76a53c39c1a912fc0ff0d84f7fdb494b20 100644
--- a/net/minecraft/world/entity/animal/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/WaterAnimal.java
@@ -70,6 +70,10 @@ public abstract class WaterAnimal extends PathfinderMob {
     ) {
         int seaLevel = level.getSeaLevel();
         int i = seaLevel - 13;
+        // Paper start - Make water animal spawn height configurable
+        seaLevel = level.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.maximum.or(seaLevel);
+        i = level.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.minimum.or(i);
+        // Paper end - Make water animal spawn height configurable
         return pos.getY() >= i
             && pos.getY() <= seaLevel
             && level.getFluidState(pos.below()).is(FluidTags.WATER)
