--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -102,7 +_,7 @@
     }
 
     private static boolean isLocatorBarEnabledFor(ServerPlayer player) {
-        return player.level().getServer().getGameRules().getBoolean(GameRules.RULE_LOCATOR_BAR);
+        return player.level().getGameRules().getBoolean(GameRules.RULE_LOCATOR_BAR); // Paper - per world gamerules
     }
 
     private void createConnection(ServerPlayer player, WaypointTransmitter waypoint) {
