From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Wed, 2 Mar 2016 00:32:25 -0600
Subject: [PATCH] Always tick falling blocks


diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index af66bb6b3733eaf722838de8b9f45396ffddc2f3..c3b7b1bc978eb0d40587682795e86b82981aed82 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -90,6 +90,7 @@ public class ActivationRange
                 || entity instanceof AbstractHurtingProjectile
                 || entity instanceof LightningBolt
                 || entity instanceof PrimedTnt
+                || entity instanceof net.minecraft.world.entity.item.FallingBlockEntity // Paper - Always tick falling blocks
                 || entity instanceof EndCrystal
                 || entity instanceof FireworkRocketEntity
                 || entity instanceof ThrownTrident )
