From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 21 Jun 2015 15:07:20 -0400
Subject: [PATCH] Custom replacement for eaten items


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index f513e9abb929fbabe12e40bb56af708745ea1ad9..d038c65e8a3d4a0000465e9d3ac7c4915b1b1015 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3614,10 +3614,11 @@ public abstract class LivingEntity extends Entity {
                     this.triggerItemUseEffects(this.useItem, 16);
                     // CraftBukkit start - fire PlayerItemConsumeEvent
                     ItemStack itemstack;
+                    PlayerItemConsumeEvent event = null; // Paper
                     if (this instanceof ServerPlayer) {
                         org.bukkit.inventory.ItemStack craftItem = CraftItemStack.asBukkitCopy(this.useItem);
                         org.bukkit.inventory.EquipmentSlot hand = org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(enumhand);
-                        PlayerItemConsumeEvent event = new PlayerItemConsumeEvent((Player) this.getBukkitEntity(), craftItem, hand);
+                        event = new PlayerItemConsumeEvent((Player) this.getBukkitEntity(), craftItem, hand); // Paper
                         level.getCraftServer().getPluginManager().callEvent(event);
 
                         if (event.isCancelled()) {
@@ -3631,6 +3632,13 @@ public abstract class LivingEntity extends Entity {
                     } else {
                         itemstack = this.useItem.finishUsingItem(this.level, this);
                     }
+
+                // Paper start - save the default replacement item and change it if necessary
+                final ItemStack defaultReplacement = itemstack;
+                if (event != null && event.getReplacement() != null) {
+                    itemstack = CraftItemStack.asNMSCopy(event.getReplacement());
+                }
+                // Paper end
                     // CraftBukkit end
 
                     if (itemstack != this.useItem) {
@@ -3638,6 +3646,11 @@ public abstract class LivingEntity extends Entity {
                     }
 
                     this.stopUsingItem();
+                // Paper start - if the replacement is anything but the default, update the client inventory
+                if (this instanceof ServerPlayer && !com.google.common.base.Objects.equal(defaultReplacement, itemstack)) {
+                    ((ServerPlayer) this).getBukkitEntity().updateInventory();
+                }
+                // Paper end
                 }
 
             }
