From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 12 Nov 2016 23:25:22 -0600
Subject: [PATCH] Filter bad data from ArmorStand and SpawnEgg items


diff --git a/src/main/java/io/papermc/paper/PaperWorldConfig.java b/src/main/java/io/papermc/paper/PaperWorldConfig.java
index 62ad155a33ee47939d0834d440dda47a57d02f12..06c04316c70351228231e0dd269a397d3fe1b2e0 100644
--- a/src/main/java/io/papermc/paper/PaperWorldConfig.java
+++ b/src/main/java/io/papermc/paper/PaperWorldConfig.java
@@ -306,4 +306,12 @@ public class PaperWorldConfig {
     private void removeCorruptTEs() {
         removeCorruptTEs = getBoolean("remove-corrupt-tile-entities", false);
     }
+
+    public boolean filterNBTFromSpawnEgg = true;
+    private void fitlerNBTFromSpawnEgg() {
+        filterNBTFromSpawnEgg = getBoolean("filter-nbt-data-from-spawn-eggs-and-related", true);
+        if (!filterNBTFromSpawnEgg) {
+            Bukkit.getLogger().warning("Spawn Egg and Armor Stand NBT filtering disabled, this is a potential security risk");
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 6c262832ba5259ec92d336114c203c254a39924c..0d476aa50170ad3623462306769020518c55cffb 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -308,6 +308,18 @@ public class FallingBlockEntity extends Entity {
     @Override
     protected void readAdditionalSaveData(CompoundTag nbt) {
         this.blockState = NbtUtils.readBlockState(nbt.getCompound("BlockState"));
+        // Paper start - Block FallingBlocks with Command Blocks
+        final Block b = this.blockState.getBlock();
+        if (this.level.paperConfig.filterNBTFromSpawnEgg
+            && (b == Blocks.COMMAND_BLOCK
+            || b == Blocks.REPEATING_COMMAND_BLOCK
+            || b == Blocks.CHAIN_COMMAND_BLOCK
+            || b == Blocks.JIGSAW
+            || b == Blocks.STRUCTURE_BLOCK
+            || b instanceof net.minecraft.world.level.block.GameMasterBlock)) {
+            this.blockState = Blocks.STONE.defaultBlockState();
+        }
+        // Paper end
         this.time = nbt.getInt("Time");
         if (nbt.contains("HurtEntities", 99)) {
             this.hurtEntities = nbt.getBoolean("HurtEntities");
