From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 22 Nov 2016 00:40:42 -0500
Subject: [PATCH] Remove unnecessary itemmeta handling


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 1db50e72ad8e04e54bed9b462bd7276feb06ce4c..1e16482830556cf34f3b68fc2af8216663b075e4 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -166,7 +166,7 @@ public final class ItemStack implements DataComponentHolder {
 
                 // CraftBukkit start
                 ItemStack itemstack = new ItemStack(holder, i, datacomponentpatch);
-                if (!datacomponentpatch.isEmpty()) {
+                if (false && !datacomponentpatch.isEmpty()) { // Paper - This is no longer needed with raw NBT being handled in metadata
                     CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
                 }
                 return itemstack;
@@ -180,8 +180,8 @@ public final class ItemStack implements DataComponentHolder {
             } else {
                 registryfriendlybytebuf.writeVarInt(itemstack.getCount());
                 // Spigot start - filter
-                itemstack = itemstack.copy();
-                CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
+                // itemstack = itemstack.copy();
+                // CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack)); // Paper - This is no longer with raw NBT being handled in metadata
                 // Spigot end
                 ITEM_STREAM_CODEC.encode(registryfriendlybytebuf, itemstack.getItemHolder()); // CraftBukkit - decompile error
                 // Paper start - adventure; conditionally render translatable components
