From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GreenMeanie <GreenMeanieMC@gmail.com>
Date: Sat, 20 Oct 2018 22:34:02 -0400
Subject: [PATCH] Reset players airTicks on respawn


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 72d17bfbe0830721caa7d1fd0aa50916e75b1ed7..085bdf17a0e844ff813bf1107b686b33943659ba 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2210,6 +2210,7 @@ public class ServerPlayer extends Player {
         }
 
         this.setHealth(this.getMaxHealth());
+        this.setAirSupply(this.getMaxAirSupply()); // Paper
         this.remainingFireTicks = 0;
         this.fallDistance = 0;
         this.foodData = new FoodData(this);
