From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 18 Nov 2018 19:49:56 +0000
Subject: [PATCH] Make the default permission message configurable

TODO: Change the message in PaperCommand

diff --git a/src/main/java/io/papermc/paper/PaperConfig.java b/src/main/java/io/papermc/paper/PaperConfig.java
index afa9f61ab258419e9c8cb3e6d4860ab4d9eac8e6..864d1b22c6b9b26ea2d828104422d49751b8e31f 100644
--- a/src/main/java/io/papermc/paper/PaperConfig.java
+++ b/src/main/java/io/papermc/paper/PaperConfig.java
@@ -20,6 +20,7 @@ import java.util.regex.Pattern;
 import com.google.common.collect.Lists;
 import net.minecraft.server.MinecraftServer;
 import org.bukkit.Bukkit;
+import org.bukkit.ChatColor;
 import org.bukkit.command.Command;
 import org.bukkit.configuration.ConfigurationSection;
 import org.bukkit.configuration.InvalidConfigurationException;
@@ -288,6 +289,11 @@ public class PaperConfig {
         connectionThrottleKickMessage = getString("messages.kick.connection-throttle", connectionThrottleKickMessage);
     }
 
+    public static String noPermissionMessage = "&cI'm sorry, but you do not have permission to perform this command. Please contact the server administrators if you believe that this is in error.";
+    private static void noPermissionMessage() {
+        noPermissionMessage = ChatColor.translateAlternateColorCodes('&', getString("messages.no-permission", noPermissionMessage));
+    }
+
     private static void savePlayerData() {
         Object val = config.get("settings.save-player-data");
         if (val instanceof Boolean) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e199ed8963d19d9bd55578d1633cdfce0203c155..f86c1d946d21a566cba6ae02d8d1d9611d01dced 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2340,6 +2340,11 @@ public final class CraftServer implements Server {
         return io.papermc.paper.PaperConfig.suggestPlayersWhenNullTabCompletions;
     }
 
+    @Override
+    public String getPermissionMessage() {
+        return io.papermc.paper.PaperConfig.noPermissionMessage;
+    }
+
     @Override
     public io.papermc.paper.profile.PlayerProfile createProfile(@Nonnull UUID uuid) {
         return createProfile(uuid, null);
