From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 20 Apr 2019 19:47:34 -0500
Subject: [PATCH] Expose the internal current tick


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index cdc209d56b80fa6dd94b2e1d45918baf96aad153..d4ea341e555ae55e8521aaaa4031ef61007b46e0 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2656,5 +2656,10 @@ public final class CraftServer implements Server {
         profile.getProperties().putAll(((CraftPlayer)player).getHandle().getGameProfile().getProperties());
         return new com.destroystokyo.paper.profile.CraftPlayerProfile(profile);
     }
+
+    @Override
+    public int getCurrentTick() {
+        return net.minecraft.server.MinecraftServer.currentTick;
+    }
     // Paper end
 }
