From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 24 Dec 2019 00:35:42 +0000
Subject: [PATCH] PlayerDeathEvent#shouldDropExperience


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 02e81ba08b84f845b57a5148825ceb63313213a6..5dea577aeb4fa627e4c207b8a84933c1f2b63516 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -895,7 +895,7 @@ public class ServerPlayer extends Player {
             this.tellNeutralMobsThatIDied();
         }
         // SPIGOT-5478 must be called manually now
-        this.dropExperience();
+        if (event.shouldDropExperience()) this.dropExperience(); // Paper - tie to event
         // we clean the player's inventory after the EntityDeathEvent is called so plugins can get the exact state of the inventory.
         if (!event.getKeepInventory()) {
             // Paper start - replace logic
