From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Wed, 9 Oct 2019 21:51:43 -0500
Subject: [PATCH] Fix stuck in sneak when changing worlds (MC-10657)


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d37eeeac45fc524abb51b3765400ad8b4e0df7e6..cb0b97cdcb64d65013e45aa13d1b196d8989a0b2 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1096,6 +1096,8 @@ public class ServerPlayer extends Player {
                 this.lastSentHealth = -1.0F;
                 this.lastSentFood = -1;
 
+                setShiftKeyDown(false); // Paper - fix MC-10657
+
                 // CraftBukkit start
                 PlayerChangedWorldEvent changeEvent = new PlayerChangedWorldEvent(this.getBukkitEntity(), worldserver1.getWorld());
                 this.level.getCraftServer().getPluginManager().callEvent(changeEvent);
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 18fc4e0e79978d1a8ee8b45d5814caaa10948f9a..8135caaf93ac4e1dfb47001cb735b00726868e9f 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -829,6 +829,8 @@ public abstract class PlayerList {
             entityplayer.connection.send(new ClientboundUpdateMobEffectPacket(entityplayer.getId(), mobEffect));
         }
 
+        entityplayer.setShiftKeyDown(false); // Paper - fix MC-10657
+
         // Fire advancement trigger
         entityplayer.triggerDimensionChangeTriggers(((CraftWorld) fromWorld).getHandle());
 
