From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <nassim@njahnke.dev>
Date: Wed, 16 Mar 2022 13:58:16 +0100
Subject: [PATCH] Remove garbage Java version check


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index c79afabed432ca9094967ae0e48b04133dc4c51b..77911e26af9ec468c8a0c1fe8161b79c67b2303c 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -205,10 +205,6 @@ public class Main {
                 System.err.println("Unsupported Java detected (" + javaVersion + "). This version of Minecraft requires at least Java 17. Check your Java version with the command 'java -version'.");
                 return;
             }
-            if (javaVersion > 64.0) {
-                System.err.println("Unsupported Java detected (" + javaVersion + "). Only up to Java 20 is supported.");
-                return;
-            }
             String javaVersionName = System.getProperty("java.version");
             // J2SE SDK/JRE Version String Naming Convention
             boolean isPreRelease = javaVersionName.contains("-");
