From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 31 Mar 2020 03:50:42 -0400
Subject: [PATCH] Remote Connections shouldn't hold up shutdown

Bugs in the connection logic appears to leave stale connections even, preventing shutdown

diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 4357d5c7c66d9490a34f2d591ddc3d58d36cadd1..ab26a26ac553071ecd943dc8a3b41debf8b797ef 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -434,11 +434,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
 
         if (this.rconThread != null) {
-            this.rconThread.stop();
+            //this.remoteControlListener.b(); // Paper - don't wait for remote connections
         }
 
         if (this.queryThreadGs4 != null) {
-            this.queryThreadGs4.stop();
+            //this.remoteStatusListener.b(); // Paper - don't wait for remote connections
         }
 
         System.exit(0); // CraftBukkit
