From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kickash32 <kickash32@gmail.com>
Date: Fri, 15 May 2020 01:10:03 -0400
Subject: [PATCH] Ensure safe gateway teleport

== AT ==
public net.minecraft.world.entity.Entity teleportPassengers()V

diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 85914124014b4e6f0a561cf560918af68682b6f5..4bd3279f750e5d40d7aee935a4c285ef8764c0f8 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -219,6 +219,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
 
                     entity1.setPortalCooldown();
                     ((ServerPlayer) entity1).connection.teleport(teleEvent.getTo());
+                    entity1.teleportPassengers(); // Paper - teleport passengers as well, preventing invisible passengers post teleport.
                     TheEndGatewayBlockEntity.triggerCooldown(world, pos, state, blockEntity); // CraftBukkit - call at end of method
                     return;
 
