From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 26 Jul 2020 12:11:39 +0100
Subject: [PATCH] Add missing strikeLighting call to
 World#spigot()#strikeLightningEffect


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f22457cf9dd5ceb6630c1f96a2050c350c4fbbc9..cec02e3b47ccc99ba8343dafa7a760ebca64ffc1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2133,6 +2133,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
             lightning.moveTo( loc.getX(), loc.getY(), loc.getZ() );
             lightning.visualOnly = true;
             lightning.isSilent = isSilent;
+            world.strikeLightning( lightning );
             return (LightningStrike) lightning.getBukkitEntity();
         }
     };
