From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 28 Jun 2020 19:36:55 -0400
Subject: [PATCH] Don't allow null UUID's for chat


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundChatPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundChatPacket.java
index 26a229f7aa3f4425ed572e2d50730b4e978bf33e..a9fdfa7ec2a022e8adaa62721fb56748884686f5 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundChatPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundChatPacket.java
@@ -18,7 +18,7 @@ public class ClientboundChatPacket implements Packet<ClientGamePacketListener> {
     public ClientboundChatPacket(Component message, ChatType location, UUID sender) {
         this.message = message;
         this.type = location;
-        this.sender = sender;
+        this.sender = sender != null ? sender : net.minecraft.Util.NIL_UUID;
     }
 
     public ClientboundChatPacket(FriendlyByteBuf buf) {
