From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 26 Jul 2020 12:11:39 +0100
Subject: [PATCH] Add missing strikeLighting call to
 World#spigot()#strikeLightningEffect


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 3bc98586792960f50ca25929f1d91fc60a8577ce..5bf060d5b4a22d0fea0c70513d9b0914614e5bbb 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1994,6 +1994,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
             lightning.moveTo( loc.getX(), loc.getY(), loc.getZ() );
             lightning.visualOnly = true;
             lightning.isSilent = isSilent;
+            world.strikeLightning( lightning );
             return (LightningStrike) lightning.getBukkitEntity();
         }
     };
