From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 26 Jul 2020 12:11:39 +0100
Subject: [PATCH] Add missing strikeLighting call to
 World#spigot()#strikeLightningEffect


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index fe200e3694bd80da4e8715ee72247a5e04a47e41..9e7a85441890d71a3ac6037a5444d62d112acd28 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2054,6 +2054,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
             lightning.moveTo( loc.getX(), loc.getY(), loc.getZ() );
             lightning.visualOnly = true;
             lightning.isSilent = isSilent;
+            world.strikeLightning( lightning );
             return (LightningStrike) lightning.getBukkitEntity();
         }
     };
