From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anrza <andrzejrzeczycki314@gmail.com>
Date: Wed, 15 Jul 2020 12:08:49 +0200
Subject: [PATCH] Add LivingEntity#clearActiveItem


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 8ed8bd2c539af20aa97bff1c0f719c4c2f793fc5..66a4d3d3626e4382e07617d0af464e6fbd8a9427 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -887,6 +887,13 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return getHandle().getUseItem().asBukkitMirror();
     }
 
+    // Paper start
+    @Override
+    public void clearActiveItem() {
+        getHandle().stopUsingItem();
+    }
+    // Paper end
+
     @Override
     public int getItemUseRemainingTime() {
         return getHandle().getUseItemRemainingTicks();
