From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Sun, 30 Aug 2020 15:30:29 +0800
Subject: [PATCH] Fix CraftTeam null check


diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index 18d5a26c34c848241c306241b3ad9825b5a0b9a9..60b1cffdccde4715832546d6edbf206fbab4e82f 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -261,7 +261,7 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
 
     @Override
     public boolean hasEntry(String entry) throws IllegalArgumentException, IllegalStateException {
-        Validate.notNull("Entry cannot be null");
+        Validate.notNull(entry, "Entry cannot be null"); // Paper
 
         CraftScoreboard scoreboard = this.checkState();
 
