From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 3 Jan 2021 21:04:03 -0800
Subject: [PATCH] Configurable max leash distance


diff --git a/src/main/java/net/minecraft/world/entity/PathfinderMob.java b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
index b5964614823059261c28eb2e998ad54249f07e7b..08245016186c47077294f758409f8bf14398199f 100644
--- a/src/main/java/net/minecraft/world/entity/PathfinderMob.java
+++ b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
@@ -56,7 +56,7 @@ public abstract class PathfinderMob extends Mob {
             float f = this.distanceTo(entity);
 
             if (this instanceof TamableAnimal && ((TamableAnimal) this).isInSittingPose()) {
-                if (f > 10.0F) {
+                if (f > entity.level().paperConfig().misc.maxLeashDistance) { // Paper - Configurable max leash distance
                     this.level().getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                     this.dropLeash(true, true);
                 }
@@ -65,7 +65,7 @@ public abstract class PathfinderMob extends Mob {
             }
 
             this.onLeashDistance(f);
-            if (f > 10.0F) {
+            if (f > entity.level().paperConfig().misc.maxLeashDistance) { // Paper - Configurable max leash distance
                 this.level().getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                 this.dropLeash(true, true);
                 this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
