From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Mon, 24 Aug 2020 08:39:06 -0700
Subject: [PATCH] Fix nerfed slime when splitting


diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index 8d03466e2f546aae705c1d4dd2d562d3db2f2671..85edba5de3ce6c1fce8872855544863de84e7759 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -239,6 +239,7 @@ public class Slime extends Mob implements Enemy {
                     entityslime.setPersistenceRequired();
                 }
 
+                entityslime.aware = this.aware; // Paper
                 entityslime.setCustomName(ichatbasecomponent);
                 entityslime.setNoAi(flag);
                 entityslime.setInvulnerable(this.isInvulnerable());
