From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Wed, 6 Jan 2021 23:38:43 +0100
Subject: [PATCH] Empty commands shall not be dispatched


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index a8f3a84615a640d2949d268ef1ac81a051f5a38b..2dc41adc4e915f98451d79f0826b75560b716d21 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -236,6 +236,7 @@ public class Commands {
         command = event.getCommand();
 
         String[] args = command.split(" ");
+        if (args.length == 0) return 0; // Paper - empty commands shall not be dispatched
 
         String cmd = args[0];
         if (cmd.startsWith("minecraft:")) cmd = cmd.substring("minecraft:".length());
