From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Mon, 24 Aug 2020 08:39:06 -0700
Subject: [PATCH] Fix nerfed slime when splitting


diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index 6a199e148929d8df9b1f39733dd821c503380973..31ca6e693acb7f2db9ea75a38ec6286ea1ed2bd4 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -241,6 +241,7 @@ public class Slime extends Mob implements Enemy {
                     entityslime.setPersistenceRequired();
                 }
 
+                entityslime.aware = this.aware; // Paper
                 entityslime.setCustomName(ichatbasecomponent);
                 entityslime.setNoAi(flag);
                 entityslime.setInvulnerable(this.isInvulnerable());
