From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 3 Jan 2021 21:04:03 -0800
Subject: [PATCH] Configurable max leash distance


diff --git a/src/main/java/net/minecraft/world/entity/PathfinderMob.java b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
index 3db309e709cd72e3aae184ff2f8b1a7b98f2c7a8..15f56707a1f06f5c33f231a15a6c5f4b4a85cb4e 100644
--- a/src/main/java/net/minecraft/world/entity/PathfinderMob.java
+++ b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
@@ -48,7 +48,7 @@ public abstract class PathfinderMob extends Mob {
             float f = this.distanceTo(entity);
 
             if (this instanceof TamableAnimal && ((TamableAnimal) this).isInSittingPose()) {
-                if (f > 10.0F) {
+                if (f > entity.level.paperConfig().misc.maxLeashDistance) { // Paper
                     this.level.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                     this.dropLeash(true, true);
                 }
@@ -57,7 +57,7 @@ public abstract class PathfinderMob extends Mob {
             }
 
             this.onLeashDistance(f);
-            if (f > 10.0F) {
+            if (f > entity.level.paperConfig().misc.maxLeashDistance) { // Paper
                 this.level.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE)); // CraftBukkit
                 this.dropLeash(true, true);
                 this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
