From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CDFN <codefun@protonmail.com>
Date: Tue, 7 Jul 2020 17:53:23 +0200
Subject: [PATCH] Return chat component with empty text instead of throwing
 exception


diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index 823ead0fd1942db0219968cd383439b324b16f6e..49b063655dfc09e30d446dbf07503fdda04a7e30 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -88,7 +88,12 @@ public abstract class AbstractContainerMenu {
     }
     private Component title;
     public final Component getTitle() {
-        Preconditions.checkState(this.title != null, "Title not set");
+        // Paper start - return chat component with empty text instead of throwing error
+        // Preconditions.checkState(this.title != null, "Title not set");
+        if(this.title == null){
+            return Component.literal("");
+        }
+        // Paper end
         return this.title;
     }
     public final void setTitle(Component title) {
