From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <admin@yive.me>
Date: Sun, 24 Jan 2021 08:55:19 -0800
Subject: [PATCH] Reset shield blocking on dimension change


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 345091d6ba617c26ac7acfd88eb023f2ec8fae11..6b6e0c408d22994a29bc562da523e0dc0575639f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1181,6 +1181,11 @@ public class ServerPlayer extends Player {
                 this.level.getCraftServer().getPluginManager().callEvent(changeEvent);
                 // CraftBukkit end
             }
+            // Paper start
+            if (this.isBlocking()) {
+                this.stopUsingItem();
+            }
+            // Paper end
 
             return this;
         }
