From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 12 Mar 2021 17:09:42 -0800
Subject: [PATCH] Item Rarity API

== AT ==
public net.minecraft.world.item.Item rarity

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftMaterial.java b/src/main/java/org/bukkit/craftbukkit/CraftMaterial.java
index d90441aa2b111659668c04bfe556bf3584e94052..4439356d9eb889a5ff0ab877510666887c95da34 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftMaterial.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftMaterial.java
@@ -391,4 +391,11 @@ public class CraftMaterial<B extends BlockData> implements BlockType<B>, ItemTyp
         return this.item.getDescriptionId();
     }
     // Paper end - translation
+
+    // Paper start - item rarity
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemRarity() {
+        return io.papermc.paper.inventory.ItemRarity.values()[this.item.rarity.ordinal()];
+    }
+    // Paper end - item rarity
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockType.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockType.java
index ee9493d40753bba51adc1f9ae729401c05587827..fbea6e34a660bc7c05d0865dd663e611779c0285 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockType.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockType.java
@@ -278,4 +278,11 @@ public class CraftBlockType<B extends BlockData> implements BlockType<B> {
         return this.block.getDescriptionId();
     }
     // Paper end - translation
+
+    // Paper start - item rarity
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemRarity() {
+        throw new IllegalArgumentException("The Material " + this + " is not an item");
+    }
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java
index fdf33146ac74480cb7c037309a213d355579b986..5ef109b8f4f035ecbbf134fdc62ec7a4e01858dd 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java
@@ -283,4 +283,11 @@ public class CraftItemType implements ItemType {
         return this.item.getDescriptionId();
     }
     // Paper end - translation
+
+    // Paper start - item rarity
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemRarity() {
+        return io.papermc.paper.inventory.ItemRarity.values()[this.item.rarity.ordinal()];
+    }
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacyMaterial.java b/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacyMaterial.java
index 33bfda5b5257529c24fdc66e96637702c361af64..21200bf4af264360bd32db9353cbc98d4c7dcbb8 100644
--- a/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacyMaterial.java
+++ b/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacyMaterial.java
@@ -990,4 +990,11 @@ public class CraftLegacyMaterial implements Material {
         throw new IllegalArgumentException("Cannot get translation key of Legacy Material");
     }
     // Paper end - translation
+
+    // Paper start - item rarity
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemRarity() {
+        throw new IllegalArgumentException("Cannot get item rarity of Legacy Material");
+    }
+    // Paper end - item rarity
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 0b4d7d004a4156547fc2ffcbbc3f1228b32bfa8e..16cc1b9544d3aff176a9a5a8da1a34c0cabc5d69 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -496,6 +496,11 @@ public final class CraftMagicNumbers implements UnsafeValues {
     public String getMainLevelName() {
         return ((net.minecraft.server.dedicated.DedicatedServer) net.minecraft.server.MinecraftServer.getServer()).getProperties().levelName;
     }
+
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemStackRarity(org.bukkit.inventory.ItemStack itemStack) {
+        return io.papermc.paper.inventory.ItemRarity.values()[getItem(itemStack.getType()).getRarity(CraftItemStack.asNMSCopy(itemStack)).ordinal()];
+    }
     // Paper end
 
     /**
diff --git a/src/test/java/io/papermc/paper/inventory/ItemRarityTest.java b/src/test/java/io/papermc/paper/inventory/ItemRarityTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..38e6d42098f216b1d24f50386e7be98181122d8d
--- /dev/null
+++ b/src/test/java/io/papermc/paper/inventory/ItemRarityTest.java
@@ -0,0 +1,24 @@
+package io.papermc.paper.inventory;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.minecraft.world.item.Rarity;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class ItemRarityTest {
+
+    @Test
+    public void testConvertFromNmsToBukkit() {
+        for (Rarity nmsRarity : Rarity.values()) {
+            assertEquals("rarity names are mis-matched", ItemRarity.values()[nmsRarity.ordinal()].name(), nmsRarity.name());
+        }
+    }
+
+    @Test
+    public void testRarityFormatting() {
+        for (Rarity nmsRarity : Rarity.values()) {
+            assertEquals("rarity formatting is mis-matched", nmsRarity.color, PaperAdventure.asVanilla(ItemRarity.values()[nmsRarity.ordinal()].color));
+        }
+    }
+}
