From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Mon, 24 Aug 2020 08:39:06 -0700
Subject: [PATCH] Fix nerfed slime when splitting


diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index 82747a12d4977ed2d26d203f690ecc9bdbfadfc1..b9df0e2986afd2b238fabd26d83d2cff923a2401 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -241,6 +241,7 @@ public class Slime extends Mob implements Enemy {
                     entityslime.setPersistenceRequired();
                 }
 
+                entityslime.aware = this.aware; // Paper
                 entityslime.setCustomName(ichatbasecomponent);
                 entityslime.setNoAi(flag);
                 entityslime.setInvulnerable(this.isInvulnerable());
