From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 23 Mar 2021 06:43:30 +0000
Subject: [PATCH] copy TESign#isEditable from snapshots


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index a0950f5902c3719dc31205ec43dca9482278c744..dfd5d05830d3892e1d640149d1aa9d70d6f8a66b 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -115,6 +115,7 @@ public class CraftSign extends CraftBlockEntityState<SignBlockEntity> implements
             }
             // Paper end
         }
+        sign.isEditable = getSnapshot().isEditable; // Paper - copy manually
     }
 
     public static void openSign(Sign sign, org.bukkit.entity.HumanEntity player) { // Paper - change move open sign to HumanEntity
