From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 23 Mar 2021 06:43:30 +0000
Subject: [PATCH] copy TESign#isEditable from snapshots


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index b0a7f558cfe0f2ff859ab7b2db38ac303e9ae842..32e92c3ce1b50ec9eba53908070ea94763880479 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -113,6 +113,7 @@ public class CraftSign extends CraftBlockEntityState<SignBlockEntity> implements
             }
             // Paper end
         }
+        sign.isEditable = getSnapshot().isEditable; // Paper - copy manually
     }
 
     // Paper start
