From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Wed, 21 Apr 2021 15:58:19 -0700
Subject: [PATCH] Added PlayerDeepSleepEvent


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index f1a1a251256646d0d3a457f1fc8af20ae085f5ef..2c65305037583d55fa20783d0f264382467f8c3f 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -253,6 +253,11 @@ public abstract class Player extends LivingEntity {
 
         if (this.isSleeping()) {
             ++this.sleepCounter;
+            // Paper start
+            if (this.sleepCounter == 100) {
+                if (!new io.papermc.paper.event.player.PlayerDeepSleepEvent((org.bukkit.entity.Player) getBukkitEntity()).callEvent()) { this.sleepCounter = Integer.MIN_VALUE; }
+            }
+            // Paper end
             if (this.sleepCounter > 100) {
                 this.sleepCounter = 100;
             }
