From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 12 Mar 2021 17:09:42 -0800
Subject: [PATCH] Item Rarity API


diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index f70d7b1c9b10f59b4d9c8575b9cc35409fec11e4..6816c2a5487d96e65acacef438fa43a0e3808b99 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -470,6 +470,20 @@ public final class CraftMagicNumbers implements UnsafeValues {
     public int nextEntityId() {
         return net.minecraft.world.entity.Entity.nextEntityId();
     }
+
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemRarity(org.bukkit.Material material) {
+        Item item = getItem(material);
+        if (item == null) {
+            throw new IllegalArgumentException(material + " is not an item, and rarity does not apply to blocks");
+        }
+        return io.papermc.paper.inventory.ItemRarity.values()[item.rarity.ordinal()];
+    }
+
+    @Override
+    public io.papermc.paper.inventory.ItemRarity getItemStackRarity(org.bukkit.inventory.ItemStack itemStack) {
+        return io.papermc.paper.inventory.ItemRarity.values()[getItem(itemStack.getType()).getRarity(CraftItemStack.asNMSCopy(itemStack)).ordinal()];
+    }
     // Paper end
 
     /**
