From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <admin@yive.me>
Date: Wed, 26 May 2021 15:09:33 -0700
Subject: [PATCH] Limit item frame cursors on maps


diff --git a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
index 4acbcafc158cf11af51d9518ba5b83aaa75f52a1..67b88da702b780f79c0496cb17f1e6f1f8dd6c2b 100644
--- a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -295,8 +295,12 @@ public class MapItemSavedData extends SavedData {
 
             MapFrame worldmapframe1 = new MapFrame(blockposition, entityitemframe.getDirection().get2DDataValue() * 90, entityitemframe.getId());
 
+            // Paper start
+            if (this.decorations.size() < player.level.paperConfig().maps.itemFrameCursorLimit) {
             this.addDecoration(MapDecoration.Type.FRAME, player.level, "frame-" + entityitemframe.getId(), (double) blockposition.getX(), (double) blockposition.getZ(), (double) (entityitemframe.getDirection().get2DDataValue() * 90), (Component) null);
             this.frameMarkers.put(worldmapframe1.getId(), worldmapframe1);
+            }
+            // Paper end
         }
 
         CompoundTag nbttagcompound = stack.getTag();
