From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Thu, 22 Apr 2021 16:45:28 -0700
Subject: [PATCH] add consumeFuel to FurnaceBurnEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 5c1283c15c7e2b42b06b5f37c276cd0752e352e6..242d9b2ecb852d812d4887883cb2aabfd21e3715 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -343,7 +343,7 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
                 if (blockEntity.isLit() && furnaceBurnEvent.isBurning()) {
                     // CraftBukkit end
                     flag1 = true;
-                    if (!itemstack.isEmpty()) {
+                    if (!itemstack.isEmpty() && furnaceBurnEvent.willConsumeFuel()) { // Paper
                         Item item = itemstack.getItem();
 
                         itemstack.shrink(1);
