From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Thu, 14 Oct 2021 12:36:58 -0500
Subject: [PATCH] Left handed API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index c90f9b519c82327b787d2920b3bb079efb0c7e8a..c526304d31fd9be1b3c895187f7d9fc45edf4866 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -147,5 +147,15 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
     public int getMaxHeadPitch() {
         return getHandle().getMaxHeadXRot();
     }
+
+    @Override
+    public boolean isLeftHanded() {
+        return getHandle().isLeftHanded();
+    }
+
+    @Override
+    public void setLeftHanded(boolean leftHanded) {
+        getHandle().setLeftHanded(leftHanded);
+    }
     // Paper end
 }
