From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 12 Mar 2021 19:22:21 -0800
Subject: [PATCH] Adds PlayerArmSwingEvent


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 61569455eeb264f82096f563078967b2f4908f88..de02aa3c97b68f03c1f09c5b56eed3ac9ec96227 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2234,7 +2234,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         }
 
         // Arm swing animation
-        PlayerAnimationEvent event = new PlayerAnimationEvent(this.getCraftPlayer());
+        io.papermc.paper.event.player.PlayerArmSwingEvent event = new io.papermc.paper.event.player.PlayerArmSwingEvent(this.getCraftPlayer(), packet.getHand() == InteractionHand.MAIN_HAND ? org.bukkit.inventory.EquipmentSlot.HAND : org.bukkit.inventory.EquipmentSlot.OFF_HAND); // Paper
         this.cserver.getPluginManager().callEvent(event);
 
         if (event.isCancelled()) return;
