From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Sun, 8 Aug 2021 00:52:54 -0400
Subject: [PATCH] Optimize entity tracker passenger checks


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index ddc5b4849939a96b76611cfa1cd34c06c7acc0f8..5246d427973f34843046e59c198785c73fccec33 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -71,7 +71,7 @@ public class ServerEntity {
         this.trackedPlayers = trackedPlayers;
         // CraftBukkit end
         this.ap = Vec3.ZERO;
-        this.lastPassengers = Collections.emptyList();
+        this.lastPassengers = com.google.common.collect.ImmutableList.of(); // Paper - optimize passenger checks
         this.level = worldserver;
         this.broadcast = consumer;
         this.entity = entity;
