From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 13 Aug 2021 15:00:06 -0700
Subject: [PATCH] Clear bucket NBT after dispense


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 15fc100e468e68cbb0c43363c0eb25dc2ef8c6e0..3d2b5f040715a0e4fac0e6786bd11a4d715330ce 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -642,8 +642,7 @@ public interface DispenseItemBehavior {
                     Item item = Items.BUCKET;
                     stack.shrink(1);
                     if (stack.isEmpty()) {
-                        stack.setItem(Items.BUCKET);
-                        stack.setCount(1);
+                        stack = new ItemStack(item); // Paper - clear tag
                     } else if (((DispenserBlockEntity) pointer.getEntity()).addItem(new ItemStack(item)) < 0) {
                         this.defaultDispenseItemBehavior.dispense(pointer, new ItemStack(item));
                     }
