From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chase <chasewhip20@gmail.com>
Date: Wed, 2 Dec 2020 22:43:39 -0800
Subject: [PATCH] add per world spawn limits

Taken from #2982. Credit to Chasewhip8

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index ba8ae4d1c86293ca5c95d830d145023b4fe21d71..fc9b69d87364b578327dffb657746c08ab505be3 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -681,6 +681,19 @@ public class PaperWorldConfig {
         zombieVillagerInfectionChance = getDouble("zombie-villager-infection-chance", zombieVillagerInfectionChance);
     }
 
+    public int spawnLimitMonsters = -1;
+    public int spawnLimitAnimals = -1;
+    public int spawnLimitWaterAnimals = -1;
+    public int spawnLimitWaterAmbient = -1;
+    public int spawnLimitAmbient = -1;
+    private void perWorldSpawnLimits() {
+        spawnLimitMonsters = getInt("spawn-limits.monsters", spawnLimitMonsters);
+        spawnLimitAnimals = getInt("spawn-limits.animals", spawnLimitAnimals);
+        spawnLimitWaterAnimals = getInt("spawn-limits.water-animals", spawnLimitWaterAnimals);
+        spawnLimitWaterAmbient = getInt("spawn-limits.water-ambient", spawnLimitWaterAmbient);
+        spawnLimitAmbient = getInt("spawn-limits.ambient", spawnLimitAmbient);
+    }
+
     public int lightQueueSize = 20;
     private void lightQueueSize() {
         lightQueueSize = getInt("light-queue-size", lightQueueSize);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index a2a8080314966fe2e5d2692d6ee2f6ca65207161..00522af9b6c4938e9b0b169178243be53aab98d8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -209,6 +209,13 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         this.biomeProvider = biomeProvider;
 
         this.environment = env;
+        // Paper start - per world spawn limits
+        this.monsterSpawn = this.world.paperConfig.spawnLimitMonsters;
+        this.animalSpawn = this.world.paperConfig.spawnLimitAnimals;
+        this.waterAnimalSpawn = this.world.paperConfig.spawnLimitWaterAnimals;
+        this.waterAmbientSpawn = this.world.paperConfig.spawnLimitWaterAmbient;
+        this.ambientSpawn = this.world.paperConfig.spawnLimitAmbient;
+        // Paper end
     }
 
     @Override
