From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <nassim@njahnke.dev>
Date: Mon, 3 Oct 2022 20:48:19 +0200
Subject: [PATCH] Remove unnecessary onTrackingStart during navigation warning


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 03959bd6f40db9fb8f5da83377b8427ed0341f8a..ee1b0bb6c6681c32b32ebf50e15e261a123beb1a 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -2248,7 +2248,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             if (entity instanceof Mob entityinsentient) {
-                if (ServerLevel.this.isUpdatingNavigations) {
+                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
                     String s = "onTrackingStart called during navigation iteration";
 
                     Util.logAndPauseIfInIde("onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration"));
@@ -2328,7 +2328,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             if (entity instanceof Mob entityinsentient) {
-                if (ServerLevel.this.isUpdatingNavigations) {
+                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
                     String s = "onTrackingStart called during navigation iteration";
 
                     Util.logAndPauseIfInIde("onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration"));
