From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Doc <nachito94@msn.com>
Date: Fri, 15 Apr 2022 17:40:30 -0400
Subject: [PATCH] Use username instead of display name in
 PlayerList#getPlayerStats


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index d0037eaf1a1a256d3d7e2e309c476593a93f8273..4602f8a4f22eed417bf3c56f0147ed697e349f37 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1447,7 +1447,7 @@ public abstract class PlayerList {
     // CraftBukkit start
     public ServerStatsCounter getPlayerStats(ServerPlayer entityhuman) {
         ServerStatsCounter serverstatisticmanager = entityhuman.getStats();
-        return serverstatisticmanager == null ? this.getPlayerStats(entityhuman.getUUID(), entityhuman.getDisplayName().getString()) : serverstatisticmanager;
+        return serverstatisticmanager == null ? this.getPlayerStats(entityhuman.getUUID(), entityhuman.getGameProfile().getName()) : serverstatisticmanager; // Paper - use username and not display name
     }
 
     public ServerStatsCounter getPlayerStats(UUID uuid, String displayName) {
