From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Wed, 8 Jun 2022 11:04:47 -0400
Subject: [PATCH] Dont resent entity on art update


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
index dd02cec9a82794a6b001c3b64f031f78d5fbb812..bcac1359c667ef1ee46384f9c7a5adf4010d2b08 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
@@ -36,7 +36,7 @@ public class CraftPainting extends CraftHanging implements Painting {
             painting.setDirection(painting.getDirection());
             return false;
         }
-        this.update();
+        //this.update(); Paper - Don't resent entity on art update
         return true;
     }
 
