From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Sat, 12 Feb 2022 12:40:50 -0700
Subject: [PATCH] Add missing Validate calls to CraftServer#getSpawnLimit

Copies appropriate checks from CraftWorld#getSpawnLimit

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1033e997653c844cafe4cdb7985191526f4a86ca..0de3faf1f29608e7a387b4e4d2332d3321934963 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2187,6 +2187,8 @@ public final class CraftServer implements Server {
     @Override
     public int getSpawnLimit(SpawnCategory spawnCategory) {
         // Paper start
+        Preconditions.checkArgument(spawnCategory != null, "SpawnCategory cannot be null");
+        Preconditions.checkArgument(CraftSpawnCategory.isValidForLimits(spawnCategory), "SpawnCategory." + spawnCategory + " does not have a spawn limit.");
         return this.getSpawnLimitUnsafe(spawnCategory);
     }
     public int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) {
