From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Fri, 10 Sep 2021 15:49:12 -0700
Subject: [PATCH] VanillaCommandWrapper didnt account for entity senders


diff --git a/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java b/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
index 8dca2ad7d25f740941187698d77819af8ebc2805..6df44aa60d2b41b95fe79ed4cf92a6d3369400ea 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
@@ -64,8 +64,10 @@ public final class VanillaCommandWrapper extends BukkitCommand {
     }
 
     public static CommandSourceStack getListener(CommandSender sender) {
-        if (sender instanceof Player) {
-            return ((CraftPlayer) sender).getHandle().createCommandSourceStack();
+        // Paper start - account for other entity command senders
+        if (sender instanceof org.bukkit.craftbukkit.entity.CraftEntity craftEntity) {
+            return craftEntity.getHandle().createCommandSourceStack();
+        // Paper end
         }
         if (sender instanceof BlockCommandSender) {
             return ((CraftBlockCommandSender) sender).getWrapper();
