From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Noah van der Aa <ndvdaa@gmail.com>
Date: Tue, 14 Sep 2021 16:24:45 +0200
Subject: [PATCH] Don't log debug logging being disabled


diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 8ef09a81543d3b4fc5e766d4f9c30af25b083942..dd502e7a0ed270b9b85d814e3a60e164bcee0cd3 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -406,7 +406,7 @@ public class SpigotConfig
             Bukkit.getLogger().info( "Debug logging is enabled" );
         } else
         {
-            Bukkit.getLogger().info( "Debug logging is disabled" );
+            // Bukkit.getLogger().info( "Debug logging is disabled" ); // Paper - Don't log if debug logging isn't enabled.
         }
     }
 
