From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <nassim@njahnke.dev>
Date: Thu, 16 Mar 2023 16:27:50 +0100
Subject: [PATCH] Fix text display error on spawn


diff --git a/src/main/java/net/minecraft/world/entity/Display.java b/src/main/java/net/minecraft/world/entity/Display.java
index d7d940e6b397e52cb7d78bc6f5fc26e8096f9228..e6cbf4506c75046a89fad778e138b448fb4a29a9 100644
--- a/src/main/java/net/minecraft/world/entity/Display.java
+++ b/src/main/java/net/minecraft/world/entity/Display.java
@@ -903,7 +903,7 @@ public abstract class Display extends Entity {
             b = loadFlag(b, nbt, "default_background", (byte)4);
             Optional<Display.TextDisplay.Align> optional = Display.TextDisplay.Align.CODEC
                 .decode(NbtOps.INSTANCE, nbt.get("alignment"))
-                .resultOrPartial(Util.prefix("Display entity", Display.LOGGER::error))
+                .result() // Paper - Hide text display error on spawn
                 .map(Pair::getFirst);
             if (optional.isPresent()) {
                 b = switch ((Display.TextDisplay.Align)optional.get()) {
