From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Warrior <50800980+Warriorrrr@users.noreply.github.com>
Date: Fri, 7 Apr 2023 20:11:17 +0200
Subject: [PATCH] Fix demo flag not enabling demo mode

https://github.com/PaperMC/Paper/issues/9046

diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 47cd7c66ac37efebf2f63c49d78dd8fe44a70ef8..fc6ce3485dc890f5105a37fe3e344a1460867556 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -325,7 +325,9 @@ public class Main {
 
                 /*
                 dedicatedserver1.setPort((Integer) optionset.valueOf(optionspec11));
-                dedicatedserver1.setDemo(optionset.has(optionspec2));
+                */
+                dedicatedserver1.setDemo(optionset.has("demo")); // Paper
+                /*
                 dedicatedserver1.setId((String) optionset.valueOf(optionspec12));
                 */
                 boolean flag2 = !optionset.has("nogui") && !optionset.nonOptionArguments().contains("nogui");
