From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <jahnke.nassim@gmail.com>
Date: Thu, 16 Sep 2021 19:32:23 +0200
Subject: [PATCH] Fix GlobalPalette serialized size


diff --git a/src/main/java/net/minecraft/world/level/chunk/GlobalPalette.java b/src/main/java/net/minecraft/world/level/chunk/GlobalPalette.java
index ebca8dd8828767a71e06fcd1cf54ebb993c3feb0..72a8fb1007b298834c4d53ef78bb23b9388e7ee1 100644
--- a/src/main/java/net/minecraft/world/level/chunk/GlobalPalette.java
+++ b/src/main/java/net/minecraft/world/level/chunk/GlobalPalette.java
@@ -41,7 +41,7 @@ public class GlobalPalette<T> implements Palette<T> {
 
     @Override
     public int getSerializedSize() {
-        return FriendlyByteBuf.getVarIntSize(0);
+        return 0; // Paper - fix serialized size
     }
 
     @Override
