From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sat, 28 Aug 2021 15:16:21 -0700
Subject: [PATCH] fixup! Paper config files


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 5fd8f682d0a76bf804137018e1d3d4b44a31d9f7..f1e0335ce7170fc36aa9f3438b720028c472a2a6 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -71,8 +71,8 @@ public class PaperConfig {
         commands.put("paper", new PaperCommand("paper"));
         commands.put("mspt", new MSPTCommand("mspt"));
 
-        version = getInt("config-version", 21);
-        set("config-version", 21);
+        version = getInt("config-version", 22);
+        set("config-version", 22);
         readConfig(PaperConfig.class, null);
     }
 
