From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: darbyjack <admin@glaremasters.me>
Date: Thu, 9 Dec 2021 19:17:02 -0600
Subject: [PATCH] Update Log4j


diff --git a/build.gradle.kts b/build.gradle.kts
index 4715d1386afd939628c392a7d8062e5887d7a507..0282e3b75470e1a68ea1fc228082483514ba432e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,10 +20,11 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - implementation
-    annotationProcessor("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - Needed to generate meta for our Log4j plugins
+    implementation("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - implementation
+    annotationProcessor("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - Needed to generate meta for our Log4j plugins
     // Paper end
     implementation("org.apache.logging.log4j:log4j-iostreams:2.17.1") // Paper
+    implementation("org.apache.logging.log4j:log4j-slf4j18-impl:2.17.1") // Paper
     implementation("org.ow2.asm:asm:9.2")
     implementation("org.ow2.asm:asm-commons:9.2") // Paper - ASM event executor generation
     runtimeOnly("org.xerial:sqlite-jdbc:3.36.0.3")
