From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Tue, 28 Dec 2021 12:19:30 -0500
Subject: [PATCH] PlayerEatItemEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 6ba1916421ab9cfc042dd4fa0c1df90c8c9cc55b..a1f0aa569cdb38090238a168942715486be76639 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1237,6 +1237,17 @@ public class CraftEventFactory {
 
     // Paper start - Expose Saturation
     public static void handleFoodChange(net.minecraft.world.entity.player.Player player, int level, float saturationModifier, @Nullable ItemStack item) {
+        if (item != null) {
+            var eatItemEvent = new io.papermc.paper.event.packet.PlayerEatItemEvent((Player) player.getBukkitEntity(), level, saturationModifier, CraftItemStack.asBukkitCopy(item));
+
+            if (eatItemEvent.callEvent()) {
+                level = eatItemEvent.getFoodLevel();
+                saturationModifier = eatItemEvent.getSaturationModifier();
+            } else {
+                return; // Ignore eat
+            }
+        }
+
         var foodData = player.getFoodData();
 
         float oldSaturationLevel = foodData.saturationLevel;
