From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Seggan <segganew@gmail.com>
Date: Thu, 5 Aug 2021 13:10:27 -0400
Subject: [PATCH] Goat ram API


diff --git a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
index 32d0387b6c66462ca965add78a562dec3c4b95a9..c7e24da48aaff9fbd1a8272483231744326e3a8e 100644
--- a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
@@ -279,6 +279,17 @@ public class Goat extends Animal {
         return new Goat.GoatPathNavigation(this, world);
     }
 
+    // Paper start - Goat ram API
+    public void ram(LivingEntity entity) {
+        Brain<Goat> brain = this.getBrain();
+        brain.setMemory(MemoryModuleType.RAM_TARGET, entity.position());
+        brain.eraseMemory(MemoryModuleType.RAM_COOLDOWN_TICKS);
+        brain.eraseMemory(MemoryModuleType.BREED_TARGET);
+        brain.eraseMemory(MemoryModuleType.TEMPTING_PLAYER);
+        brain.setActiveActivityIfPossible(net.minecraft.world.entity.schedule.Activity.RAM);
+    }
+    // Paper end
+
     private static class GoatPathNavigation extends GroundPathNavigation {
 
         GoatPathNavigation(Goat goat, Level world) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
index ae74df5c9845ac125968a52897f4343b0f348217..436aa41563b8fab112d03c8cc516cf6ff37587bd 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
@@ -34,4 +34,11 @@ public class CraftGoat extends CraftAnimals implements Goat {
     public void setScreaming(boolean screaming) {
         this.getHandle().setScreamingGoat(screaming);
     }
+
+    // Paper start - Goat ram API
+    @Override
+    public void ram(@org.jetbrains.annotations.NotNull org.bukkit.entity.LivingEntity entity) {
+        this.getHandle().ram(((CraftLivingEntity) entity).getHandle());
+    }
+    // Paper end
 }
