From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: thamid-gamer <60953955+thamid-gamer@users.noreply.github.com>
Date: Sun, 2 Jan 2022 12:50:03 -0500
Subject: [PATCH] Correct Remaining Items for Non-Recipe Crafting


diff --git a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
index cc23533f77e3fa54a1075f637a2044bcb7729536..1a4199fbb7a46b8662e12a494129292c73e251a4 100644
--- a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
@@ -138,7 +138,12 @@ public class RecipeManager extends SimpleJsonResourceReloadListener {
             NonNullList<ItemStack> nonnulllist = NonNullList.withSize(inventory.getContainerSize(), ItemStack.EMPTY);
 
             for (int i = 0; i < nonnulllist.size(); ++i) {
-                nonnulllist.set(i, inventory.getItem(i));
+                // Paper start - use remaining items for non-recipe crafting (copied from default Recipe#getRemainingItems(Container))
+                net.minecraft.world.item.Item item = inventory.getItem(i).getItem();
+                if (item.hasCraftingRemainingItem()) {
+                    nonnulllist.set(i, new ItemStack(item.getCraftingRemainingItem()));
+                }
+                // Paper end
             }
 
             return nonnulllist;
