From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: darbyjack <admin@glaremasters.me>
Date: Thu, 9 Dec 2021 21:05:20 -0600
Subject: [PATCH] Update Log4J


diff --git a/build.gradle.kts b/build.gradle.kts
index e067c69fea1dec79ef8e97cad1c74020b6a3a235..612c4ed60406a69832609a9bdb3e45a7f3de6b74 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -33,11 +33,12 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - implementation
-    annotationProcessor("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - Needed to generate meta for out hex color converter plugin
+    implementation("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - implementation
+    annotationProcessor("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - Needed to generate meta for out hex color converter plugin
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.14.1") // Paper
-    implementation("org.apache.logging.log4j:log4j-api:2.14.1") // Paper
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.17.1") // Paper
+    implementation("org.apache.logging.log4j:log4j-slf4j18-impl:2.17.1") // Paper
+    implementation("org.apache.logging.log4j:log4j-api:2.17.1") // Paper
     implementation("org.ow2.asm:asm:9.2")
     implementation("org.ow2.asm:asm-commons:9.2") // Paper - ASM event executor generation
     implementation("com.googlecode.json-simple:json-simple:1.1.1") {
