From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NeonOverflow <nneonoverflow@protonmail.com>
Date: Fri, 14 Jan 2022 18:17:40 -0500
Subject: [PATCH] Only overworld spawn chunks get a ticket


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c4e8e6af67b57406012612b617a7dcaa6e391d09..2b6d273cc4a250ed83575e1f1dd0e54bee4a8fd1 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -870,7 +870,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         totalChunks *= totalChunks;
         worldloadlistener.setChunkRadius(radiusBlocks / 16);
 
-        worldserver.addTicketsForSpawn(radiusBlocks, blockposition);
+        if(worldserver.getTypeKey() == LevelStem.OVERWORLD) { // Paper - spawn chunks should get a ticket in the overworld ONLY.
+            worldserver.addTicketsForSpawn(radiusBlocks, blockposition);
+        }
         // Paper end
 
         // CraftBukkit start
