From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Fri, 14 Jan 2022 13:48:09 -0800
Subject: [PATCH] fixup! Optimise ArraySetSorted#removeIf


diff --git a/src/main/java/net/minecraft/server/level/DistanceManager.java b/src/main/java/net/minecraft/server/level/DistanceManager.java
index 846b2588e1506738e5073f1866805a02d90c557f..2befb8058a0acf3298c43094d08a1882eb13d892 100644
--- a/src/main/java/net/minecraft/server/level/DistanceManager.java
+++ b/src/main/java/net/minecraft/server/level/DistanceManager.java
@@ -126,13 +126,27 @@ public abstract class DistanceManager {
     protected void purgeStaleTickets() {
         ++this.ticketTickCounter;
         ObjectIterator objectiterator = this.tickets.long2ObjectEntrySet().fastIterator();
+        // Paper start - use optimised removeIf
+        long[] currChunk = new long[1];
+        long ticketCounter = DistanceManager.this.ticketTickCounter;
+        java.util.function.Predicate<Ticket<?>> removeIf = (ticket) -> {
+            final boolean ret = ticket.timedOut(ticketCounter);
+            if (ret) {
+                this.tickingTicketsTracker.removeTicket(currChunk[0], ticket);
+            }
+            return ret;
+        };
+        // Paper end - use optimised removeIf
 
         while (objectiterator.hasNext()) {
             Entry<SortedArraySet<Ticket<?>>> entry = (Entry) objectiterator.next();
-            Iterator<Ticket<?>> iterator = ((SortedArraySet) entry.getValue()).iterator();
-            boolean flag = false;
+            // Paper start - use optimised removeIf
+            Iterator<Ticket<?>> iterator = null;
+            currChunk[0] = entry.getLongKey();
+            boolean flag = entry.getValue().removeIf(removeIf);
 
-            while (iterator.hasNext()) {
+            while (false && iterator.hasNext()) {
+                // Paper end - use optimised removeIf
                 Ticket<?> ticket = (Ticket) iterator.next();
 
                 if (ticket.timedOut(this.ticketTickCounter)) {
