From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Sun, 16 Oct 2022 12:10:17 -0700
Subject: [PATCH] Add entity knockback API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 8c8431f40e9415b32e43acea4cd91a20d82d8d08..bdaa21becdc17823628542d612daf2e055a273c8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1016,5 +1016,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         }
         throw new IllegalArgumentException(entityCategory + " is an unrecognized entity category");
     }
+
+    @Override
+    public void knockback(double strength, double directionX, double directionZ) {
+        Preconditions.checkArgument(strength > 0, "Knockback strength must be > 0");
+        getHandle().knockback(strength, directionX, directionZ);
+    };
     // Paper end
 }
