From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Thu, 31 Mar 2022 05:11:37 -0700
Subject: [PATCH] Ensure entity passenger world matches ridden entity

Bad plugins doing this would cause some obvious problems...

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 473b7c0bd9c49192041e3a6e4c8a9d760344a44a..82ccf8eacd5cac82ae2f44d38fc05f606ff50f8e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2670,6 +2670,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     protected boolean addPassenger(Entity entity) { // CraftBukkit
+        // Paper start
+        if (entity.level != this.level) {
+            LOGGER.error("Entity passenger world must match, cannot add " + entity + " as passenger to " + this, new Throwable());
+            return false;
+        }
+        // Paper end
         if (entity == this) throw new IllegalArgumentException("Entities cannot become a passenger of themselves"); // Paper - issue 572
         if (entity.getVehicle() != this) {
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
