From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Florian Schmidt <florian@f012.dev>
Date: Fri, 28 Jul 2023 14:14:35 +0200
Subject: [PATCH] Use correct seed on api world load


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index fab25844aca60c9da96ab160a3aa445800789e7d..d17909ec98aff3222600862fe67fe866ee87e9de 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1360,7 +1360,7 @@ public final class CraftServer implements Server {
             net.minecraft.server.Main.forceUpgrade(worldSession, DataFixers.getDataFixer(), this.console.options.has("eraseCache"), () -> true, iregistrycustom_dimension, this.console.options.has("recreateRegionFiles"));
         }
 
-        long j = BiomeManager.obfuscateSeed(creator.seed());
+        long j = BiomeManager.obfuscateSeed(worlddata.worldGenOptions().seed()); // Paper - use world seed
         List<CustomSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(worlddata));
         LevelStem worlddimension = iregistry.get(actualDimension);
 
