From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GodOfPro <1387ilia@gmail.com>
Date: Tue, 11 Apr 2023 16:31:39 +0430
Subject: [PATCH] Add Mob Experience reward API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index ad293ad9d5dcd7836a4fa29b4b687346e3ae516c..31f62fbb54ec7b270fbd8afba7bd7c4bfafa09e3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -170,4 +170,11 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         this.getHandle().setAggressive(aggressive);
     }
     // Paper end
+
+    // Paper start
+    @Override
+    public int getPossibleExperienceReward() {
+        return getHandle().getExperienceReward();
+    }
+    // Paper end
 }
