From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sat, 18 Feb 2023 16:23:18 +0100
Subject: [PATCH] Update the flag when a captured block state is outdated


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index e853c9fdc8ebde6a653aacbb1db6679ce4d9d50b..8d887a07bd2ef11667ccb34872d629656859fb6c 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -923,6 +923,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
                 blockstate = CapturedBlockState.getTreeBlockState(this, pos, flags);
                 this.capturedBlockStates.put(pos.immutable(), blockstate);
             }
+            blockstate.setFlag(flags); // Paper - update the flag also
             blockstate.setData(state);
             return true;
         }
