From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lama06 <andreasprues36@gmail.com>
Date: Sat, 21 Jan 2023 13:53:23 +0100
Subject: [PATCH] Win Screen API

== AT ==
public net.minecraft.server.level.ServerPlayer seenCredits

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index b8b23105d37ac2d461d87d0c8e3c83c6fd09c960..b1d12c30dd23e092531c0d4d3af367262ccb5402 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1226,6 +1226,25 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(packet);
     }
 
+    // Paper start
+    @Override
+    public void showWinScreen() {
+        if (getHandle().connection == null) return;
+        var packet = new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, 1);
+        getHandle().connection.send(packet);
+    }
+
+    @Override
+    public boolean hasSeenWinScreen() {
+        return getHandle().seenCredits;
+    }
+
+    @Override
+    public void setHasSeenWinScreen(boolean hasSeenWinScreen) {
+        getHandle().seenCredits = hasSeenWinScreen;
+    }
+    // Paper end
+
     @Override
     public void setRotation(float yaw, float pitch) {
         // Paper start - Teleport API
