From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Minh Nguyen <minhh2792.work@gmail.com>
Date: Thu, 21 Apr 2022 02:46:47 +0000
Subject: [PATCH] Faster map update interval


diff --git a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
index 6a78374b6ab3198571df7e17023d371667ebb24b..77d332cf103dbe07be326d486a43d4d39a565bd1 100644
--- a/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/src/main/java/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -624,7 +624,7 @@ public class MapItemSavedData extends SavedData {
         @Nullable
         Packet<?> nextUpdatePacket(int mapId) {
             MapItemSavedData.MapPatch worldmap_b;
-            if (!this.dirtyData && this.tick % 5 != 0) { this.tick++; return null; } // Paper - this won't end up sending, so don't render it!
+            if (!this.dirtyData && this.tick % 1 != 0) { this.tick++; return null; } // Paper - this won't end up sending, so don't render it!, faster map update interval
             boolean vanillaMaps = shouldUseVanillaMap(); // Paper
             org.bukkit.craftbukkit.map.RenderData render = !vanillaMaps ? MapItemSavedData.this.mapView.render((org.bukkit.craftbukkit.entity.CraftPlayer) this.player.getBukkitEntity()) : MapItemSavedData.this.vanillaRender; // CraftBukkit // Paper
 
@@ -637,7 +637,7 @@ public class MapItemSavedData extends SavedData {
 
             Collection collection;
 
-            if ((true || this.dirtyDecorations) && this.tick++ % 5 == 0) { // CraftBukkit - custom maps don't update this yet
+            if ((true || this.dirtyDecorations) && this.tick++ % 1 == 0) { // CraftBukkit - custom maps don't update this yet // Paper - faster map update interval
                 this.dirtyDecorations = false;
                 // CraftBukkit start
                 java.util.Collection<MapDecoration> icons = new java.util.ArrayList<MapDecoration>();
