From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Togira <70365614+Togira123@users.noreply.github.com>
Date: Sun, 22 May 2022 13:28:47 +0200
Subject: [PATCH] Add BrainMemoryChangeEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/Brain.java b/src/main/java/net/minecraft/world/entity/ai/Brain.java
index 91d266bd1fc37c9e284a99bdbb465d0e3100305a..c9ac16107f219264e9422c41f385164704a76e55 100644
--- a/src/main/java/net/minecraft/world/entity/ai/Brain.java
+++ b/src/main/java/net/minecraft/world/entity/ai/Brain.java
@@ -163,7 +163,38 @@ public class Brain<E extends LivingEntity> {
             if (memory.isPresent() && this.isEmptyCollection(memory.get().getValue())) {
                 this.eraseMemory(type);
             } else {
-                this.memories.put(type, memory);
+                // Paper start - check whether memory is being added/changed/deleted
+                if (hasMemoryValue(type)) {
+                    // already has memory
+                    if (memory.isPresent()) {
+                        if (new io.papermc.paper.event.entity.BrainMemoryChangeEvent(
+                                org.bukkit.craftbukkit.entity.memory.CraftMemoryKey.toMemoryKey(type),
+                                io.papermc.paper.event.entity.BrainMemoryChangeEvent.CHANGE).callEvent()) {
+                            this.memories.put(type, memory);
+                        }
+                    } else {
+                        if (new io.papermc.paper.event.entity.BrainMemoryChangeEvent(
+                                org.bukkit.craftbukkit.entity.memory.CraftMemoryKey.toMemoryKey(type),
+                                io.papermc.paper.event.entity.BrainMemoryChangeEvent.DELETE).callEvent()) {
+                            this.memories.put(type, memory);
+                        }
+                    }
+                } else {
+                    if (memory.isPresent()) {
+                        if (new io.papermc.paper.event.entity.BrainMemoryChangeEvent(
+                                org.bukkit.craftbukkit.entity.memory.CraftMemoryKey.toMemoryKey(type),
+                                io.papermc.paper.event.entity.BrainMemoryChangeEvent.ADD).callEvent()) {
+                            this.memories.put(type, memory);
+                        }
+                    } else {
+                        if (new io.papermc.paper.event.entity.BrainMemoryChangeEvent(
+                                org.bukkit.craftbukkit.entity.memory.CraftMemoryKey.toMemoryKey(type),
+                                io.papermc.paper.event.entity.BrainMemoryChangeEvent.DELETE).callEvent()) {
+                            this.memories.put(type, memory);
+                        }
+                    }
+                }
+                // Paper end
             }
         }
 
