From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Sat, 28 May 2022 23:44:17 +1000
Subject: [PATCH] implement real multiplication operation


diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeInstance.java b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
index 9c58d1190377ca887aa703cb8a96a2b028b63af1..c57578d3b8f66ed1bf355939c6a9340b9ec7ebcc 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -147,7 +147,11 @@ public class AttributeInstance {
         for(AttributeModifier attributeModifier3 : this.getModifiersOrEmpty(AttributeModifier.Operation.MULTIPLY_TOTAL)) {
             e *= 1.0D + attributeModifier3.getAmount();
         }
-
+        // Paper start - implement proper multiplication operation
+        for(AttributeModifier attributeModifier4 : this.getModifiersOrEmpty(AttributeModifier.Operation.MULTIPLY_TOTAL_REAL)) {
+            e *= attributeModifier4.getAmount();
+        }
+        // Paper end
         return this.attribute.sanitizeValue(e);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeModifier.java b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeModifier.java
index 09b8fa2cee63e6bcb92ffc3694f67d0f90a3805d..ecc109d7463f903a72bc69bd3f462cde7641c197 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeModifier.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeModifier.java
@@ -98,9 +98,12 @@ public class AttributeModifier {
     public static enum Operation {
         ADDITION(0),
         MULTIPLY_BASE(1),
-        MULTIPLY_TOTAL(2);
+        // Paper start - implement proper multiplication operation
+        MULTIPLY_TOTAL(2),
+        MULTIPLY_TOTAL_REAL(3);
 
-        private static final AttributeModifier.Operation[] OPERATIONS = new AttributeModifier.Operation[]{ADDITION, MULTIPLY_BASE, MULTIPLY_TOTAL};
+        private static final AttributeModifier.Operation[] OPERATIONS = new AttributeModifier.Operation[]{ADDITION, MULTIPLY_BASE, MULTIPLY_TOTAL, MULTIPLY_TOTAL_REAL};
+        // Paper end
         private final int value;
 
         private Operation(int id) {
