From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kihz17 <mitchdawdy@gmail.com>
Date: Wed, 1 Jun 2022 15:53:16 -0400
Subject: [PATCH] Add PlayerTrackEntityEvent


diff --git a/src/main/java/com/destroystokyo/paper/events/PlayerTrackEntityEvent.java b/src/main/java/com/destroystokyo/paper/events/PlayerTrackEntityEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..080d437667d521a5a7e71276a6bf8e971093d92b
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/events/PlayerTrackEntityEvent.java
@@ -0,0 +1,43 @@
+package com.destroystokyo.paper.events;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerTrackEntityEvent extends PlayerEvent {
+    private final TrackType type;
+    private final Entity tracked;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public PlayerTrackEntityEvent(@NotNull Player who, @NotNull Entity tracked, TrackType type) {
+        super(who);
+        this.tracked = tracked;
+        this.type = type;
+    }
+
+    public Entity getTracked() {
+        return tracked;
+    }
+
+    public TrackType getTrackType() {
+        return type;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    public enum TrackType {
+        START,
+        STOP
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 68c3c07a0a858c0e5464297d4ed76038d47d3513..bac13a6c16ab98929379ee73d07ab971c9f2fcb5 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -2,6 +2,7 @@ package net.minecraft.server.level;
 
 import com.destroystokyo.paper.events.BroadcastEntityMovePacketEvent;
 import com.destroystokyo.paper.events.BroadcastEntityVelocityPacketEvent;
+import com.destroystokyo.paper.events.PlayerTrackEntityEvent;
 import com.google.common.collect.Lists;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
@@ -286,6 +287,9 @@ public class ServerEntity {
         this.entity.stopSeenByPlayer(player);
         player.removeEntityInLOS(this.entity.getUUID());
         player.connection.send(new ClientboundRemoveEntitiesPacket(new int[]{this.entity.getId()}));
+
+        PlayerTrackEntityEvent evt = new PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity(), PlayerTrackEntityEvent.TrackType.STOP);
+        player.level.getCraftServer().getPluginManager().callEvent(evt);
     }
 
     public void addPairing(ServerPlayer player) {
@@ -294,6 +298,9 @@ public class ServerEntity {
         Objects.requireNonNull(player.connection);
         this.sendPairingData(playerconnection::send, player); // CraftBukkit - add player
         this.entity.startSeenByPlayer(player);
+
+        PlayerTrackEntityEvent evt = new PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity(), PlayerTrackEntityEvent.TrackType.START);
+        player.level.getCraftServer().getPluginManager().callEvent(evt);
     }
 
     public void sendPairingData(Consumer<Packet<?>> consumer, ServerPlayer entityplayer) { // CraftBukkit - add player
