From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MelnCat <melncatuwu@gmail.com>
Date: Sun, 16 Oct 2022 12:10:17 -0700
Subject: [PATCH] Add entity knockback API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index a4d916fcb12da37f9b75e844346578c160abea96..3076831b3d26c6dd3bf0c0d0618151ce6ad0df82 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -988,5 +988,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         }
         throw new IllegalArgumentException(entityCategory + " is an unrecognized entity category");
     }
+
+    @Override
+    public void knockback(double strength, double directionX, double directionZ) {
+        Preconditions.checkArgument(strength > 0, "Knockback strength must be > 0");
+        getHandle().knockback(strength, directionX, directionZ);
+    };
     // Paper end
 }
