From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kihz17 <mitchdawdy@gmail.com>
Date: Sat, 4 Jun 2022 15:46:37 -0400
Subject: [PATCH] More ebug messages


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index e23bf31d1cd20de62e6b83d53e27bb33df0b849c..7d4e2d2241bb34b7898b71d64f6c9067801ee941 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -328,8 +328,10 @@ public abstract class PlayerList {
     }
 
     private void postChunkLoadJoin(ServerPlayer player, ServerLevel worldserver1, Connection networkmanager, ServerGamePacketListenerImpl playerconnection, CompoundTag nbttagcompound, String s1, String s) {
+        LOGGER.info("Post chunk load for player " + player.getName());
         pendingPlayers.remove(player.getUUID(), player);
         if (!networkmanager.isConnected()) {
+            LOGGER.info("Not connected");
             return;
         }
         player.didPlayerJoinEvent = true;
@@ -345,6 +347,7 @@ public abstract class PlayerList {
         chatmessage.withStyle(ChatFormatting.YELLOW);
         Component joinMessage = chatmessage; // Paper - Adventure
 
+        LOGGER.info("We passed so far");
         playerconnection.teleport(player.getX(), player.getY(), player.getZ(), player.getYRot(), player.getXRot());
         this.players.add(player);
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
@@ -363,6 +366,7 @@ public abstract class PlayerList {
         // Ensure that player inventory is populated with its viewer
         player.containerMenu.transferTo(player.containerMenu, bukkitPlayer);
 
+        LOGGER.info("B4 join event");
         PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, PaperAdventure.asAdventure(chatmessage)); // Paper - Adventure
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
