From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Mon, 27 Mar 2023 10:20:00 -0700
Subject: [PATCH] Add Structure check API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 7200c336de27ffd4d37231768ff5192956ede972..cb85180daaa2b40cba2474784548d650bc4d639b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -246,6 +246,15 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         };
     }
     // Paper end
+    // Paper start - structure check API
+    @Override
+    public boolean hasStructureAt(final io.papermc.paper.math.Position position, final Structure structure) {
+        return this.world.structureManager().getStructureWithPieceAt(
+            io.papermc.paper.util.MCUtil.toBlockPos(position),
+            CraftStructure.bukkitToMinecraft(structure)
+        ).isValid();
+    }
+    // Paper end
 
     private static final Random rand = new Random();
 
