From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sun, 18 Sep 2022 13:10:18 -0400
Subject: [PATCH] Add PlayerAttackEntityEvent


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index cd9cbfa5ef94994b3f7f2ecdde843620e7e6c071..bd74c6216a71bafedd195aa1700879cd3c33bc5e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1252,6 +1252,7 @@ public abstract class Player extends LivingEntity {
     public void attack(Entity target) {
         if (target.isAttackable()) {
             if (!target.skipAttackInteraction(this)) {
+                if (!new io.papermc.paper.event.player.PlayerAttackEntityEvent((org.bukkit.entity.Player) this.getBukkitEntity(), target.getBukkitEntity()).callEvent()) return; // Paper
                 float f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
                 float f1;
 
