From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Wed, 29 Jun 2022 20:50:08 -0400
Subject: [PATCH] EnderDragon EntityExplodeEvent Fixes

diff --git a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index f872b60697525e3015979f286bfb08d29bb001df..610f5180841c44d78f1510cf8618eb169bb79f78 100644
--- a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -530,6 +530,7 @@ public class EnderDragon extends Mob implements Enemy {
                 this.level.removeBlock(new BlockPos(block.getX(), block.getY(), block.getZ()), false);
             }
         } else {
+            flag1 = false; // Paper
             for (org.bukkit.block.Block block : event.blockList()) {
                 org.bukkit.Material blockId = block.getType();
                 if (blockId.isAir()) {
@@ -549,9 +550,9 @@ public class EnderDragon extends Mob implements Enemy {
                     });
                     craftBlock.getNMS().spawnAfterBreak((ServerLevel) level, blockposition, ItemStack.EMPTY, false);
                 }
-                nmsBlock.wasExploded(level, blockposition, explosionSource);
+                //nmsBlock.wasExploded(level, blockposition, explosionSource); Paper - It doesn't make sense to have explosion logic here. This isn't technically an explosion.
 
-                this.level.removeBlock(blockposition, false);
+                flag1 = this.level.removeBlock(blockposition, false) || flag1; // Paper
             }
         }
         // CraftBukkit end
