From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjar <code.nopjar@gmail.com>
Date: Sun, 10 Jul 2022 17:06:09 +0200
Subject: [PATCH] Add PlayerItemConsumedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 279051442ac6cf288a03a16a35ddbf66d1cd1e90..420e7b14efbab0fbaf6d22900180ae488f0a1510 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3879,6 +3879,7 @@ public abstract class LivingEntity extends Entity {
                 // Paper start
                 if (this instanceof ServerPlayer) {
                     ((ServerPlayer) this).getBukkitEntity().updateInventory();
+                    new io.papermc.paper.event.player.PlayerItemConsumedEvent((Player) this.getBukkitEntity(), CraftItemStack.asBukkitCopy(itemstack)).callEvent(); // Paper - call event after item was consumed
                 }
                 // Paper end
                 }
