From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kihz17 <mitchdawdy@gmail.com>
Date: Tue, 28 Jun 2022 16:18:04 -0400
Subject: [PATCH] Remove RNG from fishing.


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 76fed0502c2b01f7b0cdf6f03ea39570b2c4a4c7..1041ca559fc59a86955b8e04724eedc5ee06a26e 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -306,15 +306,6 @@ public class FishingHook extends Projectile {
     private void catchingFish(BlockPos pos) {
         ServerLevel worldserver = (ServerLevel) this.level;
         int i = 1;
-        BlockPos blockposition1 = pos.above();
-
-        if (this.random.nextFloat() < 0.25F && this.level.isRainingAt(blockposition1)) {
-            ++i;
-        }
-
-        if (this.random.nextFloat() < 0.5F && !this.level.canSeeSky(blockposition1)) {
-            --i;
-        }
 
         if (this.nibble > 0) {
             --this.nibble;
