From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luis <luisc99@icloud.com>
Date: Thu, 11 Jan 2024 19:58:23 +0100
Subject: [PATCH] Add api for spawn egg texture colors


diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 9449fe944c7a06a8fdeaf0a0dfc7280950fdb6db..87e5ce7b30efe8ec99191208636b1420e4d002a1 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -637,6 +637,15 @@ public final class CraftMagicNumbers implements UnsafeValues {
         return CraftRegistry.get(registry, namespacedKey, ApiVersion.CURRENT);
     }
 
+    // Paper start - spawn egg color visibility
+    @Override
+    public org.bukkit.Color getSpawnEggLayerColor(final EntityType entityType, final int layer) {
+        final net.minecraft.world.entity.EntityType<?> nmsType = org.bukkit.craftbukkit.entity.CraftEntityType.bukkitToMinecraft(entityType);
+        final net.minecraft.world.item.SpawnEggItem eggItem = net.minecraft.world.item.SpawnEggItem.byId(nmsType);
+        return eggItem == null ? null : org.bukkit.Color.fromRGB(eggItem.getColor(layer));
+    }
+    // Paper end - spawn egg color visibility
+
     /**
      * This helper class represents the different NBT Tags.
      * <p>
