From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Wed, 20 Jul 2022 14:16:25 +0200
Subject: [PATCH] Fix wrong position for bed spawn loc


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 6d2ba650f53de8a460857f1846401a20b50cc43c..1599535361efd93d88aa1afb3de415e606128993 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -315,7 +315,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     }
 
     @Override
-    public Location getBedSpawnLocation() {
+    public Location getBedSpawnLocation(boolean load) { // Paper
         CompoundTag data = this.getData();
         if (data == null) return null;
 
@@ -332,7 +332,15 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
             if (spawnWorld == null) {
                 return null;
             }
-            return new Location(spawnWorld, data.getInt("SpawnX"), data.getInt("SpawnY"), data.getInt("SpawnZ"), respawnAngle, 0);
+            if(!load) {
+                return new Location(spawnWorld, data.getInt("SpawnX"), data.getInt("SpawnY"), data.getInt("SpawnZ"), respawnAngle, 0);
+            }
+            net.minecraft.core.BlockPos bed = new net.minecraft.core.BlockPos(data.getInt("SpawnX"), data.getInt("SpawnY"), data.getInt("SpawnZ"));
+            java.util.Optional<net.minecraft.world.phys.Vec3> spawnLoc = net.minecraft.world.entity.player.Player.findRespawnPositionAndUseSpawnBlock(((CraftWorld) spawnWorld).getHandle(), bed, respawnAngle, true, true);
+            if (spawnLoc.isPresent()) {
+                net.minecraft.world.phys.Vec3 vec = spawnLoc.get();
+                return new Location(spawnWorld, vec.x, vec.y, vec.z, respawnAngle, 0);
+            }
             // Paper end
         }
         return null;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ef9c3a7b15a4901e1662e6d55504b9cbbb804ad3..b4f6bf9a4537e6f153bf34b04c44069075b7846a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1354,17 +1354,22 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     @Override
-    public Location getBedSpawnLocation() {
+    public Location getBedSpawnLocation(boolean load) { // Paper
         ServerLevel world = this.getHandle().server.getLevel(this.getHandle().getRespawnDimension());
         BlockPos bed = this.getHandle().getRespawnPosition();
 
-        if (world != null && bed != null) {
+        if (load && world != null && bed != null) { // Paper
             Optional<Vec3> spawnLoc = net.minecraft.world.entity.player.Player.findRespawnPositionAndUseSpawnBlock(world, bed, this.getHandle().getRespawnAngle(), this.getHandle().isRespawnForced(), true);
             if (spawnLoc.isPresent()) {
                 Vec3 vec = spawnLoc.get();
                 return new Location(world.getWorld(), vec.x, vec.y, vec.z, this.getHandle().getRespawnAngle(), 0);
             }
         }
+        // Paper start
+        if(!load && world != null && bed != null) {
+            return new Location(world.getWorld(), bed.getX(), bed.getY(), bed.getZ(), this.getHandle().getRespawnAngle(), 0);
+        }
+        // Paper end
         return null;
     }
 
