From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zJanny <jan.trummer@hotmail.com>
Date: Fri, 5 Aug 2022 18:42:01 +0200
Subject: [PATCH] fix teleporting outside worldborder with enderpearl


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 23f77dfc6df93ef6a70e14a7e410263d39fae2cb..7a9d0500c40206f47a347c01d76c6f7f1018d511 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -65,6 +65,12 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                     location.setPitch(player.getLocation().getPitch());
                     location.setYaw(player.getLocation().getYaw());
 
+                    // Paper start - Check if enderpearl lands outside worldborder
+                    if(!location.getWorld().getWorldBorder().isInside(location)){
+                        return;
+                    }
+                    // Paper end
+
                     PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.ENDER_PEARL);
                     Bukkit.getPluginManager().callEvent(teleEvent);
 
