From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Doc <nachito94@msn.com>
Date: Wed, 3 Aug 2022 22:17:08 -0400
Subject: [PATCH] Add CreatureSpawnEvent.SpawnReason.SPAWNER_BUCKET


diff --git a/src/main/java/net/minecraft/world/item/MobBucketItem.java b/src/main/java/net/minecraft/world/item/MobBucketItem.java
index a45d93273c6ac881249f0828131e6fd6040c0d69..3a58aec5cc150af5f346ed9b6fbcbf00a42815b8 100644
--- a/src/main/java/net/minecraft/world/item/MobBucketItem.java
+++ b/src/main/java/net/minecraft/world/item/MobBucketItem.java
@@ -46,7 +46,7 @@ public class MobBucketItem extends BucketItem {
     }
 
     private void spawn(ServerLevel world, ItemStack stack, BlockPos pos) {
-        Entity entity = this.type.spawn(world, stack, (Player)null, pos, MobSpawnType.BUCKET, true, false);
+        Entity entity = this.type.spawn(world, stack == null ? null : stack.getTag(), stack != null && stack.hasCustomHoverName() ? stack.getHoverName() : null, (Player)null, pos, MobSpawnType.BUCKET, true, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_BUCKET); // Paper
         if (entity instanceof Bucketable bucketable) {
             bucketable.loadFromBucketTag(stack.getOrCreateTag());
             bucketable.setFromBucket(true);
