From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BuildTools <46540330+willkroboth@users.noreply.github.com>
Date: Fri, 19 Aug 2022 16:10:51 -0400
Subject: [PATCH] Add method isTickingWorlds() to Bukkit.


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b777c1294370bdc68bc6265b039e2a405d89f2f5..5b3b6d96fd8b63365214d365835046b7fa23c865 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -855,6 +855,11 @@ public final class CraftServer implements Server {
         return new ArrayList<World>(this.worlds.values());
     }
 
+    @Override
+    public boolean isTickingWorlds() {
+        return console.isIteratingOverLevels;
+    }
+
     public DedicatedPlayerList getHandle() {
         return this.playerList;
     }
