From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: elmital <54907162+elmital@users.noreply.github.com>
Date: Fri, 16 Sep 2022 17:44:34 +0200
Subject: [PATCH] Fix: EndDragonFight killed statuses should be false for newly
 created worlds


diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index f8d846345c1cc3c78f9ac14635b26f2affc77190..c36a543f091447ad50beb71741de3301e4fb81e3 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -112,9 +112,11 @@ public class EndDragonFight {
             if (nbt.contains("ExitPortalLocation", 10)) {
                 this.portalLocation = NbtUtils.readBlockPos(nbt.getCompound("ExitPortalLocation"));
             }
-        } else {
-            this.dragonKilled = true;
-            this.previouslyKilled = true;
+            // Paper start - Killed statuses should be false for newly created worlds
+            // } else {
+            //     this.dragonKilled = true;
+            //     this.previouslyKilled = true;
+            // Paper end
         }
 
         if (nbt.contains("Gateways", 9)) {
