From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BuildTools <46540330+willkroboth@users.noreply.github.com>
Date: Fri, 19 Aug 2022 16:10:51 -0400
Subject: [PATCH] Add method isTickingWorlds() to Bukkit.


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b443aba38258f501f8f00be6b055f07b709277c4..aef89f4e259bd5fb25de584acf1c18b5f65c2663 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -855,6 +855,11 @@ public final class CraftServer implements Server {
         return new ArrayList<World>(this.worlds.values());
     }
 
+    @Override
+    public boolean isTickingWorlds() {
+        return console.isIteratingOverLevels;
+    }
+
     public DedicatedPlayerList getHandle() {
         return this.playerList;
     }
