From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sat, 18 Feb 2023 16:23:18 +0100
Subject: [PATCH] Update the flag when a captured block state is outdated


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 174f5ff8f827dab2d85cee525429d46bbeae7183..973ecd50f9cb6b86c353586e84d15dcb118ccb60 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -532,6 +532,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
                 blockstate = CapturedBlockState.getTreeBlockState(this, pos, flags);
                 this.capturedBlockStates.put(pos.immutable(), blockstate);
             }
+            blockstate.setFlag(flags); // Paper - update the flag also
             blockstate.setData(state);
             return true;
         }
