From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Florian Schmidt <florian@f012.dev>
Date: Fri, 28 Jul 2023 14:14:35 +0200
Subject: [PATCH] Use correct seed on api world load


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index cce8cc4006c1e7ddae495822fa22771b751e4d20..4a5052d8ba952a386e8a5ace4821848371397a9b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1347,7 +1347,7 @@ public final class CraftServer implements Server {
 
         // Paper - move down
 
-        long j = BiomeManager.obfuscateSeed(creator.seed());
+        long j = BiomeManager.obfuscateSeed(worlddata.worldGenOptions().seed()); // Paper - use world seed
         List<CustomSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(worlddata));
         LevelStem worlddimension = iregistry.get(actualDimension);
 
