From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kenox <muranelp@gmail.com>
Date: Fri, 30 Dec 2022 20:00:01 +0100
Subject: [PATCH] Fix warden spawn reason from DEFAULT to NATURAL


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
index 0554afa72ae6d2f01886916031c8b90b0d45eb52..a96f66e7a79078163214409c29709b0e331482cb 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -36,6 +36,7 @@ import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.gameevent.GameEventListener;
 import net.minecraft.world.level.gameevent.vibrations.VibrationListener;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 import org.slf4j.Logger;
 
 public class SculkShriekerBlockEntity extends BlockEntity implements VibrationListener.VibrationListenerConfig {
@@ -198,7 +199,7 @@ public class SculkShriekerBlockEntity extends BlockEntity implements VibrationLi
     }
 
     private boolean trySummonWarden(ServerLevel world) {
-        return this.warningLevel < 4 ? false : SpawnUtil.trySpawnMob(EntityType.WARDEN, MobSpawnType.TRIGGERED, world, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER).isPresent();
+        return this.warningLevel < 4 ? false : SpawnUtil.trySpawnMob(EntityType.WARDEN, MobSpawnType.TRIGGERED, world, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER, SpawnReason.NATURAL, null).isPresent(); // Paper
     }
 
     @Override
