From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jakubk15 <77227023+Jakubk15@users.noreply.github.com>
Date: Tue, 21 Feb 2023 18:20:50 +0100
Subject: [PATCH] Don't send plugin headers when no plugins of the given type
 are present.


diff --git a/src/main/java/io/papermc/paper/command/PaperPluginsCommand.java b/src/main/java/io/papermc/paper/command/PaperPluginsCommand.java
index 72096a66a4046633de73a12f5a043ac6dff169b1..f0fce4113fb07c64adbec029d177c236cbdcbae8 100644
--- a/src/main/java/io/papermc/paper/command/PaperPluginsCommand.java
+++ b/src/main/java/io/papermc/paper/command/PaperPluginsCommand.java
@@ -186,11 +186,19 @@ public class PaperPluginsCommand extends BukkitCommand {
             //.append(INFO_ICON_START.hoverEvent(SERVER_PLUGIN_INFO)); TODO: Add docs
 
         sender.sendMessage(infoMessage);
-        sender.sendMessage(PAPER_HEADER);
+
+        if (!paperPlugins.isEmpty()) {
+            sender.sendMessage(PAPER_HEADER);
+        }
+
         for (Component component : formatProviders(paperPlugins)) {
             sender.sendMessage(component);
         }
-        sender.sendMessage(BUKKIT_HEADER);
+
+        if (!spigotPlugins.isEmpty()) {
+            sender.sendMessage(BUKKIT_HEADER);
+        }
+        
         for (Component component : formatProviders(spigotPlugins)) {
             sender.sendMessage(component);
         }
