From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GodOfPro <1387ilia@gmail.com>
Date: Tue, 11 Apr 2023 16:31:39 +0430
Subject: [PATCH] Add Mob Experience reward API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index 90c9f75278b3dfae461a98d3034ac0711b6df2c6..6075840f5aa09e4f62d615ccc65ea13e8ba04690 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -159,4 +159,11 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         getHandle().setLeftHanded(leftHanded);
     }
     // Paper end
+
+    // Paper start
+    @Override
+    public int getPossibleExperienceReward() {
+        return getHandle().getExperienceReward();
+    }
+    // Paper end
 }
