From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Wed, 19 Apr 2023 09:43:46 +0400
Subject: [PATCH] Implement event calling for AttemptItemMergeEvent


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index d47b3ac633e7936d30abfda6fc46c2c7412d76fe..4cea06579d2ce6f32af798636249f8a90da5ca81 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -290,6 +290,8 @@ public class ItemEntity extends Entity implements TraceableEntity {
         ItemStack itemstack = this.getItem();
         ItemStack itemstack1 = other.getItem();
 
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.callAttemptItemMergeEvent(other, this).isCancelled()) return; // Paper - AttemptItemMergeEvent
+
         if (Objects.equals(this.target, other.target) && ItemEntity.areMergable(itemstack, itemstack1)) {
             if (true || itemstack1.getCount() < itemstack.getCount()) { // Spigot
                 ItemEntity.merge(this, itemstack, other, itemstack1);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 819c9c020f4d5a1373f68850134960d24b8fc308..c3446f4f8003383764985cb5a1befe2ed510de4f 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -795,6 +795,20 @@ public class CraftEventFactory {
         entity.getServer().getPluginManager().callEvent(event);
         return event;
     }
+    // Paper start - AttemptItemMergeEvent
+    /**
+     * AttemptItemMergeEvent
+     */
+    public static io.papermc.paper.event.entity.AttemptItemMergeEvent callAttemptItemMergeEvent(ItemEntity merging, ItemEntity mergingWith) {
+        org.bukkit.entity.Item entityMerging = (org.bukkit.entity.Item) merging.getBukkitEntity();
+        org.bukkit.entity.Item entityMergingWith = (org.bukkit.entity.Item) mergingWith.getBukkitEntity();
+
+        io.papermc.paper.event.entity.AttemptItemMergeEvent event = new io.papermc.paper.event.entity.AttemptItemMergeEvent(entityMerging, entityMergingWith);
+
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
+    // Paper end
 
     /**
      * ItemMergeEvent
