From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 9 Jun 2023 22:49:38 -0400
Subject: [PATCH] Player standing on position API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index ea056babe2f8123f20dc608d8a636da1de634b8c..146019c3021439d5738fe0b32cb05b28ad8ce78f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1442,4 +1442,24 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !this.getHandle().level().noCollision(this.getHandle(), aabb);
     }
     // Paper End - Collision API
+    // Paper start - Player standing on position API
+    @Override
+    public io.papermc.paper.math.BlockPosition getBlockPosBelowThatAffectsMyMovement() {
+        net.minecraft.core.BlockPos blockPos = this.getHandle().getBlockPosBelowThatAffectsMyMovement();
+        return io.papermc.paper.math.Position.block(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+    }
+
+    @Override
+    public io.papermc.paper.math.BlockPosition getOnPos() {
+        net.minecraft.core.BlockPos blockPos = this.getHandle().getOnPos();
+        return io.papermc.paper.math.Position.block(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+    }
+
+    @Override
+    public io.papermc.paper.math.BlockPosition getSupportingPos() {
+        return this.getHandle().mainSupportingBlockPos
+            .map((func) -> io.papermc.paper.math.Position.block(func.getX(), func.getY(), func.getZ()))
+            .orElse(null);
+    }
+    // Paper end - Player standing on position API
 }
