From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NonSwag <mrminecraft00@gmail.com>
Date: Sun, 7 May 2023 22:00:41 +0200
Subject: [PATCH] removed unnecessary double check


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 24af8a5763851daf8a2952697328a0c26702ff2b..03f895c65d3d516bfdc40ef323794d2d567ca594 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -639,8 +639,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     public boolean enforceSecureProfile() {
         // Paper start - fix secure profile with proxy online mode
         var globalConfiguration = io.papermc.paper.configuration.GlobalConfiguration.get();
-        return this.getProperties().enforceSecureProfile && this.getProperties().onlineMode
-            || globalConfiguration.proxies.isProxyOnlineMode();
+        return this.getProperties().enforceSecureProfile && globalConfiguration.proxies.isProxyOnlineMode();
         // Paper end
     }
 
