From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Trevor Bedson <78997566+Prorickey@users.noreply.github.com>
Date: Fri, 14 Jul 2023 20:47:02 -0400
Subject: [PATCH] Fire entity death event for ender dragon


diff --git a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index ccc68b1a1b1b087c52d91591ba4c63b075bfdc66..f8cdc76561852fd32d9c079ceb3567e6d49c892e 100644
--- a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -642,6 +642,15 @@ public class EnderDragon extends Mob implements Enemy {
 
     @Override
     public void kill() {
+        // Paper start
+        this.silentDeath = true;
+        org.bukkit.event.entity.EntityDeathEvent deathEvent =  org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this);
+        if (deathEvent.isCancelled()) {
+            this.silentDeath = false; // Reset to default if event was cancelled
+            return;
+        }
+        // Paper end
+
         this.remove(Entity.RemovalReason.KILLED);
         this.gameEvent(GameEvent.ENTITY_DIE);
         if (this.dragonFight != null) {
