From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: roro1506HD <roro1506mcpro@gmail.com>
Date: Fri, 25 Aug 2023 14:42:54 +0200
Subject: [PATCH] Fix MC-255756 shield break/block sounds not playing


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index e11d7283662834047b2ff81a2fd25a4263792deb..caa89f4a667615fa6b725dc83f7030af7b10ef08 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1482,7 +1482,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
             boolean knockbackCancelled = this.level().paperConfig().environment.disableExplosionKnockback && source.is(DamageTypeTags.IS_EXPLOSION) && this instanceof net.minecraft.world.entity.player.Player; // Paper - Disable explosion knockback
             if (flag1) {
                 if (flag) {
-                    this.level().broadcastEntityEvent(this, (byte) 29);
+                    this.level().playSound(null, this, SoundEvents.SHIELD_BLOCK, SoundSource.PLAYERS, 1.0F, 0.8F + this.level().random.nextFloat() * 0.4F); // Paper - Fixes MC-255756: Play the sound instead of the entity event. Sound taken from LivingEntity#handleEntityEvent(29)
                 } else {
                     if (!knockbackCancelled) // Paper - Disable explosion knockback
                     this.level().broadcastDamageEvent(this, source);
@@ -2225,7 +2225,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             // Apply blocking code // PAIL: steal from above
             if (event.getDamage(DamageModifier.BLOCKING) < 0) {
-                this.level().broadcastEntityEvent(this, (byte) 29); // SPIGOT-4635 - shield damage sound
+                this.level().playSound(null, this, SoundEvents.SHIELD_BLOCK, SoundSource.PLAYERS, 1.0F, 0.8F + this.level().random.nextFloat() * 0.4F); // Paper - Fixes MC-255756: Play the sound instead of the entity event. Sound taken from LivingEntity#handleEntityEvent(29)
                 this.hurtCurrentlyUsedShield((float) -event.getDamage(DamageModifier.BLOCKING));
                 Entity entity = damagesource.getDirectEntity();
 
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index e2a25c29ec74147b3e66aa0b3deb85a8f6ee53a5..0d5d13f53a6605e735b88c57d275a3a1e6b1d51f 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1689,7 +1689,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
             if (this.random.nextFloat() < f) {
                 player.getCooldowns().addCooldown(Items.SHIELD, 100);
-                this.level().broadcastEntityEvent(player, (byte) 30);
+                this.level().playSound(null, player, net.minecraft.sounds.SoundEvents.SHIELD_BREAK, net.minecraft.sounds.SoundSource.PLAYERS, 0.8F, 0.8F + this.level().random.nextFloat() * 0.4F); // Paper - Fixes MC-255756: Play the sound instead of the entity event. Sound taken from LivingEntity#handleEntityEvent(30)
             }
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 58152160d609d0e9d105153aeb166a56a7955603..b6d1b16e01892581aa4a73a2923ddb51c5a1ab7a 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1471,7 +1471,7 @@ public abstract class Player extends LivingEntity {
         if (this.random.nextFloat() < f) {
             this.getCooldowns().addCooldown(Items.SHIELD, 100);
             this.stopUsingItem();
-            this.level().broadcastEntityEvent(this, (byte) 30);
+            this.level().playSound(null, this, SoundEvents.SHIELD_BREAK, SoundSource.PLAYERS, 0.8F, 0.8F + this.level().random.nextFloat() * 0.4F); // Paper - Fixes MC-255756: Play the sound instead of the entity event. Sound taken from LivingEntity#handleEntityEvent(30)
         }
 
     }
