From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tamion <70228790+notTamion@users.noreply.github.com>
Date: Sat, 30 Sep 2023 12:36:14 +0200
Subject: [PATCH] Fix strikeLightningEffect powers lightning rods and clears
 copper


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 868951dc21aff541765b1f58f08cdf3c47446d25..155082d45fc7ad31c25672116afb63a39d0fd97f 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -997,7 +997,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
                 if (entitylightning != null) {
                     entitylightning.moveTo(Vec3.atBottomCenterOf(blockposition));
-                    entitylightning.setVisualOnly(flag1);
+                    entitylightning.setSkeletonTrap(flag1); // Paper
                     this.strikeLightning(entitylightning, org.bukkit.event.weather.LightningStrikeEvent.Cause.WEATHER); // CraftBukkit
                 }
             }
diff --git a/src/main/java/net/minecraft/world/entity/LightningBolt.java b/src/main/java/net/minecraft/world/entity/LightningBolt.java
index 255fb5e922c63130708e4bcab208b4db52a58387..a0d0746f688c820f3cdfe5daf7140e5628a5be68 100644
--- a/src/main/java/net/minecraft/world/entity/LightningBolt.java
+++ b/src/main/java/net/minecraft/world/entity/LightningBolt.java
@@ -46,6 +46,7 @@ public class LightningBolt extends Entity {
     private final Set<Entity> hitEntities = Sets.newHashSet();
     private int blocksSetOnFire;
     public boolean isSilent = false; // Spigot
+    public boolean skeletonTrap; // Paper
 
     public LightningBolt(EntityType<? extends LightningBolt> type, Level world) {
         super(type, world);
@@ -59,6 +60,12 @@ public class LightningBolt extends Entity {
         this.visualOnly = cosmetic;
     }
 
+    // Paper start
+    public void setSkeletonTrap(boolean skeletonTrap) {
+        this.skeletonTrap = skeletonTrap;
+    }
+    //Paper end
+
     @Override
     public SoundSource getSoundSource() {
         return SoundSource.WEATHER;
@@ -86,7 +93,7 @@ public class LightningBolt extends Entity {
     @Override
     public void tick() {
         super.tick();
-        if (!this.isSilent && this.life == 2) { // Spigot
+        if (!this.visualOnly && !this.isSilent && this.life == 2) { // Spigot // Paper
             if (this.level().isClientSide()) {
                 this.level().playLocalSound(this.getX(), this.getY(), this.getZ(), SoundEvents.LIGHTNING_BOLT_THUNDER, SoundSource.WEATHER, 10000.0F, 0.8F + this.random.nextFloat() * 0.2F, false);
                 this.level().playLocalSound(this.getX(), this.getY(), this.getZ(), SoundEvents.LIGHTNING_BOLT_IMPACT, SoundSource.WEATHER, 2.0F, 0.5F + this.random.nextFloat() * 0.2F, false);
@@ -133,10 +140,10 @@ public class LightningBolt extends Entity {
             }
         }
 
-        if (this.life >= 0 && !this.visualOnly) { // CraftBukkit - add !this.visualOnly
+        if (this.life >= 0 && !this.visualOnly && !this.skeletonTrap) { // CraftBukkit - add !this.visualOnly // Paper
             if (!(this.level() instanceof ServerLevel)) {
                 this.level().setSkyFlashTime(2);
-            } else if (!this.visualOnly) {
+            } else { // Paper
                 list = this.level().getEntities((Entity) this, new AABB(this.getX() - 3.0D, this.getY() - 3.0D, this.getZ() - 3.0D, this.getX() + 3.0D, this.getY() + 6.0D + 3.0D, this.getZ() + 3.0D), Entity::isAlive);
                 iterator = list.iterator();
 
@@ -162,13 +169,13 @@ public class LightningBolt extends Entity {
     }
 
     private void spawnFire(int spreadAttempts) {
-        if (!this.visualOnly && !this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_DOFIRETICK)) {
+        if (!this.visualOnly && !this.skeletonTrap && !this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_DOFIRETICK)) { // Paper
             BlockPos blockposition = this.blockPosition();
             BlockState iblockdata = BaseFireBlock.getState(this.level(), blockposition);
 
             if (this.level().getBlockState(blockposition).isAir() && iblockdata.canSurvive(this.level(), blockposition)) {
                 // CraftBukkit start - add "!visualOnly"
-                if (!this.visualOnly && !CraftEventFactory.callBlockIgniteEvent(this.level(), blockposition, this).isCancelled()) {
+                if (!CraftEventFactory.callBlockIgniteEvent(this.level(), blockposition, this).isCancelled()) { // Paper
                     this.level().setBlockAndUpdate(blockposition, iblockdata);
                     ++this.blocksSetOnFire;
                 }
@@ -181,7 +188,7 @@ public class LightningBolt extends Entity {
                 iblockdata = BaseFireBlock.getState(this.level(), blockposition1);
                 if (this.level().getBlockState(blockposition1).isAir() && iblockdata.canSurvive(this.level(), blockposition1)) {
                     // CraftBukkit start - add "!visualOnly"
-                    if (!this.visualOnly && !CraftEventFactory.callBlockIgniteEvent(this.level(), blockposition1, this).isCancelled()) {
+                    if (!CraftEventFactory.callBlockIgniteEvent(this.level(), blockposition1, this).isCancelled()) { // Paper
                         this.level().setBlockAndUpdate(blockposition1, iblockdata);
                         ++this.blocksSetOnFire;
                     }
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
index d00fb16ae3b94dfcb10fd1a7c1671595e2ff1855..bb9309e8ac7f13da5fdf80b9111ce0477cad59ed 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
@@ -42,7 +42,7 @@ public class SkeletonTrapGoal extends Goal {
 
         if (entitylightning != null) {
             entitylightning.moveTo(this.horse.getX(), this.horse.getY(), this.horse.getZ());
-            entitylightning.setVisualOnly(true);
+            entitylightning.setSkeletonTrap(true); // Paper
             worldserver.strikeLightning(entitylightning, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRAP); // CraftBukkit
             Skeleton entityskeleton = this.createSkeleton(difficultydamagescaler, this.horse);
 
