From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Scandurra <david.scandurra@check24.de>
Date: Wed, 25 Oct 2023 20:36:25 +0200
Subject: [PATCH] Allow null itemstack for Player#sendEquipmentChange


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 57b67e42c1e5784ecd752cf1a5165a4d7eccdb15..00bae5df87bcc1c75d4e2f430241579d3be82c11 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1062,7 +1062,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public void sendEquipmentChange(LivingEntity entity, EquipmentSlot slot, ItemStack item) {
-        this.sendEquipmentChange(entity, Map.of(slot, item));
+        this.sendEquipmentChange(entity, java.util.Collections.singletonMap(slot, item)); // Paper - replace Map.of to allow null values
     }
 
     @Override
