From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 4 Nov 2023 18:39:18 -0400
Subject: [PATCH] Add MaterialTagsTest


diff --git a/src/test/java/io/papermc/paper/inventory/item/MaterialTagsTest.java b/src/test/java/io/papermc/paper/inventory/item/MaterialTagsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..4fde1bd735007d9efec9b81f68ea0b7e9c9d0c4d
--- /dev/null
+++ b/src/test/java/io/papermc/paper/inventory/item/MaterialTagsTest.java
@@ -0,0 +1,35 @@
+package io.papermc.paper.inventory.item;
+
+import com.destroystokyo.paper.MaterialTags;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.enchantment.EnchantmentCategory;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.support.AbstractTestingBase;
+import org.junit.jupiter.api.Test;
+import java.util.HashSet;
+import java.util.Set;
+
+public class MaterialTagsTest extends AbstractTestingBase {
+
+
+    @Test
+    public void testEnchantables() {
+        for (final Item item : BuiltInRegistries.ITEM) {
+            boolean hasMatch = false;
+            Set<EnchantmentCategory> matched = new HashSet<>();
+
+            for (final EnchantmentCategory value : EnchantmentCategory.values()) {
+                if (value.canEnchant(item)) {
+                    hasMatch = true;
+                    matched.add(value);
+                }
+            }
+
+            boolean isTagged = MaterialTags.ENCHANTABLE.isTagged(CraftMagicNumbers.getMaterial(item));
+            if (hasMatch != isTagged) {
+                System.out.println("Item: " + item.getDescriptionId() + " matchedEnch: " + hasMatch + " tagged: " + isTagged + " :" + matched);
+            }
+        }
+    }
+}
