From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dueris <jedimastertoothless@hotmail.com>
Date: Mon, 22 Jul 2024 00:56:21 -0700
Subject: [PATCH] Early Dispatcher Access


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 1d1e76de60e40224f5cb81893f9ee50fe987badb..62a1255031ce8f50fbd2f95774a04d76b39987f2 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -152,7 +152,10 @@ public class Commands {
     public static final int LEVEL_GAMEMASTERS = 2;
     public static final int LEVEL_ADMINS = 3;
     public static final int LEVEL_OWNERS = 4;
-    private final com.mojang.brigadier.CommandDispatcher<CommandSourceStack> dispatcher = new com.mojang.brigadier.CommandDispatcher();
+    // Paper start - Early access to dispatcher
+    public static final com.mojang.brigadier.CommandDispatcher<CommandSourceStack> DISPATCHER = new com.mojang.brigadier.CommandDispatcher();
+    private final com.mojang.brigadier.CommandDispatcher<CommandSourceStack> dispatcher = DISPATCHER;
+    // Paper end
 
     public Commands(Commands.CommandSelection environment, CommandBuildContext commandRegistryAccess) {
         // Paper
