From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 19 Dec 2023 23:40:01 -0800
Subject: [PATCH] Split Position from Location type hierarchy


diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/PaperClassloaderBytecodeModifier.java b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/PaperClassloaderBytecodeModifier.java
index 9bd330f3a3afcb8f693920517a5ad29c3c18098b..bde91e90f0b55a9de5d3557f72e4ba55276d4aeb 100644
--- a/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/PaperClassloaderBytecodeModifier.java
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/PaperClassloaderBytecodeModifier.java
@@ -3,6 +3,7 @@ package io.papermc.paper.plugin.entrypoint.classloader.bytecode;
 import com.google.common.collect.Iterators;
 import io.papermc.paper.plugin.configuration.PluginMeta;
 import io.papermc.paper.plugin.entrypoint.classloader.ClassloaderBytecodeModifier;
+import io.papermc.paper.plugin.entrypoint.classloader.bytecode.versions.API_1_20_4;
 import io.papermc.paper.plugin.provider.configuration.serializer.constraints.PluginConfigConstraints;
 import java.util.Iterator;
 import java.util.LinkedHashMap;
@@ -15,6 +16,7 @@ import org.objectweb.asm.Opcodes;
 public class PaperClassloaderBytecodeModifier implements ClassloaderBytecodeModifier {
 
     private static final Map<String, List<ModifierFactory>> MODIFIERS = Util.make(new LinkedHashMap<>(), map -> {
+        map.put("1.20.4", List.of(API_1_20_4::new));
     });
 
     private final Map<String, List<VersionedClassloaderBytecodeModifier>> constructedModifiers = MODIFIERS.entrySet().stream()
diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/versions/API_1_20_4.java b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/versions/API_1_20_4.java
new file mode 100644
index 0000000000000000000000000000000000000000..0240a1c1e707d45eef749f4a6ccefb27e745252d
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/classloader/bytecode/versions/API_1_20_4.java
@@ -0,0 +1,70 @@
+package io.papermc.paper.plugin.entrypoint.classloader.bytecode.versions;
+
+import io.papermc.asm.rules.RewriteRule;
+import io.papermc.asm.rules.builder.ConfiguredRuleFactory;
+import io.papermc.paper.event.world.StructuresLocateEvent;
+import io.papermc.paper.math.Position;
+import io.papermc.paper.plugin.entrypoint.classloader.bytecode.VersionedClassloaderBytecodeModifier;
+import java.lang.constant.ClassDesc;
+import java.lang.reflect.Method;
+import java.util.Set;
+import java.util.function.Consumer;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.Server;
+import org.bukkit.World;
+import org.bukkit.block.SculkCatalyst;
+import org.bukkit.block.Sign;
+import org.bukkit.boss.DragonBattle;
+import org.bukkit.entity.Player;
+
+import static io.papermc.asm.rules.RewriteRule.chain;
+import static io.papermc.asm.util.DescriptorUtils.desc;
+
+public class API_1_20_4 extends VersionedClassloaderBytecodeModifier {
+
+    private static final ClassDesc POSITION = desc(Position.class);
+
+    private static final Method POSITION_FUZZY_HANDLER;
+
+    static {
+        try {
+            POSITION_FUZZY_HANDLER = API_1_20_4.class.getDeclaredMethod("toPos", Object.class);
+        } catch (final NoSuchMethodException throwable) {
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    // for removing FinePosition from superclasses of Location
+    private static Consumer<ConfiguredRuleFactory> locationSplit(final String... methods) {
+        return r -> r.changeParamFuzzy(Position.class, POSITION_FUZZY_HANDLER, m -> m.names(methods).containsParam(POSITION));
+    }
+
+    public API_1_20_4(final int api) {
+        super(api);
+    }
+
+    @Override
+    protected RewriteRule createRule() {
+        return chain(
+            this.createSplitRule()
+        );
+    }
+
+    private RewriteRule createSplitRule() {
+        // For all methods with Position parameters prior to the removal of Position from Location type hierarchy
+        return chain(
+            this.forOwners(Set.of(Bukkit.class, Server.class), locationSplit("isOwnedByCurrentRegion")),
+            this.forOwner(World.class, locationSplit("hasStructureAt", "isPositionLoaded", "rayTraceEntities", "rayTraceBlocks", "rayTrace")),
+            this.forOwner(SculkCatalyst.class, locationSplit("bloom")),
+            this.forOwner(Sign.class, locationSplit("getInteractableSideFor")),
+            this.forOwner(DragonBattle.class, locationSplit("spawnNewGateway")),
+            this.forOwner(Player.class, locationSplit("lookAt")),
+            this.forOwner(StructuresLocateEvent.Result.class, locationSplit("<init>"))
+        );
+    }
+
+    public static Position toPos(final Object object) { // for split rule
+        return object instanceof final Location loc ? loc.asPosition() : (Position) object;
+    }
+}
