From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tamion <70228790+notTamion@users.noreply.github.com>
Date: Tue, 14 May 2024 16:33:08 +0200
Subject: [PATCH] Add ItemInHands API to Allay


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftAllay.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftAllay.java
index c64918175ec08d20cde2bda9e0cac8b474385fe0..03f3ace338a28062eeedc3505c1e0fd1c7f93533 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftAllay.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftAllay.java
@@ -91,4 +91,16 @@ public class CraftAllay extends CraftCreature implements org.bukkit.entity.Allay
         BlockPos nmsJukeboxPos = this.getHandle().jukeboxPos;
         return (nmsJukeboxPos != null) ? CraftLocation.toBukkit(nmsJukeboxPos, this.getWorld()) : null;
     }
+
+    // Paper start - Add ItemInHands API to Allay
+    @Override
+    public void setItemInHands(org.bukkit.inventory.ItemStack item) {
+       this.getHandle().setItemInHand(net.minecraft.world.InteractionHand.MAIN_HAND, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(item));
+    }
+
+    @Override
+    public org.bukkit.inventory.ItemStack getItemInHands() {
+        return this.getHandle().getItemInHand(net.minecraft.world.InteractionHand.MAIN_HAND).asBukkitCopy();
+    }
+    // Paper end - Add ItemInHands API to Allay
 }
