From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Tue, 19 Dec 2023 12:46:26 -0800
Subject: [PATCH] Make a test fail


diff --git a/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
index af4b7704ab4690138e3188a45d41b890369f6237..6e70390916a2666bd27a724bf1e571de2d332af6 100644
--- a/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
+++ b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
@@ -71,7 +71,7 @@ class AdventureCodecsTest extends AbstractTestingBase {
         final Tag result = TEXT_COLOR_CODEC.encodeStart(NbtOps.INSTANCE, color).result().orElseThrow();
         assertEquals(color.asHexString(), result.getAsString());
         final net.minecraft.network.chat.TextColor nms = net.minecraft.network.chat.TextColor.CODEC.decode(NbtOps.INSTANCE, result).result().orElseThrow().getFirst();
-        assertEquals(color.value(), nms.getValue());
+        assertEquals(color.value() + 1, nms.getValue());
     }
 
     @Test
