From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: maxcom1 <46265094+maxcom1@users.noreply.github.com>
Date: Fri, 8 Mar 2024 18:12:15 +0100
Subject: [PATCH] Add BeehiveNectarDepositEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
index 55b0e2bf98a285cdcd30e40d94192b7a1802efd8..bdb03c2b5ae28af1bb900d8421eee06107318965 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -255,7 +255,8 @@ public class BeehiveBlockEntity extends BlockEntity {
                                 if (iblockdata.is(BlockTags.BEEHIVES, (blockbase_blockdata) -> {
                                     return blockbase_blockdata.hasProperty(BeehiveBlock.HONEY_LEVEL);
                                 })) {
-                                    int i = BeehiveBlockEntity.getHoneyLevel(iblockdata);
+                                    // Paper start - Beehive nectar deposit event
+                                    int i = BeehiveBlockEntity.getHoneyLevel(iblockdata); final int oldHoneyLevel = i; // Paper - OBFHELPER
 
                                     if (i < 5) {
                                         int j = world.random.nextInt(100) == 0 ? 2 : 1;
@@ -264,7 +265,14 @@ public class BeehiveBlockEntity extends BlockEntity {
                                             --j;
                                         }
 
-                                        world.setBlockAndUpdate(blockposition, (BlockState) iblockdata.setValue(BeehiveBlock.HONEY_LEVEL, i + j));
+                                        final int depositedNectar = j; // Paper - OBFHELPER
+
+                                        io.papermc.paper.event.block.BeehiveNectarDepositEvent nectarDepositEvent = new io.papermc.paper.event.block.BeehiveNectarDepositEvent((org.bukkit.entity.Bee) entitybee.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, blockposition), depositedNectar, oldHoneyLevel, oldHoneyLevel + depositedNectar);
+                                        boolean isDepositCancelled = !nectarDepositEvent.callEvent();
+                                        int newHoneyLevel = nectarDepositEvent.getNewHoneyLevel();
+
+                                        if (!isDepositCancelled) world.setBlockAndUpdate(blockposition, (BlockState) iblockdata.setValue(BeehiveBlock.HONEY_LEVEL, newHoneyLevel));
+                                        // Paper end - Beehive nectar deposit event
                                     }
                                 }
                             }
@@ -304,7 +312,6 @@ public class BeehiveBlockEntity extends BlockEntity {
 
             compound.remove(s);
         }
-
     }
 
     private static void setBeeReleaseData(int ticks, Bee bee) {
