From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: roro1506HD <roro1506mcpro@gmail.com>
Date: Fri, 17 Mar 2023 21:20:55 +0100
Subject: [PATCH] Make banners modifiable with both BannerMeta and
 BlockStateMeta


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
index 9469b0d5d8a46ac17c3998a4b537a4feb1deb3b0..dfdf95aac56408d302277ea99259205df943b24f 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
@@ -162,7 +162,7 @@ public final class CraftItemFactory implements ItemFactory {
         case WHITE_WALL_BANNER:
         case YELLOW_BANNER:
         case YELLOW_WALL_BANNER:
-            return meta instanceof CraftMetaBanner ? meta : new CraftMetaBanner(meta);
+            return meta instanceof CraftMetaBanner ? meta : new CraftMetaBanner(meta, material); // Paper - Provide the material
         case ALLAY_SPAWN_EGG:
         case AXOLOTL_SPAWN_EGG:
         case BAT_SPAWN_EGG:
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 654694515b4b9257a41c8623675fa3abc51a1cb7..de9b940eb1220d9062a0f0bfcadab51d7dd678b2 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -414,7 +414,7 @@ public final class CraftItemStack extends ItemStack {
             case WHITE_WALL_BANNER:
             case YELLOW_BANNER:
             case YELLOW_WALL_BANNER:
-                return new CraftMetaBanner(item.getTag());
+                return new CraftMetaBanner(item.getTag(), material); // Paper - Provide the material
             case ALLAY_SPAWN_EGG:
             case AXOLOTL_SPAWN_EGG:
             case BAT_SPAWN_EGG:
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
index d8039afde8a44cbb0112256b3400ced479626331..bbd9985fe6cf2a87dfe4d208c769f1f3f77d0add 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
@@ -19,7 +19,7 @@ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.inventory.meta.BannerMeta;
 
 @DelegateDeserialization(CraftMetaItem.SerializableMeta.class)
-public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+public class CraftMetaBanner extends CraftMetaBlockState implements BannerMeta { // Paper - Make banners also modifiable with BlockStateMeta
 
     private static final Set<Material> BANNER_MATERIALS = Sets.newHashSet(
             Material.BLACK_BANNER,
@@ -61,13 +61,21 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
     static final ItemMetaKey COLOR = new ItemMetaKey("Color", "color");
     static final ItemMetaKey PATTERN = new ItemMetaKey("Pattern", "pattern");
 
+    // Paper start - Add banner materials to CraftMetaBlockState's material list
+    static {
+        CraftMetaBlockState.BLOCK_STATE_MATERIALS.addAll(CraftMetaBanner.BANNER_MATERIALS);
+    }
+    // Paper end - Add banner materials to CraftMetaBlockState's material list
+
     private DyeColor base;
     private List<Pattern> patterns = new ArrayList<Pattern>();
 
-    CraftMetaBanner(CraftMetaItem meta) {
-        super(meta);
+    // Paper start - Create new constructor to properly handle different banners
+    CraftMetaBanner(CraftMetaItem meta, Material material) {
+        super(meta, material);
 
-        if (!(meta instanceof CraftMetaBanner)) {
+        if (true || !(meta instanceof CraftMetaBanner)) {
+            // Paper end - Create new constructor to properly handle different banners
             return;
         }
 
@@ -76,10 +84,12 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
         this.patterns = new ArrayList<Pattern>(banner.patterns);
     }
 
-    CraftMetaBanner(CompoundTag tag) {
-        super(tag);
+    // Paper start - Create new constructor to properly handle different banners
+    CraftMetaBanner(CompoundTag tag, Material material) {
+        super(tag, material);
 
-        if (!tag.contains("BlockEntityTag")) {
+        if (true || !tag.contains("BlockEntityTag")) {
+            // Paper end - Create new constructor to properly handle different banners
             return;
         }
 
@@ -101,12 +111,27 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
         }
     }
 
+    // Paper start - Correctly deserialize old BannerMeta to new BlockStateMeta
+    private static Map<String, Object> ensureDefaultMaterialPresence(Map<String, Object> map) {
+        if (map.containsKey("blockMaterial")) {
+            return map;
+        } else {
+            final Map<String, Object> newMap = new java.util.HashMap<>(map);
+            newMap.put("blockMaterial", "minecraft:white_banner");
+            return newMap;
+        }
+    }
+
     CraftMetaBanner(Map<String, Object> map) {
-        super(map);
+        super(CraftMetaBanner.ensureDefaultMaterialPresence(map));
+        if (map.containsKey("internal")) {
+            return;
+        }
+        // Paper end - Correctly deserialize old BannerMeta to new BlockStateMeta
 
         String baseStr = SerializableMeta.getString(map, CraftMetaBanner.BASE.BUKKIT, true);
         if (baseStr != null) {
-            this.base = DyeColor.legacyValueOf(baseStr);
+            this.setBaseColor(DyeColor.legacyValueOf(baseStr)); // Paper - Use setter to correctly update block state
         }
 
         Iterable<?> rawPatternList = SerializableMeta.getObject(Iterable.class, map, CraftMetaBanner.PATTERNS.BUKKIT, true);
@@ -122,6 +147,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
     @Override
     void applyToItem(CompoundTag tag) {
         super.applyToItem(tag);
+        if (true) return; // Paper - CraftMetaBlockState already save the banner for us
 
         CompoundTag entityTag = new CompoundTag();
         if (this.base != null) {
@@ -143,46 +169,88 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
 
     @Override
     public DyeColor getBaseColor() {
+        if (true) return ((org.bukkit.block.Banner) this.getBlockState()).getBaseColor(); // Paper - Use underlying block state
         return this.base;
     }
 
+    // Paper start - Use underlying block state
+    private void processModification(java.util.function.Consumer<org.bukkit.block.Banner> consumer) {
+        org.bukkit.block.Banner banner = (org.bukkit.block.Banner) this.getBlockState();
+        consumer.accept(banner);
+        this.setBlockState(banner);
+    }
+
     @Override
     public void setBaseColor(DyeColor color) {
+        if (true) {
+            this.processModification(banner -> banner.setBaseColor(color));
+            return;
+        }
+        // Paper end - Use underlying block state
         this.base = color;
     }
 
     @Override
     public List<Pattern> getPatterns() {
+        if (true) return ((org.bukkit.block.Banner) this.getBlockState()).getPatterns(); // Paper - Use underlying block state
         return new ArrayList<Pattern>(this.patterns);
     }
 
     @Override
     public void setPatterns(List<Pattern> patterns) {
+        // Paper start - Use underlying block state
+        if (true) {
+            this.processModification(banner -> banner.setPatterns(patterns));
+            return;
+        }
+        // Paper end - Use underlying block state
         this.patterns = new ArrayList<Pattern>(patterns);
     }
 
     @Override
     public void addPattern(Pattern pattern) {
+        // Paper start - Use underlying block state
+        if (true) {
+            this.processModification(banner -> banner.addPattern(pattern));
+            return;
+        }
+        // Paper end - Use underlying block state
         this.patterns.add(pattern);
     }
 
     @Override
     public Pattern getPattern(int i) {
+        if (true) return ((org.bukkit.block.Banner) this.getBlockState()).getPattern(i); // Paper - Use underlying block state
         return this.patterns.get(i);
     }
 
     @Override
     public Pattern removePattern(int i) {
+        // Paper start - Use underlying block state
+        if (true) {
+            org.bukkit.block.Banner banner = (org.bukkit.block.Banner) this.getBlockState();
+            Pattern removedPattern = banner.removePattern(i);
+            this.setBlockState(banner);
+            return removedPattern;
+        }
+        // Paper end - Use underlying block state
         return this.patterns.remove(i);
     }
 
     @Override
     public void setPattern(int i, Pattern pattern) {
+        // Paper start - Use underlying block state
+        if (true) {
+            this.processModification(banner -> banner.setPattern(i, pattern));
+            return;
+        }
+        // Paper end - Use underlying block state
         this.patterns.set(i, pattern);
     }
 
     @Override
     public int numberOfPatterns() {
+        if (true) return ((org.bukkit.block.Banner) this.getBlockState()).numberOfPatterns(); // Paper - Use underlying block state
         return this.patterns.size();
     }
 
@@ -190,6 +258,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
     ImmutableMap.Builder<String, Object> serialize(ImmutableMap.Builder<String, Object> builder) {
         super.serialize(builder);
 
+        if (true) return builder; // Paper - Let CraftMetaBlockState serialize for us
         if (this.base != null) {
             builder.put(CraftMetaBanner.BASE.BUKKIT, this.base.toString());
         }
@@ -205,6 +274,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
     int applyHash() {
         final int original;
         int hash = original = super.applyHash();
+        if (true) return original; // Paper - CraftMetaBlockState's hash is good enough
         if (this.base != null) {
             hash = 31 * hash + this.base.hashCode();
         }
@@ -216,6 +286,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
 
     @Override
     public boolean equalsCommon(CraftMetaItem meta) {
+        if (true) return super.equalsCommon(meta); // Paper - Only check CraftMetaBlockState's data
         if (!super.equalsCommon(meta)) {
             return false;
         }
@@ -229,11 +300,13 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
 
     @Override
     boolean notUncommon(CraftMetaItem meta) {
+        if (true) return super.notUncommon(meta);  // Paper - Only check CraftMetaBlockState's data
         return super.notUncommon(meta) && (meta instanceof CraftMetaBanner || (this.patterns.isEmpty() && this.base == null));
     }
 
     @Override
     boolean isEmpty() {
+        if (true) return super.isEmpty(); // Paper - Only check CraftMetaBlockState's data
         return super.isEmpty() && this.patterns.isEmpty() && this.base == null;
     }
 
@@ -245,7 +318,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
     @Override
     public CraftMetaBanner clone() {
         CraftMetaBanner meta = (CraftMetaBanner) super.clone();
-        meta.patterns = new ArrayList<>(this.patterns);
+        // meta.patterns = new ArrayList<>(this.patterns); // Paper - Only copy CraftMetaBlockState's data
         return meta;
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
index cab86a4fe1e4abc4cbf4e6a053465e578b34be67..17c97a34c04f085618e143e1c0c04aff6c46c555 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
@@ -40,7 +40,7 @@ public class CraftMetaBlockState extends CraftMetaItem implements BlockStateMeta
             Material.BLACK_SHULKER_BOX
     );
 
-    private static final Set<Material> BLOCK_STATE_MATERIALS = Sets.newHashSet(
+    static final Set<Material> BLOCK_STATE_MATERIALS = Sets.newHashSet( // Paper - Change it to package-private, so we can use that in CraftMetaBanner
             Material.FURNACE,
             Material.CHEST,
             Material.TRAPPED_CHEST,
