From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: YoungSoulluoS <70120313+YoungSoulluoS@users.noreply.github.com>
Date: Tue, 27 Aug 2024 14:32:55 +0500
Subject: [PATCH] Fix-MC-195278-universal_anger_gamerule


diff --git a/src/main/java/net/minecraft/world/entity/NeutralMob.java b/src/main/java/net/minecraft/world/entity/NeutralMob.java
index b55633bf4e26dc00e27ab28c7739c274e48b74c3..d917a2495a01e0b8be9ea8d70ca576dee2fac763 100644
--- a/src/main/java/net/minecraft/world/entity/NeutralMob.java
+++ b/src/main/java/net/minecraft/world/entity/NeutralMob.java
@@ -81,7 +81,7 @@ public interface NeutralMob {
     }
 
     default boolean isAngryAtAllPlayers(Level world) {
-        return world.getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER) && this.isAngry() && this.getPersistentAngerTarget() == null;
+        return world.getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER) && this.isAngry();// && this.getPersistentAngerTarget() == null; // Paper - fix MC-195278
     }
 
     default boolean isAngry() {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
index 80ef4b6649da3049f21624926fa38595d76c5da5..97d47db01649bd2c8149ff98329c4d6b399cd59f 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
@@ -38,7 +38,7 @@ public class HurtByTargetGoal extends TargetGoal {
         LivingEntity entityliving = this.mob.getLastHurtByMob();
 
         if (i != this.timestamp && entityliving != null) {
-            if (entityliving.getType() == EntityType.PLAYER && this.mob.level().getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER)) {
+            if (entityliving.getType() == EntityType.PLAYER && this.mob.level().getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER) && this.targetMob != null) { // Paper - fix MC-195278
                 return false;
             } else {
                 Class[] aclass = this.toIgnoreDamage;
