From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Sun, 17 Mar 2024 20:53:42 +0100
Subject: [PATCH] Add permission check for spawn protection


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 68d268b6fff126e8645b6deec3fb549ea2286b77..2c923159a4acca65f637f8dd93731bf2213589c8 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -575,6 +575,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             return false;
         } else if (this.getPlayerList().isOp(player.getGameProfile())) {
             return false;
+        } else if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.allowOverrideOpBehaviorForSpawnProtection && player.getBukkitEntity().hasPermission("paper.environment.spawnprotection.bypass")) { // Paper start - Add permission check for spawn protection
+            return false; // Paper end - Add permission check for spawn protection
         } else if (this.getSpawnProtectionRadius() <= 0) {
             return false;
         } else {
