From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gero <gecam59@gmail.com>
Date: Sat, 9 Nov 2024 22:27:58 +0100
Subject: [PATCH] Call CraftPlayer#onEntityRemove for all online players


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 5c9d15dbc55aec3ec99143a255e42b5e28a9e95e..b3e709c9ce116bf41cd050674ee1fb31b08e684f 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -2804,7 +2804,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             // CraftBukkit start
             entity.valid = false;
             if (!(entity instanceof ServerPlayer)) {
-                for (ServerPlayer player : ServerLevel.this.players) {
+                for (ServerPlayer player : ServerLevel.this.server.getPlayerList().players) { // Paper - call onEntityRemove for all online players
                     player.getBukkitEntity().onEntityRemove(entity);
                 }
             }
