From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Abel <abelvanhulst@gmail.com>
Date: Sun, 10 Nov 2024 16:32:34 +0100
Subject: [PATCH] API to check if the server is sleeping


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 64b56abf8900d0424100da460fc68ac964394793..f0f2111b5610e493baf718a8a2257f9fb8c679e7 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1614,9 +1614,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     public void onServerExit() {}
 
+    // Paper start - API to check if the server is sleeping
     public boolean isPaused() {
-        return false;
+        return this.emptyTicks >= this.pauseWhileEmptySeconds() * 20;
     }
+    // Paper end - API to check if the server is sleeping
 
     public void tickServer(BooleanSupplier shouldKeepTicking) {
         org.spigotmc.WatchdogThread.tick(); // Spigot
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index dfddcfb1fe1679adaecf75375757dca720e76ce1..f98b9ae05618d8a4199aaf004f26a61be9d26008 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3246,4 +3246,11 @@ public final class CraftServer implements Server {
         return this.potionBrewer;
     }
     // Paper end
+
+    // Paper start - API to check if the server is sleeping
+    @Override
+    public boolean isPaused() {
+        return console.isPaused();
+    }
+    // Paper end - API to check if the server is sleeping
 }
