From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: masmc05 <masmc05@gmail.com>
Date: Sat, 5 Oct 2024 04:23:13 +0300
Subject: [PATCH] Add ItemStack#effectiveName


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index ccad5a7018f6eaacb011818e9da990c75e0d06df..427e0fb8e932cf401a5e0ae7c310fe338554dd27 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -469,6 +469,13 @@ public final class CraftItemStack extends ItemStack {
         return true;
     }
 
+    // Paper start - add ItemStack#effectiveName
+    @Override
+    public net.kyori.adventure.text.Component effectiveName() {
+        return this.handle == null ? net.kyori.adventure.text.Component.empty() : io.papermc.paper.adventure.PaperAdventure.asAdventure(this.handle.getStyledHoverName());
+    }
+    // Paper end - add ItemStack#effectiveName
+
     @Override
     public boolean isSimilar(ItemStack stack) {
         if (stack == null) {
