From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 28 Jun 2020 19:36:55 -0400
Subject: [PATCH] Don't allow null UUID's for chat


diff --git a/src/main/java/net/minecraft/network/chat/ChatSender.java b/src/main/java/net/minecraft/network/chat/ChatSender.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/net/minecraft/network/chat/ChatSender.java
+++ b/src/main/java/net/minecraft/network/chat/ChatSender.java
@@ -0,0 +0,0 @@ import net.minecraft.world.entity.player.ProfilePublicKey;
 public record ChatSender(UUID profileId, @Nullable ProfilePublicKey profilePublicKey) {
     public static final ChatSender SYSTEM = new ChatSender(Util.NIL_UUID, (ProfilePublicKey)null);
 
+    // Paper start
+    public ChatSender {
+        com.google.common.base.Preconditions.checkNotNull(profileId, "uuid cannot be null");
+    }
+    // Paper end
+
     public boolean isSystem() {
         return SYSTEM.equals(this);
     }
