From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 23 Mar 2021 06:43:30 +0000
Subject: [PATCH] copy TESign#isEditable from snapshots


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index 09ea221665bd17bc853274d0146987ebadefe46f..e0bb5146059ad767feca38f3816f29069239ffda 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -116,6 +116,7 @@ public class CraftSign extends CraftBlockEntityState<SignBlockEntity> implements
             }
             // Paper end
         }
+        sign.isEditable = getSnapshot().isEditable; // Paper - copy manually
     }
 
     public static void openSign(Sign sign, org.bukkit.entity.HumanEntity player) { // Paper - change move open sign to HumanEntity
