From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 22 Mar 2022 09:50:40 -0700
Subject: [PATCH] fix player loottables running when mob loot gamerule is false


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index c2b883f47829b7b14d1fdd69d2e9bc563f708e53..f01586da79b5e3965d90ed143204f9636d35ecaa 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -866,12 +866,14 @@ public class ServerPlayer extends Player {
                 }
             }
         }
+        if (this.shouldDropLoot() && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Paper - preserve this check from vanilla
         // SPIGOT-5071: manually add player loot tables (SPIGOT-5195 - ignores keepInventory rule)
         this.dropFromLootTable(damageSource, this.lastHurtByPlayerTime > 0);
         for (org.bukkit.inventory.ItemStack item : this.drops) {
             loot.add(item);
         }
         this.drops.clear(); // SPIGOT-5188: make sure to clear
+        } // Paper
 
         Component defaultMessage = this.getCombatTracker().getDeathMessage();
 
