From 6a0255d07a09045c7b0d3db3fa88ccfd06092226 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/commands/arguments/blocks/BlockStateParser.java b/net/minecraft/commands/arguments/blocks/BlockStateParser.java
index 762aa2a5ee9775cf54182954a8e608555d37254c..57eb5bffaae6282cd9ee798aa8d0ad0eff54dfa2 100644
--- a/net/minecraft/commands/arguments/blocks/BlockStateParser.java
+++ b/net/minecraft/commands/arguments/blocks/BlockStateParser.java
@@ -69,7 +69,7 @@ public class BlockStateParser {
     private final StringReader reader;
     private final boolean forTesting;
     private final boolean allowNbt;
-    private final Map<Property<?>, Comparable<?>> properties = Maps.newHashMap();
+    private final Map<Property<?>, Comparable<?>> properties = Maps.newLinkedHashMap(); // CraftBukkit - stable
     private final Map<String, String> vagueProperties = Maps.newHashMap();
     private ResourceLocation id = ResourceLocation.withDefaultNamespace("");
     @Nullable
