From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 6 Oct 2018 21:47:44 -0500
Subject: [PATCH] Allow setting the vex's summoner


diff --git a/src/main/java/net/minecraft/world/entity/monster/Vex.java b/src/main/java/net/minecraft/world/entity/monster/Vex.java
index ec4f6d96360e759ffc19de838fdbf3027164a424..a4be2bddc5f51601d419647a280c89f7101371f2 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vex.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vex.java
@@ -165,6 +165,7 @@ public class Vex extends Monster {
         this.setVexFlag(1, charging);
     }
 
+    public void setOwner(Mob entityinsentient) { setOwner(entityinsentient); } // Paper - OBFHELPER
     public void setOwner(Mob owner) {
         this.owner = owner;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
index 56bec4350f36a94d4dfa71a234872a795c2dcb3f..07c470f8b049bea930337abc1cc87f4669d2f11a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
@@ -20,6 +20,10 @@ public class CraftVex extends CraftMonster implements Vex {
         net.minecraft.world.entity.Mob owner = getHandle().getOwner();
         return owner != null ? (org.bukkit.entity.Mob) owner.getBukkitEntity() : null;
     }
+
+    public void setSummoner(org.bukkit.entity.Mob summoner) {
+        getHandle().setOwner(summoner == null ? null : ((CraftMob) summoner).getHandle());
+    }
     // Paper end
 
     @Override
