From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 23 Mar 2021 06:43:30 +0000
Subject: [PATCH] copy TESign#isEditable from snapshots


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index 65e8a349c80a700f63dd27b11bb2099f65cbc069..eb0739c0927e821a5080d14e762225fd4936b82d 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -109,6 +109,7 @@ public class CraftSign extends CraftBlockEntityState<SignBlockEntity> implements
             }
             // Paper end
         }
+        sign.isEditable = getSnapshot().isEditable; // Paper - copy manually
     }
 
     // Paper start
