From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 3 Aug 2018 00:04:54 -0400
Subject: [PATCH] MC-135506: Experience should save as Integers


diff --git a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
+++ b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
@@ -0,0 +0,0 @@ public class ExperienceOrb extends Entity {
     public void addAdditionalSaveData(CompoundTag tag) {
         tag.putShort("Health", (short) this.health);
         tag.putShort("Age", (short) this.age);
-        tag.putShort("Value", (short) this.value);
+        tag.putInt("Value", this.value); // Paper - save as Integer
         this.savePaperNBT(tag); // Paper
     }
 
@@ -0,0 +0,0 @@ public class ExperienceOrb extends Entity {
     public void readAdditionalSaveData(CompoundTag tag) {
         this.health = tag.getShort("Health");
         this.age = tag.getShort("Age");
-        this.value = tag.getShort("Value");
+        this.value = tag.getInt("Value"); // Paper - load as Integer
         this.loadPaperNBT(tag); // Paper
     }
 
