--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
@@ -210,7 +210,7 @@
     }
 
     public void resetAfterBecomingOminous(TrialSpawner trialspawner, WorldServer worldserver) {
-        Stream stream = this.currentMobs.stream();
+        Stream<UUID> stream = this.currentMobs.stream(); // CraftBukkit - decompile error
 
         Objects.requireNonNull(worldserver);
         stream.map(worldserver::getEntity).forEach((entity) -> {
@@ -222,7 +222,7 @@
                     entityinsentient.dropPreservedEquipment(worldserver);
                 }
 
-                entity.remove(Entity.RemovalReason.DISCARDED);
+                entity.remove(Entity.RemovalReason.DISCARDED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - Add bukkit remove cause;
             }
         });
         if (!trialspawner.getOminousConfig().spawnPotentialsDefinition().isEmpty()) {
