--- a/net/minecraft/world/level/levelgen/DebugLevelSource.java
+++ b/net/minecraft/world/level/levelgen/DebugLevelSource.java
@@ -48,6 +_,13 @@
         super(new FixedBiomeSource(plains));
     }
 
+    // Paper start - Add way to get world type key
+    @Override
+    public org.bukkit.NamespacedKey getKey() {
+        return org.bukkit.craftbukkit.util.CraftNamespacedKey.fromResourceKey(net.minecraft.world.level.levelgen.presets.WorldPresets.DEBUG);
+    }
+    // Paper end - Add way to get world type key
+
     @Override
     protected MapCodec<? extends ChunkGenerator> codec() {
         return CODEC;
