--- a/net/minecraft/world/level/levelgen/PatrolSpawner.java
+++ b/net/minecraft/world/level/levelgen/PatrolSpawner.java
@@ -18,17 +_,47 @@
 
     @Override
     public void tick(final ServerLevel level, final boolean spawnEnemies) {
+        if (level.paperConfig().entities.behavior.pillagerPatrols.disable || level.paperConfig().entities.behavior.pillagerPatrols.spawnChance == 0) return; // Paper - Add option to disable pillager patrols & Pillager patrol spawn settings and per player options
         if (spawnEnemies) {
             if (level.getGameRules().get(GameRules.SPAWN_PATROLS)) {
                 RandomSource random = level.getRandom();
-                this.nextTick--;
-                if (this.nextTick <= 0) {
-                    this.nextTick = this.nextTick + 12000 + random.nextInt(1200);
-                    if (level.isBrightOutside()) {
-                        if (random.nextInt(5) == 0) {
-                            int playerCount = level.players().size();
+                // this.nextTick--;
+                // Paper start - Pillager patrol spawn settings and per player options
+                int playerCount = level.players().size();
+                if (playerCount < 1) {
+                    return;
+                }
+
+                net.minecraft.server.level.ServerPlayer player = level.players().get(random.nextInt(playerCount));
+                if (player.isSpectator()) {
+                    return;
+                }
+
+                int patrolSpawnDelay;
+                if (level.paperConfig().entities.behavior.pillagerPatrols.spawnDelay.perPlayer) {
+                    --player.patrolSpawnDelay;
+                    patrolSpawnDelay = player.patrolSpawnDelay;
+                } else {
+                    this.nextTick--;
+                    patrolSpawnDelay = this.nextTick;
+                }
+                if (patrolSpawnDelay <= 0) {
+                    long dayCount;
+                    if (level.paperConfig().entities.behavior.pillagerPatrols.start.perPlayer) {
+                        dayCount = player.getStats().getValue(net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.PLAY_TIME)) / 24000L; // PLAY_TIME is counting in ticks
+                    } else {
+                        dayCount = level.getDayCount();
+                    }
+                    if (level.paperConfig().entities.behavior.pillagerPatrols.spawnDelay.perPlayer) {
+                        player.patrolSpawnDelay += level.paperConfig().entities.behavior.pillagerPatrols.spawnDelay.ticks + random.nextInt(1200);
+                    } else {
+                        this.nextTick += level.paperConfig().entities.behavior.pillagerPatrols.spawnDelay.ticks + random.nextInt(1200);
+                    }
+
+                    if (dayCount >= level.paperConfig().entities.behavior.pillagerPatrols.start.day && level.isBrightOutside()) {
+                        if (random.nextDouble() < level.paperConfig().entities.behavior.pillagerPatrols.spawnChance) {
+                            // Paper end - Pillager patrol spawn settings and per player options
                             if (playerCount >= 1) {
-                                Player player = level.players().get(random.nextInt(playerCount));
                                 if (!player.isSpectator()) {
                                     if (!level.isCloseToVillage(player.blockPosition(), 2)) {
                                         int x = (24 + random.nextInt(24)) * (random.nextBoolean() ? -1 : 1);
@@ -80,7 +_,7 @@
 
                 mob.setPos(pos.getX(), pos.getY(), pos.getZ());
                 mob.finalizeSpawn(level, level.getCurrentDifficultyAt(pos), EntitySpawnReason.PATROL, null);
-                level.addFreshEntityWithPassengers(mob);
+                level.addFreshEntityWithPassengers(mob, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PATROL); // CraftBukkit
                 return true;
             } else {
                 return false;
