--- a/net/minecraft/world/level/levelgen/WorldgenRandom.java
+++ b/net/minecraft/world/level/levelgen/WorldgenRandom.java
@@ -64,7 +_,7 @@
     }
 
     public void setLargeFeatureWithSalt(final long seed, final int x, final int z, final int blend) {
-        long result = x * 341873128712L + z * 132897987541L + seed + blend;
+        long result = x * 341873128712L + z * 132897987541L + seed + blend; // Paper - diff on change for CustomChunkGenerator
         this.setSeed(result);
     }
 
