--- a/net/minecraft/core/RegistrySetBuilder.java
+++ b/net/minecraft/core/RegistrySetBuilder.java
@@ -42,6 +_,11 @@
     ) {
         return new RegistrySetBuilder.EmptyTagRegistryLookup<T>(owner) {
             @Override
+            public Optional<T> getValueForCopying(final ResourceKey<T> resourceKey) {
+                return this.get(resourceKey).map(Holder.Reference::value);
+            }
+
+            @Override
             public ResourceKey<? extends Registry<? extends T>> key() {
                 return registryKey;
             }
