--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -48,7 +_,7 @@
     private static final String FRAME_PREFIX = "frame-";
     public static final Codec<MapItemSavedData> CODEC = RecordCodecBuilder.create(
         i -> i.group(
-                Level.RESOURCE_KEY_CODEC.fieldOf("dimension").forGetter(m -> m.dimension),
+                createUUIDBackedDimensionKeyCodec().forGetter(MapItemSavedData::packUUIDBackedDimension), // Paper - store target world by uuid in addition to dimension
                 Codec.INT.fieldOf("xCenter").forGetter(m -> m.centerX),
                 Codec.INT.fieldOf("zCenter").forGetter(m -> m.centerZ),
                 Codec.BYTE.optionalFieldOf("scale", (byte)0).forGetter(m -> m.scale),
@@ -69,6 +_,7 @@
     public byte scale;
     public byte[] colors = new byte[16384];
     public boolean locked;
+    private final org.bukkit.craftbukkit.map.RenderData vanillaRender = new org.bukkit.craftbukkit.map.RenderData(); // Paper - Use Vanilla map renderer when possible
     public final List<MapItemSavedData.HoldingPlayer> carriedBy = Lists.newArrayList();
     public final Map<Player, MapItemSavedData.HoldingPlayer> carriedByPlayers = Maps.newHashMap();
     private final Map<String, MapBanner> bannerMarkers = Maps.newHashMap();
@@ -76,6 +_,13 @@
     private final Map<String, MapFrame> frameMarkers = Maps.newHashMap();
     private int trackedDecorationCount;
 
+    // CraftBukkit start
+    public final org.bukkit.craftbukkit.map.CraftMapView mapView;
+    private final org.bukkit.craftbukkit.CraftServer server;
+    public java.util.UUID uniqueId;
+    public MapId id;
+    // CraftBukkit end
+
     public static SavedDataType<MapItemSavedData> type(final MapId id) {
         return new SavedDataType<>(Identifier.withDefaultNamespace(id.key()), () -> {
             throw new IllegalStateException("Should never create an empty map saved data");
@@ -98,7 +_,29 @@
         this.trackingPosition = trackingPosition;
         this.unlimitedTracking = unlimitedTracking;
         this.locked = locked;
-    }
+        // CraftBukkit start
+        this.mapView = new org.bukkit.craftbukkit.map.CraftMapView(this);
+        this.server = (org.bukkit.craftbukkit.CraftServer) org.bukkit.Bukkit.getServer();
+        this.vanillaRender.buffer = colors; // Paper - Use Vanilla map renderer when possible
+        // CraftBukkit end
+    }
+
+    // Paper start - store target world by uuid in addition to dimension
+    private MapItemSavedData(
+        UUIDBackedDimension dimension,
+        int x,
+        int z,
+        byte scale,
+        ByteBuffer colors,
+        boolean trackingPosition,
+        boolean unlimitedTracking,
+        boolean locked,
+        List<MapBanner> banners,
+        List<MapFrame> frames
+    ) {
+        this(dimension.resolveOrThrow(), x, z, scale, colors, trackingPosition, unlimitedTracking, locked, banners, frames);
+    }
+    // Paper end - store target world by uuid in addition to dimension
 
     private MapItemSavedData(
         final ResourceKey<Level> dimension,
@@ -126,6 +_,8 @@
             this.frameMarkers.put(frame.getId(), frame);
             this.addDecoration(MapDecorationTypes.FRAME, null, getFrameKey(frame.entityId()), frame.pos().getX(), frame.pos().getZ(), frame.rotation(), null);
         }
+
+        this.vanillaRender.buffer = colors.array(); // Paper - Use Vanilla map renderer when possible
     }
 
     public static MapItemSavedData createFresh(
@@ -209,6 +_,7 @@
             }
 
             MapFrame mapFrame = new MapFrame(pos, placedInFrame.getDirection().get2DDataValue() * 90, placedInFrame.getId());
+            if (this.decorations.size() < tickingPlayer.level().paperConfig().maps.itemFrameCursorLimit) { // Paper - Limit item frame cursors on maps
             this.addDecoration(
                 MapDecorationTypes.FRAME,
                 tickingPlayer.level(),
@@ -222,6 +_,7 @@
             if (!mapFrame.equals(oldFrame)) {
                 this.setDirty();
             }
+            } // Paper - Limit item frame cursors on maps
         }
 
         MapDecorations staticDecorations = itemStack.getOrDefault(DataComponents.MAP_DECORATIONS, MapDecorations.EMPTY);
@@ -252,7 +_,7 @@
             this.trackedDecorationCount--;
         }
 
-        this.setDecorationsDirty();
+        if (decoration != null) this.setDecorationsDirty(); // Paper - only mark dirty if a change occurs
     }
 
     public static void addTargetDecoration(final ItemStack itemStack, final BlockPos position, final String key, final Holder<MapDecorationType> decorationType) {
@@ -375,7 +_,12 @@
     }
 
     public void setColorsDirty(final int x, final int y) {
-        this.setDirty();
+    // Paper start - Fix unnecessary map data saves
+        this.setColorsDirty(x, y, true);
+    }
+    public void setColorsDirty(final int x, final int y, final boolean markFileDirty) {
+        if (markFileDirty) this.setDirty();
+    // Paper end - Fix unnecessary map data saves
 
         for (MapItemSavedData.HoldingPlayer holdingPlayer : this.carriedBy) {
             holdingPlayer.markColorsDirty(x, y);
@@ -416,7 +_,7 @@
                 return true;
             }
 
-            if (!this.isTrackedCountOverLimit(256)) {
+            if (!this.isTrackedCountOverLimit(((Level) level).paperConfig().maps.itemFrameCursorLimit)) { // Paper - Limit item frame cursors on maps
                 this.bannerMarkers.put(banner.getId(), banner);
                 this.addDecoration(banner.getDecoration(), level, banner.getId(), xPos, zPos, 180.0, banner.name().orElse(null));
                 this.setDirty();
@@ -524,7 +_,7 @@
             this.player = player;
         }
 
-        private MapItemSavedData.MapPatch createPatch() {
+        private MapItemSavedData.MapPatch createPatch(final byte[] buffer) { // CraftBukkit
             int startX = this.minDirtyX;
             int startY = this.minDirtyY;
             int width = this.maxDirtyX + 1 - this.minDirtyX;
@@ -533,7 +_,7 @@
 
             for (int x = 0; x < width; x++) {
                 for (int y = 0; y < height; y++) {
-                    patch[x + y * width] = MapItemSavedData.this.colors[startX + x + (startY + y) * 128];
+                    patch[x + y * width] = buffer[startX + x + (startY + y) * 128]; // CraftBukkit
                 }
             }
 
@@ -542,17 +_,38 @@
 
         private @Nullable Packet<?> nextUpdatePacket(final MapId id) {
             MapItemSavedData.MapPatch patch;
+            // Paper start
+            if (!this.dirtyData && this.tick % 5 != 0) {
+                // this won't end up sending, so don't render it!
+                this.tick++;
+                return null;
+            }
+
+            final boolean vanillaMaps = this.shouldUseVanillaMap();
+            // Use Vanilla map renderer when possible - much simpler/faster than the CB rendering
+            org.bukkit.craftbukkit.map.RenderData render = !vanillaMaps ? MapItemSavedData.this.mapView.render((org.bukkit.craftbukkit.entity.CraftPlayer) this.player.getBukkitEntity()) : MapItemSavedData.this.vanillaRender;
+            // Paper end
             if (this.dirtyData) {
                 this.dirtyData = false;
-                patch = this.createPatch();
+                patch = this.createPatch(render.buffer); // CraftBukkit
             } else {
                 patch = null;
             }
 
             Collection<MapDecoration> decorations;
-            if (this.dirtyDecorations && this.tick++ % 5 == 0) {
+            if ((!vanillaMaps || this.dirtyDecorations) && this.tick++ % 5 == 0) { // Paper - bypass dirtyDecorations for custom maps
                 this.dirtyDecorations = false;
-                decorations = MapItemSavedData.this.decorations.values();
+                // CraftBukkit start
+                Collection<MapDecoration> icons = new java.util.ArrayList<>();
+                if (vanillaMaps) this.addSeenPlayers(icons); // Paper
+
+                for (org.bukkit.map.MapCursor cursor : render.cursors) {
+                    if (cursor.isVisible()) {
+                        icons.add(new MapDecoration(org.bukkit.craftbukkit.map.CraftMapCursor.CraftType.bukkitToMinecraftHolder(cursor.getType()), cursor.getX(), cursor.getY(), cursor.getDirection(), Optional.ofNullable(io.papermc.paper.adventure.PaperAdventure.asVanilla(cursor.caption()))));
+                    }
+                }
+                decorations = icons;
+                // CraftBukkit end
             } else {
                 decorations = null;
             }
@@ -580,6 +_,23 @@
         private void markDecorationsDirty() {
             this.dirtyDecorations = true;
         }
+
+        // Paper start
+        private void addSeenPlayers(java.util.Collection<MapDecoration> icons) {
+            org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.player.getBukkitEntity();
+            MapItemSavedData.this.decorations.forEach((name, mapIcon) -> {
+                // If this cursor is for a player check visibility with vanish system
+                org.bukkit.entity.Player other = org.bukkit.Bukkit.getPlayerExact(name); // Spigot
+                if (other == null || player.canSee(other)) {
+                    icons.add(mapIcon);
+                }
+            });
+        }
+
+        private boolean shouldUseVanillaMap() {
+            return mapView.getRenderers().size() == 1 && mapView.getRenderers().getFirst().getClass() == org.bukkit.craftbukkit.map.CraftMapRenderer.class;
+        }
+        // Paper end
     }
 
     private record MapDecorationLocation(Holder<MapDecorationType> type, byte x, byte y, byte rot) {
@@ -624,4 +_,71 @@
             }
         }
     }
+
+    // Paper start - store target world by uuid in addition to dimension
+    record UUIDAndError(java.util.UUID uuid, String faultyDimension) {
+
+    }
+    record UUIDBackedDimension(@Nullable ResourceKey<Level> resourceKey, @Nullable UUIDAndError uuid) {
+        public UUIDBackedDimension(final ResourceKey<Level> resourceKey) {
+            this(resourceKey, null);
+        }
+        public UUIDBackedDimension {
+            com.google.common.base.Preconditions.checkArgument(resourceKey != null || uuid != null, "Created uuid backed dimension with null level and uuid. This is a bug");
+        }
+
+        public ResourceKey<Level> resolveOrThrow() {
+            if (resourceKey != null) return resourceKey;
+
+            final org.bukkit.World worldByUUID = org.bukkit.Bukkit.getWorld(uuid.uuid());
+            if (worldByUUID != null) return ((org.bukkit.craftbukkit.CraftWorld) worldByUUID).getHandle().dimension();
+
+            throw new IllegalArgumentException("Invalid dimension " + uuid.faultyDimension() + " and unknown world uuid " + uuid.uuid);
+        }
+    }
+
+    private UUIDBackedDimension packUUIDBackedDimension() {
+        final net.minecraft.server.level.ServerLevel mappedLevel = net.minecraft.server.MinecraftServer.getServer().getLevel(this.dimension);
+        return new UUIDBackedDimension(this.dimension, mappedLevel == null ? null : new UUIDAndError(mappedLevel.uuid, ""));
+    }
+
+    private static com.mojang.serialization.MapCodec<UUIDBackedDimension> createUUIDBackedDimensionKeyCodec() {
+        return new com.mojang.serialization.MapCodec<>() {
+            @Override
+            public <T> java.util.stream.Stream<T> keys(final com.mojang.serialization.DynamicOps<T> ops) {
+                return java.util.stream.Stream.of("dimension", "UUIDLeast", "UUIDMost").map(ops::createString);
+            }
+
+            @Override
+            public <T> com.mojang.serialization.DataResult<UUIDBackedDimension> decode(final com.mojang.serialization.DynamicOps<T> ops,
+                                                                                       final com.mojang.serialization.MapLike<T> input) {
+                final com.mojang.serialization.DataResult<UUIDBackedDimension> foundDimension = Level.RESOURCE_KEY_CODEC.decode(ops, input.get("dimension"))
+                    .map(Pair::getFirst)
+                    .map(UUIDBackedDimension::new); // Do not pack uuid when reading as the level itself might reference an unloaded world. UUID lookup would be faulty + should be re-generated when written.
+                if (foundDimension.isSuccess()) return foundDimension;
+
+                // Fallback attempt at parsing the uuid
+                final com.mojang.serialization.DataResult<UUIDBackedDimension> fromUUID = Codec.LONG.decode(ops, input.get("UUIDMost")).map(Pair::getFirst).apply2(
+                    java.util.UUID::new,
+                    Codec.LONG.decode(ops, input.get("UUIDLeast")).map(Pair::getFirst)
+                ).map(uuid -> new UUIDBackedDimension(null, new UUIDAndError(uuid, String.valueOf(input.get("dimension")))));
+                if (fromUUID.isSuccess()) return fromUUID;
+
+                return foundDimension; // Return the found dimension instead, it's error is more "accurate" over the bukkit added uuids.
+            }
+
+            @Override
+            public <T> com.mojang.serialization.RecordBuilder<T> encode(final UUIDBackedDimension input,
+                                                                        final com.mojang.serialization.DynamicOps<T> ops,
+                                                                        final com.mojang.serialization.RecordBuilder<T> prefix) {
+                prefix.add("dimension", input.resourceKey(), Level.RESOURCE_KEY_CODEC);
+                if (input.uuid != null) {
+                    prefix.add("UUIDMost", input.uuid.uuid().getMostSignificantBits(), Codec.LONG);
+                    prefix.add("UUIDLeast", input.uuid.uuid().getLeastSignificantBits(), Codec.LONG);
+                }
+                return prefix;
+            }
+        };
+    }
+    // Paper end - store target world by uuid in addition to dimension
 }
