From 7634e42ab037ea62b1f418d5cb8b9df246f8cba0 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 14:37:42 +0100
Subject: [PATCH] paper File Patches


diff --git a/net/minecraft/server/commands/GameRuleCommand.java b/net/minecraft/server/commands/GameRuleCommand.java
index d9bfc914a65837a35c6a6e530b9217976d85a55b..ea6517b152417575be7c24bf06a75141a6061c1d 100644
--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -34,8 +34,7 @@ public class GameRuleCommand {
 
     private static <T> int setRule(final CommandContext<CommandSourceStack> context, final GameRule<T> gameRule) {
         CommandSourceStack source = context.getSource();
-        T value = context.getArgument("value", gameRule.valueClass());
-        source.getLevel().getGameRules().set(gameRule, value, context.getSource().getServer());
+        T value = org.bukkit.craftbukkit.event.CraftEventFactory.handleGameRuleSet(gameRule, context.getArgument("value", gameRule.valueClass()), source.getLevel(), source.getBukkitSender()).value(); // Paper - per-world game rules and event
         source.sendSuccess(() -> Component.translatable("commands.gamerule.set", gameRule.id(), gameRule.serialize(value)), true);
         return gameRule.getCommandResult(value);
     }
